<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>5.2 Evolutionary biology: FST | R tutorials in BIOS1140</title>
  <meta name="description" content="This is a set of R tutorials connected to the course BIOS1140 at the University of Oslo, and are meant as a companion to the book Evolutionary Genetics (Ravinet &amp; Sætre, 2018). They do not assume any prior R knowledge." />
  <meta name="generator" content="bookdown 0.21 and GitBook 2.6.7" />

  <meta property="og:title" content="5.2 Evolutionary biology: FST | R tutorials in BIOS1140" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="This is a set of R tutorials connected to the course BIOS1140 at the University of Oslo, and are meant as a companion to the book Evolutionary Genetics (Ravinet &amp; Sætre, 2018). They do not assume any prior R knowledge." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="5.2 Evolutionary biology: FST | R tutorials in BIOS1140" />
  
  <meta name="twitter:description" content="This is a set of R tutorials connected to the course BIOS1140 at the University of Oslo, and are meant as a companion to the book Evolutionary Genetics (Ravinet &amp; Sætre, 2018). They do not assume any prior R knowledge." />
  

<meta name="author" content="Mark Ravinet, Even Garvang, Solveig Brochmann" />


<meta name="date" content="2021-06-04" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="r-programming-vectorisation.html"/>
<link rel="next" href="study-questions-2.html"/>
<script src="libs/header-attrs-2.6/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">R tutorials for the course BIOS1140</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a></li>
<li class="chapter" data-level="1" data-path="introduction-to-r.html"><a href="introduction-to-r.html"><i class="fa fa-check"></i><b>1</b> Introduction to R</a>
<ul>
<li class="chapter" data-level="" data-path="introduction-to-r.html"><a href="introduction-to-r.html#what-to-expect"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="1.1" data-path="start-using-r.html"><a href="start-using-r.html"><i class="fa fa-check"></i><b>1.1</b> Start using R!</a>
<ul>
<li class="chapter" data-level="1.1.1" data-path="start-using-r.html"><a href="start-using-r.html#install-r-and-rstudio"><i class="fa fa-check"></i><b>1.1.1</b> Install R and Rstudio</a></li>
<li class="chapter" data-level="1.1.2" data-path="start-using-r.html"><a href="start-using-r.html#getting-familiar-with-r-and-the-console"><i class="fa fa-check"></i><b>1.1.2</b> Getting familiar with R and the console</a></li>
<li class="chapter" data-level="1.1.3" data-path="start-using-r.html"><a href="start-using-r.html#scripts"><i class="fa fa-check"></i><b>1.1.3</b> Scripts</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="r-essentials.html"><a href="r-essentials.html"><i class="fa fa-check"></i><b>1.2</b> R essentials</a>
<ul>
<li class="chapter" data-level="1.2.1" data-path="r-essentials.html"><a href="r-essentials.html#assigning-to-objects"><i class="fa fa-check"></i><b>1.2.1</b> Assigning to objects</a></li>
<li class="chapter" data-level="1.2.2" data-path="r-essentials.html"><a href="r-essentials.html#vectors"><i class="fa fa-check"></i><b>1.2.2</b> Vectors</a></li>
<li class="chapter" data-level="1.2.3" data-path="r-essentials.html"><a href="r-essentials.html#strings"><i class="fa fa-check"></i><b>1.2.3</b> Strings</a></li>
<li class="chapter" data-level="1.2.4" data-path="r-essentials.html"><a href="r-essentials.html#logical"><i class="fa fa-check"></i><b>1.2.4</b> Logical values</a></li>
<li class="chapter" data-level="1.2.5" data-path="r-essentials.html"><a href="r-essentials.html#functions"><i class="fa fa-check"></i><b>1.2.5</b> Functions</a></li>
<li class="chapter" data-level="1.2.6" data-path="r-essentials.html"><a href="r-essentials.html#data-frames"><i class="fa fa-check"></i><b>1.2.6</b> Data frames</a></li>
</ul></li>
<li class="chapter" data-level="1.3" data-path="dataimport.html"><a href="dataimport.html"><i class="fa fa-check"></i><b>1.3</b> Data import</a>
<ul>
<li class="chapter" data-level="1.3.1" data-path="dataimport.html"><a href="dataimport.html#data-formats"><i class="fa fa-check"></i><b>1.3.1</b> Data formats</a></li>
<li class="chapter" data-level="1.3.2" data-path="dataimport.html"><a href="dataimport.html#working-directories"><i class="fa fa-check"></i><b>1.3.2</b> Working directories</a></li>
<li class="chapter" data-level="1.3.3" data-path="dataimport.html"><a href="dataimport.html#importing"><i class="fa fa-check"></i><b>1.3.3</b> Importing</a></li>
</ul></li>
<li class="chapter" data-level="1.4" data-path="plotting.html"><a href="plotting.html"><i class="fa fa-check"></i><b>1.4</b> Plotting</a>
<ul>
<li class="chapter" data-level="1.4.1" data-path="plotting.html"><a href="plotting.html#plotting-vectors"><i class="fa fa-check"></i><b>1.4.1</b> Plotting vectors</a></li>
<li class="chapter" data-level="1.4.2" data-path="plotting.html"><a href="plotting.html#customizing-your-plots"><i class="fa fa-check"></i><b>1.4.2</b> Customizing your plots</a></li>
</ul></li>
<li class="chapter" data-level="1.5" data-path="installing-and-loading-packages.html"><a href="installing-and-loading-packages.html"><i class="fa fa-check"></i><b>1.5</b> Installing and loading packages</a></li>
<li class="chapter" data-level="1.6" data-path="going-further.html"><a href="going-further.html"><i class="fa fa-check"></i><b>1.6</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="building-on-your-foundations-going-further-with-r.html"><a href="building-on-your-foundations-going-further-with-r.html"><i class="fa fa-check"></i><b>2</b> Building on your foundations: going further with R</a>
<ul>
<li class="chapter" data-level="" data-path="building-on-your-foundations-going-further-with-r.html"><a href="building-on-your-foundations-going-further-with-r.html#what-to-expect-1"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="2.1" data-path="intro-to-data-manipulation-with-tidyverse.html"><a href="intro-to-data-manipulation-with-tidyverse.html"><i class="fa fa-check"></i><b>2.1</b> Intro to data manipulation with tidyverse</a>
<ul>
<li class="chapter" data-level="2.1.1" data-path="intro-to-data-manipulation-with-tidyverse.html"><a href="intro-to-data-manipulation-with-tidyverse.html#what-is-the-tidyverse"><i class="fa fa-check"></i><b>2.1.1</b> What is the tidyverse?</a></li>
<li class="chapter" data-level="2.1.2" data-path="intro-to-data-manipulation-with-tidyverse.html"><a href="intro-to-data-manipulation-with-tidyverse.html#the-dplyr-package"><i class="fa fa-check"></i><b>2.1.2</b> The dplyr package</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="using-dplyr-to-work-with-your-data.html"><a href="using-dplyr-to-work-with-your-data.html"><i class="fa fa-check"></i><b>2.2</b> Using dplyr to work with your data</a>
<ul>
<li class="chapter" data-level="2.2.1" data-path="using-dplyr-to-work-with-your-data.html"><a href="using-dplyr-to-work-with-your-data.html#the-pipe"><i class="fa fa-check"></i><b>2.2.1</b> The pipe</a></li>
<li class="chapter" data-level="2.2.2" data-path="using-dplyr-to-work-with-your-data.html"><a href="using-dplyr-to-work-with-your-data.html#selecting-colums-with-select"><i class="fa fa-check"></i><b>2.2.2</b> Selecting colums with <code>select()</code></a></li>
<li class="chapter" data-level="2.2.3" data-path="using-dplyr-to-work-with-your-data.html"><a href="using-dplyr-to-work-with-your-data.html#filtering-colums-using-filter"><i class="fa fa-check"></i><b>2.2.3</b> Filtering colums using <code>filter()</code></a></li>
<li class="chapter" data-level="2.2.4" data-path="using-dplyr-to-work-with-your-data.html"><a href="using-dplyr-to-work-with-your-data.html#creating-new-columns-with-mutate"><i class="fa fa-check"></i><b>2.2.4</b> Creating new columns with <code>mutate()</code></a></li>
<li class="chapter" data-level="2.2.5" data-path="using-dplyr-to-work-with-your-data.html"><a href="using-dplyr-to-work-with-your-data.html#grouped-summaries-with-group_by-and-summarise"><i class="fa fa-check"></i><b>2.2.5</b> Grouped summaries with <code>group_by()</code> and <code>summarise()</code></a></li>
<li class="chapter" data-level="2.2.6" data-path="using-dplyr-to-work-with-your-data.html"><a href="using-dplyr-to-work-with-your-data.html#using-everything-weve-learned-in-a-single-pipe-and-a-dplyr-exercise"><i class="fa fa-check"></i><b>2.2.6</b> Using everything we’ve learned in a single pipe, and a <code>dplyr</code> exercise</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="plotting-your-data-with-ggplot2.html"><a href="plotting-your-data-with-ggplot2.html"><i class="fa fa-check"></i><b>2.3</b> Plotting your data with <code>ggplot2</code></a>
<ul>
<li class="chapter" data-level="2.3.1" data-path="plotting-your-data-with-ggplot2.html"><a href="plotting-your-data-with-ggplot2.html#the-three-things-you-need-in-a-ggplot"><i class="fa fa-check"></i><b>2.3.1</b> The three things you need in a ggplot</a></li>
<li class="chapter" data-level="2.3.2" data-path="plotting-your-data-with-ggplot2.html"><a href="plotting-your-data-with-ggplot2.html#storing-ggplots-in-objects"><i class="fa fa-check"></i><b>2.3.2</b> Storing ggplots in objects</a></li>
<li class="chapter" data-level="2.3.3" data-path="plotting-your-data-with-ggplot2.html"><a href="plotting-your-data-with-ggplot2.html#customizing-your-plots-1"><i class="fa fa-check"></i><b>2.3.3</b> Customizing your plots</a></li>
<li class="chapter" data-level="2.3.4" data-path="plotting-your-data-with-ggplot2.html"><a href="plotting-your-data-with-ggplot2.html#more-ggplot-examples"><i class="fa fa-check"></i><b>2.3.4</b> More ggplot examples</a></li>
<li class="chapter" data-level="2.3.5" data-path="plotting-your-data-with-ggplot2.html"><a href="plotting-your-data-with-ggplot2.html#saving-your-plots"><i class="fa fa-check"></i><b>2.3.5</b> Saving your plots</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="reshaping-data-with-pivot-longer.html"><a href="reshaping-data-with-pivot-longer.html"><i class="fa fa-check"></i><b>2.4</b> Reshaping data with pivot_longer()</a>
<ul>
<li class="chapter" data-level="2.4.1" data-path="reshaping-data-with-pivot-longer.html"><a href="reshaping-data-with-pivot-longer.html#wide-and-long-format"><i class="fa fa-check"></i><b>2.4.1</b> Wide and long format</a></li>
<li class="chapter" data-level="2.4.2" data-path="reshaping-data-with-pivot-longer.html"><a href="reshaping-data-with-pivot-longer.html#import-example-data"><i class="fa fa-check"></i><b>2.4.2</b> Import example data</a></li>
<li class="chapter" data-level="2.4.3" data-path="reshaping-data-with-pivot-longer.html"><a href="reshaping-data-with-pivot-longer.html#reshape-the-data"><i class="fa fa-check"></i><b>2.4.3</b> Reshape the data</a></li>
<li class="chapter" data-level="2.4.4" data-path="reshaping-data-with-pivot-longer.html"><a href="reshaping-data-with-pivot-longer.html#plot-the-data"><i class="fa fa-check"></i><b>2.4.4</b> Plot the data</a></li>
</ul></li>
<li class="chapter" data-level="2.5" data-path="going-further-1.html"><a href="going-further-1.html"><i class="fa fa-check"></i><b>2.5</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="changes-in-allele-and-genotype-frequency.html"><a href="changes-in-allele-and-genotype-frequency.html"><i class="fa fa-check"></i><b>3</b> Changes in Allele and Genotype Frequency</a>
<ul>
<li class="chapter" data-level="" data-path="changes-in-allele-and-genotype-frequency.html"><a href="changes-in-allele-and-genotype-frequency.html#what-to-expect-2"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="3.1" data-path="r-programming-for-loops.html"><a href="r-programming-for-loops.html"><i class="fa fa-check"></i><b>3.1</b> R programming: for-loops</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="r-programming-for-loops.html"><a href="r-programming-for-loops.html#motivation-why-loop"><i class="fa fa-check"></i><b>3.1.1</b> Motivation: why loop?</a></li>
<li class="chapter" data-level="3.1.2" data-path="r-programming-for-loops.html"><a href="r-programming-for-loops.html#how-a-for-loop-works"><i class="fa fa-check"></i><b>3.1.2</b> How a for-loop works</a></li>
<li class="chapter" data-level="3.1.3" data-path="r-programming-for-loops.html"><a href="r-programming-for-loops.html#indexing"><i class="fa fa-check"></i><b>3.1.3</b> Indexing with for-loops</a></li>
<li class="chapter" data-level="3.1.4" data-path="r-programming-for-loops.html"><a href="r-programming-for-loops.html#solving-our-problem"><i class="fa fa-check"></i><b>3.1.4</b> Solving our problem</a></li>
<li class="chapter" data-level="3.1.5" data-path="r-programming-for-loops.html"><a href="r-programming-for-loops.html#storing-values-from-a-for-loop"><i class="fa fa-check"></i><b>3.1.5</b> Storing values from a for-loop</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="evolutionary-biology.html"><a href="evolutionary-biology.html"><i class="fa fa-check"></i><b>3.2</b> Evolutionary biology</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="evolutionary-biology.html"><a href="evolutionary-biology.html#the-hardy-weinberg-model"><i class="fa fa-check"></i><b>3.2.1</b> The Hardy-Weinberg Model</a></li>
<li class="chapter" data-level="3.2.2" data-path="evolutionary-biology.html"><a href="evolutionary-biology.html#testing-for-deviations-from-the-hardy-weinberg-expectation"><i class="fa fa-check"></i><b>3.2.2</b> Testing for deviations from the Hardy Weinberg Expectation</a></li>
<li class="chapter" data-level="3.2.3" data-path="evolutionary-biology.html"><a href="evolutionary-biology.html#simulating-genetic-drift"><i class="fa fa-check"></i><b>3.2.3</b> Simulating genetic drift</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="study-questions.html"><a href="study-questions.html"><i class="fa fa-check"></i><b>3.3</b> Study questions</a></li>
<li class="chapter" data-level="3.4" data-path="going-further-2.html"><a href="going-further-2.html"><i class="fa fa-check"></i><b>3.4</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="the-theory-of-natural-selection.html"><a href="the-theory-of-natural-selection.html"><i class="fa fa-check"></i><b>4</b> The Theory of Natural Selection</a>
<ul>
<li class="chapter" data-level="" data-path="the-theory-of-natural-selection.html"><a href="the-theory-of-natural-selection.html#what-to-expect-3"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="" data-path="the-theory-of-natural-selection.html"><a href="the-theory-of-natural-selection.html#getting-started"><i class="fa fa-check"></i>Getting started</a></li>
<li class="chapter" data-level="4.1" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html"><i class="fa fa-check"></i><b>4.1</b> R programming: Making custom functions</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#motivation"><i class="fa fa-check"></i><b>4.1.1</b> Motivation</a></li>
<li class="chapter" data-level="4.1.2" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#function-basics"><i class="fa fa-check"></i><b>4.1.2</b> Function basics</a></li>
<li class="chapter" data-level="4.1.3" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#a-simple-example"><i class="fa fa-check"></i><b>4.1.3</b> A simple example</a></li>
<li class="chapter" data-level="4.1.4" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#some-important-function-properties"><i class="fa fa-check"></i><b>4.1.4</b> Some important function properties</a></li>
<li class="chapter" data-level="4.1.5" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#a-slightly-more-useful-example"><i class="fa fa-check"></i><b>4.1.5</b> A slightly more useful example</a></li>
<li class="chapter" data-level="4.1.6" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#example-calculating-genotype-frequencies"><i class="fa fa-check"></i><b>4.1.6</b> Example: calculating genotype frequencies</a></li>
<li class="chapter" data-level="4.1.7" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#creating-a-function-of-the-drift-simulation"><i class="fa fa-check"></i><b>4.1.7</b> Creating a function of the drift simulation</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="evolutionary-biology-fitness.html"><a href="evolutionary-biology-fitness.html"><i class="fa fa-check"></i><b>4.2</b> Evolutionary biology: fitness</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="evolutionary-biology-fitness.html"><a href="evolutionary-biology-fitness.html#understanding-fitness"><i class="fa fa-check"></i><b>4.2.1</b> Understanding fitness</a></li>
<li class="chapter" data-level="4.2.2" data-path="evolutionary-biology-fitness.html"><a href="evolutionary-biology-fitness.html#one-locus-model-of-viability-selection"><i class="fa fa-check"></i><b>4.2.2</b> One-locus model of viability selection</a></li>
<li class="chapter" data-level="4.2.3" data-path="evolutionary-biology-fitness.html"><a href="evolutionary-biology-fitness.html#directional-selection"><i class="fa fa-check"></i><b>4.2.3</b> Directional selection</a></li>
<li class="chapter" data-level="4.2.4" data-path="evolutionary-biology-fitness.html"><a href="evolutionary-biology-fitness.html#over-and-underdominance"><i class="fa fa-check"></i><b>4.2.4</b> Over and underdominance</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="study-questions-1.html"><a href="study-questions-1.html"><i class="fa fa-check"></i><b>4.3</b> Study questions</a></li>
<li class="chapter" data-level="4.4" data-path="going-further-3.html"><a href="going-further-3.html"><i class="fa fa-check"></i><b>4.4</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="detecting-natural-selection.html"><a href="detecting-natural-selection.html"><i class="fa fa-check"></i><b>5</b> Detecting Natural Selection</a>
<ul>
<li class="chapter" data-level="" data-path="detecting-natural-selection.html"><a href="detecting-natural-selection.html#what-to-expect-4"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="" data-path="detecting-natural-selection.html"><a href="detecting-natural-selection.html#getting-started-1"><i class="fa fa-check"></i>Getting started</a></li>
<li class="chapter" data-level="5.1" data-path="r-programming-vectorisation.html"><a href="r-programming-vectorisation.html"><i class="fa fa-check"></i><b>5.1</b> R-programming: Vectorisation</a>
<ul>
<li class="chapter" data-level="5.1.1" data-path="r-programming-vectorisation.html"><a href="r-programming-vectorisation.html#more-on-functions-vectorisation"><i class="fa fa-check"></i><b>5.1.1</b> More on functions: vectorisation</a></li>
<li class="chapter" data-level="5.1.2" data-path="r-programming-vectorisation.html"><a href="r-programming-vectorisation.html#the-apply-function"><i class="fa fa-check"></i><b>5.1.2</b> The <code>apply()</code> function</a></li>
<li class="chapter" data-level="5.1.3" data-path="r-programming-vectorisation.html"><a href="r-programming-vectorisation.html#ifelse"><i class="fa fa-check"></i><b>5.1.3</b> The <code>ifelse()</code> function</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="evolutionary-biology-fst.html"><a href="evolutionary-biology-fst.html"><i class="fa fa-check"></i><b>5.2</b> Evolutionary biology: F<sub>ST</sub></a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="evolutionary-biology-fst.html"><a href="evolutionary-biology-fst.html#understanding-fst---the-fixation-index"><i class="fa fa-check"></i><b>5.2.1</b> Understanding <em>F</em><sub>ST</sub> - the fixation index</a></li>
<li class="chapter" data-level="5.2.2" data-path="evolutionary-biology-fst.html"><a href="evolutionary-biology-fst.html#visualising-fst-along-a-chromosome"><i class="fa fa-check"></i><b>5.2.2</b> Visualising <em>F</em><sub>ST</sub> along a chromosome</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="study-questions-2.html"><a href="study-questions-2.html"><i class="fa fa-check"></i><b>5.3</b> Study questions</a></li>
<li class="chapter" data-level="5.4" data-path="going-further-4.html"><a href="going-further-4.html"><i class="fa fa-check"></i><b>5.4</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="multilocus-evolution.html"><a href="multilocus-evolution.html"><i class="fa fa-check"></i><b>6</b> Multilocus Evolution</a>
<ul>
<li class="chapter" data-level="" data-path="multilocus-evolution.html"><a href="multilocus-evolution.html#what-to-expect-5"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="" data-path="multilocus-evolution.html"><a href="multilocus-evolution.html#getting-started-2"><i class="fa fa-check"></i>Getting started</a></li>
<li class="chapter" data-level="6.1" data-path="tools-for-investigating-differences-between-groups.html"><a href="tools-for-investigating-differences-between-groups.html"><i class="fa fa-check"></i><b>6.1</b> Tools for investigating differences between groups</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="tools-for-investigating-differences-between-groups.html"><a href="tools-for-investigating-differences-between-groups.html#exploring-group-differences"><i class="fa fa-check"></i><b>6.1.1</b> Exploring group differences</a></li>
<li class="chapter" data-level="6.1.2" data-path="tools-for-investigating-differences-between-groups.html"><a href="tools-for-investigating-differences-between-groups.html#analysis-of-variance-anova"><i class="fa fa-check"></i><b>6.1.2</b> Analysis of variance (ANOVA)</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="performing-a-qtl-analysis-in-bedbugs.html"><a href="performing-a-qtl-analysis-in-bedbugs.html"><i class="fa fa-check"></i><b>6.2</b> Performing a QTL analysis in bedbugs</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="performing-a-qtl-analysis-in-bedbugs.html"><a href="performing-a-qtl-analysis-in-bedbugs.html#bedbugs-pesticide-resistance-and-study-design"><i class="fa fa-check"></i><b>6.2.1</b> Bedbugs, pesticide resistance and study design</a></li>
<li class="chapter" data-level="6.2.2" data-path="performing-a-qtl-analysis-in-bedbugs.html"><a href="performing-a-qtl-analysis-in-bedbugs.html#linking-anova-to-our-qtl-analysis"><i class="fa fa-check"></i><b>6.2.2</b> Linking ANOVA to our QTL analysis</a></li>
<li class="chapter" data-level="6.2.3" data-path="performing-a-qtl-analysis-in-bedbugs.html"><a href="performing-a-qtl-analysis-in-bedbugs.html#testing-the-same-genotype-phenotype-association-with-anova"><i class="fa fa-check"></i><b>6.2.3</b> Testing the same genotype phenotype association with ANOVA</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="study-questions-3.html"><a href="study-questions-3.html"><i class="fa fa-check"></i><b>6.3</b> Study questions</a></li>
<li class="chapter" data-level="6.4" data-path="going-further-5.html"><a href="going-further-5.html"><i class="fa fa-check"></i><b>6.4</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="inferring-evolutionary-processes-from-sequence-data.html"><a href="inferring-evolutionary-processes-from-sequence-data.html"><i class="fa fa-check"></i><b>7</b> Inferring Evolutionary Processes from Sequence Data</a>
<ul>
<li class="chapter" data-level="" data-path="inferring-evolutionary-processes-from-sequence-data.html"><a href="inferring-evolutionary-processes-from-sequence-data.html#what-to-expect-6"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="" data-path="inferring-evolutionary-processes-from-sequence-data.html"><a href="inferring-evolutionary-processes-from-sequence-data.html#getting-started-3"><i class="fa fa-check"></i>Getting started</a></li>
<li class="chapter" data-level="7.1" data-path="working-with-dna-sequence-data.html"><a href="working-with-dna-sequence-data.html"><i class="fa fa-check"></i><b>7.1</b> Working with DNA sequence data</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="working-with-dna-sequence-data.html"><a href="working-with-dna-sequence-data.html#reading-sequence-data-into-r."><i class="fa fa-check"></i><b>7.1.1</b> Reading sequence data into R.</a></li>
<li class="chapter" data-level="7.1.2" data-path="working-with-dna-sequence-data.html"><a href="working-with-dna-sequence-data.html#exploring-dna-sequence-data"><i class="fa fa-check"></i><b>7.1.2</b> Exploring DNA sequence data</a></li>
<li class="chapter" data-level="7.1.3" data-path="working-with-dna-sequence-data.html"><a href="working-with-dna-sequence-data.html#calculating-basic-sequence-statistics"><i class="fa fa-check"></i><b>7.1.3</b> Calculating basic sequence statistics</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="working-with-a-larger-dataset.html"><a href="working-with-a-larger-dataset.html"><i class="fa fa-check"></i><b>7.2</b> Working with a larger dataset</a>
<ul>
<li class="chapter" data-level="7.2.1" data-path="working-with-a-larger-dataset.html"><a href="working-with-a-larger-dataset.html#sample-size-and-sequence-statistics"><i class="fa fa-check"></i><b>7.2.1</b> Sample size and sequence statistics</a></li>
<li class="chapter" data-level="7.2.2" data-path="working-with-a-larger-dataset.html"><a href="working-with-a-larger-dataset.html#inferring-evolutionary-processes-using-tajimas-d"><i class="fa fa-check"></i><b>7.2.2</b> Inferring evolutionary processes using Tajima’s <em>D</em></a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="calculating-statistics-at-the-whole-genome-level.html"><a href="calculating-statistics-at-the-whole-genome-level.html"><i class="fa fa-check"></i><b>7.3</b> Calculating statistics at the whole genome level</a>
<ul>
<li class="chapter" data-level="7.3.1" data-path="calculating-statistics-at-the-whole-genome-level.html"><a href="calculating-statistics-at-the-whole-genome-level.html#reading-in-variant-data"><i class="fa fa-check"></i><b>7.3.1</b> Reading in variant data</a></li>
<li class="chapter" data-level="7.3.2" data-path="calculating-statistics-at-the-whole-genome-level.html"><a href="calculating-statistics-at-the-whole-genome-level.html#calculating-nucleotide-diversity-statistics"><i class="fa fa-check"></i><b>7.3.2</b> Calculating nucleotide diversity statistics</a></li>
<li class="chapter" data-level="7.3.3" data-path="calculating-statistics-at-the-whole-genome-level.html"><a href="calculating-statistics-at-the-whole-genome-level.html#visualising-nucleotide-diversity-along-the-chromosome"><i class="fa fa-check"></i><b>7.3.3</b> Visualising nucleotide diversity along the chromosome</a></li>
<li class="chapter" data-level="7.3.4" data-path="calculating-statistics-at-the-whole-genome-level.html"><a href="calculating-statistics-at-the-whole-genome-level.html#performing-a-sliding-window-analysis"><i class="fa fa-check"></i><b>7.3.4</b> Performing a sliding window analysis</a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="study-questions-4.html"><a href="study-questions-4.html"><i class="fa fa-check"></i><b>7.4</b> Study questions</a></li>
<li class="chapter" data-level="7.5" data-path="going-further-6.html"><a href="going-further-6.html"><i class="fa fa-check"></i><b>7.5</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="speciation-genomics.html"><a href="speciation-genomics.html"><i class="fa fa-check"></i><b>8</b> Speciation Genomics</a>
<ul>
<li class="chapter" data-level="" data-path="speciation-genomics.html"><a href="speciation-genomics.html#what-to-expect-7"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="" data-path="speciation-genomics.html"><a href="speciation-genomics.html#getting-started-4"><i class="fa fa-check"></i>Getting started</a></li>
<li class="chapter" data-level="8.1" data-path="r-programming-visualizing-complex-data.html"><a href="r-programming-visualizing-complex-data.html"><i class="fa fa-check"></i><b>8.1</b> R programming: Visualizing complex data</a>
<ul>
<li class="chapter" data-level="8.1.1" data-path="r-programming-visualizing-complex-data.html"><a href="r-programming-visualizing-complex-data.html#faceting"><i class="fa fa-check"></i><b>8.1.1</b> Faceting</a></li>
</ul></li>
<li class="chapter" data-level="8.2" data-path="evolutionary-genetics-returning-to-the-sparrow-dataset.html"><a href="evolutionary-genetics-returning-to-the-sparrow-dataset.html"><i class="fa fa-check"></i><b>8.2</b> Evolutionary genetics: Returning to the sparrow dataset</a>
<ul>
<li class="chapter" data-level="8.2.1" data-path="evolutionary-genetics-returning-to-the-sparrow-dataset.html"><a href="evolutionary-genetics-returning-to-the-sparrow-dataset.html#reading-in-the-sparrow-vcf"><i class="fa fa-check"></i><b>8.2.1</b> Reading in the sparrow vcf</a></li>
<li class="chapter" data-level="8.2.2" data-path="evolutionary-genetics-returning-to-the-sparrow-dataset.html"><a href="evolutionary-genetics-returning-to-the-sparrow-dataset.html#examining-the-variant-data"><i class="fa fa-check"></i><b>8.2.2</b> Examining the variant data</a></li>
<li class="chapter" data-level="8.2.3" data-path="evolutionary-genetics-returning-to-the-sparrow-dataset.html"><a href="evolutionary-genetics-returning-to-the-sparrow-dataset.html#setting-up-sliding-windows"><i class="fa fa-check"></i><b>8.2.3</b> Setting up sliding windows</a></li>
<li class="chapter" data-level="8.2.4" data-path="evolutionary-genetics-returning-to-the-sparrow-dataset.html"><a href="evolutionary-genetics-returning-to-the-sparrow-dataset.html#calculating-sliding-window-estimates-of-nucleotide-diversity-and-differentiation"><i class="fa fa-check"></i><b>8.2.4</b> Calculating sliding window estimates of nucleotide diversity and differentiation</a></li>
<li class="chapter" data-level="8.2.5" data-path="evolutionary-genetics-returning-to-the-sparrow-dataset.html"><a href="evolutionary-genetics-returning-to-the-sparrow-dataset.html#visualising-the-data"><i class="fa fa-check"></i><b>8.2.5</b> Visualising the data</a></li>
<li class="chapter" data-level="8.2.6" data-path="evolutionary-genetics-returning-to-the-sparrow-dataset.html"><a href="evolutionary-genetics-returning-to-the-sparrow-dataset.html#interlude-relative-vs.-absolute-measures-of-nucleotide-diversity"><i class="fa fa-check"></i><b>8.2.6</b> Interlude: relative vs. absolute measures of nucleotide diversity</a></li>
<li class="chapter" data-level="8.2.7" data-path="evolutionary-genetics-returning-to-the-sparrow-dataset.html"><a href="evolutionary-genetics-returning-to-the-sparrow-dataset.html#investigating-recombination-rate-variation"><i class="fa fa-check"></i><b>8.2.7</b> Investigating recombination rate variation</a></li>
</ul></li>
<li class="chapter" data-level="8.3" data-path="study-questions-5.html"><a href="study-questions-5.html"><i class="fa fa-check"></i><b>8.3</b> Study questions</a></li>
<li class="chapter" data-level="8.4" data-path="going-further-7.html"><a href="going-further-7.html"><i class="fa fa-check"></i><b>8.4</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="reconstructing-the-past.html"><a href="reconstructing-the-past.html"><i class="fa fa-check"></i><b>9</b> Reconstructing the Past</a>
<ul>
<li class="chapter" data-level="" data-path="reconstructing-the-past.html"><a href="reconstructing-the-past.html#what-to-expect-8"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="" data-path="reconstructing-the-past.html"><a href="reconstructing-the-past.html#getting-started-5"><i class="fa fa-check"></i>Getting started</a></li>
<li class="chapter" data-level="9.1" data-path="phylogenetics-in-r.html"><a href="phylogenetics-in-r.html"><i class="fa fa-check"></i><b>9.1</b> Phylogenetics in R</a>
<ul>
<li class="chapter" data-level="9.1.1" data-path="phylogenetics-in-r.html"><a href="phylogenetics-in-r.html#storing-trees-in-r"><i class="fa fa-check"></i><b>9.1.1</b> Storing trees in R</a></li>
<li class="chapter" data-level="9.1.2" data-path="phylogenetics-in-r.html"><a href="phylogenetics-in-r.html#plotting-trees"><i class="fa fa-check"></i><b>9.1.2</b> Plotting trees</a></li>
<li class="chapter" data-level="9.1.4" data-path="phylogenetics-in-r.html"><a href="phylogenetics-in-r.html#a-simple-example-with-real-data---avian-phylogenetics"><i class="fa fa-check"></i><b>9.1.4</b> A simple example with real data - avian phylogenetics</a></li>
<li class="chapter" data-level="9.1.5" data-path="phylogenetics-in-r.html"><a href="phylogenetics-in-r.html#constructing-trees-with-r"><i class="fa fa-check"></i><b>9.1.5</b> Constructing trees with R</a></li>
</ul></li>
<li class="chapter" data-level="9.2" data-path="population-structure.html"><a href="population-structure.html"><i class="fa fa-check"></i><b>9.2</b> Population structure</a>
<ul>
<li class="chapter" data-level="9.2.1" data-path="population-structure.html"><a href="population-structure.html#village-dogs-as-an-insight-to-dog-domestication"><i class="fa fa-check"></i><b>9.2.1</b> Village dogs as an insight to dog domestication</a></li>
<li class="chapter" data-level="9.2.2" data-path="population-structure.html"><a href="population-structure.html#reading-the-data-into-r"><i class="fa fa-check"></i><b>9.2.2</b> Reading the data into R</a></li>
<li class="chapter" data-level="9.2.3" data-path="population-structure.html"><a href="population-structure.html#performing-a-pca"><i class="fa fa-check"></i><b>9.2.3</b> Performing a PCA</a></li>
<li class="chapter" data-level="9.2.4" data-path="population-structure.html"><a href="population-structure.html#visualising-the-pca"><i class="fa fa-check"></i><b>9.2.4</b> Visualising the PCA</a></li>
</ul></li>
<li class="chapter" data-level="9.3" data-path="study-questions-6.html"><a href="study-questions-6.html"><i class="fa fa-check"></i><b>9.3</b> Study questions</a></li>
<li class="chapter" data-level="9.4" data-path="going-further-8.html"><a href="going-further-8.html"><i class="fa fa-check"></i><b>9.4</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="advancing-further-in-r.html"><a href="advancing-further-in-r.html"><i class="fa fa-check"></i><b>10</b> Advancing Further in R</a>
<ul>
<li class="chapter" data-level="" data-path="advancing-further-in-r.html"><a href="advancing-further-in-r.html#what-to-expect-9"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="" data-path="advancing-further-in-r.html"><a href="advancing-further-in-r.html#getting-started-6"><i class="fa fa-check"></i>Getting started</a></li>
<li class="chapter" data-level="10.1" data-path="advanced-features-of-rstudio.html"><a href="advanced-features-of-rstudio.html"><i class="fa fa-check"></i><b>10.1</b> Advanced features of RStudio</a>
<ul>
<li class="chapter" data-level="10.1.1" data-path="advanced-features-of-rstudio.html"><a href="advanced-features-of-rstudio.html#projects-projects-projects"><i class="fa fa-check"></i><b>10.1.1</b> Projects, projects, projects</a></li>
<li class="chapter" data-level="10.1.2" data-path="advanced-features-of-rstudio.html"><a href="advanced-features-of-rstudio.html#everyone-has-a-history"><i class="fa fa-check"></i><b>10.1.2</b> Everyone has a history</a></li>
<li class="chapter" data-level="10.1.3" data-path="advanced-features-of-rstudio.html"><a href="advanced-features-of-rstudio.html#tab-complete-and-other-hotkeys"><i class="fa fa-check"></i><b>10.1.3</b> Tab complete and other hotkeys</a></li>
<li class="chapter" data-level="10.1.4" data-path="advanced-features-of-rstudio.html"><a href="advanced-features-of-rstudio.html#code-checking"><i class="fa fa-check"></i><b>10.1.4</b> Code checking</a></li>
</ul></li>
<li class="chapter" data-level="10.2" data-path="control-flow-for-else-if-and-ifelse.html"><a href="control-flow-for-else-if-and-ifelse.html"><i class="fa fa-check"></i><b>10.2</b> Control flow - <code>for</code>, <code>else</code>, <code>if</code> and <code>ifelse</code></a>
<ul>
<li class="chapter" data-level="10.2.1" data-path="control-flow-for-else-if-and-ifelse.html"><a href="control-flow-for-else-if-and-ifelse.html#if-and-else"><i class="fa fa-check"></i><b>10.2.1</b> <code>if</code> and <code>else</code></a></li>
<li class="chapter" data-level="10.2.2" data-path="control-flow-for-else-if-and-ifelse.html"><a href="control-flow-for-else-if-and-ifelse.html#ifelse---a-single-line-alternative"><i class="fa fa-check"></i><b>10.2.2</b> <code>ifelse</code> - a single line alternative</a></li>
<li class="chapter" data-level="10.2.3" data-path="control-flow-for-else-if-and-ifelse.html"><a href="control-flow-for-else-if-and-ifelse.html#for-loops"><i class="fa fa-check"></i><b>10.2.3</b> <code>for</code> loops</a></li>
</ul></li>
<li class="chapter" data-level="10.3" data-path="vectorisation.html"><a href="vectorisation.html"><i class="fa fa-check"></i><b>10.3</b> Vectorisation</a>
<ul>
<li class="chapter" data-level="10.3.1" data-path="vectorisation.html"><a href="vectorisation.html#why-is-vectorisation-preferable-in-r"><i class="fa fa-check"></i><b>10.3.1</b> Why is vectorisation preferable in R?</a></li>
<li class="chapter" data-level="10.3.2" data-path="vectorisation.html"><a href="vectorisation.html#using-sapply-instead-of-for-on-vectors"><i class="fa fa-check"></i><b>10.3.2</b> Using <code>sapply</code> instead of <code>for</code> on vectors</a></li>
<li class="chapter" data-level="10.3.3" data-path="vectorisation.html"><a href="vectorisation.html#using-apply-instead-of-for-on-matrices-and-dataframes"><i class="fa fa-check"></i><b>10.3.3</b> Using apply instead of <code>for</code> on matrices and dataframes</a></li>
</ul></li>
<li class="chapter" data-level="10.4" data-path="returning-to-writing-your-own-functions.html"><a href="returning-to-writing-your-own-functions.html"><i class="fa fa-check"></i><b>10.4</b> Returning to writing your own functions</a></li>
<li class="chapter" data-level="10.5" data-path="concluding-remarks.html"><a href="concluding-remarks.html"><i class="fa fa-check"></i><b>10.5</b> Concluding remarks</a></li>
<li class="chapter" data-level="10.6" data-path="study-questions-7.html"><a href="study-questions-7.html"><i class="fa fa-check"></i><b>10.6</b> Study questions</a></li>
<li class="chapter" data-level="10.7" data-path="going-further-9.html"><a href="going-further-9.html"><i class="fa fa-check"></i><b>10.7</b> Going further</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Made with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">R tutorials in BIOS1140</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="evolutionary-biology-fst" class="section level2" number="5.2">
<h2><span class="header-section-number">5.2</span> Evolutionary biology: F<sub>ST</sub></h2>
<script src="js/hideOutput.js"></script>
<div id="understanding-fst---the-fixation-index" class="section level3" number="5.2.1">
<h3><span class="header-section-number">5.2.1</span> Understanding <em>F</em><sub>ST</sub> - the fixation index</h3>
<div id="what-is-fst" class="section level4" number="5.2.1.1">
<h4><span class="header-section-number">5.2.1.1</span> What is <em>F</em><sub>ST</sub>?</h4>
<p><em>F</em><sub>ST</sub>, also known as the <strong>fixation index</strong>, is an extremely important statistic in population genetics, molecular ecology and evolutionary biology. It is also arguably one of the most famous population genetic statistics you will encounter. <em>F</em>~ST essentially measures the level of <strong>genetic differentiation</strong> between two or more populations. It ranges from 0 (i.e. no genetic differentiation) to 1 (complete genetic differentiation)</p>
<p>Ultimately, it is quite a simple statistic to understand but it can sometimes take time to properly grasp. So we will go over the basics properly here. One of the most confusing things about <em>F</em>~ST is that are several different ways to define it. For ease of undersanding, we will use a simple formulation:</p>
<p><span class="math inline">\(F_{ST} = \displaystyle \frac{H_T - H_S}{H_T}\)</span></p>
<p>For simplicity, imagine we are examining two populations only. With this formulate, <span class="math inline">\(H_T\)</span> is the <strong>expected heterozygosity</strong> when the two populations are considered as one large meta-population. <span class="math inline">\(H_S\)</span> is the <strong>average expected heterozygosity</strong> in the two populations.</p>
<p>You might be thinking, hang on a minute… what do we mean by expected heterozygosity? To appreciate this, we need to think back to the Hardy-Weinberg expectation we learned about in <a href="https://evolutionarygenetics.github.io/Chapter3.html">Chapter 3</a>. Remember that at a simple bi-allelic locus, <span class="math inline">\(p\)</span> and <span class="math inline">\(q\)</span> are the frequencies of the two alleles. We can calculate the expected frequency of heterozygotes with <span class="math inline">\(2pq\)</span> - this is the <strong>expected heterozygosity</strong>.</p>
</div>
<div id="a-worked-example-of-fst-in-humans" class="section level4" number="5.2.1.2">
<h4><span class="header-section-number">5.2.1.2</span> A worked example of <em>F</em><sub>ST</sub> in humans</h4>
<p>As an illustrative example, we will calculate <em>F</em><sub>ST</sub> for the SNP rs4988235 associated with lactase persistence in humans. This SNP is located ~14 Kb upstream of the <em>LCT</em> gene on Chromosome 2 and is biallelic for C/T; a T at this position is strongly associated with the ability to digest milk in adulthood. We sample 80 people each from two populations which differ in the frequency of lactase persistence - Americans of European descent and Druze people from Israel. The counts of genotypes are shown in the table below. Note that these data are modified from <a href="https://www.sciencedirect.com/science/article/pii/S0002929707628389">Bersaglieri et al. 2002</a>.</p>
<table>
<thead>
<tr class="header">
<th align="left">Population</th>
<th align="left">TT</th>
<th align="left">CT</th>
<th align="left">CC</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">American</td>
<td align="left">48</td>
<td align="left">28</td>
<td align="left">4</td>
</tr>
<tr class="even">
<td align="left">Druze</td>
<td align="left">0</td>
<td align="left">3</td>
<td align="left">77</td>
</tr>
</tbody>
</table>
<p>Knowing these numbers, we will first calculate the allele frequences in each population. We will use <span class="math inline">\(p\)</span> to denote the frequency of the T allele at this locus.</p>
<div class="sourceCode" id="cb236"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb236-1"><a href="evolutionary-biology-fst.html#cb236-1" aria-hidden="true" tabindex="-1"></a><span class="co"># set up genotype counts</span></span>
<span id="cb236-2"><a href="evolutionary-biology-fst.html#cb236-2" aria-hidden="true" tabindex="-1"></a>a <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">48</span>, <span class="dv">28</span>, <span class="dv">4</span>) <span class="co"># americans</span></span>
<span id="cb236-3"><a href="evolutionary-biology-fst.html#cb236-3" aria-hidden="true" tabindex="-1"></a>d <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">3</span>, <span class="dv">77</span>) <span class="co"># druze</span></span>
<span id="cb236-4"><a href="evolutionary-biology-fst.html#cb236-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb236-5"><a href="evolutionary-biology-fst.html#cb236-5" aria-hidden="true" tabindex="-1"></a><span class="co"># get the number of people sampled (same for both)</span></span>
<span id="cb236-6"><a href="evolutionary-biology-fst.html#cb236-6" aria-hidden="true" tabindex="-1"></a>n <span class="ot">&lt;-</span> <span class="fu">sum</span>(a)</span>
<span id="cb236-7"><a href="evolutionary-biology-fst.html#cb236-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb236-8"><a href="evolutionary-biology-fst.html#cb236-8" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate the frequency of the T allele - or p</span></span>
<span id="cb236-9"><a href="evolutionary-biology-fst.html#cb236-9" aria-hidden="true" tabindex="-1"></a><span class="co"># for americans</span></span>
<span id="cb236-10"><a href="evolutionary-biology-fst.html#cb236-10" aria-hidden="true" tabindex="-1"></a>p_a <span class="ot">&lt;-</span> ((a[<span class="dv">1</span>]<span class="sc">*</span><span class="dv">2</span>) <span class="sc">+</span> a[<span class="dv">2</span>])<span class="sc">/</span>(<span class="dv">2</span><span class="sc">*</span>n)</span>
<span id="cb236-11"><a href="evolutionary-biology-fst.html#cb236-11" aria-hidden="true" tabindex="-1"></a><span class="co"># for druze</span></span>
<span id="cb236-12"><a href="evolutionary-biology-fst.html#cb236-12" aria-hidden="true" tabindex="-1"></a>p_d <span class="ot">&lt;-</span> ((d[<span class="dv">1</span>]<span class="sc">*</span><span class="dv">2</span>) <span class="sc">+</span> d[<span class="dv">2</span>])<span class="sc">/</span>(<span class="dv">2</span><span class="sc">*</span>n)</span>
<span id="cb236-13"><a href="evolutionary-biology-fst.html#cb236-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb236-14"><a href="evolutionary-biology-fst.html#cb236-14" aria-hidden="true" tabindex="-1"></a><span class="co"># calculating the frequency of C (or q) is then trivial</span></span>
<span id="cb236-15"><a href="evolutionary-biology-fst.html#cb236-15" aria-hidden="true" tabindex="-1"></a>q_a <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">-</span> p_a</span>
<span id="cb236-16"><a href="evolutionary-biology-fst.html#cb236-16" aria-hidden="true" tabindex="-1"></a>q_d <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">-</span> p_d</span></code></pre></div>
<p>Next we can calculate the allele frequencies for the metapopulation - i.e. Americans of European descent and Druze considered as a single population. This is as simple as taking the mean of the two allele frequencies.</p>
<div class="sourceCode" id="cb237"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb237-1"><a href="evolutionary-biology-fst.html#cb237-1" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate total allele frequency</span></span>
<span id="cb237-2"><a href="evolutionary-biology-fst.html#cb237-2" aria-hidden="true" tabindex="-1"></a>p_t <span class="ot">&lt;-</span> (p_a <span class="sc">+</span> p_d)<span class="sc">/</span><span class="dv">2</span></span>
<span id="cb237-3"><a href="evolutionary-biology-fst.html#cb237-3" aria-hidden="true" tabindex="-1"></a>q_t <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">-</span> p_t</span></code></pre></div>
<p>With these allele frequencies calculated, we can very easily calculate expected heterozygosities - remember this is just <span class="math inline">\(2pq\)</span>.</p>
<div class="sourceCode" id="cb238"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb238-1"><a href="evolutionary-biology-fst.html#cb238-1" aria-hidden="true" tabindex="-1"></a><span class="co"># first calculate expected heterozygosity for the two populations</span></span>
<span id="cb238-2"><a href="evolutionary-biology-fst.html#cb238-2" aria-hidden="true" tabindex="-1"></a><span class="co"># americans</span></span>
<span id="cb238-3"><a href="evolutionary-biology-fst.html#cb238-3" aria-hidden="true" tabindex="-1"></a>hs_a <span class="ot">&lt;-</span> <span class="dv">2</span><span class="sc">*</span>p_a<span class="sc">*</span>q_a</span>
<span id="cb238-4"><a href="evolutionary-biology-fst.html#cb238-4" aria-hidden="true" tabindex="-1"></a><span class="co"># druze</span></span>
<span id="cb238-5"><a href="evolutionary-biology-fst.html#cb238-5" aria-hidden="true" tabindex="-1"></a>hs_d <span class="ot">&lt;-</span> <span class="dv">2</span><span class="sc">*</span>p_d<span class="sc">*</span>q_d</span>
<span id="cb238-6"><a href="evolutionary-biology-fst.html#cb238-6" aria-hidden="true" tabindex="-1"></a><span class="co"># then take the mean of this</span></span>
<span id="cb238-7"><a href="evolutionary-biology-fst.html#cb238-7" aria-hidden="true" tabindex="-1"></a>hs <span class="ot">&lt;-</span> (hs_a <span class="sc">+</span> hs_d)<span class="sc">/</span><span class="dv">2</span></span>
<span id="cb238-8"><a href="evolutionary-biology-fst.html#cb238-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb238-9"><a href="evolutionary-biology-fst.html#cb238-9" aria-hidden="true" tabindex="-1"></a><span class="co"># next calculate expected heterozygosity for the metapopulations</span></span>
<span id="cb238-10"><a href="evolutionary-biology-fst.html#cb238-10" aria-hidden="true" tabindex="-1"></a>ht <span class="ot">&lt;-</span> <span class="dv">2</span><span class="sc">*</span>p_t<span class="sc">*</span>q_t</span></code></pre></div>
<p>With all the relevant expected heterozygosities in place, we are now ready to calculate <em>F</em><sub>ST</sub> which we can do like so:</p>
<div class="sourceCode" id="cb239"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb239-1"><a href="evolutionary-biology-fst.html#cb239-1" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate fst</span></span>
<span id="cb239-2"><a href="evolutionary-biology-fst.html#cb239-2" aria-hidden="true" tabindex="-1"></a>fst <span class="ot">&lt;-</span> (ht <span class="sc">-</span> hs)<span class="sc">/</span>ht</span></code></pre></div>
<p>If your calculations were correct, then you should have an <em>F</em><sub>ST</sub> estimate of 0.59 - this is very high for between two human populations. One way to interpret the <em>F</em><sub>ST</sub> value we have here is that 59% of genetic variance we observe differs between populations. Since population can explain such a large difference in this case, we might expect selection to be responsible…</p>
</div>
<div id="writing-a-set-of-fst-functions" class="section level4" number="5.2.1.3">
<h4><span class="header-section-number">5.2.1.3</span> Writing a set of <em>F</em><sub>ST</sub> functions</h4>
<p>The code in the previous section was useful to demonstrate how we can calculate <em>F</em><sub>ST</sub>, but it would be a lot of work to run through this every single time we want estimate the statistic for a locus. This being R, we can of course easily create a function that will do all of the leg work for us! We will take the code we wrote out in the last section and use it here to write two functions that we can use when we want to calculate <em>F</em><sub>ST</sub>. Note that for simplicity, we will only write functions that work for <strong>two populations</strong>.</p>
<p>First, we will write a function called <code>calc_af</code> which will take genotype counts from a population and calculate allele frequencies. This will probably be similar to the function you made in last week’s assignment.</p>
<div class="sourceCode" id="cb240"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb240-1"><a href="evolutionary-biology-fst.html#cb240-1" aria-hidden="true" tabindex="-1"></a><span class="co"># a simple function to calculate allele frequencies in two populations</span></span>
<span id="cb240-2"><a href="evolutionary-biology-fst.html#cb240-2" aria-hidden="true" tabindex="-1"></a>calc_af <span class="ot">&lt;-</span> <span class="cf">function</span>(counts){</span>
<span id="cb240-3"><a href="evolutionary-biology-fst.html#cb240-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># get the number of samples</span></span>
<span id="cb240-4"><a href="evolutionary-biology-fst.html#cb240-4" aria-hidden="true" tabindex="-1"></a>  n <span class="ot">&lt;-</span> <span class="fu">sum</span>(counts)</span>
<span id="cb240-5"><a href="evolutionary-biology-fst.html#cb240-5" aria-hidden="true" tabindex="-1"></a>  <span class="co"># calculate frequency of 1st allele - p</span></span>
<span id="cb240-6"><a href="evolutionary-biology-fst.html#cb240-6" aria-hidden="true" tabindex="-1"></a>  p <span class="ot">&lt;-</span> ((counts[<span class="dv">1</span>]<span class="sc">*</span><span class="dv">2</span>) <span class="sc">+</span> counts[<span class="dv">2</span>])<span class="sc">/</span>(<span class="dv">2</span><span class="sc">*</span>n)</span>
<span id="cb240-7"><a href="evolutionary-biology-fst.html#cb240-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(p)</span>
<span id="cb240-8"><a href="evolutionary-biology-fst.html#cb240-8" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<p>Since it is very straightforward for use to calculate the frequency of the second allele once we have the frequency of the first (i.e. <span class="math inline">\(q = 1- p\)</span>), our <code>calc_af</code> function will only calculate <span class="math inline">\(p\)</span> for both populations. Let’s test it on the data from our previous example.</p>
<div class="sourceCode" id="cb241"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb241-1"><a href="evolutionary-biology-fst.html#cb241-1" aria-hidden="true" tabindex="-1"></a><span class="co"># testing our function on the american/druze example</span></span>
<span id="cb241-2"><a href="evolutionary-biology-fst.html#cb241-2" aria-hidden="true" tabindex="-1"></a>af_american <span class="ot">&lt;-</span> <span class="fu">calc_af</span>(<span class="fu">c</span>(<span class="dv">48</span>, <span class="dv">28</span>, <span class="dv">4</span>))</span>
<span id="cb241-3"><a href="evolutionary-biology-fst.html#cb241-3" aria-hidden="true" tabindex="-1"></a>af_druze <span class="ot">&lt;-</span> <span class="fu">calc_af</span>(<span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">3</span>, <span class="dv">77</span>))</span></code></pre></div>
<p>So now that we have a function that calculates allele frequencies in the two populations, we can write our <code>calc_fst</code> function to take these frequencies and calculate <em>F</em><sub>ST</sub> from them.</p>
<div class="sourceCode" id="cb242"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb242-1"><a href="evolutionary-biology-fst.html#cb242-1" aria-hidden="true" tabindex="-1"></a><span class="co"># a function to calculate fst</span></span>
<span id="cb242-2"><a href="evolutionary-biology-fst.html#cb242-2" aria-hidden="true" tabindex="-1"></a>calc_fst <span class="ot">&lt;-</span> <span class="cf">function</span>(p_1, p_2){</span>
<span id="cb242-3"><a href="evolutionary-biology-fst.html#cb242-3" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb242-4"><a href="evolutionary-biology-fst.html#cb242-4" aria-hidden="true" tabindex="-1"></a>  <span class="co"># calculate q1 and q2</span></span>
<span id="cb242-5"><a href="evolutionary-biology-fst.html#cb242-5" aria-hidden="true" tabindex="-1"></a>  q_1 <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">-</span> p_1</span>
<span id="cb242-6"><a href="evolutionary-biology-fst.html#cb242-6" aria-hidden="true" tabindex="-1"></a>  q_2 <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">-</span> p_2</span>
<span id="cb242-7"><a href="evolutionary-biology-fst.html#cb242-7" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb242-8"><a href="evolutionary-biology-fst.html#cb242-8" aria-hidden="true" tabindex="-1"></a>  <span class="co"># calculate total allele frequency</span></span>
<span id="cb242-9"><a href="evolutionary-biology-fst.html#cb242-9" aria-hidden="true" tabindex="-1"></a>  p_t <span class="ot">&lt;-</span> (p_1 <span class="sc">+</span> p_2)<span class="sc">/</span><span class="dv">2</span></span>
<span id="cb242-10"><a href="evolutionary-biology-fst.html#cb242-10" aria-hidden="true" tabindex="-1"></a>  q_t <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">-</span> p_t</span>
<span id="cb242-11"><a href="evolutionary-biology-fst.html#cb242-11" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb242-12"><a href="evolutionary-biology-fst.html#cb242-12" aria-hidden="true" tabindex="-1"></a>  <span class="co"># calculate expected heterozygosity</span></span>
<span id="cb242-13"><a href="evolutionary-biology-fst.html#cb242-13" aria-hidden="true" tabindex="-1"></a>  <span class="co"># first calculate expected heterozygosity for the two populations</span></span>
<span id="cb242-14"><a href="evolutionary-biology-fst.html#cb242-14" aria-hidden="true" tabindex="-1"></a>  <span class="co"># pop1</span></span>
<span id="cb242-15"><a href="evolutionary-biology-fst.html#cb242-15" aria-hidden="true" tabindex="-1"></a>  hs_1 <span class="ot">&lt;-</span> <span class="dv">2</span><span class="sc">*</span>p_1<span class="sc">*</span>q_1</span>
<span id="cb242-16"><a href="evolutionary-biology-fst.html#cb242-16" aria-hidden="true" tabindex="-1"></a>  <span class="co"># pop2</span></span>
<span id="cb242-17"><a href="evolutionary-biology-fst.html#cb242-17" aria-hidden="true" tabindex="-1"></a>  hs_2 <span class="ot">&lt;-</span> <span class="dv">2</span><span class="sc">*</span>p_2<span class="sc">*</span>q_2</span>
<span id="cb242-18"><a href="evolutionary-biology-fst.html#cb242-18" aria-hidden="true" tabindex="-1"></a>  <span class="co"># then take the mean of this</span></span>
<span id="cb242-19"><a href="evolutionary-biology-fst.html#cb242-19" aria-hidden="true" tabindex="-1"></a>  hs <span class="ot">&lt;-</span> (hs_1 <span class="sc">+</span> hs_2)<span class="sc">/</span><span class="dv">2</span></span>
<span id="cb242-20"><a href="evolutionary-biology-fst.html#cb242-20" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb242-21"><a href="evolutionary-biology-fst.html#cb242-21" aria-hidden="true" tabindex="-1"></a>  <span class="co"># next calculate expected heterozygosity for the metapopulations</span></span>
<span id="cb242-22"><a href="evolutionary-biology-fst.html#cb242-22" aria-hidden="true" tabindex="-1"></a>  ht <span class="ot">&lt;-</span> <span class="dv">2</span><span class="sc">*</span>p_t<span class="sc">*</span>q_t</span>
<span id="cb242-23"><a href="evolutionary-biology-fst.html#cb242-23" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb242-24"><a href="evolutionary-biology-fst.html#cb242-24" aria-hidden="true" tabindex="-1"></a>  <span class="co"># calculate fst</span></span>
<span id="cb242-25"><a href="evolutionary-biology-fst.html#cb242-25" aria-hidden="true" tabindex="-1"></a>  fst <span class="ot">&lt;-</span> (ht <span class="sc">-</span> hs)<span class="sc">/</span>ht</span>
<span id="cb242-26"><a href="evolutionary-biology-fst.html#cb242-26" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb242-27"><a href="evolutionary-biology-fst.html#cb242-27" aria-hidden="true" tabindex="-1"></a>  <span class="co"># return output</span></span>
<span id="cb242-28"><a href="evolutionary-biology-fst.html#cb242-28" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(fst)</span>
<span id="cb242-29"><a href="evolutionary-biology-fst.html#cb242-29" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<p>Let’s test our function on the allele frequencies we calculated with our <code>calc_af</code> function.</p>
<div class="sourceCode" id="cb243"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb243-1"><a href="evolutionary-biology-fst.html#cb243-1" aria-hidden="true" tabindex="-1"></a><span class="co"># testing our function on the american/druze example</span></span>
<span id="cb243-2"><a href="evolutionary-biology-fst.html#cb243-2" aria-hidden="true" tabindex="-1"></a><span class="fu">calc_fst</span>(af_american, af_druze)</span></code></pre></div>
<p>This should be the same as you got before, but with a lot less work. Next, we’ll look at applying a function to a bigger dataset with <code>apply()</code>.</p>
</div>
<div id="applying-functions-to-matrices-and-data-frames" class="section level4" number="5.2.1.4">
<h4><span class="header-section-number">5.2.1.4</span> Applying functions to matrices and data frames</h4>
<p>Extending our <em>LCT</em> and lactase persistence example, let’s get some data from multiple human populations. You can download the data <a href="https://evolutionarygenetics.github.io/lct_count.tsv">here</a></p>
<p><strong>Import the data into R using the <code>read.table()</code> function. If you’re unsure about how you do this, remember that you can go back and check in the tutorial from the first week</strong></p>
<details>
<summary>
<p>Show hint</p>
</summary>
<p>The data is separated by tabulator (<code>"\t"</code>), and has a header.</p>
</details>
<div class="sourceCode" id="cb244"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb244-1"><a href="evolutionary-biology-fst.html#cb244-1" aria-hidden="true" tabindex="-1"></a>lct_counts <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">&quot;lct_count.tsv&quot;</span>, <span class="at">header =</span> <span class="cn">TRUE</span>, <span class="at">sep =</span> <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>)</span></code></pre></div>
<p>You should now have a data frame in your R environment with allele counts for the SNP rs4988235 for 53 populations. Again, these data are all from <a href="https://www.sciencedirect.com/science/article/pii/S0002929707628389">Bersaglieri et al. 2002</a>.</p>
<p>What we have is the counts of allleles but what we actually want is the allele frequency for T - that is how we can calculate <em>F</em><sub>ST</sub>. We can use our <code>calc_af()</code> function for this, so let’s try this function out on counts for a single population. We use indexing here to select the first row and only columns 2:4, since our function is only expecting the count data, not the population name.</p>
<div class="sourceCode" id="cb245"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb245-1"><a href="evolutionary-biology-fst.html#cb245-1" aria-hidden="true" tabindex="-1"></a><span class="fu">calc_af</span>(lct_counts[<span class="dv">1</span>, <span class="dv">2</span><span class="sc">:</span><span class="dv">4</span>])</span></code></pre></div>
<p>Great! So this works well. Now let’s get <span class="math inline">\(p\)</span> (the frequency of the T allele) for all the populations. We can do this extremely fast and easily using the <code>apply()</code> function that you learned about in the R-section. Note that the data needs to be columns 2 through 4 of the data, and the second argument needs to be <code>1</code> since we’re working on rows.</p>
<div class="sourceCode" id="cb246"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb246-1"><a href="evolutionary-biology-fst.html#cb246-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">apply</span>(lct_counts[,<span class="dv">2</span><span class="sc">:</span><span class="dv">4</span>], <span class="dv">1</span>, calc_af)</span></code></pre></div>
<p>We can now combine our vector of allele frequencies with the population names to create a data.frame of frequencies. Like so</p>
<div class="sourceCode" id="cb247"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb247-1"><a href="evolutionary-biology-fst.html#cb247-1" aria-hidden="true" tabindex="-1"></a>lct_freq <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">pop =</span> lct_counts<span class="sc">$</span>pop, p)</span></code></pre></div>
<p>Now we can easily calculate a pairwise <em>F</em><sub>ST</sub> with our <code>calc_fst</code> function. For example, let’s calculate <em>F</em><sub>ST</sub> for European Americans and East Asians. We will use <code>dplyr</code> commands for this. First we make subsets of the data with <code>filter()</code>. We then end up with a single row of our data. We provide the <code>p</code>-column of that subset to <code>calc_fst</code> using <code>$</code><a href="#fn21" class="footnote-ref" id="fnref21"><sup>21</sup></a>.</p>
<div class="sourceCode" id="cb248"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb248-1"><a href="evolutionary-biology-fst.html#cb248-1" aria-hidden="true" tabindex="-1"></a><span class="co"># extract frequencies</span></span>
<span id="cb248-2"><a href="evolutionary-biology-fst.html#cb248-2" aria-hidden="true" tabindex="-1"></a>af_euram <span class="ot">&lt;-</span> <span class="fu">filter</span>(lct_freq, pop <span class="sc">==</span> <span class="st">&quot;European_American&quot;</span>)</span>
<span id="cb248-3"><a href="evolutionary-biology-fst.html#cb248-3" aria-hidden="true" tabindex="-1"></a>af_eastasian <span class="ot">&lt;-</span> <span class="fu">filter</span>(lct_freq, pop <span class="sc">==</span> <span class="st">&quot;East_Asian&quot;</span>)</span>
<span id="cb248-4"><a href="evolutionary-biology-fst.html#cb248-4" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate fst</span></span>
<span id="cb248-5"><a href="evolutionary-biology-fst.html#cb248-5" aria-hidden="true" tabindex="-1"></a><span class="fu">calc_fst</span>(af_euram<span class="sc">$</span>p, af_eastasian<span class="sc">$</span>p)</span></code></pre></div>
<p>As with our previous example, we can see <em>F</em><sub>ST</sub> is actually pretty high between these populations for this SNP. What about if we compared East Asians with the Bedouin people from Israel?</p>
<div class="sourceCode" id="cb249"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb249-1"><a href="evolutionary-biology-fst.html#cb249-1" aria-hidden="true" tabindex="-1"></a><span class="co"># extract frequencies</span></span>
<span id="cb249-2"><a href="evolutionary-biology-fst.html#cb249-2" aria-hidden="true" tabindex="-1"></a>af_bedouin <span class="ot">&lt;-</span> <span class="fu">filter</span>(lct_freq, pop <span class="sc">==</span> <span class="st">&quot;Bedouin_Negev_Israel&quot;</span>)</span>
<span id="cb249-3"><a href="evolutionary-biology-fst.html#cb249-3" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate fst</span></span>
<span id="cb249-4"><a href="evolutionary-biology-fst.html#cb249-4" aria-hidden="true" tabindex="-1"></a><span class="fu">calc_fst</span>(af_eastasian<span class="sc">$</span>p, af_bedouin<span class="sc">$</span>p)</span></code></pre></div>
<p>Here we see <em>F</em><sub>ST</sub> is substantially lower. Allele frequency differences are lower between these populations.</p>
</div>
</div>
<div id="visualising-fst-along-a-chromosome" class="section level3" number="5.2.2">
<h3><span class="header-section-number">5.2.2</span> Visualising <em>F</em><sub>ST</sub> along a chromosome</h3>
<p>Next, we will combine vectorisation and our custom functions to calculate <em>F</em><sub>ST</sub> for a series of SNPs in the vicinity of the <em>LCT</em> gene on chromosome 2. This is essentially a genome scan, an approach that can be used to detect signatures of selection in the genome. You can download the data <a href="https://evolutionarygenetics.github.io/LCT_snps.tsv">here</a></p>
<p><strong>Read in the file <code>LCT_snps.tsv</code>.</strong></p>
<div class="fold s">
<div class="sourceCode" id="cb250"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb250-1"><a href="evolutionary-biology-fst.html#cb250-1" aria-hidden="true" tabindex="-1"></a>lct_snps <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">&quot;LCT_snps.tsv&quot;</span>, <span class="at">header =</span> <span class="cn">TRUE</span>, <span class="at">sep =</span> <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>)</span></code></pre></div>
</div>
<p>This data is also from from <a href="https://www.sciencedirect.com/science/article/pii/S0002929707628389">Bersaglieri et al. 2002</a>. It is the allele frequency in various human populations for one allele at a set of 101 biallelic SNP markers close to the <em>LCT</em> gene on chromosome 2 in the human gene. Each row is a SNP and there are three frequencies - one for North Americans of European descent, one for African Americans and one for East Asians.</p>
<p>Since we have the allele frequencies, we can easily calculate <em>F</em><sub>ST</sub> for each of these SNPs. For our example here, we will do this between <code>european_americans</code> and <code>east_asians</code>. First of all, let’s use our <code>calc_fst</code> function on just a single SNP.</p>
<div class="sourceCode" id="cb251"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb251-1"><a href="evolutionary-biology-fst.html#cb251-1" aria-hidden="true" tabindex="-1"></a><span class="fu">calc_fst</span>(lct_snps[<span class="dv">1</span>, <span class="st">&quot;european_americans&quot;</span>], lct_snps[<span class="dv">1</span>, <span class="st">&quot;east_asians&quot;</span>])</span></code></pre></div>
<p>So it works great for a single row. Actually, <code>calc_fst()</code> will work with vectors of values as well, so to calculate the F<sub>ST</sub> for all loci simultaneously, we can supply the entire columns as arguments with the <code>$</code> operator.</p>
<div class="sourceCode" id="cb252"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb252-1"><a href="evolutionary-biology-fst.html#cb252-1" aria-hidden="true" tabindex="-1"></a><span class="fu">calc_fst</span>(lct_snps<span class="sc">$</span>european_americans, lct_snps<span class="sc">$</span>east_asians)</span></code></pre></div>
<p>It works equally great for the entire data set! Let’s add these F<sub>ST</sub>s as a column in the <code>lct_snps</code> data frame.</p>
<div class="sourceCode" id="cb253"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb253-1"><a href="evolutionary-biology-fst.html#cb253-1" aria-hidden="true" tabindex="-1"></a><span class="co"># make an fst column</span></span>
<span id="cb253-2"><a href="evolutionary-biology-fst.html#cb253-2" aria-hidden="true" tabindex="-1"></a>lct_snps<span class="sc">$</span>fst <span class="ot">&lt;-</span> <span class="fu">calc_fst</span>(lct_snps<span class="sc">$</span>european_americans, lct_snps<span class="sc">$</span>east_asians)</span></code></pre></div>
<p>Now that we have <em>F</em><sub>ST</sub> estimates for each of our SNPs, we can visualise the variation along the chromosome with <code>ggplot2</code>.</p>
<div class="sourceCode" id="cb254"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb254-1"><a href="evolutionary-biology-fst.html#cb254-1" aria-hidden="true" tabindex="-1"></a>a <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(lct_snps, <span class="fu">aes</span>(coord, fst)) <span class="sc">+</span> <span class="fu">geom_point</span>()</span>
<span id="cb254-2"><a href="evolutionary-biology-fst.html#cb254-2" aria-hidden="true" tabindex="-1"></a>a <span class="ot">&lt;-</span> a <span class="sc">+</span> <span class="fu">xlab</span>(<span class="st">&quot;Position (Mb)&quot;</span>) <span class="sc">+</span> <span class="fu">ylab</span>(<span class="fu">expression</span>(<span class="fu">italic</span>(F)[ST]))</span>
<span id="cb254-3"><a href="evolutionary-biology-fst.html#cb254-3" aria-hidden="true" tabindex="-1"></a>a <span class="ot">&lt;-</span> a <span class="sc">+</span> <span class="fu">theme_light</span>()</span>
<span id="cb254-4"><a href="evolutionary-biology-fst.html#cb254-4" aria-hidden="true" tabindex="-1"></a>a</span></code></pre></div>
<p><img src="Exercise5_files/figure-html/unnamed-chunk-35-1.png" width="672" /></p>
<p>What are we seeing here? Quite clearly, there is a an increase in <em>F</em><sub>ST</sub> along the chromosome, with a few SNPs showing extremely high values. It might make things a bit clearer if we mark on our plot the midpoint of the <em>LCT</em> gene. We know the gene occurs between 136,261,885 bp and 136,311,220 bp on Chromsome 2 (<a href="https://genome.ucsc.edu/cgi-bin/hgTracks?db=hg18&amp;lastVirtModeType=default&amp;lastVirtModeExtraState=&amp;virtModeType=default&amp;virtMode=0&amp;nonVirtPosition=&amp;position=chr2%3A136261885-136311220&amp;hgsid=690775201_Cttqx4oVutDVGra183ZtAUFm6z2W">from the UCSC Genome Browser</a>). So first we will find the midpoint of the gene.</p>
<div class="sourceCode" id="cb255"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb255-1"><a href="evolutionary-biology-fst.html#cb255-1" aria-hidden="true" tabindex="-1"></a><span class="co"># define the start and stop positions of the gene</span></span>
<span id="cb255-2"><a href="evolutionary-biology-fst.html#cb255-2" aria-hidden="true" tabindex="-1"></a>lct_start <span class="ot">&lt;-</span> <span class="dv">136261885</span></span>
<span id="cb255-3"><a href="evolutionary-biology-fst.html#cb255-3" aria-hidden="true" tabindex="-1"></a>lct_stop <span class="ot">&lt;-</span> <span class="dv">136311220</span></span>
<span id="cb255-4"><a href="evolutionary-biology-fst.html#cb255-4" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate the midpoint</span></span>
<span id="cb255-5"><a href="evolutionary-biology-fst.html#cb255-5" aria-hidden="true" tabindex="-1"></a>lct_mid <span class="ot">&lt;-</span> (lct_start <span class="sc">+</span> lct_stop)<span class="sc">/</span><span class="dv">2</span></span></code></pre></div>
<p>All we need to do to add it to our plot is add the <code>geom_vline()</code> geom.</p>
<div class="sourceCode" id="cb256"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb256-1"><a href="evolutionary-biology-fst.html#cb256-1" aria-hidden="true" tabindex="-1"></a>a <span class="ot">&lt;-</span> a <span class="sc">+</span> <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> lct_mid, <span class="at">lty =</span> <span class="dv">2</span>, <span class="at">col =</span> <span class="st">&quot;blue&quot;</span>)</span>
<span id="cb256-2"><a href="evolutionary-biology-fst.html#cb256-2" aria-hidden="true" tabindex="-1"></a>a</span></code></pre></div>
<p><img src="Exercise5_files/figure-html/unnamed-chunk-37-1.png" width="672" /></p>
<p>When the mid point of the gene is marked, it is clear that there is an increase in <em>F</em><sub>ST</sub> just upstream from the <em>LCT</em> gene. Perhaps we want to highlight the SNP that we calculated <em>F</em><sub>ST</sub> for in our first example?</p>
<p>This is a perfect opportunity to use the <code>ifelse()</code> function we learned about in Section <a href="r-programming-vectorisation.html#ifelse">5.1.3</a>. We can make a new column in our data based on whether or not <code>lct_snps$snp_id</code> is exactly equal to “rs4988235”.</p>
<p><strong>Exercise: add a column named <code>status</code> to <code>lct_snps</code> using <code>ifelse()</code>. The column should contain “Yes” if the id is rs4988235, and “No” if it’s not.</strong></p>
<details>
<summary>
<p>Show hint</p>
</summary>
<p>Start with <code>lct_snps$status &lt;-</code> to assign the result to a new column.</p>
</details>
<details>
<summary>
<p>Show another hint</p>
</summary>
<p>The logical statement you need to use is <code>lct_snps$snp_id == "rs4988235"</code></p>
</details>
<div class="fold s">
<div class="sourceCode" id="cb257"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb257-1"><a href="evolutionary-biology-fst.html#cb257-1" aria-hidden="true" tabindex="-1"></a>lct_snps<span class="sc">$</span>status <span class="ot">&lt;-</span> <span class="fu">if_else</span>(lct_snps<span class="sc">$</span>snp_id <span class="sc">==</span> <span class="st">&quot;rs4988235&quot;</span>, <span class="st">&quot;Yes&quot;</span>, <span class="st">&quot;No&quot;</span>)</span></code></pre></div>
</div>
<p>Now to highlight the SNP on our plot. We can make it big and colored this by mapping it to both the <code>col</code> and <code>size</code> aesthetic.</p>
<div class="sourceCode" id="cb258"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb258-1"><a href="evolutionary-biology-fst.html#cb258-1" aria-hidden="true" tabindex="-1"></a>a <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(lct_snps, <span class="fu">aes</span>(coord, fst, <span class="at">col =</span> status, <span class="at">size =</span> status)) <span class="sc">+</span> <span class="fu">geom_point</span>()</span>
<span id="cb258-2"><a href="evolutionary-biology-fst.html#cb258-2" aria-hidden="true" tabindex="-1"></a>a <span class="ot">&lt;-</span> a <span class="sc">+</span> <span class="fu">xlab</span>(<span class="st">&quot;Position (Mb)&quot;</span>) <span class="sc">+</span> <span class="fu">ylab</span>(<span class="fu">expression</span>(<span class="fu">italic</span>(F)[ST]))</span>
<span id="cb258-3"><a href="evolutionary-biology-fst.html#cb258-3" aria-hidden="true" tabindex="-1"></a>a <span class="ot">&lt;-</span> a <span class="sc">+</span> <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> lct_mid, <span class="at">lty =</span> <span class="dv">2</span>, <span class="at">col =</span> <span class="st">&quot;blue&quot;</span>)</span>
<span id="cb258-4"><a href="evolutionary-biology-fst.html#cb258-4" aria-hidden="true" tabindex="-1"></a>a <span class="ot">&lt;-</span> a <span class="sc">+</span> <span class="fu">theme_light</span>() <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">&quot;none&quot;</span>)</span>
<span id="cb258-5"><a href="evolutionary-biology-fst.html#cb258-5" aria-hidden="true" tabindex="-1"></a>a</span></code></pre></div>
<p><img src="Exercise5_files/figure-html/unnamed-chunk-39-1.png" width="672" /></p>
<p>Now we see, our focal SNP is highlighted in the plot. We’ll change the colours to make it a little bit clearer.</p>
<div class="sourceCode" id="cb259"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb259-1"><a href="evolutionary-biology-fst.html#cb259-1" aria-hidden="true" tabindex="-1"></a>a <span class="sc">+</span> <span class="fu">scale_colour_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">&quot;black&quot;</span>, <span class="st">&quot;red&quot;</span>))</span></code></pre></div>
<p><img src="Exercise5_files/figure-html/unnamed-chunk-40-1.png" width="672" /></p>
<p>In the next section, we’ll demonstrate how we can use the distribution of <em>F</em><sub>ST</sub> to identify <strong>outliers</strong> as potential targets of selection.</p>
<div id="identifying-outliers-in-our-fst-distribution" class="section level4" number="5.2.2.1">
<h4><span class="header-section-number">5.2.2.1</span> Identifying outliers in our <em>F</em><sub>ST</sub> distribution</h4>
<p>How can we identify outliers in our <em>F</em><sub>ST</sub> data? First of all, we can look at the distribution of our data by making a histogram of <code>fst</code>.</p>
<div class="sourceCode" id="cb260"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb260-1"><a href="evolutionary-biology-fst.html#cb260-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(lct_snps, <span class="fu">aes</span>(fst)) <span class="sc">+</span> <span class="fu">geom_histogram</span>(<span class="at">binwidth =</span> <span class="fl">0.05</span>)</span></code></pre></div>
<p><img src="Exercise5_files/figure-html/unnamed-chunk-41-1.png" width="672" /></p>
<p>Now, it’s apparent that some values are way larger than the rest, but where do we set the threshold? One way to do it is to set some arbitrary value, and say that all values larger than this should be considered outliers. This can for instance be that we mark the highest 5% as outliers. In R, we can get this value with the <code>quantile()</code> function.</p>
<p>In the following example, we make a vector of numbers from 0 to 100, and use the quantile function to find the highest 5% (or lowest 95%).</p>
<div class="sourceCode" id="cb261"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb261-1"><a href="evolutionary-biology-fst.html#cb261-1" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="dv">0</span><span class="sc">:</span><span class="dv">200</span></span>
<span id="cb261-2"><a href="evolutionary-biology-fst.html#cb261-2" aria-hidden="true" tabindex="-1"></a><span class="fu">quantile</span>(x, <span class="fl">0.95</span>)</span>
<span id="cb261-3"><a href="evolutionary-biology-fst.html#cb261-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 95% </span></span>
<span id="cb261-4"><a href="evolutionary-biology-fst.html#cb261-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 190</span></span></code></pre></div>
<p>The function returns <code>190</code>, which means that 95% of the values are below 190, and 5% of the values are above 190 (which shouldn’t be all that surprising).</p>
<p>Now we use the <code>quantile()</code> function on our data to set the outlier threshold. Note that this time we need to add <code>na.rm = T</code> in order to ignore some loci which have no <em>F</em><sub>ST</sub> estimates.</p>
<div class="sourceCode" id="cb262"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb262-1"><a href="evolutionary-biology-fst.html#cb262-1" aria-hidden="true" tabindex="-1"></a><span class="co"># set threshold</span></span>
<span id="cb262-2"><a href="evolutionary-biology-fst.html#cb262-2" aria-hidden="true" tabindex="-1"></a>threshold <span class="ot">&lt;-</span> <span class="fu">quantile</span>(lct_snps<span class="sc">$</span>fst, <span class="fl">0.95</span>, <span class="at">na.rm =</span> T)</span>
<span id="cb262-3"><a href="evolutionary-biology-fst.html#cb262-3" aria-hidden="true" tabindex="-1"></a><span class="co"># plot histogram with threshold marked</span></span>
<span id="cb262-4"><a href="evolutionary-biology-fst.html#cb262-4" aria-hidden="true" tabindex="-1"></a>a <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(lct_snps, <span class="fu">aes</span>(fst)) <span class="sc">+</span> <span class="fu">geom_histogram</span>(<span class="at">binwidth =</span> <span class="fl">0.05</span>)</span>
<span id="cb262-5"><a href="evolutionary-biology-fst.html#cb262-5" aria-hidden="true" tabindex="-1"></a>a <span class="sc">+</span> <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> threshold, <span class="at">colour =</span> <span class="st">&quot;red&quot;</span>, <span class="at">lty =</span> <span class="dv">2</span>, <span class="at">size =</span> <span class="dv">1</span>)</span></code></pre></div>
<p><img src="Exercise5_files/figure-html/unnamed-chunk-43-1.png" width="672" /></p>
<p>Now what if we want to visualise this on our chromosome-wide plot? Once again, we need to use the <code>ifelse()</code> function.</p>
<div class="sourceCode" id="cb263"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb263-1"><a href="evolutionary-biology-fst.html#cb263-1" aria-hidden="true" tabindex="-1"></a>lct_snps<span class="sc">$</span>outlier <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(lct_snps<span class="sc">$</span>fst <span class="sc">&gt;</span> threshold, <span class="st">&quot;Outlier&quot;</span>, <span class="st">&quot;Non-outlier&quot;</span>)</span></code></pre></div>
<p>Take a look at the <code>lct_snps</code> data frame - you should now see an additional column which is a character vector with the status of each locus as either outlier or non-outlier. Next we can incorporate this into our plotting:</p>
<div class="sourceCode" id="cb264"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb264-1"><a href="evolutionary-biology-fst.html#cb264-1" aria-hidden="true" tabindex="-1"></a>a <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(lct_snps, <span class="fu">aes</span>(coord, fst, <span class="at">colour =</span> outlier)) <span class="sc">+</span> <span class="fu">geom_point</span>()</span>
<span id="cb264-2"><a href="evolutionary-biology-fst.html#cb264-2" aria-hidden="true" tabindex="-1"></a>a <span class="ot">&lt;-</span> a <span class="sc">+</span> <span class="fu">xlab</span>(<span class="st">&quot;Position (Mb)&quot;</span>) <span class="sc">+</span> <span class="fu">ylab</span>(<span class="fu">expression</span>(<span class="fu">italic</span>(F)[ST]))</span>
<span id="cb264-3"><a href="evolutionary-biology-fst.html#cb264-3" aria-hidden="true" tabindex="-1"></a>a <span class="ot">&lt;-</span> a <span class="sc">+</span> <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> lct_mid, <span class="at">lty =</span> <span class="dv">2</span>, <span class="at">col =</span> <span class="st">&quot;blue&quot;</span>)</span>
<span id="cb264-4"><a href="evolutionary-biology-fst.html#cb264-4" aria-hidden="true" tabindex="-1"></a>a <span class="ot">&lt;-</span> a <span class="sc">+</span> <span class="fu">theme_light</span>() <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">&quot;bottom&quot;</span>)</span>
<span id="cb264-5"><a href="evolutionary-biology-fst.html#cb264-5" aria-hidden="true" tabindex="-1"></a>a</span></code></pre></div>
<p><img src="Exercise5_files/figure-html/unnamed-chunk-45-1.png" width="672" /></p>
<p>So now our potential outlier SNPs are marked on the figure. There are only 5 of them but they all occur just upstream from the <em>LCT</em> locus.</p>
<p>We cannot say for certain that these SNPs have increased <em>F</em><sub>ST</sub> values because of selection - other processes such as genetic drift or demographic history (i.e. a bottleneck in one of the two populations) might be responsible. However, given our knowledge that <em>LCT</em> is involved in lactase persistence, we can at least hypothesise that this is the case.</p>
<p>One important point to note here is that the threshold we set to identify a SNP as being potentially under selection is <strong>entirely arbitrary</strong>. In a way this line of thinking forces us to think of selection acting in some binary way on some SNPs and not others. This is obviously not the case. Still, for SNP data like this an <em>F</em><sub>ST</sub> scan can be a very useful tool.</p>
</div>
</div>
</div>
<div class="footnotes">
<hr />
<ol start="21">
<li id="fn21"><p>Tip: to better understand this code, try printing the objects <code>af_euram</code> and <code>af_eastasian</code>, as well as the columns <code>af_euram$p</code> and <code>af_eastasian$p</code>. This way you can follow what the code is doing.<a href="evolutionary-biology-fst.html#fnref21" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="r-programming-vectorisation.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="study-questions-2.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
