<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>4.1 R programming: Making custom functions | R tutorials in BIOS1140</title>
  <meta name="description" content="This is a set of R tutorials connected to the course BIOS1140 at the University of Oslo, and are meant as a companion to the book Evolutionary Genetics (Ravinet &amp; Sætre, 2018). They do not assume any prior R knowledge." />
  <meta name="generator" content="bookdown 0.21 and GitBook 2.6.7" />

  <meta property="og:title" content="4.1 R programming: Making custom functions | R tutorials in BIOS1140" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="This is a set of R tutorials connected to the course BIOS1140 at the University of Oslo, and are meant as a companion to the book Evolutionary Genetics (Ravinet &amp; Sætre, 2018). They do not assume any prior R knowledge." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="4.1 R programming: Making custom functions | R tutorials in BIOS1140" />
  
  <meta name="twitter:description" content="This is a set of R tutorials connected to the course BIOS1140 at the University of Oslo, and are meant as a companion to the book Evolutionary Genetics (Ravinet &amp; Sætre, 2018). They do not assume any prior R knowledge." />
  

<meta name="author" content="Mark Ravinet, Even Garvang, Solveig Brochmann" />


<meta name="date" content="2021-06-04" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="the-theory-of-natural-selection.html"/>
<link rel="next" href="evolutionary-biology-fitness.html"/>
<script src="libs/header-attrs-2.6/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">R tutorials for the course BIOS1140</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a></li>
<li class="chapter" data-level="1" data-path="introduction-to-r.html"><a href="introduction-to-r.html"><i class="fa fa-check"></i><b>1</b> Introduction to R</a>
<ul>
<li class="chapter" data-level="" data-path="introduction-to-r.html"><a href="introduction-to-r.html#what-to-expect"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="1.1" data-path="start-using-r.html"><a href="start-using-r.html"><i class="fa fa-check"></i><b>1.1</b> Start using R!</a>
<ul>
<li class="chapter" data-level="1.1.1" data-path="start-using-r.html"><a href="start-using-r.html#install-r-and-rstudio"><i class="fa fa-check"></i><b>1.1.1</b> Install R and Rstudio</a></li>
<li class="chapter" data-level="1.1.2" data-path="start-using-r.html"><a href="start-using-r.html#getting-familiar-with-r-and-the-console"><i class="fa fa-check"></i><b>1.1.2</b> Getting familiar with R and the console</a></li>
<li class="chapter" data-level="1.1.3" data-path="start-using-r.html"><a href="start-using-r.html#scripts"><i class="fa fa-check"></i><b>1.1.3</b> Scripts</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="r-essentials.html"><a href="r-essentials.html"><i class="fa fa-check"></i><b>1.2</b> R essentials</a>
<ul>
<li class="chapter" data-level="1.2.1" data-path="r-essentials.html"><a href="r-essentials.html#assigning-to-objects"><i class="fa fa-check"></i><b>1.2.1</b> Assigning to objects</a></li>
<li class="chapter" data-level="1.2.2" data-path="r-essentials.html"><a href="r-essentials.html#vectors"><i class="fa fa-check"></i><b>1.2.2</b> Vectors</a></li>
<li class="chapter" data-level="1.2.3" data-path="r-essentials.html"><a href="r-essentials.html#strings"><i class="fa fa-check"></i><b>1.2.3</b> Strings</a></li>
<li class="chapter" data-level="1.2.4" data-path="r-essentials.html"><a href="r-essentials.html#logical"><i class="fa fa-check"></i><b>1.2.4</b> Logical values</a></li>
<li class="chapter" data-level="1.2.5" data-path="r-essentials.html"><a href="r-essentials.html#functions"><i class="fa fa-check"></i><b>1.2.5</b> Functions</a></li>
<li class="chapter" data-level="1.2.6" data-path="r-essentials.html"><a href="r-essentials.html#data-frames"><i class="fa fa-check"></i><b>1.2.6</b> Data frames</a></li>
</ul></li>
<li class="chapter" data-level="1.3" data-path="dataimport.html"><a href="dataimport.html"><i class="fa fa-check"></i><b>1.3</b> Data import</a>
<ul>
<li class="chapter" data-level="1.3.1" data-path="dataimport.html"><a href="dataimport.html#data-formats"><i class="fa fa-check"></i><b>1.3.1</b> Data formats</a></li>
<li class="chapter" data-level="1.3.2" data-path="dataimport.html"><a href="dataimport.html#working-directories"><i class="fa fa-check"></i><b>1.3.2</b> Working directories</a></li>
<li class="chapter" data-level="1.3.3" data-path="dataimport.html"><a href="dataimport.html#importing"><i class="fa fa-check"></i><b>1.3.3</b> Importing</a></li>
</ul></li>
<li class="chapter" data-level="1.4" data-path="plotting.html"><a href="plotting.html"><i class="fa fa-check"></i><b>1.4</b> Plotting</a>
<ul>
<li class="chapter" data-level="1.4.1" data-path="plotting.html"><a href="plotting.html#plotting-vectors"><i class="fa fa-check"></i><b>1.4.1</b> Plotting vectors</a></li>
<li class="chapter" data-level="1.4.2" data-path="plotting.html"><a href="plotting.html#customizing-your-plots"><i class="fa fa-check"></i><b>1.4.2</b> Customizing your plots</a></li>
</ul></li>
<li class="chapter" data-level="1.5" data-path="installing-and-loading-packages.html"><a href="installing-and-loading-packages.html"><i class="fa fa-check"></i><b>1.5</b> Installing and loading packages</a></li>
<li class="chapter" data-level="1.6" data-path="going-further.html"><a href="going-further.html"><i class="fa fa-check"></i><b>1.6</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="building-on-your-foundations-going-further-with-r.html"><a href="building-on-your-foundations-going-further-with-r.html"><i class="fa fa-check"></i><b>2</b> Building on your foundations: going further with R</a>
<ul>
<li class="chapter" data-level="" data-path="building-on-your-foundations-going-further-with-r.html"><a href="building-on-your-foundations-going-further-with-r.html#what-to-expect-1"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="2.1" data-path="intro-to-data-manipulation-with-tidyverse.html"><a href="intro-to-data-manipulation-with-tidyverse.html"><i class="fa fa-check"></i><b>2.1</b> Intro to data manipulation with tidyverse</a>
<ul>
<li class="chapter" data-level="2.1.1" data-path="intro-to-data-manipulation-with-tidyverse.html"><a href="intro-to-data-manipulation-with-tidyverse.html#what-is-the-tidyverse"><i class="fa fa-check"></i><b>2.1.1</b> What is the tidyverse?</a></li>
<li class="chapter" data-level="2.1.2" data-path="intro-to-data-manipulation-with-tidyverse.html"><a href="intro-to-data-manipulation-with-tidyverse.html#the-dplyr-package"><i class="fa fa-check"></i><b>2.1.2</b> The dplyr package</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="using-dplyr-to-work-with-your-data.html"><a href="using-dplyr-to-work-with-your-data.html"><i class="fa fa-check"></i><b>2.2</b> Using dplyr to work with your data</a>
<ul>
<li class="chapter" data-level="2.2.1" data-path="using-dplyr-to-work-with-your-data.html"><a href="using-dplyr-to-work-with-your-data.html#the-pipe"><i class="fa fa-check"></i><b>2.2.1</b> The pipe</a></li>
<li class="chapter" data-level="2.2.2" data-path="using-dplyr-to-work-with-your-data.html"><a href="using-dplyr-to-work-with-your-data.html#selecting-colums-with-select"><i class="fa fa-check"></i><b>2.2.2</b> Selecting colums with <code>select()</code></a></li>
<li class="chapter" data-level="2.2.3" data-path="using-dplyr-to-work-with-your-data.html"><a href="using-dplyr-to-work-with-your-data.html#filtering-colums-using-filter"><i class="fa fa-check"></i><b>2.2.3</b> Filtering colums using <code>filter()</code></a></li>
<li class="chapter" data-level="2.2.4" data-path="using-dplyr-to-work-with-your-data.html"><a href="using-dplyr-to-work-with-your-data.html#creating-new-columns-with-mutate"><i class="fa fa-check"></i><b>2.2.4</b> Creating new columns with <code>mutate()</code></a></li>
<li class="chapter" data-level="2.2.5" data-path="using-dplyr-to-work-with-your-data.html"><a href="using-dplyr-to-work-with-your-data.html#grouped-summaries-with-group_by-and-summarise"><i class="fa fa-check"></i><b>2.2.5</b> Grouped summaries with <code>group_by()</code> and <code>summarise()</code></a></li>
<li class="chapter" data-level="2.2.6" data-path="using-dplyr-to-work-with-your-data.html"><a href="using-dplyr-to-work-with-your-data.html#using-everything-weve-learned-in-a-single-pipe-and-a-dplyr-exercise"><i class="fa fa-check"></i><b>2.2.6</b> Using everything we’ve learned in a single pipe, and a <code>dplyr</code> exercise</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="plotting-your-data-with-ggplot2.html"><a href="plotting-your-data-with-ggplot2.html"><i class="fa fa-check"></i><b>2.3</b> Plotting your data with <code>ggplot2</code></a>
<ul>
<li class="chapter" data-level="2.3.1" data-path="plotting-your-data-with-ggplot2.html"><a href="plotting-your-data-with-ggplot2.html#the-three-things-you-need-in-a-ggplot"><i class="fa fa-check"></i><b>2.3.1</b> The three things you need in a ggplot</a></li>
<li class="chapter" data-level="2.3.2" data-path="plotting-your-data-with-ggplot2.html"><a href="plotting-your-data-with-ggplot2.html#storing-ggplots-in-objects"><i class="fa fa-check"></i><b>2.3.2</b> Storing ggplots in objects</a></li>
<li class="chapter" data-level="2.3.3" data-path="plotting-your-data-with-ggplot2.html"><a href="plotting-your-data-with-ggplot2.html#customizing-your-plots-1"><i class="fa fa-check"></i><b>2.3.3</b> Customizing your plots</a></li>
<li class="chapter" data-level="2.3.4" data-path="plotting-your-data-with-ggplot2.html"><a href="plotting-your-data-with-ggplot2.html#more-ggplot-examples"><i class="fa fa-check"></i><b>2.3.4</b> More ggplot examples</a></li>
<li class="chapter" data-level="2.3.5" data-path="plotting-your-data-with-ggplot2.html"><a href="plotting-your-data-with-ggplot2.html#saving-your-plots"><i class="fa fa-check"></i><b>2.3.5</b> Saving your plots</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="reshaping-data-with-pivot-longer.html"><a href="reshaping-data-with-pivot-longer.html"><i class="fa fa-check"></i><b>2.4</b> Reshaping data with pivot_longer()</a>
<ul>
<li class="chapter" data-level="2.4.1" data-path="reshaping-data-with-pivot-longer.html"><a href="reshaping-data-with-pivot-longer.html#wide-and-long-format"><i class="fa fa-check"></i><b>2.4.1</b> Wide and long format</a></li>
<li class="chapter" data-level="2.4.2" data-path="reshaping-data-with-pivot-longer.html"><a href="reshaping-data-with-pivot-longer.html#import-example-data"><i class="fa fa-check"></i><b>2.4.2</b> Import example data</a></li>
<li class="chapter" data-level="2.4.3" data-path="reshaping-data-with-pivot-longer.html"><a href="reshaping-data-with-pivot-longer.html#reshape-the-data"><i class="fa fa-check"></i><b>2.4.3</b> Reshape the data</a></li>
<li class="chapter" data-level="2.4.4" data-path="reshaping-data-with-pivot-longer.html"><a href="reshaping-data-with-pivot-longer.html#plot-the-data"><i class="fa fa-check"></i><b>2.4.4</b> Plot the data</a></li>
</ul></li>
<li class="chapter" data-level="2.5" data-path="going-further-1.html"><a href="going-further-1.html"><i class="fa fa-check"></i><b>2.5</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="changes-in-allele-and-genotype-frequency.html"><a href="changes-in-allele-and-genotype-frequency.html"><i class="fa fa-check"></i><b>3</b> Changes in Allele and Genotype Frequency</a>
<ul>
<li class="chapter" data-level="" data-path="changes-in-allele-and-genotype-frequency.html"><a href="changes-in-allele-and-genotype-frequency.html#what-to-expect-2"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="3.1" data-path="r-programming-for-loops.html"><a href="r-programming-for-loops.html"><i class="fa fa-check"></i><b>3.1</b> R programming: for-loops</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="r-programming-for-loops.html"><a href="r-programming-for-loops.html#motivation-why-loop"><i class="fa fa-check"></i><b>3.1.1</b> Motivation: why loop?</a></li>
<li class="chapter" data-level="3.1.2" data-path="r-programming-for-loops.html"><a href="r-programming-for-loops.html#how-a-for-loop-works"><i class="fa fa-check"></i><b>3.1.2</b> How a for-loop works</a></li>
<li class="chapter" data-level="3.1.3" data-path="r-programming-for-loops.html"><a href="r-programming-for-loops.html#indexing"><i class="fa fa-check"></i><b>3.1.3</b> Indexing with for-loops</a></li>
<li class="chapter" data-level="3.1.4" data-path="r-programming-for-loops.html"><a href="r-programming-for-loops.html#solving-our-problem"><i class="fa fa-check"></i><b>3.1.4</b> Solving our problem</a></li>
<li class="chapter" data-level="3.1.5" data-path="r-programming-for-loops.html"><a href="r-programming-for-loops.html#storing-values-from-a-for-loop"><i class="fa fa-check"></i><b>3.1.5</b> Storing values from a for-loop</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="evolutionary-biology.html"><a href="evolutionary-biology.html"><i class="fa fa-check"></i><b>3.2</b> Evolutionary biology</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="evolutionary-biology.html"><a href="evolutionary-biology.html#the-hardy-weinberg-model"><i class="fa fa-check"></i><b>3.2.1</b> The Hardy-Weinberg Model</a></li>
<li class="chapter" data-level="3.2.2" data-path="evolutionary-biology.html"><a href="evolutionary-biology.html#testing-for-deviations-from-the-hardy-weinberg-expectation"><i class="fa fa-check"></i><b>3.2.2</b> Testing for deviations from the Hardy Weinberg Expectation</a></li>
<li class="chapter" data-level="3.2.3" data-path="evolutionary-biology.html"><a href="evolutionary-biology.html#simulating-genetic-drift"><i class="fa fa-check"></i><b>3.2.3</b> Simulating genetic drift</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="study-questions.html"><a href="study-questions.html"><i class="fa fa-check"></i><b>3.3</b> Study questions</a></li>
<li class="chapter" data-level="3.4" data-path="going-further-2.html"><a href="going-further-2.html"><i class="fa fa-check"></i><b>3.4</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="the-theory-of-natural-selection.html"><a href="the-theory-of-natural-selection.html"><i class="fa fa-check"></i><b>4</b> The Theory of Natural Selection</a>
<ul>
<li class="chapter" data-level="" data-path="the-theory-of-natural-selection.html"><a href="the-theory-of-natural-selection.html#what-to-expect-3"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="" data-path="the-theory-of-natural-selection.html"><a href="the-theory-of-natural-selection.html#getting-started"><i class="fa fa-check"></i>Getting started</a></li>
<li class="chapter" data-level="4.1" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html"><i class="fa fa-check"></i><b>4.1</b> R programming: Making custom functions</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#motivation"><i class="fa fa-check"></i><b>4.1.1</b> Motivation</a></li>
<li class="chapter" data-level="4.1.2" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#function-basics"><i class="fa fa-check"></i><b>4.1.2</b> Function basics</a></li>
<li class="chapter" data-level="4.1.3" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#a-simple-example"><i class="fa fa-check"></i><b>4.1.3</b> A simple example</a></li>
<li class="chapter" data-level="4.1.4" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#some-important-function-properties"><i class="fa fa-check"></i><b>4.1.4</b> Some important function properties</a></li>
<li class="chapter" data-level="4.1.5" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#a-slightly-more-useful-example"><i class="fa fa-check"></i><b>4.1.5</b> A slightly more useful example</a></li>
<li class="chapter" data-level="4.1.6" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#example-calculating-genotype-frequencies"><i class="fa fa-check"></i><b>4.1.6</b> Example: calculating genotype frequencies</a></li>
<li class="chapter" data-level="4.1.7" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#creating-a-function-of-the-drift-simulation"><i class="fa fa-check"></i><b>4.1.7</b> Creating a function of the drift simulation</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="evolutionary-biology-fitness.html"><a href="evolutionary-biology-fitness.html"><i class="fa fa-check"></i><b>4.2</b> Evolutionary biology: fitness</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="evolutionary-biology-fitness.html"><a href="evolutionary-biology-fitness.html#understanding-fitness"><i class="fa fa-check"></i><b>4.2.1</b> Understanding fitness</a></li>
<li class="chapter" data-level="4.2.2" data-path="evolutionary-biology-fitness.html"><a href="evolutionary-biology-fitness.html#one-locus-model-of-viability-selection"><i class="fa fa-check"></i><b>4.2.2</b> One-locus model of viability selection</a></li>
<li class="chapter" data-level="4.2.3" data-path="evolutionary-biology-fitness.html"><a href="evolutionary-biology-fitness.html#directional-selection"><i class="fa fa-check"></i><b>4.2.3</b> Directional selection</a></li>
<li class="chapter" data-level="4.2.4" data-path="evolutionary-biology-fitness.html"><a href="evolutionary-biology-fitness.html#over-and-underdominance"><i class="fa fa-check"></i><b>4.2.4</b> Over and underdominance</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="study-questions-1.html"><a href="study-questions-1.html"><i class="fa fa-check"></i><b>4.3</b> Study questions</a></li>
<li class="chapter" data-level="4.4" data-path="going-further-3.html"><a href="going-further-3.html"><i class="fa fa-check"></i><b>4.4</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="detecting-natural-selection.html"><a href="detecting-natural-selection.html"><i class="fa fa-check"></i><b>5</b> Detecting Natural Selection</a>
<ul>
<li class="chapter" data-level="" data-path="detecting-natural-selection.html"><a href="detecting-natural-selection.html#what-to-expect-4"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="" data-path="detecting-natural-selection.html"><a href="detecting-natural-selection.html#getting-started-1"><i class="fa fa-check"></i>Getting started</a></li>
<li class="chapter" data-level="5.1" data-path="r-programming-vectorisation.html"><a href="r-programming-vectorisation.html"><i class="fa fa-check"></i><b>5.1</b> R-programming: Vectorisation</a>
<ul>
<li class="chapter" data-level="5.1.1" data-path="r-programming-vectorisation.html"><a href="r-programming-vectorisation.html#more-on-functions-vectorisation"><i class="fa fa-check"></i><b>5.1.1</b> More on functions: vectorisation</a></li>
<li class="chapter" data-level="5.1.2" data-path="r-programming-vectorisation.html"><a href="r-programming-vectorisation.html#the-apply-function"><i class="fa fa-check"></i><b>5.1.2</b> The <code>apply()</code> function</a></li>
<li class="chapter" data-level="5.1.3" data-path="r-programming-vectorisation.html"><a href="r-programming-vectorisation.html#ifelse"><i class="fa fa-check"></i><b>5.1.3</b> The <code>ifelse()</code> function</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="evolutionary-biology-fst.html"><a href="evolutionary-biology-fst.html"><i class="fa fa-check"></i><b>5.2</b> Evolutionary biology: F<sub>ST</sub></a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="evolutionary-biology-fst.html"><a href="evolutionary-biology-fst.html#understanding-fst---the-fixation-index"><i class="fa fa-check"></i><b>5.2.1</b> Understanding <em>F</em><sub>ST</sub> - the fixation index</a></li>
<li class="chapter" data-level="5.2.2" data-path="evolutionary-biology-fst.html"><a href="evolutionary-biology-fst.html#visualising-fst-along-a-chromosome"><i class="fa fa-check"></i><b>5.2.2</b> Visualising <em>F</em><sub>ST</sub> along a chromosome</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="study-questions-2.html"><a href="study-questions-2.html"><i class="fa fa-check"></i><b>5.3</b> Study questions</a></li>
<li class="chapter" data-level="5.4" data-path="going-further-4.html"><a href="going-further-4.html"><i class="fa fa-check"></i><b>5.4</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="multilocus-evolution.html"><a href="multilocus-evolution.html"><i class="fa fa-check"></i><b>6</b> Multilocus Evolution</a>
<ul>
<li class="chapter" data-level="" data-path="multilocus-evolution.html"><a href="multilocus-evolution.html#what-to-expect-5"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="" data-path="multilocus-evolution.html"><a href="multilocus-evolution.html#getting-started-2"><i class="fa fa-check"></i>Getting started</a></li>
<li class="chapter" data-level="6.1" data-path="tools-for-investigating-differences-between-groups.html"><a href="tools-for-investigating-differences-between-groups.html"><i class="fa fa-check"></i><b>6.1</b> Tools for investigating differences between groups</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="tools-for-investigating-differences-between-groups.html"><a href="tools-for-investigating-differences-between-groups.html#exploring-group-differences"><i class="fa fa-check"></i><b>6.1.1</b> Exploring group differences</a></li>
<li class="chapter" data-level="6.1.2" data-path="tools-for-investigating-differences-between-groups.html"><a href="tools-for-investigating-differences-between-groups.html#analysis-of-variance-anova"><i class="fa fa-check"></i><b>6.1.2</b> Analysis of variance (ANOVA)</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="performing-a-qtl-analysis-in-bedbugs.html"><a href="performing-a-qtl-analysis-in-bedbugs.html"><i class="fa fa-check"></i><b>6.2</b> Performing a QTL analysis in bedbugs</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="performing-a-qtl-analysis-in-bedbugs.html"><a href="performing-a-qtl-analysis-in-bedbugs.html#bedbugs-pesticide-resistance-and-study-design"><i class="fa fa-check"></i><b>6.2.1</b> Bedbugs, pesticide resistance and study design</a></li>
<li class="chapter" data-level="6.2.2" data-path="performing-a-qtl-analysis-in-bedbugs.html"><a href="performing-a-qtl-analysis-in-bedbugs.html#linking-anova-to-our-qtl-analysis"><i class="fa fa-check"></i><b>6.2.2</b> Linking ANOVA to our QTL analysis</a></li>
<li class="chapter" data-level="6.2.3" data-path="performing-a-qtl-analysis-in-bedbugs.html"><a href="performing-a-qtl-analysis-in-bedbugs.html#testing-the-same-genotype-phenotype-association-with-anova"><i class="fa fa-check"></i><b>6.2.3</b> Testing the same genotype phenotype association with ANOVA</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="study-questions-3.html"><a href="study-questions-3.html"><i class="fa fa-check"></i><b>6.3</b> Study questions</a></li>
<li class="chapter" data-level="6.4" data-path="going-further-5.html"><a href="going-further-5.html"><i class="fa fa-check"></i><b>6.4</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="inferring-evolutionary-processes-from-sequence-data.html"><a href="inferring-evolutionary-processes-from-sequence-data.html"><i class="fa fa-check"></i><b>7</b> Inferring Evolutionary Processes from Sequence Data</a>
<ul>
<li class="chapter" data-level="" data-path="inferring-evolutionary-processes-from-sequence-data.html"><a href="inferring-evolutionary-processes-from-sequence-data.html#what-to-expect-6"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="" data-path="inferring-evolutionary-processes-from-sequence-data.html"><a href="inferring-evolutionary-processes-from-sequence-data.html#getting-started-3"><i class="fa fa-check"></i>Getting started</a></li>
<li class="chapter" data-level="7.1" data-path="working-with-dna-sequence-data.html"><a href="working-with-dna-sequence-data.html"><i class="fa fa-check"></i><b>7.1</b> Working with DNA sequence data</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="working-with-dna-sequence-data.html"><a href="working-with-dna-sequence-data.html#reading-sequence-data-into-r."><i class="fa fa-check"></i><b>7.1.1</b> Reading sequence data into R.</a></li>
<li class="chapter" data-level="7.1.2" data-path="working-with-dna-sequence-data.html"><a href="working-with-dna-sequence-data.html#exploring-dna-sequence-data"><i class="fa fa-check"></i><b>7.1.2</b> Exploring DNA sequence data</a></li>
<li class="chapter" data-level="7.1.3" data-path="working-with-dna-sequence-data.html"><a href="working-with-dna-sequence-data.html#calculating-basic-sequence-statistics"><i class="fa fa-check"></i><b>7.1.3</b> Calculating basic sequence statistics</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="working-with-a-larger-dataset.html"><a href="working-with-a-larger-dataset.html"><i class="fa fa-check"></i><b>7.2</b> Working with a larger dataset</a>
<ul>
<li class="chapter" data-level="7.2.1" data-path="working-with-a-larger-dataset.html"><a href="working-with-a-larger-dataset.html#sample-size-and-sequence-statistics"><i class="fa fa-check"></i><b>7.2.1</b> Sample size and sequence statistics</a></li>
<li class="chapter" data-level="7.2.2" data-path="working-with-a-larger-dataset.html"><a href="working-with-a-larger-dataset.html#inferring-evolutionary-processes-using-tajimas-d"><i class="fa fa-check"></i><b>7.2.2</b> Inferring evolutionary processes using Tajima’s <em>D</em></a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="calculating-statistics-at-the-whole-genome-level.html"><a href="calculating-statistics-at-the-whole-genome-level.html"><i class="fa fa-check"></i><b>7.3</b> Calculating statistics at the whole genome level</a>
<ul>
<li class="chapter" data-level="7.3.1" data-path="calculating-statistics-at-the-whole-genome-level.html"><a href="calculating-statistics-at-the-whole-genome-level.html#reading-in-variant-data"><i class="fa fa-check"></i><b>7.3.1</b> Reading in variant data</a></li>
<li class="chapter" data-level="7.3.2" data-path="calculating-statistics-at-the-whole-genome-level.html"><a href="calculating-statistics-at-the-whole-genome-level.html#calculating-nucleotide-diversity-statistics"><i class="fa fa-check"></i><b>7.3.2</b> Calculating nucleotide diversity statistics</a></li>
<li class="chapter" data-level="7.3.3" data-path="calculating-statistics-at-the-whole-genome-level.html"><a href="calculating-statistics-at-the-whole-genome-level.html#visualising-nucleotide-diversity-along-the-chromosome"><i class="fa fa-check"></i><b>7.3.3</b> Visualising nucleotide diversity along the chromosome</a></li>
<li class="chapter" data-level="7.3.4" data-path="calculating-statistics-at-the-whole-genome-level.html"><a href="calculating-statistics-at-the-whole-genome-level.html#performing-a-sliding-window-analysis"><i class="fa fa-check"></i><b>7.3.4</b> Performing a sliding window analysis</a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="study-questions-4.html"><a href="study-questions-4.html"><i class="fa fa-check"></i><b>7.4</b> Study questions</a></li>
<li class="chapter" data-level="7.5" data-path="going-further-6.html"><a href="going-further-6.html"><i class="fa fa-check"></i><b>7.5</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="speciation-genomics.html"><a href="speciation-genomics.html"><i class="fa fa-check"></i><b>8</b> Speciation Genomics</a>
<ul>
<li class="chapter" data-level="" data-path="speciation-genomics.html"><a href="speciation-genomics.html#what-to-expect-7"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="" data-path="speciation-genomics.html"><a href="speciation-genomics.html#getting-started-4"><i class="fa fa-check"></i>Getting started</a></li>
<li class="chapter" data-level="8.1" data-path="r-programming-visualizing-complex-data.html"><a href="r-programming-visualizing-complex-data.html"><i class="fa fa-check"></i><b>8.1</b> R programming: Visualizing complex data</a>
<ul>
<li class="chapter" data-level="8.1.1" data-path="r-programming-visualizing-complex-data.html"><a href="r-programming-visualizing-complex-data.html#faceting"><i class="fa fa-check"></i><b>8.1.1</b> Faceting</a></li>
</ul></li>
<li class="chapter" data-level="8.2" data-path="evolutionary-genetics-returning-to-the-sparrow-dataset.html"><a href="evolutionary-genetics-returning-to-the-sparrow-dataset.html"><i class="fa fa-check"></i><b>8.2</b> Evolutionary genetics: Returning to the sparrow dataset</a>
<ul>
<li class="chapter" data-level="8.2.1" data-path="evolutionary-genetics-returning-to-the-sparrow-dataset.html"><a href="evolutionary-genetics-returning-to-the-sparrow-dataset.html#reading-in-the-sparrow-vcf"><i class="fa fa-check"></i><b>8.2.1</b> Reading in the sparrow vcf</a></li>
<li class="chapter" data-level="8.2.2" data-path="evolutionary-genetics-returning-to-the-sparrow-dataset.html"><a href="evolutionary-genetics-returning-to-the-sparrow-dataset.html#examining-the-variant-data"><i class="fa fa-check"></i><b>8.2.2</b> Examining the variant data</a></li>
<li class="chapter" data-level="8.2.3" data-path="evolutionary-genetics-returning-to-the-sparrow-dataset.html"><a href="evolutionary-genetics-returning-to-the-sparrow-dataset.html#setting-up-sliding-windows"><i class="fa fa-check"></i><b>8.2.3</b> Setting up sliding windows</a></li>
<li class="chapter" data-level="8.2.4" data-path="evolutionary-genetics-returning-to-the-sparrow-dataset.html"><a href="evolutionary-genetics-returning-to-the-sparrow-dataset.html#calculating-sliding-window-estimates-of-nucleotide-diversity-and-differentiation"><i class="fa fa-check"></i><b>8.2.4</b> Calculating sliding window estimates of nucleotide diversity and differentiation</a></li>
<li class="chapter" data-level="8.2.5" data-path="evolutionary-genetics-returning-to-the-sparrow-dataset.html"><a href="evolutionary-genetics-returning-to-the-sparrow-dataset.html#visualising-the-data"><i class="fa fa-check"></i><b>8.2.5</b> Visualising the data</a></li>
<li class="chapter" data-level="8.2.6" data-path="evolutionary-genetics-returning-to-the-sparrow-dataset.html"><a href="evolutionary-genetics-returning-to-the-sparrow-dataset.html#interlude-relative-vs.-absolute-measures-of-nucleotide-diversity"><i class="fa fa-check"></i><b>8.2.6</b> Interlude: relative vs. absolute measures of nucleotide diversity</a></li>
<li class="chapter" data-level="8.2.7" data-path="evolutionary-genetics-returning-to-the-sparrow-dataset.html"><a href="evolutionary-genetics-returning-to-the-sparrow-dataset.html#investigating-recombination-rate-variation"><i class="fa fa-check"></i><b>8.2.7</b> Investigating recombination rate variation</a></li>
</ul></li>
<li class="chapter" data-level="8.3" data-path="study-questions-5.html"><a href="study-questions-5.html"><i class="fa fa-check"></i><b>8.3</b> Study questions</a></li>
<li class="chapter" data-level="8.4" data-path="going-further-7.html"><a href="going-further-7.html"><i class="fa fa-check"></i><b>8.4</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="reconstructing-the-past.html"><a href="reconstructing-the-past.html"><i class="fa fa-check"></i><b>9</b> Reconstructing the Past</a>
<ul>
<li class="chapter" data-level="" data-path="reconstructing-the-past.html"><a href="reconstructing-the-past.html#what-to-expect-8"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="" data-path="reconstructing-the-past.html"><a href="reconstructing-the-past.html#getting-started-5"><i class="fa fa-check"></i>Getting started</a></li>
<li class="chapter" data-level="9.1" data-path="phylogenetics-in-r.html"><a href="phylogenetics-in-r.html"><i class="fa fa-check"></i><b>9.1</b> Phylogenetics in R</a>
<ul>
<li class="chapter" data-level="9.1.1" data-path="phylogenetics-in-r.html"><a href="phylogenetics-in-r.html#storing-trees-in-r"><i class="fa fa-check"></i><b>9.1.1</b> Storing trees in R</a></li>
<li class="chapter" data-level="9.1.2" data-path="phylogenetics-in-r.html"><a href="phylogenetics-in-r.html#plotting-trees"><i class="fa fa-check"></i><b>9.1.2</b> Plotting trees</a></li>
<li class="chapter" data-level="9.1.4" data-path="phylogenetics-in-r.html"><a href="phylogenetics-in-r.html#a-simple-example-with-real-data---avian-phylogenetics"><i class="fa fa-check"></i><b>9.1.4</b> A simple example with real data - avian phylogenetics</a></li>
<li class="chapter" data-level="9.1.5" data-path="phylogenetics-in-r.html"><a href="phylogenetics-in-r.html#constructing-trees-with-r"><i class="fa fa-check"></i><b>9.1.5</b> Constructing trees with R</a></li>
</ul></li>
<li class="chapter" data-level="9.2" data-path="population-structure.html"><a href="population-structure.html"><i class="fa fa-check"></i><b>9.2</b> Population structure</a>
<ul>
<li class="chapter" data-level="9.2.1" data-path="population-structure.html"><a href="population-structure.html#village-dogs-as-an-insight-to-dog-domestication"><i class="fa fa-check"></i><b>9.2.1</b> Village dogs as an insight to dog domestication</a></li>
<li class="chapter" data-level="9.2.2" data-path="population-structure.html"><a href="population-structure.html#reading-the-data-into-r"><i class="fa fa-check"></i><b>9.2.2</b> Reading the data into R</a></li>
<li class="chapter" data-level="9.2.3" data-path="population-structure.html"><a href="population-structure.html#performing-a-pca"><i class="fa fa-check"></i><b>9.2.3</b> Performing a PCA</a></li>
<li class="chapter" data-level="9.2.4" data-path="population-structure.html"><a href="population-structure.html#visualising-the-pca"><i class="fa fa-check"></i><b>9.2.4</b> Visualising the PCA</a></li>
</ul></li>
<li class="chapter" data-level="9.3" data-path="study-questions-6.html"><a href="study-questions-6.html"><i class="fa fa-check"></i><b>9.3</b> Study questions</a></li>
<li class="chapter" data-level="9.4" data-path="going-further-8.html"><a href="going-further-8.html"><i class="fa fa-check"></i><b>9.4</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="advancing-further-in-r.html"><a href="advancing-further-in-r.html"><i class="fa fa-check"></i><b>10</b> Advancing Further in R</a>
<ul>
<li class="chapter" data-level="" data-path="advancing-further-in-r.html"><a href="advancing-further-in-r.html#what-to-expect-9"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="" data-path="advancing-further-in-r.html"><a href="advancing-further-in-r.html#getting-started-6"><i class="fa fa-check"></i>Getting started</a></li>
<li class="chapter" data-level="10.1" data-path="advanced-features-of-rstudio.html"><a href="advanced-features-of-rstudio.html"><i class="fa fa-check"></i><b>10.1</b> Advanced features of RStudio</a>
<ul>
<li class="chapter" data-level="10.1.1" data-path="advanced-features-of-rstudio.html"><a href="advanced-features-of-rstudio.html#projects-projects-projects"><i class="fa fa-check"></i><b>10.1.1</b> Projects, projects, projects</a></li>
<li class="chapter" data-level="10.1.2" data-path="advanced-features-of-rstudio.html"><a href="advanced-features-of-rstudio.html#everyone-has-a-history"><i class="fa fa-check"></i><b>10.1.2</b> Everyone has a history</a></li>
<li class="chapter" data-level="10.1.3" data-path="advanced-features-of-rstudio.html"><a href="advanced-features-of-rstudio.html#tab-complete-and-other-hotkeys"><i class="fa fa-check"></i><b>10.1.3</b> Tab complete and other hotkeys</a></li>
<li class="chapter" data-level="10.1.4" data-path="advanced-features-of-rstudio.html"><a href="advanced-features-of-rstudio.html#code-checking"><i class="fa fa-check"></i><b>10.1.4</b> Code checking</a></li>
</ul></li>
<li class="chapter" data-level="10.2" data-path="control-flow-for-else-if-and-ifelse.html"><a href="control-flow-for-else-if-and-ifelse.html"><i class="fa fa-check"></i><b>10.2</b> Control flow - <code>for</code>, <code>else</code>, <code>if</code> and <code>ifelse</code></a>
<ul>
<li class="chapter" data-level="10.2.1" data-path="control-flow-for-else-if-and-ifelse.html"><a href="control-flow-for-else-if-and-ifelse.html#if-and-else"><i class="fa fa-check"></i><b>10.2.1</b> <code>if</code> and <code>else</code></a></li>
<li class="chapter" data-level="10.2.2" data-path="control-flow-for-else-if-and-ifelse.html"><a href="control-flow-for-else-if-and-ifelse.html#ifelse---a-single-line-alternative"><i class="fa fa-check"></i><b>10.2.2</b> <code>ifelse</code> - a single line alternative</a></li>
<li class="chapter" data-level="10.2.3" data-path="control-flow-for-else-if-and-ifelse.html"><a href="control-flow-for-else-if-and-ifelse.html#for-loops"><i class="fa fa-check"></i><b>10.2.3</b> <code>for</code> loops</a></li>
</ul></li>
<li class="chapter" data-level="10.3" data-path="vectorisation.html"><a href="vectorisation.html"><i class="fa fa-check"></i><b>10.3</b> Vectorisation</a>
<ul>
<li class="chapter" data-level="10.3.1" data-path="vectorisation.html"><a href="vectorisation.html#why-is-vectorisation-preferable-in-r"><i class="fa fa-check"></i><b>10.3.1</b> Why is vectorisation preferable in R?</a></li>
<li class="chapter" data-level="10.3.2" data-path="vectorisation.html"><a href="vectorisation.html#using-sapply-instead-of-for-on-vectors"><i class="fa fa-check"></i><b>10.3.2</b> Using <code>sapply</code> instead of <code>for</code> on vectors</a></li>
<li class="chapter" data-level="10.3.3" data-path="vectorisation.html"><a href="vectorisation.html#using-apply-instead-of-for-on-matrices-and-dataframes"><i class="fa fa-check"></i><b>10.3.3</b> Using apply instead of <code>for</code> on matrices and dataframes</a></li>
</ul></li>
<li class="chapter" data-level="10.4" data-path="returning-to-writing-your-own-functions.html"><a href="returning-to-writing-your-own-functions.html"><i class="fa fa-check"></i><b>10.4</b> Returning to writing your own functions</a></li>
<li class="chapter" data-level="10.5" data-path="concluding-remarks.html"><a href="concluding-remarks.html"><i class="fa fa-check"></i><b>10.5</b> Concluding remarks</a></li>
<li class="chapter" data-level="10.6" data-path="study-questions-7.html"><a href="study-questions-7.html"><i class="fa fa-check"></i><b>10.6</b> Study questions</a></li>
<li class="chapter" data-level="10.7" data-path="going-further-9.html"><a href="going-further-9.html"><i class="fa fa-check"></i><b>10.7</b> Going further</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Made with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">R tutorials in BIOS1140</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="r-programming-making-custom-functions" class="section level2" number="4.1">
<h2><span class="header-section-number">4.1</span> R programming: Making custom functions</h2>
<script src="js/hideOutput.js"></script>
<div id="motivation" class="section level3" number="4.1.1">
<h3><span class="header-section-number">4.1.1</span> Motivation</h3>
<p>In last week’s tutorial and assignment, you used a for-loop to simulate genetic drift. If you did the assignment, you are probably tired of seeing code-snippets looking like this or similar:</p>
<div class="sourceCode" id="cb176"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb176-1"><a href="r-programming-making-custom-functions.html#cb176-1" aria-hidden="true" tabindex="-1"></a>N <span class="ot">&lt;-</span> <span class="dv">10</span></span>
<span id="cb176-2"><a href="r-programming-making-custom-functions.html#cb176-2" aria-hidden="true" tabindex="-1"></a>ngen <span class="ot">&lt;-</span> <span class="dv">1000</span></span>
<span id="cb176-3"><a href="r-programming-making-custom-functions.html#cb176-3" aria-hidden="true" tabindex="-1"></a>p_init <span class="ot">&lt;-</span> <span class="fl">0.5</span></span>
<span id="cb176-4"><a href="r-programming-making-custom-functions.html#cb176-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb176-5"><a href="r-programming-making-custom-functions.html#cb176-5" aria-hidden="true" tabindex="-1"></a>n10 <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="cn">NA</span>, ngen)</span>
<span id="cb176-6"><a href="r-programming-making-custom-functions.html#cb176-6" aria-hidden="true" tabindex="-1"></a>n10[<span class="dv">1</span>] <span class="ot">&lt;-</span> p_init</span>
<span id="cb176-7"><a href="r-programming-making-custom-functions.html#cb176-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb176-8"><a href="r-programming-making-custom-functions.html#cb176-8" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">2</span><span class="sc">:</span>ngen){</span>
<span id="cb176-9"><a href="r-programming-making-custom-functions.html#cb176-9" aria-hidden="true" tabindex="-1"></a>  nA1 <span class="ot">&lt;-</span> <span class="fu">rbinom</span>(<span class="dv">1</span>, <span class="dv">2</span><span class="sc">*</span>N, n10[i<span class="dv">-1</span>])</span>
<span id="cb176-10"><a href="r-programming-making-custom-functions.html#cb176-10" aria-hidden="true" tabindex="-1"></a>  n10[i] <span class="ot">&lt;-</span> nA1 <span class="sc">/</span> (<span class="dv">2</span><span class="sc">*</span>N)</span>
<span id="cb176-11"><a href="r-programming-making-custom-functions.html#cb176-11" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<p>Although the for-loop saved you thousands of lines of code compared to writing out the simulation manually, you still had to copy and paste the for loop itself many times. You had to write it 3 times for the tutorial and another 3 times for the assignment, just changing a single number each time! You may have thought that there has to be a better way of doing this, and there is! You can make this simpler by making your own custom functions, which you will learn about in the R-part of this tutorial.</p>
</div>
<div id="function-basics" class="section level3" number="4.1.2">
<h3><span class="header-section-number">4.1.2</span> Function basics</h3>
<p>You have already used a lot of functions in R, with names like <code>mean()</code>, <code>sum()</code> and <code>ggplot()</code>. Common for these is that they take some input—arguments—does something with that input and returns the result. For making your own function, you need three things:</p>
<ul>
<li>A name for your function. This could be anything, but it’s best to give them a sensible name.</li>
<li>The arguments to your function. These are the values that you want to be able to change each time you run the function.</li>
<li>What the function does with the arguments and what it returns. This is called the “body” of the function.</li>
</ul>
<p>A basic custom function typically looks like this:</p>
<div class="sourceCode" id="cb177"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb177-1"><a href="r-programming-making-custom-functions.html#cb177-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb177-2"><a href="r-programming-making-custom-functions.html#cb177-2" aria-hidden="true" tabindex="-1"></a>function_name <span class="ot">&lt;-</span> <span class="cf">function</span>(argument){</span>
<span id="cb177-3"><a href="r-programming-making-custom-functions.html#cb177-3" aria-hidden="true" tabindex="-1"></a>  <span class="co">#the body of the function is inside curly brackets</span></span>
<span id="cb177-4"><a href="r-programming-making-custom-functions.html#cb177-4" aria-hidden="true" tabindex="-1"></a>  <span class="co">#here you do something with the arguments, for example:</span></span>
<span id="cb177-5"><a href="r-programming-making-custom-functions.html#cb177-5" aria-hidden="true" tabindex="-1"></a>  result <span class="ot">&lt;-</span> argument<span class="sc">^</span><span class="dv">2</span> </span>
<span id="cb177-6"><a href="r-programming-making-custom-functions.html#cb177-6" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb177-7"><a href="r-programming-making-custom-functions.html#cb177-7" aria-hidden="true" tabindex="-1"></a>  <span class="co"># functions typically end with returning some value</span></span>
<span id="cb177-8"><a href="r-programming-making-custom-functions.html#cb177-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(result)</span>
<span id="cb177-9"><a href="r-programming-making-custom-functions.html#cb177-9" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<p>The above function will take an input value, square it, and return the result. You can now use the function by writing <code>function_name(somevalue)</code>, like you would with the functions you’ve encountered so far. The arguments go into the parentheses of <code>function()</code>, and the body is inside curly brackets. The value to return is wrapped inside <code>return()</code> Note that the name of the function and arguments can be whatever you’d like. You can also have as many arguments as you want, separated by comma.</p>
<div class="green">
<p><strong>Important concept:</strong><br />
Custom functions makes a piece of code reusable, either for doing something many times, or for using for other purposes later. A function takes one or more <em>arguments</em>, does something with the arguments in the <em>body</em> and <em>returns</em> the result.</p>
</div>
</div>
<div id="a-simple-example" class="section level3" number="4.1.3">
<h3><span class="header-section-number">4.1.3</span> A simple example</h3>
<p>Let’s make a simple function that allows you to see clearly what the function is doing. The following function takes an argument, and prints it in a sentence:</p>
<div class="sourceCode" id="cb178"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb178-1"><a href="r-programming-making-custom-functions.html#cb178-1" aria-hidden="true" tabindex="-1"></a>print_arg <span class="ot">&lt;-</span> <span class="cf">function</span>(argument){</span>
<span id="cb178-2"><a href="r-programming-making-custom-functions.html#cb178-2" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb178-3"><a href="r-programming-making-custom-functions.html#cb178-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># make a string explaining what the argument is</span></span>
<span id="cb178-4"><a href="r-programming-making-custom-functions.html#cb178-4" aria-hidden="true" tabindex="-1"></a>  sentence <span class="ot">&lt;-</span> <span class="fu">paste</span>(<span class="st">&quot;the argument is:&quot;</span>, argument)</span>
<span id="cb178-5"><a href="r-programming-making-custom-functions.html#cb178-5" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb178-6"><a href="r-programming-making-custom-functions.html#cb178-6" aria-hidden="true" tabindex="-1"></a>  <span class="co"># return the string</span></span>
<span id="cb178-7"><a href="r-programming-making-custom-functions.html#cb178-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(sentence)</span>
<span id="cb178-8"><a href="r-programming-making-custom-functions.html#cb178-8" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<p>When you run this code, nothing is printed, but the function has been saved. You can now call your new function with any argument.</p>
<div class="sourceCode" id="cb179"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb179-1"><a href="r-programming-making-custom-functions.html#cb179-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print_arg</span>(<span class="dv">5</span>)</span>
<span id="cb179-2"><a href="r-programming-making-custom-functions.html#cb179-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] &quot;the argument is: 5&quot;</span></span>
<span id="cb179-3"><a href="r-programming-making-custom-functions.html#cb179-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb179-4"><a href="r-programming-making-custom-functions.html#cb179-4" aria-hidden="true" tabindex="-1"></a><span class="co"># you can store the returned value to an object and print it</span></span>
<span id="cb179-5"><a href="r-programming-making-custom-functions.html#cb179-5" aria-hidden="true" tabindex="-1"></a>horse_sentence <span class="ot">&lt;-</span> <span class="fu">print_arg</span>(<span class="st">&quot;horse&quot;</span>)</span>
<span id="cb179-6"><a href="r-programming-making-custom-functions.html#cb179-6" aria-hidden="true" tabindex="-1"></a>horse_sentence</span>
<span id="cb179-7"><a href="r-programming-making-custom-functions.html#cb179-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] &quot;the argument is: horse&quot;</span></span>
<span id="cb179-8"><a href="r-programming-making-custom-functions.html#cb179-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb179-9"><a href="r-programming-making-custom-functions.html#cb179-9" aria-hidden="true" tabindex="-1"></a><span class="co"># you can put any expression as the argument,</span></span>
<span id="cb179-10"><a href="r-programming-making-custom-functions.html#cb179-10" aria-hidden="true" tabindex="-1"></a><span class="co"># for example a logical statement</span></span>
<span id="cb179-11"><a href="r-programming-making-custom-functions.html#cb179-11" aria-hidden="true" tabindex="-1"></a><span class="fu">print_arg</span>(pi <span class="sc">&gt;</span> <span class="fu">exp</span>(<span class="dv">1</span>))</span>
<span id="cb179-12"><a href="r-programming-making-custom-functions.html#cb179-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] &quot;the argument is: TRUE&quot;</span></span></code></pre></div>
<p>You can also reference the argument by name when calling the function, like you learned about in week 1.</p>
<div class="sourceCode" id="cb180"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb180-1"><a href="r-programming-making-custom-functions.html#cb180-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print_arg</span>(<span class="at">argument =</span> <span class="st">&quot;zebra&quot;</span>)</span>
<span id="cb180-2"><a href="r-programming-making-custom-functions.html#cb180-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] &quot;the argument is: zebra&quot;</span></span></code></pre></div>
<p>While this particular function is quite useless, notice how much typing you have saved already. We will move on to some more useful functions shortly, after an exercise.</p>
<p><strong>Exercise: create a new function called <code>print_args()</code> that takes two arguments as input, and returns the text “the first argument is [argument1] and the second is [argument2]”.</strong></p>
<div class="fold s">
<div class="sourceCode" id="cb181"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb181-1"><a href="r-programming-making-custom-functions.html#cb181-1" aria-hidden="true" tabindex="-1"></a>print_args <span class="ot">&lt;-</span> <span class="cf">function</span>(argument1, argument2){</span>
<span id="cb181-2"><a href="r-programming-making-custom-functions.html#cb181-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># make a string</span></span>
<span id="cb181-3"><a href="r-programming-making-custom-functions.html#cb181-3" aria-hidden="true" tabindex="-1"></a>  sentence <span class="ot">&lt;-</span> <span class="fu">paste</span>(<span class="st">&quot;the first argument is:&quot;</span>, argument1, <span class="st">&quot;and the second is&quot;</span>, argument2)</span>
<span id="cb181-4"><a href="r-programming-making-custom-functions.html#cb181-4" aria-hidden="true" tabindex="-1"></a>  <span class="co"># return the string</span></span>
<span id="cb181-5"><a href="r-programming-making-custom-functions.html#cb181-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(sentence)</span>
<span id="cb181-6"><a href="r-programming-making-custom-functions.html#cb181-6" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb181-7"><a href="r-programming-making-custom-functions.html#cb181-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb181-8"><a href="r-programming-making-custom-functions.html#cb181-8" aria-hidden="true" tabindex="-1"></a><span class="co"># test the code</span></span>
<span id="cb181-9"><a href="r-programming-making-custom-functions.html#cb181-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb181-10"><a href="r-programming-making-custom-functions.html#cb181-10" aria-hidden="true" tabindex="-1"></a><span class="fu">print_args</span>(<span class="dv">5</span>, <span class="st">&quot;horse&quot;</span>)</span>
<span id="cb181-11"><a href="r-programming-making-custom-functions.html#cb181-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] &quot;the first argument is: 5 and the second is horse&quot;</span></span></code></pre></div>
</div>
</div>
<div id="some-important-function-properties" class="section level3" number="4.1.4">
<h3><span class="header-section-number">4.1.4</span> Some important function properties</h3>
<p>One important thing about functions is that any objects you create <strong>only exist inside the function</strong>. This means that even though the above function creates an object named <code>sentence</code>, and we’ve run the function many times, there is no object called sentence in your R session:</p>
<div class="sourceCode" id="cb182"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb182-1"><a href="r-programming-making-custom-functions.html#cb182-1" aria-hidden="true" tabindex="-1"></a>sentence</span>
<span id="cb182-2"><a href="r-programming-making-custom-functions.html#cb182-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Error in eval(expr, envir, enclos): object &#39;sentence&#39; not found</span></span></code></pre></div>
<p>This means that you don’t need to worry about naming conflicts when running a function, which is convenient. However, it also means that you can’t access any objects that are made inside your function, only whatever you put inside <code>return()</code></p>
<p>Another important thing is that <strong>you can only return 1 thing from a function</strong>. This means that if you want to return multiple objects, you need to make these into a vector. In the following three examples, only the last function will work as intended:</p>
<div class="sourceCode" id="cb183"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb183-1"><a href="r-programming-making-custom-functions.html#cb183-1" aria-hidden="true" tabindex="-1"></a>return_2 <span class="ot">&lt;-</span> <span class="cf">function</span>(x, y){</span>
<span id="cb183-2"><a href="r-programming-making-custom-functions.html#cb183-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(x, y) <span class="co"># will produce an error, you can only return a single object!</span></span>
<span id="cb183-3"><a href="r-programming-making-custom-functions.html#cb183-3" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb183-4"><a href="r-programming-making-custom-functions.html#cb183-4" aria-hidden="true" tabindex="-1"></a><span class="fu">return_2</span>(<span class="dv">2</span>, <span class="dv">4</span>)</span>
<span id="cb183-5"><a href="r-programming-making-custom-functions.html#cb183-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Error in return(x, y): multi-argument returns are not permitted</span></span>
<span id="cb183-6"><a href="r-programming-making-custom-functions.html#cb183-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb183-7"><a href="r-programming-making-custom-functions.html#cb183-7" aria-hidden="true" tabindex="-1"></a>return_twice <span class="ot">&lt;-</span> <span class="cf">function</span>(x, y){</span>
<span id="cb183-8"><a href="r-programming-making-custom-functions.html#cb183-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(x) <span class="co"># x is returned, and the function stops</span></span>
<span id="cb183-9"><a href="r-programming-making-custom-functions.html#cb183-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(y) <span class="co"># the function stops before this line, so y is lost forever</span></span>
<span id="cb183-10"><a href="r-programming-making-custom-functions.html#cb183-10" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb183-11"><a href="r-programming-making-custom-functions.html#cb183-11" aria-hidden="true" tabindex="-1"></a><span class="fu">return_twice</span>(<span class="dv">2</span>, <span class="dv">4</span>)</span>
<span id="cb183-12"><a href="r-programming-making-custom-functions.html#cb183-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 2</span></span>
<span id="cb183-13"><a href="r-programming-making-custom-functions.html#cb183-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb183-14"><a href="r-programming-making-custom-functions.html#cb183-14" aria-hidden="true" tabindex="-1"></a>return_vector <span class="ot">&lt;-</span> <span class="cf">function</span>(x, y){</span>
<span id="cb183-15"><a href="r-programming-making-custom-functions.html#cb183-15" aria-hidden="true" tabindex="-1"></a>  return_vec <span class="ot">&lt;-</span> <span class="fu">c</span>(x, y)</span>
<span id="cb183-16"><a href="r-programming-making-custom-functions.html#cb183-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(return_vec) <span class="co"># this is OK, only one object is returned</span></span>
<span id="cb183-17"><a href="r-programming-making-custom-functions.html#cb183-17" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb183-18"><a href="r-programming-making-custom-functions.html#cb183-18" aria-hidden="true" tabindex="-1"></a><span class="fu">return_vector</span>(<span class="dv">2</span>, <span class="dv">4</span>)</span>
<span id="cb183-19"><a href="r-programming-making-custom-functions.html#cb183-19" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 2 4</span></span></code></pre></div>
<div class="green">
<p><strong>Important concept:</strong></p>
<ul>
<li><p>Any variables you create inside a function only exists within that function.</p></li>
<li><p>If you want to return more than 1 value, you need to bind the values together in e.g. a vector.</p></li>
</ul>
</div>
</div>
<div id="a-slightly-more-useful-example" class="section level3" number="4.1.5">
<h3><span class="header-section-number">4.1.5</span> A slightly more useful example</h3>
<p>A more useful function is often one that does some calculations and returns the result. Here’s an example of a function that takes two arguments and divides the first by the other.</p>
<div class="sourceCode" id="cb184"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb184-1"><a href="r-programming-making-custom-functions.html#cb184-1" aria-hidden="true" tabindex="-1"></a>divide <span class="ot">&lt;-</span> <span class="cf">function</span>(numerator, denominator){</span>
<span id="cb184-2"><a href="r-programming-making-custom-functions.html#cb184-2" aria-hidden="true" tabindex="-1"></a>  result <span class="ot">&lt;-</span> numerator<span class="sc">/</span>denominator</span>
<span id="cb184-3"><a href="r-programming-making-custom-functions.html#cb184-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(result)</span>
<span id="cb184-4"><a href="r-programming-making-custom-functions.html#cb184-4" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb184-5"><a href="r-programming-making-custom-functions.html#cb184-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb184-6"><a href="r-programming-making-custom-functions.html#cb184-6" aria-hidden="true" tabindex="-1"></a><span class="fu">divide</span>(<span class="dv">5</span>, <span class="dv">3</span>)</span>
<span id="cb184-7"><a href="r-programming-making-custom-functions.html#cb184-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 1.666667</span></span>
<span id="cb184-8"><a href="r-programming-making-custom-functions.html#cb184-8" aria-hidden="true" tabindex="-1"></a><span class="fu">divide</span>(<span class="dv">10</span>, <span class="dv">0</span>)</span>
<span id="cb184-9"><a href="r-programming-making-custom-functions.html#cb184-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] Inf</span></span>
<span id="cb184-10"><a href="r-programming-making-custom-functions.html#cb184-10" aria-hidden="true" tabindex="-1"></a><span class="co"># oops</span></span></code></pre></div>
<p><strong>Exercise: create a function that takes two arguments and returns the sum of those arguments</strong></p>
<div class="fold s">
<div class="sourceCode" id="cb185"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb185-1"><a href="r-programming-making-custom-functions.html#cb185-1" aria-hidden="true" tabindex="-1"></a>add <span class="ot">&lt;-</span> <span class="cf">function</span>(x, y){</span>
<span id="cb185-2"><a href="r-programming-making-custom-functions.html#cb185-2" aria-hidden="true" tabindex="-1"></a>  result <span class="ot">&lt;-</span> x <span class="sc">+</span> y</span>
<span id="cb185-3"><a href="r-programming-making-custom-functions.html#cb185-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(result)</span>
<span id="cb185-4"><a href="r-programming-making-custom-functions.html#cb185-4" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
</div>
</div>
<div id="example-calculating-genotype-frequencies" class="section level3" number="4.1.6">
<h3><span class="header-section-number">4.1.6</span> Example: calculating genotype frequencies</h3>
<p>In last week’s tutorial, you used the following snippet of code to calculate genotype frequencies from allele frequencies:</p>
<div class="sourceCode" id="cb186"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb186-1"><a href="r-programming-making-custom-functions.html#cb186-1" aria-hidden="true" tabindex="-1"></a><span class="co"># first we set the frequencies</span></span>
<span id="cb186-2"><a href="r-programming-making-custom-functions.html#cb186-2" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fl">0.8</span></span>
<span id="cb186-3"><a href="r-programming-making-custom-functions.html#cb186-3" aria-hidden="true" tabindex="-1"></a>q <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">-</span> p </span>
<span id="cb186-4"><a href="r-programming-making-custom-functions.html#cb186-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb186-5"><a href="r-programming-making-custom-functions.html#cb186-5" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate the expected genotype frequencies (_e denotes expected)</span></span>
<span id="cb186-6"><a href="r-programming-making-custom-functions.html#cb186-6" aria-hidden="true" tabindex="-1"></a>A1A1_e <span class="ot">&lt;-</span> p<span class="sc">^</span><span class="dv">2</span></span>
<span id="cb186-7"><a href="r-programming-making-custom-functions.html#cb186-7" aria-hidden="true" tabindex="-1"></a>A1A2_e <span class="ot">&lt;-</span> <span class="dv">2</span> <span class="sc">*</span> (p <span class="sc">*</span> q)</span>
<span id="cb186-8"><a href="r-programming-making-custom-functions.html#cb186-8" aria-hidden="true" tabindex="-1"></a>A2A2_e <span class="ot">&lt;-</span> q<span class="sc">^</span><span class="dv">2</span></span>
<span id="cb186-9"><a href="r-programming-making-custom-functions.html#cb186-9" aria-hidden="true" tabindex="-1"></a><span class="co"># show the genotype frequencies in the console</span></span>
<span id="cb186-10"><a href="r-programming-making-custom-functions.html#cb186-10" aria-hidden="true" tabindex="-1"></a><span class="fu">c</span>(A1A1_e, A1A2_e, A2A2_e)</span>
<span id="cb186-11"><a href="r-programming-making-custom-functions.html#cb186-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 0.64 0.32 0.04</span></span></code></pre></div>
<p>This is something I imagine you will do often in this course and later in life if you pursue evolutionary biology, so it would make sense to make it into a function. I will let you try to do it on your own before showing you how to do this. Use the hints if you get stuck!</p>
<p><strong>Exercise: create a function that takes a single argument, <code>p</code>, and returns the expected genotype frequencies according to the Hardy-Weinberg expectation.</strong></p>
<details>
<summary>
<p>Show hint</p>
</summary>
<p>Start by wrapping the whole code snippet above inside the curly brackets of a function.</p>
</details>
<details>
<summary>
<p>Show another hint</p>
</summary>
<p>Remember that you can only return 1 object, so you need to make the genotype frequencies into a vector. Also, remember that you should be able to change p, so you should not have the line <code>p &lt;- 0.8</code> in the body of the function.</p>
</details>
<div class="fold s">
<div class="sourceCode" id="cb187"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb187-1"><a href="r-programming-making-custom-functions.html#cb187-1" aria-hidden="true" tabindex="-1"></a>calc_geno <span class="ot">&lt;-</span> <span class="cf">function</span>(p){</span>
<span id="cb187-2"><a href="r-programming-making-custom-functions.html#cb187-2" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb187-3"><a href="r-programming-making-custom-functions.html#cb187-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># calculate q from p</span></span>
<span id="cb187-4"><a href="r-programming-making-custom-functions.html#cb187-4" aria-hidden="true" tabindex="-1"></a>  q <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">-</span> p </span>
<span id="cb187-5"><a href="r-programming-making-custom-functions.html#cb187-5" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb187-6"><a href="r-programming-making-custom-functions.html#cb187-6" aria-hidden="true" tabindex="-1"></a>  <span class="co"># calculate the expected genotype frequencies (_e denotes expected)</span></span>
<span id="cb187-7"><a href="r-programming-making-custom-functions.html#cb187-7" aria-hidden="true" tabindex="-1"></a>  A1A1_e <span class="ot">&lt;-</span> p<span class="sc">^</span><span class="dv">2</span></span>
<span id="cb187-8"><a href="r-programming-making-custom-functions.html#cb187-8" aria-hidden="true" tabindex="-1"></a>  A1A2_e <span class="ot">&lt;-</span> <span class="dv">2</span> <span class="sc">*</span> (p <span class="sc">*</span> q)</span>
<span id="cb187-9"><a href="r-programming-making-custom-functions.html#cb187-9" aria-hidden="true" tabindex="-1"></a>  A2A2_e <span class="ot">&lt;-</span> q<span class="sc">^</span><span class="dv">2</span></span>
<span id="cb187-10"><a href="r-programming-making-custom-functions.html#cb187-10" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb187-11"><a href="r-programming-making-custom-functions.html#cb187-11" aria-hidden="true" tabindex="-1"></a>  <span class="co"># return the genotype frequencies</span></span>
<span id="cb187-12"><a href="r-programming-making-custom-functions.html#cb187-12" aria-hidden="true" tabindex="-1"></a>  geno_freq <span class="ot">&lt;-</span> <span class="fu">c</span>(A1A1_e, A1A2_e, A2A2_e)</span>
<span id="cb187-13"><a href="r-programming-making-custom-functions.html#cb187-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(geno_freq)</span>
<span id="cb187-14"><a href="r-programming-making-custom-functions.html#cb187-14" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
</div>
<p>When testing your function, it should show the following output:</p>
<div class="sourceCode" id="cb188"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb188-1"><a href="r-programming-making-custom-functions.html#cb188-1" aria-hidden="true" tabindex="-1"></a><span class="fu">calc_geno</span>(<span class="fl">0.2</span>)</span>
<span id="cb188-2"><a href="r-programming-making-custom-functions.html#cb188-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 0.04 0.32 0.64</span></span>
<span id="cb188-3"><a href="r-programming-making-custom-functions.html#cb188-3" aria-hidden="true" tabindex="-1"></a><span class="fu">calc_geno</span>(<span class="fl">0.5</span>)</span>
<span id="cb188-4"><a href="r-programming-making-custom-functions.html#cb188-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 0.25 0.50 0.25</span></span>
<span id="cb188-5"><a href="r-programming-making-custom-functions.html#cb188-5" aria-hidden="true" tabindex="-1"></a><span class="fu">calc_geno</span>(<span class="fl">0.7</span>)</span>
<span id="cb188-6"><a href="r-programming-making-custom-functions.html#cb188-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 0.49 0.42 0.09</span></span></code></pre></div>
<p>Congratulations, you have now made your first truly useful function! Any time you need to calculate genotype frequencies from allele frequencies from now on, it will be a breeze!</p>
<div class="green">
<p><strong>Important concept:</strong><br />
Making a function isn’t all that hard when you get used to it. A general recipe for converting code into a function is:</p>
<ol style="list-style-type: decimal">
<li>Paste your code inside the function body.</li>
<li>Make any objects you want to be able to change into an argument of the function (<code>p</code> in this example). Remember to remove those objects from the body of the function to avoid overwriting your arguments.</li>
<li>Wrap your result in <code>return()</code>.</li>
</ol>
</div>
</div>
<div id="creating-a-function-of-the-drift-simulation" class="section level3" number="4.1.7">
<h3><span class="header-section-number">4.1.7</span> Creating a function of the drift simulation</h3>
<p>Now we’re ready to make a function to simulate drift. Just to remind you, this is the code we used last week for the simulation:</p>
<div class="sourceCode" id="cb189"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb189-1"><a href="r-programming-making-custom-functions.html#cb189-1" aria-hidden="true" tabindex="-1"></a><span class="co"># set population size</span></span>
<span id="cb189-2"><a href="r-programming-making-custom-functions.html#cb189-2" aria-hidden="true" tabindex="-1"></a>N <span class="ot">&lt;-</span> <span class="dv">8</span></span>
<span id="cb189-3"><a href="r-programming-making-custom-functions.html#cb189-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb189-4"><a href="r-programming-making-custom-functions.html#cb189-4" aria-hidden="true" tabindex="-1"></a><span class="co"># set number of generations</span></span>
<span id="cb189-5"><a href="r-programming-making-custom-functions.html#cb189-5" aria-hidden="true" tabindex="-1"></a>ngen <span class="ot">&lt;-</span> <span class="dv">100</span></span>
<span id="cb189-6"><a href="r-programming-making-custom-functions.html#cb189-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb189-7"><a href="r-programming-making-custom-functions.html#cb189-7" aria-hidden="true" tabindex="-1"></a><span class="co"># set initial frequency of A1</span></span>
<span id="cb189-8"><a href="r-programming-making-custom-functions.html#cb189-8" aria-hidden="true" tabindex="-1"></a>p_init <span class="ot">&lt;-</span> <span class="fl">0.5</span></span>
<span id="cb189-9"><a href="r-programming-making-custom-functions.html#cb189-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb189-10"><a href="r-programming-making-custom-functions.html#cb189-10" aria-hidden="true" tabindex="-1"></a><span class="co"># create vector for storing results</span></span>
<span id="cb189-11"><a href="r-programming-making-custom-functions.html#cb189-11" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="cn">NA</span>, ngen)</span>
<span id="cb189-12"><a href="r-programming-making-custom-functions.html#cb189-12" aria-hidden="true" tabindex="-1"></a><span class="co"># set first element to initial p</span></span>
<span id="cb189-13"><a href="r-programming-making-custom-functions.html#cb189-13" aria-hidden="true" tabindex="-1"></a>p[<span class="dv">1</span>] <span class="ot">&lt;-</span> p_init</span>
<span id="cb189-14"><a href="r-programming-making-custom-functions.html#cb189-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb189-15"><a href="r-programming-making-custom-functions.html#cb189-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb189-16"><a href="r-programming-making-custom-functions.html#cb189-16" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">2</span><span class="sc">:</span>ngen){</span>
<span id="cb189-17"><a href="r-programming-making-custom-functions.html#cb189-17" aria-hidden="true" tabindex="-1"></a>  <span class="co"># sample number of A1 alleles based on p in previous generation</span></span>
<span id="cb189-18"><a href="r-programming-making-custom-functions.html#cb189-18" aria-hidden="true" tabindex="-1"></a>  nA1 <span class="ot">&lt;-</span> <span class="fu">rbinom</span>(<span class="dv">1</span>, <span class="dv">2</span><span class="sc">*</span>N, p[i<span class="dv">-1</span>])</span>
<span id="cb189-19"><a href="r-programming-making-custom-functions.html#cb189-19" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb189-20"><a href="r-programming-making-custom-functions.html#cb189-20" aria-hidden="true" tabindex="-1"></a>  <span class="co"># set frequency of A1 as p in the current generation</span></span>
<span id="cb189-21"><a href="r-programming-making-custom-functions.html#cb189-21" aria-hidden="true" tabindex="-1"></a>  p[i] <span class="ot">&lt;-</span> nA1 <span class="sc">/</span> (<span class="dv">2</span><span class="sc">*</span>N)</span>
<span id="cb189-22"><a href="r-programming-making-custom-functions.html#cb189-22" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb189-23"><a href="r-programming-making-custom-functions.html#cb189-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb189-24"><a href="r-programming-making-custom-functions.html#cb189-24" aria-hidden="true" tabindex="-1"></a>p</span></code></pre></div>
<p>Notice that there are three parameters we will probably want to be able to change for each simulation: <code>N</code>, <code>ngen</code> and <code>p_init</code>. It makes sense, then, to make these into function arguments.</p>
<p><strong>Exercise: Make the above drift simulation into a function. It should take three arguments: <code>N</code>, <code>ngen</code> and <code>p_init</code>, and return a vector of p’s for each generation.</strong></p>
<details>
<summary>
<p>Show hint</p>
</summary>
<p>Use the three steps from the previous section: paste the code into the body, turn <code>N</code>, <code>ngen</code> and <code>p_init</code> into arguments and return your result.</p>
</details>
<div class="fold s">
<div class="sourceCode" id="cb190"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb190-1"><a href="r-programming-making-custom-functions.html#cb190-1" aria-hidden="true" tabindex="-1"></a>drift_sim <span class="ot">&lt;-</span> <span class="cf">function</span>(N, ngen, p_init){</span>
<span id="cb190-2"><a href="r-programming-making-custom-functions.html#cb190-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb190-3"><a href="r-programming-making-custom-functions.html#cb190-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># create vector for storing results</span></span>
<span id="cb190-4"><a href="r-programming-making-custom-functions.html#cb190-4" aria-hidden="true" tabindex="-1"></a>  p <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="cn">NA</span>, ngen)</span>
<span id="cb190-5"><a href="r-programming-making-custom-functions.html#cb190-5" aria-hidden="true" tabindex="-1"></a>  <span class="co"># set first element to initial p</span></span>
<span id="cb190-6"><a href="r-programming-making-custom-functions.html#cb190-6" aria-hidden="true" tabindex="-1"></a>  p[<span class="dv">1</span>] <span class="ot">&lt;-</span> p_init</span>
<span id="cb190-7"><a href="r-programming-making-custom-functions.html#cb190-7" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb190-8"><a href="r-programming-making-custom-functions.html#cb190-8" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb190-9"><a href="r-programming-making-custom-functions.html#cb190-9" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">2</span><span class="sc">:</span>ngen){</span>
<span id="cb190-10"><a href="r-programming-making-custom-functions.html#cb190-10" aria-hidden="true" tabindex="-1"></a>    <span class="co"># sample number of A1 alleles based on p in previous generation</span></span>
<span id="cb190-11"><a href="r-programming-making-custom-functions.html#cb190-11" aria-hidden="true" tabindex="-1"></a>    nA1 <span class="ot">&lt;-</span> <span class="fu">rbinom</span>(<span class="dv">1</span>, <span class="dv">2</span><span class="sc">*</span>N, p[i<span class="dv">-1</span>])</span>
<span id="cb190-12"><a href="r-programming-making-custom-functions.html#cb190-12" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb190-13"><a href="r-programming-making-custom-functions.html#cb190-13" aria-hidden="true" tabindex="-1"></a>    <span class="co"># set frequency of A1 as p in the current generation</span></span>
<span id="cb190-14"><a href="r-programming-making-custom-functions.html#cb190-14" aria-hidden="true" tabindex="-1"></a>    p[i] <span class="ot">&lt;-</span> nA1 <span class="sc">/</span> (<span class="dv">2</span><span class="sc">*</span>N)</span>
<span id="cb190-15"><a href="r-programming-making-custom-functions.html#cb190-15" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb190-16"><a href="r-programming-making-custom-functions.html#cb190-16" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb190-17"><a href="r-programming-making-custom-functions.html#cb190-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(p)</span>
<span id="cb190-18"><a href="r-programming-making-custom-functions.html#cb190-18" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
</div>
<p>That concludes the R-part of this tutorial! The rest of the tutorial will be about fitness and selection, but will make use of functions for exploring this. Remember to check back here if you become unsure about how functions work!</p>
<p><strong>Exercise: when working through the rest of the tutorial, think once in a while: “Will I need to do this many times? Could I make this code into a function? How would I do that?”</strong></p>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="the-theory-of-natural-selection.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="evolutionary-biology-fitness.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
