<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>3.2 Evolutionary biology | R tutorials in BIOS1140</title>
  <meta name="description" content="This is a set of R tutorials connected to the course BIOS1140 at the University of Oslo, and are meant as a companion to the book Evolutionary Genetics (Ravinet &amp; Sætre, 2018). They do not assume any prior R knowledge." />
  <meta name="generator" content="bookdown 0.21 and GitBook 2.6.7" />

  <meta property="og:title" content="3.2 Evolutionary biology | R tutorials in BIOS1140" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="This is a set of R tutorials connected to the course BIOS1140 at the University of Oslo, and are meant as a companion to the book Evolutionary Genetics (Ravinet &amp; Sætre, 2018). They do not assume any prior R knowledge." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="3.2 Evolutionary biology | R tutorials in BIOS1140" />
  
  <meta name="twitter:description" content="This is a set of R tutorials connected to the course BIOS1140 at the University of Oslo, and are meant as a companion to the book Evolutionary Genetics (Ravinet &amp; Sætre, 2018). They do not assume any prior R knowledge." />
  



<meta name="date" content="2021-06-13" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="r-programming-for-loops.html"/>
<link rel="next" href="study-questions-1.html"/>
<script src="libs/header-attrs-2.6/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">R tutorials for the course BIOS1140</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a></li>
<li class="part"><span><b>I Introductory R course</b></span></li>
<li class="chapter" data-level="1" data-path="introduction-to-r.html"><a href="introduction-to-r.html"><i class="fa fa-check"></i><b>1</b> Introduction to R</a>
<ul>
<li class="chapter" data-level="" data-path="introduction-to-r.html"><a href="introduction-to-r.html#what-to-expect"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="1.1" data-path="start-using-r.html"><a href="start-using-r.html"><i class="fa fa-check"></i><b>1.1</b> Start using R!</a>
<ul>
<li class="chapter" data-level="1.1.1" data-path="start-using-r.html"><a href="start-using-r.html#install-r-and-rstudio"><i class="fa fa-check"></i><b>1.1.1</b> Install R and Rstudio</a></li>
<li class="chapter" data-level="1.1.2" data-path="start-using-r.html"><a href="start-using-r.html#getting-familiar-with-r-and-the-console"><i class="fa fa-check"></i><b>1.1.2</b> Getting familiar with R and the console</a></li>
<li class="chapter" data-level="1.1.3" data-path="start-using-r.html"><a href="start-using-r.html#scripts"><i class="fa fa-check"></i><b>1.1.3</b> Scripts</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="r-essentials.html"><a href="r-essentials.html"><i class="fa fa-check"></i><b>1.2</b> R essentials</a>
<ul>
<li class="chapter" data-level="1.2.1" data-path="r-essentials.html"><a href="r-essentials.html#assigning-to-objects"><i class="fa fa-check"></i><b>1.2.1</b> Assigning to objects</a></li>
<li class="chapter" data-level="1.2.2" data-path="r-essentials.html"><a href="r-essentials.html#vectors"><i class="fa fa-check"></i><b>1.2.2</b> Vectors</a></li>
<li class="chapter" data-level="1.2.3" data-path="r-essentials.html"><a href="r-essentials.html#strings"><i class="fa fa-check"></i><b>1.2.3</b> Strings</a></li>
<li class="chapter" data-level="1.2.4" data-path="r-essentials.html"><a href="r-essentials.html#logical"><i class="fa fa-check"></i><b>1.2.4</b> Logical values</a></li>
<li class="chapter" data-level="1.2.5" data-path="r-essentials.html"><a href="r-essentials.html#functions"><i class="fa fa-check"></i><b>1.2.5</b> Functions</a></li>
<li class="chapter" data-level="1.2.6" data-path="r-essentials.html"><a href="r-essentials.html#data-frames"><i class="fa fa-check"></i><b>1.2.6</b> Data frames</a></li>
</ul></li>
<li class="chapter" data-level="1.3" data-path="dataimport.html"><a href="dataimport.html"><i class="fa fa-check"></i><b>1.3</b> Data import</a>
<ul>
<li class="chapter" data-level="1.3.1" data-path="dataimport.html"><a href="dataimport.html#data-formats"><i class="fa fa-check"></i><b>1.3.1</b> Data formats</a></li>
<li class="chapter" data-level="1.3.2" data-path="dataimport.html"><a href="dataimport.html#working-directories"><i class="fa fa-check"></i><b>1.3.2</b> Working directories</a></li>
<li class="chapter" data-level="1.3.3" data-path="dataimport.html"><a href="dataimport.html#importing"><i class="fa fa-check"></i><b>1.3.3</b> Importing</a></li>
</ul></li>
<li class="chapter" data-level="1.4" data-path="plotting.html"><a href="plotting.html"><i class="fa fa-check"></i><b>1.4</b> Plotting</a>
<ul>
<li class="chapter" data-level="1.4.1" data-path="plotting.html"><a href="plotting.html#plotting-vectors"><i class="fa fa-check"></i><b>1.4.1</b> Plotting vectors</a></li>
<li class="chapter" data-level="1.4.2" data-path="plotting.html"><a href="plotting.html#customizing-your-plots"><i class="fa fa-check"></i><b>1.4.2</b> Customizing your plots</a></li>
</ul></li>
<li class="chapter" data-level="1.5" data-path="installing-and-loading-packages.html"><a href="installing-and-loading-packages.html"><i class="fa fa-check"></i><b>1.5</b> Installing and loading packages</a></li>
<li class="chapter" data-level="1.6" data-path="going-further.html"><a href="going-further.html"><i class="fa fa-check"></i><b>1.6</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="building-on-your-foundations-going-further-with-r.html"><a href="building-on-your-foundations-going-further-with-r.html"><i class="fa fa-check"></i><b>2</b> Building on your foundations: going further with R</a>
<ul>
<li class="chapter" data-level="" data-path="building-on-your-foundations-going-further-with-r.html"><a href="building-on-your-foundations-going-further-with-r.html#what-to-expect-1"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="2.1" data-path="intro-to-data-manipulation-with-tidyverse.html"><a href="intro-to-data-manipulation-with-tidyverse.html"><i class="fa fa-check"></i><b>2.1</b> Intro to data manipulation with tidyverse</a>
<ul>
<li class="chapter" data-level="2.1.1" data-path="intro-to-data-manipulation-with-tidyverse.html"><a href="intro-to-data-manipulation-with-tidyverse.html#what-is-the-tidyverse"><i class="fa fa-check"></i><b>2.1.1</b> What is the tidyverse?</a></li>
<li class="chapter" data-level="2.1.2" data-path="intro-to-data-manipulation-with-tidyverse.html"><a href="intro-to-data-manipulation-with-tidyverse.html#the-dplyr-package"><i class="fa fa-check"></i><b>2.1.2</b> The dplyr package</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="using-dplyr-to-work-with-your-data.html"><a href="using-dplyr-to-work-with-your-data.html"><i class="fa fa-check"></i><b>2.2</b> Using dplyr to work with your data</a>
<ul>
<li class="chapter" data-level="2.2.1" data-path="using-dplyr-to-work-with-your-data.html"><a href="using-dplyr-to-work-with-your-data.html#the-pipe"><i class="fa fa-check"></i><b>2.2.1</b> The pipe</a></li>
<li class="chapter" data-level="2.2.2" data-path="using-dplyr-to-work-with-your-data.html"><a href="using-dplyr-to-work-with-your-data.html#selecting-colums-with-select"><i class="fa fa-check"></i><b>2.2.2</b> Selecting colums with <code>select()</code></a></li>
<li class="chapter" data-level="2.2.3" data-path="using-dplyr-to-work-with-your-data.html"><a href="using-dplyr-to-work-with-your-data.html#filtering-colums-using-filter"><i class="fa fa-check"></i><b>2.2.3</b> Filtering colums using <code>filter()</code></a></li>
<li class="chapter" data-level="2.2.4" data-path="using-dplyr-to-work-with-your-data.html"><a href="using-dplyr-to-work-with-your-data.html#creating-new-columns-with-mutate"><i class="fa fa-check"></i><b>2.2.4</b> Creating new columns with <code>mutate()</code></a></li>
<li class="chapter" data-level="2.2.5" data-path="using-dplyr-to-work-with-your-data.html"><a href="using-dplyr-to-work-with-your-data.html#grouped-summaries-with-group_by-and-summarise"><i class="fa fa-check"></i><b>2.2.5</b> Grouped summaries with <code>group_by()</code> and <code>summarise()</code></a></li>
<li class="chapter" data-level="2.2.6" data-path="using-dplyr-to-work-with-your-data.html"><a href="using-dplyr-to-work-with-your-data.html#using-everything-weve-learned-in-a-single-pipe-and-a-dplyr-exercise"><i class="fa fa-check"></i><b>2.2.6</b> Using everything we’ve learned in a single pipe, and a <code>dplyr</code> exercise</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="plotting-your-data-with-ggplot2.html"><a href="plotting-your-data-with-ggplot2.html"><i class="fa fa-check"></i><b>2.3</b> Plotting your data with <code>ggplot2</code></a>
<ul>
<li class="chapter" data-level="2.3.1" data-path="plotting-your-data-with-ggplot2.html"><a href="plotting-your-data-with-ggplot2.html#the-three-things-you-need-in-a-ggplot"><i class="fa fa-check"></i><b>2.3.1</b> The three things you need in a ggplot</a></li>
<li class="chapter" data-level="2.3.2" data-path="plotting-your-data-with-ggplot2.html"><a href="plotting-your-data-with-ggplot2.html#storing-ggplots-in-objects"><i class="fa fa-check"></i><b>2.3.2</b> Storing ggplots in objects</a></li>
<li class="chapter" data-level="2.3.3" data-path="plotting-your-data-with-ggplot2.html"><a href="plotting-your-data-with-ggplot2.html#customizing-your-plots-1"><i class="fa fa-check"></i><b>2.3.3</b> Customizing your plots</a></li>
<li class="chapter" data-level="2.3.4" data-path="plotting-your-data-with-ggplot2.html"><a href="plotting-your-data-with-ggplot2.html#saving-your-plots"><i class="fa fa-check"></i><b>2.3.4</b> Saving your plots</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="reshaping-data-with-pivot-longer.html"><a href="reshaping-data-with-pivot-longer.html"><i class="fa fa-check"></i><b>2.4</b> Reshaping data with pivot_longer()</a>
<ul>
<li class="chapter" data-level="2.4.1" data-path="reshaping-data-with-pivot-longer.html"><a href="reshaping-data-with-pivot-longer.html#wide-and-long-format"><i class="fa fa-check"></i><b>2.4.1</b> Wide and long format</a></li>
<li class="chapter" data-level="2.4.2" data-path="reshaping-data-with-pivot-longer.html"><a href="reshaping-data-with-pivot-longer.html#import-example-data"><i class="fa fa-check"></i><b>2.4.2</b> Import example data</a></li>
<li class="chapter" data-level="2.4.3" data-path="reshaping-data-with-pivot-longer.html"><a href="reshaping-data-with-pivot-longer.html#reshape-the-data"><i class="fa fa-check"></i><b>2.4.3</b> Reshape the data</a></li>
<li class="chapter" data-level="2.4.4" data-path="reshaping-data-with-pivot-longer.html"><a href="reshaping-data-with-pivot-longer.html#plot-the-data"><i class="fa fa-check"></i><b>2.4.4</b> Plot the data</a></li>
</ul></li>
<li class="chapter" data-level="2.5" data-path="study-questions.html"><a href="study-questions.html"><i class="fa fa-check"></i><b>2.5</b> Study questions</a></li>
<li class="chapter" data-level="2.6" data-path="going-further-1.html"><a href="going-further-1.html"><i class="fa fa-check"></i><b>2.6</b> Going further</a></li>
</ul></li>
<li class="part"><span><b>II Evolutionary genetics with R</b></span></li>
<li class="chapter" data-level="3" data-path="changes-in-allele-and-genotype-frequency.html"><a href="changes-in-allele-and-genotype-frequency.html"><i class="fa fa-check"></i><b>3</b> Changes in Allele and Genotype Frequency</a>
<ul>
<li class="chapter" data-level="" data-path="changes-in-allele-and-genotype-frequency.html"><a href="changes-in-allele-and-genotype-frequency.html#what-to-expect-2"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="3.1" data-path="r-programming-for-loops.html"><a href="r-programming-for-loops.html"><i class="fa fa-check"></i><b>3.1</b> R programming: for-loops</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="r-programming-for-loops.html"><a href="r-programming-for-loops.html#motivation-why-loop"><i class="fa fa-check"></i><b>3.1.1</b> Motivation: why loop?</a></li>
<li class="chapter" data-level="3.1.2" data-path="r-programming-for-loops.html"><a href="r-programming-for-loops.html#how-a-for-loop-works"><i class="fa fa-check"></i><b>3.1.2</b> How a for-loop works</a></li>
<li class="chapter" data-level="3.1.3" data-path="r-programming-for-loops.html"><a href="r-programming-for-loops.html#indexing"><i class="fa fa-check"></i><b>3.1.3</b> Indexing with for-loops</a></li>
<li class="chapter" data-level="3.1.4" data-path="r-programming-for-loops.html"><a href="r-programming-for-loops.html#solving-our-problem"><i class="fa fa-check"></i><b>3.1.4</b> Solving our problem</a></li>
<li class="chapter" data-level="3.1.5" data-path="r-programming-for-loops.html"><a href="r-programming-for-loops.html#storing-values-from-a-for-loop"><i class="fa fa-check"></i><b>3.1.5</b> Storing values from a for-loop</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="evolutionary-biology.html"><a href="evolutionary-biology.html"><i class="fa fa-check"></i><b>3.2</b> Evolutionary biology</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="evolutionary-biology.html"><a href="evolutionary-biology.html#the-hardy-weinberg-model"><i class="fa fa-check"></i><b>3.2.1</b> The Hardy-Weinberg Model</a></li>
<li class="chapter" data-level="3.2.2" data-path="evolutionary-biology.html"><a href="evolutionary-biology.html#testing-for-deviations-from-the-hardy-weinberg-expectation"><i class="fa fa-check"></i><b>3.2.2</b> Testing for deviations from the Hardy Weinberg Expectation</a></li>
<li class="chapter" data-level="3.2.3" data-path="evolutionary-biology.html"><a href="evolutionary-biology.html#simulating-genetic-drift"><i class="fa fa-check"></i><b>3.2.3</b> Simulating genetic drift</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="study-questions-1.html"><a href="study-questions-1.html"><i class="fa fa-check"></i><b>3.3</b> Study questions</a></li>
<li class="chapter" data-level="3.4" data-path="going-further-2.html"><a href="going-further-2.html"><i class="fa fa-check"></i><b>3.4</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="the-theory-of-natural-selection.html"><a href="the-theory-of-natural-selection.html"><i class="fa fa-check"></i><b>4</b> The Theory of Natural Selection</a>
<ul>
<li class="chapter" data-level="" data-path="the-theory-of-natural-selection.html"><a href="the-theory-of-natural-selection.html#what-to-expect-3"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="" data-path="the-theory-of-natural-selection.html"><a href="the-theory-of-natural-selection.html#getting-started"><i class="fa fa-check"></i>Getting started</a></li>
<li class="chapter" data-level="4.1" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html"><i class="fa fa-check"></i><b>4.1</b> R programming: Making custom functions</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#motivation"><i class="fa fa-check"></i><b>4.1.1</b> Motivation</a></li>
<li class="chapter" data-level="4.1.2" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#function-basics"><i class="fa fa-check"></i><b>4.1.2</b> Function basics</a></li>
<li class="chapter" data-level="4.1.3" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#a-simple-example"><i class="fa fa-check"></i><b>4.1.3</b> A simple example</a></li>
<li class="chapter" data-level="4.1.4" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#some-important-function-properties"><i class="fa fa-check"></i><b>4.1.4</b> Some important function properties</a></li>
<li class="chapter" data-level="4.1.5" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#a-slightly-more-useful-example"><i class="fa fa-check"></i><b>4.1.5</b> A slightly more useful example</a></li>
<li class="chapter" data-level="4.1.6" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#example-calculating-genotype-frequencies"><i class="fa fa-check"></i><b>4.1.6</b> Example: calculating genotype frequencies</a></li>
<li class="chapter" data-level="4.1.7" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#creating-a-function-of-the-drift-simulation"><i class="fa fa-check"></i><b>4.1.7</b> Creating a function of the drift simulation</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="evolutionary-biology-fitness.html"><a href="evolutionary-biology-fitness.html"><i class="fa fa-check"></i><b>4.2</b> Evolutionary biology: fitness</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="evolutionary-biology-fitness.html"><a href="evolutionary-biology-fitness.html#understanding-fitness"><i class="fa fa-check"></i><b>4.2.1</b> Understanding fitness</a></li>
<li class="chapter" data-level="4.2.2" data-path="evolutionary-biology-fitness.html"><a href="evolutionary-biology-fitness.html#one-locus-model-of-viability-selection"><i class="fa fa-check"></i><b>4.2.2</b> One-locus model of viability selection</a></li>
<li class="chapter" data-level="4.2.3" data-path="evolutionary-biology-fitness.html"><a href="evolutionary-biology-fitness.html#directional-selection"><i class="fa fa-check"></i><b>4.2.3</b> Directional selection</a></li>
<li class="chapter" data-level="4.2.4" data-path="evolutionary-biology-fitness.html"><a href="evolutionary-biology-fitness.html#over-and-underdominance"><i class="fa fa-check"></i><b>4.2.4</b> Over and underdominance</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="study-questions-2.html"><a href="study-questions-2.html"><i class="fa fa-check"></i><b>4.3</b> Study questions</a></li>
<li class="chapter" data-level="4.4" data-path="going-further-3.html"><a href="going-further-3.html"><i class="fa fa-check"></i><b>4.4</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="detecting-natural-selection.html"><a href="detecting-natural-selection.html"><i class="fa fa-check"></i><b>5</b> Detecting Natural Selection</a>
<ul>
<li class="chapter" data-level="" data-path="detecting-natural-selection.html"><a href="detecting-natural-selection.html#what-to-expect-4"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="" data-path="detecting-natural-selection.html"><a href="detecting-natural-selection.html#getting-started-1"><i class="fa fa-check"></i>Getting started</a></li>
<li class="chapter" data-level="5.1" data-path="more-on-functions-vectorisation.html"><a href="more-on-functions-vectorisation.html"><i class="fa fa-check"></i><b>5.1</b> More on functions: Vectorisation</a>
<ul>
<li class="chapter" data-level="5.1.1" data-path="more-on-functions-vectorisation.html"><a href="more-on-functions-vectorisation.html#the-apply-function"><i class="fa fa-check"></i><b>5.1.1</b> The <code>apply()</code> function</a></li>
<li class="chapter" data-level="5.1.2" data-path="more-on-functions-vectorisation.html"><a href="more-on-functions-vectorisation.html#ifelse"><i class="fa fa-check"></i><b>5.1.2</b> The <code>ifelse()</code> function</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="understanding-fst-the-fixation-index.html"><a href="understanding-fst-the-fixation-index.html"><i class="fa fa-check"></i><b>5.2</b> Understanding <em>F</em><sub>ST</sub> - the fixation index</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="understanding-fst-the-fixation-index.html"><a href="understanding-fst-the-fixation-index.html#a-worked-example-of-fst-in-humans"><i class="fa fa-check"></i><b>5.2.1</b> A worked example of <em>F</em><sub>ST</sub> in humans</a></li>
<li class="chapter" data-level="5.2.2" data-path="understanding-fst-the-fixation-index.html"><a href="understanding-fst-the-fixation-index.html#writing-a-set-of-fst-functions"><i class="fa fa-check"></i><b>5.2.2</b> Writing a set of <em>F</em><sub>ST</sub> functions</a></li>
<li class="chapter" data-level="5.2.3" data-path="understanding-fst-the-fixation-index.html"><a href="understanding-fst-the-fixation-index.html#applying-functions-to-matrices-and-data-frames"><i class="fa fa-check"></i><b>5.2.3</b> Applying functions to matrices and data frames</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="visualising-fst-along-a-chromosome.html"><a href="visualising-fst-along-a-chromosome.html"><i class="fa fa-check"></i><b>5.3</b> Visualising <em>F</em><sub>ST</sub> along a chromosome</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="visualising-fst-along-a-chromosome.html"><a href="visualising-fst-along-a-chromosome.html#identifying-outliers-in-our-fst-distribution"><i class="fa fa-check"></i><b>5.3.1</b> Identifying outliers in our <em>F</em><sub>ST</sub> distribution</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="study-questions-3.html"><a href="study-questions-3.html"><i class="fa fa-check"></i><b>5.4</b> Study questions</a></li>
<li class="chapter" data-level="5.5" data-path="going-further-4.html"><a href="going-further-4.html"><i class="fa fa-check"></i><b>5.5</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="multilocus-evolution.html"><a href="multilocus-evolution.html"><i class="fa fa-check"></i><b>6</b> Multilocus Evolution</a>
<ul>
<li class="chapter" data-level="" data-path="multilocus-evolution.html"><a href="multilocus-evolution.html#what-to-expect-5"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="" data-path="multilocus-evolution.html"><a href="multilocus-evolution.html#getting-started-2"><i class="fa fa-check"></i>Getting started</a></li>
<li class="chapter" data-level="6.1" data-path="tools-for-investigating-differences-between-groups.html"><a href="tools-for-investigating-differences-between-groups.html"><i class="fa fa-check"></i><b>6.1</b> Tools for investigating differences between groups</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="tools-for-investigating-differences-between-groups.html"><a href="tools-for-investigating-differences-between-groups.html#exploring-group-differences"><i class="fa fa-check"></i><b>6.1.1</b> Exploring group differences</a></li>
<li class="chapter" data-level="6.1.2" data-path="tools-for-investigating-differences-between-groups.html"><a href="tools-for-investigating-differences-between-groups.html#analysis-of-variance-anova"><i class="fa fa-check"></i><b>6.1.2</b> Analysis of variance (ANOVA)</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="performing-a-qtl-analysis-in-bedbugs.html"><a href="performing-a-qtl-analysis-in-bedbugs.html"><i class="fa fa-check"></i><b>6.2</b> Performing a QTL analysis in bedbugs</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="performing-a-qtl-analysis-in-bedbugs.html"><a href="performing-a-qtl-analysis-in-bedbugs.html#bedbugs-pesticide-resistance-and-study-design"><i class="fa fa-check"></i><b>6.2.1</b> Bedbugs, pesticide resistance and study design</a></li>
<li class="chapter" data-level="6.2.2" data-path="performing-a-qtl-analysis-in-bedbugs.html"><a href="performing-a-qtl-analysis-in-bedbugs.html#linking-anova-to-our-qtl-analysis"><i class="fa fa-check"></i><b>6.2.2</b> Linking ANOVA to our QTL analysis</a></li>
<li class="chapter" data-level="6.2.3" data-path="performing-a-qtl-analysis-in-bedbugs.html"><a href="performing-a-qtl-analysis-in-bedbugs.html#testing-the-same-genotype-phenotype-association-with-anova"><i class="fa fa-check"></i><b>6.2.3</b> Testing the same genotype phenotype association with ANOVA</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="study-questions-4.html"><a href="study-questions-4.html"><i class="fa fa-check"></i><b>6.3</b> Study questions</a></li>
<li class="chapter" data-level="6.4" data-path="going-further-5.html"><a href="going-further-5.html"><i class="fa fa-check"></i><b>6.4</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="inferring-evolutionary-processes-from-sequence-data.html"><a href="inferring-evolutionary-processes-from-sequence-data.html"><i class="fa fa-check"></i><b>7</b> Inferring Evolutionary Processes from Sequence Data</a>
<ul>
<li class="chapter" data-level="" data-path="inferring-evolutionary-processes-from-sequence-data.html"><a href="inferring-evolutionary-processes-from-sequence-data.html#what-to-expect-6"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="" data-path="inferring-evolutionary-processes-from-sequence-data.html"><a href="inferring-evolutionary-processes-from-sequence-data.html#getting-started-3"><i class="fa fa-check"></i>Getting started</a></li>
<li class="chapter" data-level="7.1" data-path="working-with-dna-sequence-data.html"><a href="working-with-dna-sequence-data.html"><i class="fa fa-check"></i><b>7.1</b> Working with DNA sequence data</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="working-with-dna-sequence-data.html"><a href="working-with-dna-sequence-data.html#reading-sequence-data-into-r."><i class="fa fa-check"></i><b>7.1.1</b> Reading sequence data into R.</a></li>
<li class="chapter" data-level="7.1.2" data-path="working-with-dna-sequence-data.html"><a href="working-with-dna-sequence-data.html#exploring-dna-sequence-data"><i class="fa fa-check"></i><b>7.1.2</b> Exploring DNA sequence data</a></li>
<li class="chapter" data-level="7.1.3" data-path="working-with-dna-sequence-data.html"><a href="working-with-dna-sequence-data.html#calculating-basic-sequence-statistics"><i class="fa fa-check"></i><b>7.1.3</b> Calculating basic sequence statistics</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="working-with-a-larger-dataset.html"><a href="working-with-a-larger-dataset.html"><i class="fa fa-check"></i><b>7.2</b> Working with a larger dataset</a>
<ul>
<li class="chapter" data-level="7.2.1" data-path="working-with-a-larger-dataset.html"><a href="working-with-a-larger-dataset.html#sample-size-and-sequence-statistics"><i class="fa fa-check"></i><b>7.2.1</b> Sample size and sequence statistics</a></li>
<li class="chapter" data-level="7.2.2" data-path="working-with-a-larger-dataset.html"><a href="working-with-a-larger-dataset.html#inferring-evolutionary-processes-using-tajimas-d"><i class="fa fa-check"></i><b>7.2.2</b> Inferring evolutionary processes using Tajima’s <em>D</em></a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="calculating-statistics-at-the-whole-genome-level.html"><a href="calculating-statistics-at-the-whole-genome-level.html"><i class="fa fa-check"></i><b>7.3</b> Calculating statistics at the whole genome level</a>
<ul>
<li class="chapter" data-level="7.3.1" data-path="calculating-statistics-at-the-whole-genome-level.html"><a href="calculating-statistics-at-the-whole-genome-level.html#reading-in-variant-data"><i class="fa fa-check"></i><b>7.3.1</b> Reading in variant data</a></li>
<li class="chapter" data-level="7.3.2" data-path="calculating-statistics-at-the-whole-genome-level.html"><a href="calculating-statistics-at-the-whole-genome-level.html#calculating-nucleotide-diversity-statistics"><i class="fa fa-check"></i><b>7.3.2</b> Calculating nucleotide diversity statistics</a></li>
<li class="chapter" data-level="7.3.3" data-path="calculating-statistics-at-the-whole-genome-level.html"><a href="calculating-statistics-at-the-whole-genome-level.html#visualising-nucleotide-diversity-along-the-chromosome"><i class="fa fa-check"></i><b>7.3.3</b> Visualising nucleotide diversity along the chromosome</a></li>
<li class="chapter" data-level="7.3.4" data-path="calculating-statistics-at-the-whole-genome-level.html"><a href="calculating-statistics-at-the-whole-genome-level.html#performing-a-sliding-window-analysis"><i class="fa fa-check"></i><b>7.3.4</b> Performing a sliding window analysis</a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="study-questions-5.html"><a href="study-questions-5.html"><i class="fa fa-check"></i><b>7.4</b> Study questions</a></li>
<li class="chapter" data-level="7.5" data-path="going-further-6.html"><a href="going-further-6.html"><i class="fa fa-check"></i><b>7.5</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="speciation-genomics.html"><a href="speciation-genomics.html"><i class="fa fa-check"></i><b>8</b> Speciation Genomics</a>
<ul>
<li class="chapter" data-level="" data-path="speciation-genomics.html"><a href="speciation-genomics.html#what-to-expect-7"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="" data-path="speciation-genomics.html"><a href="speciation-genomics.html#getting-started-4"><i class="fa fa-check"></i>Getting started</a></li>
<li class="chapter" data-level="8.1" data-path="r-programming-visualizing-complex-data.html"><a href="r-programming-visualizing-complex-data.html"><i class="fa fa-check"></i><b>8.1</b> R programming: Visualizing complex data</a>
<ul>
<li class="chapter" data-level="8.1.1" data-path="r-programming-visualizing-complex-data.html"><a href="r-programming-visualizing-complex-data.html#faceting"><i class="fa fa-check"></i><b>8.1.1</b> Faceting</a></li>
</ul></li>
<li class="chapter" data-level="8.2" data-path="evolutionary-genetics-returning-to-the-sparrow-dataset.html"><a href="evolutionary-genetics-returning-to-the-sparrow-dataset.html"><i class="fa fa-check"></i><b>8.2</b> Evolutionary genetics: Returning to the sparrow dataset</a>
<ul>
<li class="chapter" data-level="8.2.1" data-path="evolutionary-genetics-returning-to-the-sparrow-dataset.html"><a href="evolutionary-genetics-returning-to-the-sparrow-dataset.html#reading-in-the-sparrow-vcf"><i class="fa fa-check"></i><b>8.2.1</b> Reading in the sparrow vcf</a></li>
<li class="chapter" data-level="8.2.2" data-path="evolutionary-genetics-returning-to-the-sparrow-dataset.html"><a href="evolutionary-genetics-returning-to-the-sparrow-dataset.html#examining-the-variant-data"><i class="fa fa-check"></i><b>8.2.2</b> Examining the variant data</a></li>
<li class="chapter" data-level="8.2.3" data-path="evolutionary-genetics-returning-to-the-sparrow-dataset.html"><a href="evolutionary-genetics-returning-to-the-sparrow-dataset.html#setting-up-sliding-windows"><i class="fa fa-check"></i><b>8.2.3</b> Setting up sliding windows</a></li>
<li class="chapter" data-level="8.2.4" data-path="evolutionary-genetics-returning-to-the-sparrow-dataset.html"><a href="evolutionary-genetics-returning-to-the-sparrow-dataset.html#calculating-sliding-window-estimates-of-nucleotide-diversity-and-differentiation"><i class="fa fa-check"></i><b>8.2.4</b> Calculating sliding window estimates of nucleotide diversity and differentiation</a></li>
<li class="chapter" data-level="8.2.5" data-path="evolutionary-genetics-returning-to-the-sparrow-dataset.html"><a href="evolutionary-genetics-returning-to-the-sparrow-dataset.html#visualising-the-data"><i class="fa fa-check"></i><b>8.2.5</b> Visualising the data</a></li>
<li class="chapter" data-level="8.2.6" data-path="evolutionary-genetics-returning-to-the-sparrow-dataset.html"><a href="evolutionary-genetics-returning-to-the-sparrow-dataset.html#interlude-relative-vs.-absolute-measures-of-nucleotide-diversity"><i class="fa fa-check"></i><b>8.2.6</b> Interlude: relative vs. absolute measures of nucleotide diversity</a></li>
<li class="chapter" data-level="8.2.7" data-path="evolutionary-genetics-returning-to-the-sparrow-dataset.html"><a href="evolutionary-genetics-returning-to-the-sparrow-dataset.html#investigating-recombination-rate-variation"><i class="fa fa-check"></i><b>8.2.7</b> Investigating recombination rate variation</a></li>
</ul></li>
<li class="chapter" data-level="8.3" data-path="study-questions-6.html"><a href="study-questions-6.html"><i class="fa fa-check"></i><b>8.3</b> Study questions</a></li>
<li class="chapter" data-level="8.4" data-path="going-further-7.html"><a href="going-further-7.html"><i class="fa fa-check"></i><b>8.4</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="reconstructing-the-past.html"><a href="reconstructing-the-past.html"><i class="fa fa-check"></i><b>9</b> Reconstructing the Past</a>
<ul>
<li class="chapter" data-level="" data-path="reconstructing-the-past.html"><a href="reconstructing-the-past.html#what-to-expect-8"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="" data-path="reconstructing-the-past.html"><a href="reconstructing-the-past.html#getting-started-5"><i class="fa fa-check"></i>Getting started</a></li>
<li class="chapter" data-level="9.1" data-path="phylogenetics-in-r.html"><a href="phylogenetics-in-r.html"><i class="fa fa-check"></i><b>9.1</b> Phylogenetics in R</a>
<ul>
<li class="chapter" data-level="9.1.1" data-path="phylogenetics-in-r.html"><a href="phylogenetics-in-r.html#storing-trees-in-r"><i class="fa fa-check"></i><b>9.1.1</b> Storing trees in R</a></li>
<li class="chapter" data-level="9.1.2" data-path="phylogenetics-in-r.html"><a href="phylogenetics-in-r.html#plotting-trees"><i class="fa fa-check"></i><b>9.1.2</b> Plotting trees</a></li>
<li class="chapter" data-level="9.1.3" data-path="phylogenetics-in-r.html"><a href="phylogenetics-in-r.html#manipulating-trees"><i class="fa fa-check"></i><b>9.1.3</b> Manipulating trees</a></li>
<li class="chapter" data-level="9.1.4" data-path="phylogenetics-in-r.html"><a href="phylogenetics-in-r.html#a-simple-example-with-real-data---avian-phylogenetics"><i class="fa fa-check"></i><b>9.1.4</b> A simple example with real data - avian phylogenetics</a></li>
<li class="chapter" data-level="9.1.5" data-path="phylogenetics-in-r.html"><a href="phylogenetics-in-r.html#constructing-trees-with-r"><i class="fa fa-check"></i><b>9.1.5</b> Constructing trees with R</a></li>
</ul></li>
<li class="chapter" data-level="9.2" data-path="population-structure.html"><a href="population-structure.html"><i class="fa fa-check"></i><b>9.2</b> Population structure</a>
<ul>
<li class="chapter" data-level="9.2.1" data-path="population-structure.html"><a href="population-structure.html#village-dogs-as-an-insight-to-dog-domestication"><i class="fa fa-check"></i><b>9.2.1</b> Village dogs as an insight to dog domestication</a></li>
<li class="chapter" data-level="9.2.2" data-path="population-structure.html"><a href="population-structure.html#reading-the-data-into-r"><i class="fa fa-check"></i><b>9.2.2</b> Reading the data into R</a></li>
<li class="chapter" data-level="9.2.3" data-path="population-structure.html"><a href="population-structure.html#performing-a-pca"><i class="fa fa-check"></i><b>9.2.3</b> Performing a PCA</a></li>
<li class="chapter" data-level="9.2.4" data-path="population-structure.html"><a href="population-structure.html#visualising-the-pca"><i class="fa fa-check"></i><b>9.2.4</b> Visualising the PCA</a></li>
</ul></li>
<li class="chapter" data-level="9.3" data-path="study-questions-7.html"><a href="study-questions-7.html"><i class="fa fa-check"></i><b>9.3</b> Study questions</a></li>
<li class="chapter" data-level="9.4" data-path="going-further-8.html"><a href="going-further-8.html"><i class="fa fa-check"></i><b>9.4</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="advancing-further-in-r.html"><a href="advancing-further-in-r.html"><i class="fa fa-check"></i><b>10</b> Advancing Further in R</a>
<ul>
<li class="chapter" data-level="" data-path="advancing-further-in-r.html"><a href="advancing-further-in-r.html#what-to-expect-9"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="" data-path="advancing-further-in-r.html"><a href="advancing-further-in-r.html#getting-started-6"><i class="fa fa-check"></i>Getting started</a></li>
<li class="chapter" data-level="10.1" data-path="advanced-features-of-rstudio.html"><a href="advanced-features-of-rstudio.html"><i class="fa fa-check"></i><b>10.1</b> Advanced features of RStudio</a>
<ul>
<li class="chapter" data-level="10.1.1" data-path="advanced-features-of-rstudio.html"><a href="advanced-features-of-rstudio.html#projects-projects-projects"><i class="fa fa-check"></i><b>10.1.1</b> Projects, projects, projects</a></li>
<li class="chapter" data-level="10.1.2" data-path="advanced-features-of-rstudio.html"><a href="advanced-features-of-rstudio.html#everyone-has-a-history"><i class="fa fa-check"></i><b>10.1.2</b> Everyone has a history</a></li>
<li class="chapter" data-level="10.1.3" data-path="advanced-features-of-rstudio.html"><a href="advanced-features-of-rstudio.html#tab-complete-and-other-hotkeys"><i class="fa fa-check"></i><b>10.1.3</b> Tab complete and other hotkeys</a></li>
<li class="chapter" data-level="10.1.4" data-path="advanced-features-of-rstudio.html"><a href="advanced-features-of-rstudio.html#code-checking"><i class="fa fa-check"></i><b>10.1.4</b> Code checking</a></li>
</ul></li>
<li class="chapter" data-level="10.2" data-path="control-flow-for-else-if-and-ifelse.html"><a href="control-flow-for-else-if-and-ifelse.html"><i class="fa fa-check"></i><b>10.2</b> Control flow - <code>for</code>, <code>else</code>, <code>if</code> and <code>ifelse</code></a>
<ul>
<li class="chapter" data-level="10.2.1" data-path="control-flow-for-else-if-and-ifelse.html"><a href="control-flow-for-else-if-and-ifelse.html#if-and-else"><i class="fa fa-check"></i><b>10.2.1</b> <code>if</code> and <code>else</code></a></li>
<li class="chapter" data-level="10.2.2" data-path="control-flow-for-else-if-and-ifelse.html"><a href="control-flow-for-else-if-and-ifelse.html#ifelse---a-single-line-alternative"><i class="fa fa-check"></i><b>10.2.2</b> <code>ifelse</code> - a single line alternative</a></li>
<li class="chapter" data-level="10.2.3" data-path="control-flow-for-else-if-and-ifelse.html"><a href="control-flow-for-else-if-and-ifelse.html#for-loops"><i class="fa fa-check"></i><b>10.2.3</b> <code>for</code> loops</a></li>
</ul></li>
<li class="chapter" data-level="10.3" data-path="vectorisation.html"><a href="vectorisation.html"><i class="fa fa-check"></i><b>10.3</b> Vectorisation</a>
<ul>
<li class="chapter" data-level="10.3.1" data-path="vectorisation.html"><a href="vectorisation.html#why-is-vectorisation-preferable-in-r"><i class="fa fa-check"></i><b>10.3.1</b> Why is vectorisation preferable in R?</a></li>
<li class="chapter" data-level="10.3.2" data-path="vectorisation.html"><a href="vectorisation.html#using-sapply-instead-of-for-on-vectors"><i class="fa fa-check"></i><b>10.3.2</b> Using <code>sapply</code> instead of <code>for</code> on vectors</a></li>
<li class="chapter" data-level="10.3.3" data-path="vectorisation.html"><a href="vectorisation.html#using-apply-instead-of-for-on-matrices-and-dataframes"><i class="fa fa-check"></i><b>10.3.3</b> Using apply instead of <code>for</code> on matrices and dataframes</a></li>
</ul></li>
<li class="chapter" data-level="10.4" data-path="returning-to-writing-your-own-functions.html"><a href="returning-to-writing-your-own-functions.html"><i class="fa fa-check"></i><b>10.4</b> Returning to writing your own functions</a></li>
<li class="chapter" data-level="10.5" data-path="concluding-remarks.html"><a href="concluding-remarks.html"><i class="fa fa-check"></i><b>10.5</b> Concluding remarks</a></li>
<li class="chapter" data-level="10.6" data-path="study-questions-8.html"><a href="study-questions-8.html"><i class="fa fa-check"></i><b>10.6</b> Study questions</a></li>
<li class="chapter" data-level="10.7" data-path="going-further-9.html"><a href="going-further-9.html"><i class="fa fa-check"></i><b>10.7</b> Going further</a></li>
</ul></li>
<li class="part"><span><b>III Assignments</b></span></li>
<li class="chapter" data-level="" data-path="w01.html"><a href="w01.html"><i class="fa fa-check"></i>Week 1-2 assignment</a></li>
<li class="chapter" data-level="" data-path="w03.html"><a href="w03.html"><i class="fa fa-check"></i>Week 3 Assignment</a></li>
<li class="chapter" data-level="" data-path="w04.html"><a href="w04.html"><i class="fa fa-check"></i>Week 4 assignment</a></li>
<li class="chapter" data-level="" data-path="w05.html"><a href="w05.html"><i class="fa fa-check"></i>Week 5 assignment</a></li>
<li class="chapter" data-level="" data-path="w06.html"><a href="w06.html"><i class="fa fa-check"></i>Week 6 assignment</a></li>
<li class="chapter" data-level="" data-path="w07.html"><a href="w07.html"><i class="fa fa-check"></i>Week 7 assignment</a></li>
<li class="chapter" data-level="" data-path="w08.html"><a href="w08.html"><i class="fa fa-check"></i>Week 8 assignment</a></li>
<li class="chapter" data-level="" data-path="w09.html"><a href="w09.html"><i class="fa fa-check"></i>Week 9 assignment</a></li>
<li class="chapter" data-level="" data-path="w10.html"><a href="w10.html"><i class="fa fa-check"></i>Week 10 assignment</a></li>
<li class="appendix"><span><b>Appendix</b></span></li>
<li class="chapter" data-level="A" data-path="rmarkdown.html"><a href="rmarkdown.html"><i class="fa fa-check"></i><b>A</b> About assignments and RMarkdown</a>
<ul>
<li class="chapter" data-level="" data-path="rmarkdown.html"><a href="rmarkdown.html#rmarkdown-1"><i class="fa fa-check"></i>RMarkdown</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Made with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">R tutorials in BIOS1140</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="evolutionary-biology" class="section level2" number="3.2">
<h2><span class="header-section-number">3.2</span> Evolutionary biology</h2>
<script src="js/hideOutput.js"></script>
<p>For this part of the tutorial, we need to load the <code>tidyverse</code> package.</p>
<div class="sourceCode" id="cb142"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb142-1"><a href="evolutionary-biology.html#cb142-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code></pre></div>
<div id="the-hardy-weinberg-model" class="section level3" number="3.2.1">
<h3><span class="header-section-number">3.2.1</span> The Hardy-Weinberg Model</h3>
<p>The HW model is a way of demonstrating the relationship between allele and genotype frequencies. It is mathematically basic but very important for understanding how different processes can alter these frequencies. As you might recall from the main text, the model has five major assumptions:</p>
<ol style="list-style-type: decimal">
<li>All individuals in the population mate randomly.</li>
<li>The population is infinitely large.</li>
<li>No selection occurs.</li>
<li>No mutation occurs.</li>
<li>No gene flow occurs.</li>
</ol>
<p>Imagining our idealized population, we will focus on a single locus, <span class="math inline">\(A\)</span> which has two alleles, <span class="math inline">\(A_1\)</span> and <span class="math inline">\(A_2\)</span>. We can use <span class="math inline">\(p\)</span> to denote the frequency of the <span class="math inline">\(A_1\)</span> allele and <span class="math inline">\(q\)</span> for <span class="math inline">\(A_2\)</span>. This means that if we know <span class="math inline">\(p\)</span>, we can derive <span class="math inline">\(q\)</span> as <span class="math inline">\(q = 1- p\)</span>, since <span class="math inline">\(p + q = 1\)</span>. Remember that the HW-model is a means of representing the relationship between allele and genotype frequencies across generations. With our assumptions outlined above in place, the only thing that determines the frequency of genotypes is their probabilities, derived from the allele frequencies.</p>
<p>It is quite simple to determine these probabilities. With 2 alleles <span class="math inline">\(A_1\)</span> and <span class="math inline">\(A_2\)</span>, we have three possible genotypes and three ways to determine their probabilities as:</p>
<ul>
<li><span class="math inline">\(A_1A_1 = p^2\)</span></li>
<li><span class="math inline">\(A_1A_2 = 2pq\)</span></li>
<li><span class="math inline">\(A_2A_2 = q^2\)</span></li>
</ul>
<p>Another way to look at these probabilities is as the expected frequencies of the three genotypes at Hardy-Weinberg equilibrium given the allele frequencies. Let’s use some R code to actually demonstrate how we might calculate these expected frequencies. For our test example, we will use a simple case where the frequency of <span class="math inline">\(A_1\)</span>, <span class="math inline">\(p\)</span> is 0.8 and the frequency of <span class="math inline">\(A_2\)</span>, <span class="math inline">\(q\)</span> is 0.2.</p>
<div class="sourceCode" id="cb143"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb143-1"><a href="evolutionary-biology.html#cb143-1" aria-hidden="true" tabindex="-1"></a><span class="co"># first we set the frequencies</span></span>
<span id="cb143-2"><a href="evolutionary-biology.html#cb143-2" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fl">0.8</span></span>
<span id="cb143-3"><a href="evolutionary-biology.html#cb143-3" aria-hidden="true" tabindex="-1"></a>q <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">-</span> p </span>
<span id="cb143-4"><a href="evolutionary-biology.html#cb143-4" aria-hidden="true" tabindex="-1"></a><span class="co"># check p and q are equal to 1</span></span>
<span id="cb143-5"><a href="evolutionary-biology.html#cb143-5" aria-hidden="true" tabindex="-1"></a>(q <span class="sc">+</span> p) <span class="sc">==</span> <span class="dv">1</span></span>
<span id="cb143-6"><a href="evolutionary-biology.html#cb143-6" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate the expected genotype frequencies (_e denotes expected)</span></span>
<span id="cb143-7"><a href="evolutionary-biology.html#cb143-7" aria-hidden="true" tabindex="-1"></a>A1A1_e <span class="ot">&lt;-</span> p<span class="sc">^</span><span class="dv">2</span></span>
<span id="cb143-8"><a href="evolutionary-biology.html#cb143-8" aria-hidden="true" tabindex="-1"></a>A1A2_e <span class="ot">&lt;-</span> <span class="dv">2</span> <span class="sc">*</span> (p <span class="sc">*</span> q)</span>
<span id="cb143-9"><a href="evolutionary-biology.html#cb143-9" aria-hidden="true" tabindex="-1"></a>A2A2_e <span class="ot">&lt;-</span> q<span class="sc">^</span><span class="dv">2</span></span>
<span id="cb143-10"><a href="evolutionary-biology.html#cb143-10" aria-hidden="true" tabindex="-1"></a><span class="co"># show the genotype frequencies in the console</span></span>
<span id="cb143-11"><a href="evolutionary-biology.html#cb143-11" aria-hidden="true" tabindex="-1"></a><span class="fu">c</span>(A1A1_e, A1A2_e, A2A2_e)</span>
<span id="cb143-12"><a href="evolutionary-biology.html#cb143-12" aria-hidden="true" tabindex="-1"></a><span class="co"># since these are genotype frequencies, they should also sum to 1 - you can check this like so</span></span>
<span id="cb143-13"><a href="evolutionary-biology.html#cb143-13" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(<span class="fu">c</span>(A1A1_e, A1A2_e, A2A2_e))</span></code></pre></div>
<p>So you can see it is quite easy to calculate the frequencies you need for the HW model in R. If you want to, you can play around with the initial frequencies defined at the start of the R code to see how it alters the genotype frequencies. This is one nice thing about using R for evolutionary genetics, it is pretty simple to demonstrate things mathematically to benefit your own understanding. However, it might be even more useful to demonstrate the relationships the HW model lays out graphically. That way you can see how the expected genotype frequencies are altered over a whole range of allele frequencies. Luckily for us, R is great for this sort of visualisation <em>and</em> this is an excellent opportunity to demonstrate some programming in R.</p>
<div id="plotting-the-expected-genotype-frequencies" class="section level4" number="3.2.1.1">
<h4><span class="header-section-number">3.2.1.1</span> Plotting the expected genotype frequencies</h4>
<p>The first thing we need to do is generate a range of allele frequencies to visualize the expected genotype frequencies across. We’ll do this in the simplest way first, before going into how you might take a programming approach to the same problem. Either way, generating a range is simple - we just need to do it for a single allele, <span class="math inline">\(A_1\)</span> and then we can very easily derive the frequency of <span class="math inline">\(A_2\)</span>. We will do this like so:</p>
<div class="sourceCode" id="cb144"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb144-1"><a href="evolutionary-biology.html#cb144-1" aria-hidden="true" tabindex="-1"></a><span class="co"># generate a range for p</span></span>
<span id="cb144-2"><a href="evolutionary-biology.html#cb144-2" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="fl">0.01</span>)</span>
<span id="cb144-3"><a href="evolutionary-biology.html#cb144-3" aria-hidden="true" tabindex="-1"></a><span class="co"># and also for q</span></span>
<span id="cb144-4"><a href="evolutionary-biology.html#cb144-4" aria-hidden="true" tabindex="-1"></a>q <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">-</span> p</span></code></pre></div>
<p>All we did here was use the <code>seq</code> function, which we first saw back in <a href="https://evolutionarygenetics.github.io/Introduction.html">Chapter 1</a>. Remember, we only have to do it once because <span class="math inline">\(q\)</span> is the inverse of <span class="math inline">\(p\)</span>. Next we need to generate the expected frequencies. Because R is really great at handling vectors, the code for this is identical for if we were using a single value.</p>
<div class="sourceCode" id="cb145"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb145-1"><a href="evolutionary-biology.html#cb145-1" aria-hidden="true" tabindex="-1"></a><span class="co"># generate the expected genotype frequencies</span></span>
<span id="cb145-2"><a href="evolutionary-biology.html#cb145-2" aria-hidden="true" tabindex="-1"></a>A1A1_e <span class="ot">&lt;-</span> p<span class="sc">^</span><span class="dv">2</span></span>
<span id="cb145-3"><a href="evolutionary-biology.html#cb145-3" aria-hidden="true" tabindex="-1"></a>A1A2_e <span class="ot">&lt;-</span> <span class="dv">2</span> <span class="sc">*</span> (p <span class="sc">*</span> q)</span>
<span id="cb145-4"><a href="evolutionary-biology.html#cb145-4" aria-hidden="true" tabindex="-1"></a>A2A2_e <span class="ot">&lt;-</span> q<span class="sc">^</span><span class="dv">2</span></span></code></pre></div>
<p>So now we basically have the expected frequencies at HWE of our different genotypes across a range of allele frequencies with just a little vector calculation. However, recalling from before that we often need to get our data into a <code>data.frame</code> in order to use <code>ggplot2</code>, we need to do a bit of work on this first. We’ll now get our data into a <code>data.frame</code> ready for working on…</p>
<div class="sourceCode" id="cb146"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb146-1"><a href="evolutionary-biology.html#cb146-1" aria-hidden="true" tabindex="-1"></a><span class="co"># arrange allele frequencies into a tibble/data.frame</span></span>
<span id="cb146-2"><a href="evolutionary-biology.html#cb146-2" aria-hidden="true" tabindex="-1"></a>geno_freq <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(p, q, A1A1_e, A1A2_e, A2A2_e)</span></code></pre></div>
<p>You also might remember from the previous session that we need to sometimes do some data manipulation to get the data into a form that we can easily plot. Luckily, this is quite straightforward in this case as we can use the <code>pivot_longer</code> function to do it quickly and easily.</p>
<div class="sourceCode" id="cb147"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb147-1"><a href="evolutionary-biology.html#cb147-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Use pivot_longer to reshape the data.frame for straightforward plotting</span></span>
<span id="cb147-2"><a href="evolutionary-biology.html#cb147-2" aria-hidden="true" tabindex="-1"></a>geno_freq <span class="ot">&lt;-</span> <span class="fu">pivot_longer</span>(geno_freq, <span class="fu">c</span>(A1A1_e, A1A2_e, A2A2_e), </span>
<span id="cb147-3"><a href="evolutionary-biology.html#cb147-3" aria-hidden="true" tabindex="-1"></a>                          <span class="at">names_to =</span> <span class="st">&quot;genotype&quot;</span>, </span>
<span id="cb147-4"><a href="evolutionary-biology.html#cb147-4" aria-hidden="true" tabindex="-1"></a>                          <span class="at">values_to =</span> <span class="st">&quot;freq&quot;</span>)</span></code></pre></div>
<p><code>pivot_longer()</code> commands can take a while to get the hang of, the best thing is to continually practice them like this! If you’re unsure why we use <code>pivot_longer()</code> or how it works, you can always go back to the explanation from <a href="PLACEHOLDER">week 2</a>.</p>
<p>Now it is really straightforward to make a plot that shows the expected HW—frequencies across our allele frequencies.</p>
<div class="sourceCode" id="cb148"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb148-1"><a href="evolutionary-biology.html#cb148-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot the expected genotype frequencies</span></span>
<span id="cb148-2"><a href="evolutionary-biology.html#cb148-2" aria-hidden="true" tabindex="-1"></a>a <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(geno_freq, <span class="fu">aes</span>(p, freq, <span class="at">colour =</span> genotype)) <span class="sc">+</span> </span>
<span id="cb148-3"><a href="evolutionary-biology.html#cb148-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>() <span class="sc">+</span></span>
<span id="cb148-4"><a href="evolutionary-biology.html#cb148-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">&quot;Genotype frequency&quot;</span>, <span class="at">y =</span> <span class="st">&quot;p frequency&quot;</span>)</span>
<span id="cb148-5"><a href="evolutionary-biology.html#cb148-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb148-6"><a href="evolutionary-biology.html#cb148-6" aria-hidden="true" tabindex="-1"></a><span class="co"># print the plot with some theme options to make it pretty</span></span>
<span id="cb148-7"><a href="evolutionary-biology.html#cb148-7" aria-hidden="true" tabindex="-1"></a>a <span class="sc">+</span> <span class="fu">theme_light</span>() <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">&quot;bottom&quot;</span>)</span></code></pre></div>
<div class="figure"><span id="fig:unnamed-chunk-29"></span>
<img src="Exercise3_files/figure-html/unnamed-chunk-29-1.png" alt="Relationship between allele frequency and genotype frequency under Hardy-Weinberg equilibrium." width="672" />
<p class="caption">
Figure 3.1: Relationship between allele frequency and genotype frequency under Hardy-Weinberg equilibrium.
</p>
</div>
<p>From this graph, you can find the expected genotype frequencies for any value of <span class="math inline">\(p\)</span>. Note for instance how at <span class="math inline">\(p=0.5\)</span>, the frequencies are 0.25 for each of the homozygotes, and 0.5 for the heterozygote. At fixation of <span class="math inline">\(A_1\)</span> (<span class="math inline">\(p=1\)</span>), you can see that we (of course) only have A1A1 homozygotes, and vice versa for <span class="math inline">\(p=0\)</span>.</p>
</div>
</div>
<div id="testing-for-deviations-from-the-hardy-weinberg-expectation" class="section level3" number="3.2.2">
<h3><span class="header-section-number">3.2.2</span> Testing for deviations from the Hardy Weinberg Expectation</h3>
<p>We’ve already learned that the HW model is essentially an idealised scenario where no demographic or evolutionary processes are shaping the relationship between allele and genotype frequencies. We can therefore use it as a null model against which to test our data. Basically, we can compare our observed genotype frequencies from those expected under HWE. To get the expected frequencies, we just generate them from the allele frequencies, as we did in the previous section of the tutorial.</p>
<p>We can work through an example of this together. Once again we’ll assume a locus <span class="math inline">\(A\)</span> with two alleles, <span class="math inline">\(A_1\)</span> and <span class="math inline">\(A_2\)</span>. This means three genotypes, <span class="math inline">\(A_1A_1\)</span>, <span class="math inline">\(A_1A_2\)</span> and <span class="math inline">\(A_2A_2\)</span>. We sample 150 individuals from a population. The next block of R code shows the numbers of each genotype we collected and also combines them into an <code>observed</code> vector - for later use.</p>
<div class="sourceCode" id="cb149"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb149-1"><a href="evolutionary-biology.html#cb149-1" aria-hidden="true" tabindex="-1"></a><span class="co"># numbers of genotypes</span></span>
<span id="cb149-2"><a href="evolutionary-biology.html#cb149-2" aria-hidden="true" tabindex="-1"></a>A1A1 <span class="ot">&lt;-</span> <span class="dv">80</span></span>
<span id="cb149-3"><a href="evolutionary-biology.html#cb149-3" aria-hidden="true" tabindex="-1"></a>A1A2 <span class="ot">&lt;-</span> <span class="dv">15</span></span>
<span id="cb149-4"><a href="evolutionary-biology.html#cb149-4" aria-hidden="true" tabindex="-1"></a>A2A2 <span class="ot">&lt;-</span> <span class="dv">55</span></span>
<span id="cb149-5"><a href="evolutionary-biology.html#cb149-5" aria-hidden="true" tabindex="-1"></a><span class="co"># make into a vector</span></span>
<span id="cb149-6"><a href="evolutionary-biology.html#cb149-6" aria-hidden="true" tabindex="-1"></a>observed <span class="ot">&lt;-</span> <span class="fu">c</span>(A1A1, A1A2, A2A2)</span></code></pre></div>
<p>In order to go further here, we need to work out the allele frequencies from our observed data. This is quite straightforward - homozygotes have two copies of an allele whereas heterozygotes have only one. So we derive the number of each allele and divide it by the total:</p>
<ul>
<li><span class="math inline">\(\displaystyle p = \frac{2(A_1A_1) + A_1A_2}{n}\)</span></li>
<li><span class="math inline">\(\displaystyle q = \frac{2(A_2A_2) + A_1A_2}{n}\)</span></li>
</ul>
<p>Where the genotype notation here represents the number of genotype class and <span class="math inline">\(n\)</span> is the total number of alleles. We can calculate the frequency the alleles in R like so:</p>
<div class="sourceCode" id="cb150"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb150-1"><a href="evolutionary-biology.html#cb150-1" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate total number of alleles</span></span>
<span id="cb150-2"><a href="evolutionary-biology.html#cb150-2" aria-hidden="true" tabindex="-1"></a>n <span class="ot">&lt;-</span> <span class="dv">2</span><span class="sc">*</span><span class="fu">sum</span>(observed)</span>
<span id="cb150-3"><a href="evolutionary-biology.html#cb150-3" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate frequency of A1 or p</span></span>
<span id="cb150-4"><a href="evolutionary-biology.html#cb150-4" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> (<span class="dv">2</span><span class="sc">*</span>(A1A1) <span class="sc">+</span> A1A2)<span class="sc">/</span>n</span>
<span id="cb150-5"><a href="evolutionary-biology.html#cb150-5" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate frequency of A2 or q</span></span>
<span id="cb150-6"><a href="evolutionary-biology.html#cb150-6" aria-hidden="true" tabindex="-1"></a>q <span class="ot">&lt;-</span> (<span class="dv">2</span><span class="sc">*</span>(A2A2) <span class="sc">+</span> A1A2)<span class="sc">/</span>n</span>
<span id="cb150-7"><a href="evolutionary-biology.html#cb150-7" aria-hidden="true" tabindex="-1"></a><span class="co"># print frequencies to screen</span></span>
<span id="cb150-8"><a href="evolutionary-biology.html#cb150-8" aria-hidden="true" tabindex="-1"></a><span class="fu">c</span>(p, q)</span></code></pre></div>
<p>Note that with the code above, we summed the number of observed genotypes and then multiplied by 2 as there are 2 alleles for each individual (i.e. there are 150 individuals and 300 alleles). We now have the allele frequencies, so we can use the code we worked out in the previous section to calculate the expected Hardy-Weinberg frequencies for this population.</p>
<div class="sourceCode" id="cb151"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb151-1"><a href="evolutionary-biology.html#cb151-1" aria-hidden="true" tabindex="-1"></a><span class="co"># generate the expected genotype frequencies</span></span>
<span id="cb151-2"><a href="evolutionary-biology.html#cb151-2" aria-hidden="true" tabindex="-1"></a>A1A1_e <span class="ot">&lt;-</span> p<span class="sc">^</span><span class="dv">2</span></span>
<span id="cb151-3"><a href="evolutionary-biology.html#cb151-3" aria-hidden="true" tabindex="-1"></a>A1A2_e <span class="ot">&lt;-</span> <span class="dv">2</span> <span class="sc">*</span> (p <span class="sc">*</span> q)</span>
<span id="cb151-4"><a href="evolutionary-biology.html#cb151-4" aria-hidden="true" tabindex="-1"></a>A2A2_e <span class="ot">&lt;-</span> q<span class="sc">^</span><span class="dv">2</span></span>
<span id="cb151-5"><a href="evolutionary-biology.html#cb151-5" aria-hidden="true" tabindex="-1"></a><span class="co"># combine into a vector</span></span>
<span id="cb151-6"><a href="evolutionary-biology.html#cb151-6" aria-hidden="true" tabindex="-1"></a>expected_freq <span class="ot">&lt;-</span> <span class="fu">c</span>(A1A1_e, A1A2_e, A2A2_e)</span></code></pre></div>
<p>So now that we have the expected genotype <strong>frequencies</strong>, the last thing we need to do before testing whether there is a deviation from HWE in our data is calculate the expected <strong>number</strong> of each genotype. This is easy - we just multiply the frequencies by the number of individuals we sampled - 150.</p>
<div class="sourceCode" id="cb152"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb152-1"><a href="evolutionary-biology.html#cb152-1" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate observed genotype frequencies</span></span>
<span id="cb152-2"><a href="evolutionary-biology.html#cb152-2" aria-hidden="true" tabindex="-1"></a>expected <span class="ot">&lt;-</span> expected_freq <span class="sc">*</span> <span class="dv">150</span></span></code></pre></div>
<p>The next thing we’d like to do is see whether our observed genotypes deviate from the expected. We will test this formally, but first let’s just look at the differences between them. The simplest way to do this is combine them into a matrix:</p>
<div class="sourceCode" id="cb153"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb153-1"><a href="evolutionary-biology.html#cb153-1" aria-hidden="true" tabindex="-1"></a><span class="co"># combine observed and expected frequencies into a matrix</span></span>
<span id="cb153-2"><a href="evolutionary-biology.html#cb153-2" aria-hidden="true" tabindex="-1"></a>mydata <span class="ot">&lt;-</span> <span class="fu">cbind</span>(observed, expected)</span>
<span id="cb153-3"><a href="evolutionary-biology.html#cb153-3" aria-hidden="true" tabindex="-1"></a><span class="co"># add rownames</span></span>
<span id="cb153-4"><a href="evolutionary-biology.html#cb153-4" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(mydata) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;A1A1&quot;</span>, <span class="st">&quot;A1A2&quot;</span>, <span class="st">&quot;A2A2&quot;</span>)</span>
<span id="cb153-5"><a href="evolutionary-biology.html#cb153-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb153-6"><a href="evolutionary-biology.html#cb153-6" aria-hidden="true" tabindex="-1"></a>mydata</span>
<span id="cb153-7"><a href="evolutionary-biology.html#cb153-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;      observed expected</span></span>
<span id="cb153-8"><a href="evolutionary-biology.html#cb153-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; A1A1       80 51.04167</span></span>
<span id="cb153-9"><a href="evolutionary-biology.html#cb153-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; A1A2       15 72.91667</span></span>
<span id="cb153-10"><a href="evolutionary-biology.html#cb153-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; A2A2       55 26.04167</span></span></code></pre></div>
<p>Well, just from eyeballing this matrix, we see that there is a difference between the expected and the observed genotypes. Notably there seems to be a lower frequency of heterozygotes than we expect. Still, just looking at this is not enough, we need to test it formally. As you will probably remember from the main text, we can use a Chi-squared goodness of fit test for this purpose.</p>
<p>When we use a statistical test like the Chi-squared test, we want to find out how likely it is that we would get our observed values when we assume that our null hypothesis is true. Here, our null hypothesis is the Hardy-Weinberg expectation (<span class="math inline">\(p^2\)</span>, <span class="math inline">\(2pq\)</span>, <span class="math inline">\(q^2\)</span>). These are the genotype frequencies we would expect to see if the population was in Hardy-Weinberg equilibrium. If the difference between the observed data and the expectation from the null hypothesis is large enough, we can say that it is statistically significant. What is “large enough” is what our chi squared test will tell us – this is where the p-value comes in. If we get a low p-value, we can reject the null hypothesis and conclude that our population is not in Hardy-Weinberg equilibrium.</p>
<p>A p-value tells us the probability of getting our observed result when we assume that the null hypothesis is true. A p-value of 0.5 means there is a 50% probability that we would get these results if the null hypothesis was true. Another way to look at it is that if we sampled 150 individuals from a population in Hardy-Weinberg equilibrium, we would get results like ours (our more extreme results, meaning a larger difference between the observed and the expected genotypes) 50% of the time. The cut-off for saying that a result is statistically significant is typically a p-value below 0.05. If our p-value is 0.05, it means there is a probability of only 5% that we would get these results if our null hypothesis was true.<a href="#fn18" class="footnote-ref" id="fnref18"><sup>18</sup></a></p>
<p>In the book, we calculated our chi-squared manually. In R, we can simply use the function chisq.test():</p>
<div class="sourceCode" id="cb154"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb154-1"><a href="evolutionary-biology.html#cb154-1" aria-hidden="true" tabindex="-1"></a><span class="co"># perform a chi-squared test</span></span>
<span id="cb154-2"><a href="evolutionary-biology.html#cb154-2" aria-hidden="true" tabindex="-1"></a>mychi <span class="ot">&lt;-</span> <span class="fu">chisq.test</span>(observed, <span class="at">p =</span> expected_freq)</span></code></pre></div>
<p>Note that what we put into the function is the <strong>number</strong> of observed genotypes (stored in our <em>observed</em> variable) and the <strong>frequency</strong> of the expectation (stored in <em>expected_freq</em>).</p>
<p>Then we can have a look at the results:</p>
<div class="sourceCode" id="cb155"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb155-1"><a href="evolutionary-biology.html#cb155-1" aria-hidden="true" tabindex="-1"></a>mychi</span>
<span id="cb155-2"><a href="evolutionary-biology.html#cb155-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb155-3"><a href="evolutionary-biology.html#cb155-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  Chi-squared test for given probabilities</span></span>
<span id="cb155-4"><a href="evolutionary-biology.html#cb155-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb155-5"><a href="evolutionary-biology.html#cb155-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; data:  observed</span></span>
<span id="cb155-6"><a href="evolutionary-biology.html#cb155-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; X-squared = 94.633, df = 2, p-value &lt; 2.2e-16</span></span></code></pre></div>
<p>We see that the p-value is very low - well below our limit of 0.05. This means we have a highly significant deviation from HWE - in other words, some assumption of the Hardy-Weinberg model is violated. What process could cause this significant heterozygote deficit? Maybe non-random mating?</p>
</div>
<div id="simulating-genetic-drift" class="section level3" number="3.2.3">
<h3><span class="header-section-number">3.2.3</span> Simulating genetic drift</h3>
<p>One of the main assumptions of the HW model is that populations are infinite. This is obviously unrealistic! All populations are finite - but what does this mean for how evolution unfolds? An important concept in evolutionary genetics is how population size can influence the relative power of a process such as <strong>genetic drift</strong>. Drift essentially describes the random component of survival in a population - some individuals will survive and reproduce more than others by chance. An important factor contributing to such genetic drift is the random sampling of alleles that takes place at fertilization. For instance, just like some pairs produce an excess of sons and daughters, a heterozygous mating pair (<span class="math inline">\(A_1A_2\)</span>) can produce offspring with an excess of the genotypes <span class="math inline">\(A_1A_1\)</span> (or <span class="math inline">\(A_2A_2\)</span>) thereby causing a random change in allele frequency in the population. When a population is large, the effects of drift will be small relative to other processes such as selection. However, when a population is small, drift as a result of random chance has a much bigger effect.</p>
<p>One of the best ways to get your head around a concept like genetic drift and how it interacts with population size is to actually simulate, which is exactly what we are going to do here. We’ll first explain how we can simulate it over a single generation and then we are going to use a for-loop to do all the leg work for us. So we’ll not only be exercising your understanding in evolutionary genetics, but your R programming chops too.</p>
<div id="a-simple-case-of-genetic-drift-over-a-single-generation-using-the-binomial-distribution" class="section level4" number="3.2.3.1">
<h4><span class="header-section-number">3.2.3.1</span> A simple case of genetic drift over a single generation using the binomial distribution</h4>
<p>To demonstrate a simple case of drift, we’ll use the example from the book. We have a population of N = 8 individuals and we will focus on a single locus A with two alleles, <span class="math inline">\(A_1\)</span> and <span class="math inline">\(A_2\)</span>. We will assume both alleles are equally common in the population - i.e. <span class="math inline">\(p=q = 0.5\)</span>.</p>
<p>In the book you will recall we used a coin-flipping experiment to sample our alleles to form genotypes, since both alleles have a frequency of 0.5. Now we will do something similar, but instead of sampling alleles to form genotypes, we will sample the number of <span class="math inline">\(A_1\)</span> alleles in the next generation directly (we call this <span class="math inline">\(p’\)</span>). In the example in the book, we defined heads as <span class="math inline">\(A_1\)</span> and tails as <span class="math inline">\(A_2\)</span>. The principle here is the same, we start with a population where <span class="math inline">\(p = q\)</span>, so there is an equal probabilty of sampling one allele or the other. You can imagine each coin flip representing the passing on of one allele from a parent to the next generation. If we do this twice for each individual in the population, the sum of the heads we get will be the number of A1 alleles in the next generation.</p>
<p>This is a form of binomial sampling, and to do it ir R we can use the rbinom() function. The <code>rbinom()</code> function essentially simulates a coin flip. However, while in a regular coin flip the probability of heads and tails is always 0.5, <code>rbinom()</code> allows for changing these probabilities. Simplified, the following code says: “flip 1 coin (<code>n</code>) 16 times (<code>size</code>) with a probability of 0.5 (<code>prob</code>) for getting heads”. The output is the number of heads we got.</p>
<div class="sourceCode" id="cb156"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb156-1"><a href="evolutionary-biology.html#cb156-1" aria-hidden="true" tabindex="-1"></a><span class="co"># recreate coin flipping experiment using binomial sampling</span></span>
<span id="cb156-2"><a href="evolutionary-biology.html#cb156-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rbinom</span>(<span class="at">n =</span> <span class="dv">1</span>, <span class="at">size =</span> <span class="dv">16</span>, <span class="at">prob =</span> <span class="fl">0.5</span>)</span></code></pre></div>
<p>Run this code a few times, and observe that the results are different each time, because the <code>rbinom()</code> function samples randomly.</p>
<p>In our example, the output of <code>rbinom</code> could represent the number of <span class="math inline">\(A_1\)</span> alleles in the next generation – so we could easily work out <span class="math inline">\(p´\)</span> and <span class="math inline">\(q´\)</span>, (i.e. <span class="math inline">\(p\)</span> and <span class="math inline">\(q\)</span> in the next generation). We can use the following code to sample the number of <span class="math inline">\(A_1\)</span> in generation 2:</p>
<div class="sourceCode" id="cb157"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb157-1"><a href="evolutionary-biology.html#cb157-1" aria-hidden="true" tabindex="-1"></a><span class="co"># set population size</span></span>
<span id="cb157-2"><a href="evolutionary-biology.html#cb157-2" aria-hidden="true" tabindex="-1"></a>N <span class="ot">&lt;-</span> <span class="dv">8</span></span>
<span id="cb157-3"><a href="evolutionary-biology.html#cb157-3" aria-hidden="true" tabindex="-1"></a><span class="co"># set allele frequency</span></span>
<span id="cb157-4"><a href="evolutionary-biology.html#cb157-4" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fl">0.5</span> </span>
<span id="cb157-5"><a href="evolutionary-biology.html#cb157-5" aria-hidden="true" tabindex="-1"></a><span class="co"># recreate coin flipping experiment using binomial sampling</span></span>
<span id="cb157-6"><a href="evolutionary-biology.html#cb157-6" aria-hidden="true" tabindex="-1"></a>nA1 <span class="ot">&lt;-</span> <span class="fu">rbinom</span>(<span class="dv">1</span>, <span class="at">size =</span> <span class="dv">2</span><span class="sc">*</span>N, <span class="at">prob =</span> p)</span></code></pre></div>
<p>Remember since we are sampling alleles for diploid individuals, we need to multiply <code>N</code> by 2. The value of <code>nA1</code> will differ for everyone since this is a <strong>random</strong> sample of the binomial distribution. We can now easily work out the value of <span class="math inline">\(p´\)</span> - i.e. the new frequency of the <span class="math inline">\(A_1\)</span> in the next generation. Note that R doesn’t allow the prime notation so for the actual R code, we will use <code>p2</code>.</p>
<div class="sourceCode" id="cb158"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb158-1"><a href="evolutionary-biology.html#cb158-1" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate p in generation 2</span></span>
<span id="cb158-2"><a href="evolutionary-biology.html#cb158-2" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> nA1<span class="sc">/</span>(<span class="dv">2</span><span class="sc">*</span>N)</span></code></pre></div>
<p>OK, so now we have a value for <code>p2</code> – again it will differ for everyone because we performed random sampling. You will see though that <code>p2</code> is likely different to <code>p</code> solely because of random sampling - i.e. we have demonstrated drift over a single generation. Then we can go on to calculate <span class="math inline">\(p\)</span> in the third generation. Our new <code>p2</code> will then be the <code>prob</code> argument of <code>rbinom()</code>, and everything else can be left unchanged:</p>
<div class="sourceCode" id="cb159"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb159-1"><a href="evolutionary-biology.html#cb159-1" aria-hidden="true" tabindex="-1"></a><span class="co"># draw a new random sample</span></span>
<span id="cb159-2"><a href="evolutionary-biology.html#cb159-2" aria-hidden="true" tabindex="-1"></a>nA1 <span class="ot">&lt;-</span> <span class="fu">rbinom</span>(<span class="dv">1</span>, <span class="at">size =</span> <span class="dv">2</span><span class="sc">*</span>N, <span class="at">prob =</span> p2)</span>
<span id="cb159-3"><a href="evolutionary-biology.html#cb159-3" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate p in generation 3</span></span>
<span id="cb159-4"><a href="evolutionary-biology.html#cb159-4" aria-hidden="true" tabindex="-1"></a>p3 <span class="ot">&lt;-</span> nA1<span class="sc">/</span>(<span class="dv">2</span><span class="sc">*</span>N)</span>
<span id="cb159-5"><a href="evolutionary-biology.html#cb159-5" aria-hidden="true" tabindex="-1"></a>p3</span>
<span id="cb159-6"><a href="evolutionary-biology.html#cb159-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 0.75</span></span></code></pre></div>
<p>If we were to to properly understand drift over more than one generation, we could now have used our <code>p3</code> to find p in the fourth generation, use the new <span class="math inline">\(p\)</span> to calculate <span class="math inline">\(p\)</span> in the next generation, and so on until we have simulated for all the number of generations we want.</p>
<div class="fold c">
<div class="sourceCode" id="cb160"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb160-1"><a href="evolutionary-biology.html#cb160-1" aria-hidden="true" tabindex="-1"></a><span class="co"># draw a new random sample</span></span>
<span id="cb160-2"><a href="evolutionary-biology.html#cb160-2" aria-hidden="true" tabindex="-1"></a>nA1 <span class="ot">&lt;-</span> <span class="fu">rbinom</span>(<span class="dv">1</span>, <span class="at">size =</span> <span class="dv">2</span><span class="sc">*</span>N, <span class="at">prob =</span> p3)</span>
<span id="cb160-3"><a href="evolutionary-biology.html#cb160-3" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate p in generation 4</span></span>
<span id="cb160-4"><a href="evolutionary-biology.html#cb160-4" aria-hidden="true" tabindex="-1"></a>p4 <span class="ot">&lt;-</span> nA1<span class="sc">/</span>(<span class="dv">2</span><span class="sc">*</span>N)</span>
<span id="cb160-5"><a href="evolutionary-biology.html#cb160-5" aria-hidden="true" tabindex="-1"></a>p4</span>
<span id="cb160-6"><a href="evolutionary-biology.html#cb160-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 0.8125</span></span>
<span id="cb160-7"><a href="evolutionary-biology.html#cb160-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb160-8"><a href="evolutionary-biology.html#cb160-8" aria-hidden="true" tabindex="-1"></a><span class="co"># draw a new random sample</span></span>
<span id="cb160-9"><a href="evolutionary-biology.html#cb160-9" aria-hidden="true" tabindex="-1"></a>nA1 <span class="ot">&lt;-</span> <span class="fu">rbinom</span>(<span class="dv">1</span>, <span class="at">size =</span> <span class="dv">2</span><span class="sc">*</span>N, <span class="at">prob =</span> p4)</span>
<span id="cb160-10"><a href="evolutionary-biology.html#cb160-10" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate p in generation 5</span></span>
<span id="cb160-11"><a href="evolutionary-biology.html#cb160-11" aria-hidden="true" tabindex="-1"></a>p5 <span class="ot">&lt;-</span> nA1<span class="sc">/</span>(<span class="dv">2</span><span class="sc">*</span>N)</span>
<span id="cb160-12"><a href="evolutionary-biology.html#cb160-12" aria-hidden="true" tabindex="-1"></a>p5</span>
<span id="cb160-13"><a href="evolutionary-biology.html#cb160-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 0.75</span></span>
<span id="cb160-14"><a href="evolutionary-biology.html#cb160-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb160-15"><a href="evolutionary-biology.html#cb160-15" aria-hidden="true" tabindex="-1"></a><span class="co"># draw a new random sample</span></span>
<span id="cb160-16"><a href="evolutionary-biology.html#cb160-16" aria-hidden="true" tabindex="-1"></a>nA1 <span class="ot">&lt;-</span> <span class="fu">rbinom</span>(<span class="dv">1</span>, <span class="at">size =</span> <span class="dv">2</span><span class="sc">*</span>N, <span class="at">prob =</span> p5)</span>
<span id="cb160-17"><a href="evolutionary-biology.html#cb160-17" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate p in generation 5</span></span>
<span id="cb160-18"><a href="evolutionary-biology.html#cb160-18" aria-hidden="true" tabindex="-1"></a>p6 <span class="ot">&lt;-</span> nA1<span class="sc">/</span>(<span class="dv">2</span><span class="sc">*</span>N)</span>
<span id="cb160-19"><a href="evolutionary-biology.html#cb160-19" aria-hidden="true" tabindex="-1"></a>p6</span>
<span id="cb160-20"><a href="evolutionary-biology.html#cb160-20" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 0.875</span></span>
<span id="cb160-21"><a href="evolutionary-biology.html#cb160-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb160-22"><a href="evolutionary-biology.html#cb160-22" aria-hidden="true" tabindex="-1"></a><span class="co"># and so on ...</span></span></code></pre></div>
</div>
<p>But hopefully, you’re thinking: “that sounds like a lot of work, maybe we can use a for-loop instead?”. We can!</p>
</div>
<div id="using-a-for-loop-to-simulate-drift-over-multiple-generations" class="section level4" number="3.2.3.2">
<h4><span class="header-section-number">3.2.3.2</span> Using a for-loop to simulate drift over multiple generations</h4>
<p>Let’s do try to wrap all that tedious code in a for loop instead. If we simplify our problem by a lot, what we want to do is: “use p from the previous generation to calculate p in this generation”. If your p-values are in a vector <code>p</code>, and you are using indexing like we learned in section <a href="r-programming-for-loops.html#indexing">3.1.3</a>, “p from the previous generation” is <code>p[i-1]</code>, and “p in this generation” is <code>p[i]</code>. The contents of our for-loop would then be:</p>
<div class="sourceCode" id="cb161"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb161-1"><a href="evolutionary-biology.html#cb161-1" aria-hidden="true" tabindex="-1"></a>nA1 <span class="ot">&lt;-</span> <span class="fu">rbinom</span>(<span class="dv">1</span>, <span class="at">size =</span> <span class="dv">2</span><span class="sc">*</span>N, <span class="at">prob =</span> p[i <span class="sc">-</span> <span class="dv">1</span>])</span>
<span id="cb161-2"><a href="evolutionary-biology.html#cb161-2" aria-hidden="true" tabindex="-1"></a>p[i] <span class="ot">&lt;-</span> nA1<span class="sc">/</span>(<span class="dv">2</span><span class="sc">*</span>N)</span></code></pre></div>
<p>All that remains then is to create the for-loop itself. See if you manage to make this yourself before looking at my code!</p>
<details>
<summary>
<p>Show hint</p>
</summary>
<p>Hint: start by creating a vector <code>p</code> to store your results. The length of this vector determines how many generations you use in your simulation. Then, set the first element of <code>p</code> to 0.5</p>
</details>
<details>
<summary>
<p>Show another hint</p>
</summary>
<p>Hint: start the vector you’re looping over at 2, and end at the number of generations (i.e. the length of your <code>p</code> vector)</p>
</details>
<div class="fold c">
<div class="sourceCode" id="cb162"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb162-1"><a href="evolutionary-biology.html#cb162-1" aria-hidden="true" tabindex="-1"></a><span class="co"># set population size</span></span>
<span id="cb162-2"><a href="evolutionary-biology.html#cb162-2" aria-hidden="true" tabindex="-1"></a>N <span class="ot">&lt;-</span> <span class="dv">8</span></span>
<span id="cb162-3"><a href="evolutionary-biology.html#cb162-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb162-4"><a href="evolutionary-biology.html#cb162-4" aria-hidden="true" tabindex="-1"></a><span class="co"># set number of generations</span></span>
<span id="cb162-5"><a href="evolutionary-biology.html#cb162-5" aria-hidden="true" tabindex="-1"></a>ngen <span class="ot">&lt;-</span> <span class="dv">100</span></span>
<span id="cb162-6"><a href="evolutionary-biology.html#cb162-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb162-7"><a href="evolutionary-biology.html#cb162-7" aria-hidden="true" tabindex="-1"></a><span class="co"># set initial frequency of A1</span></span>
<span id="cb162-8"><a href="evolutionary-biology.html#cb162-8" aria-hidden="true" tabindex="-1"></a>p_init <span class="ot">&lt;-</span> <span class="fl">0.5</span></span>
<span id="cb162-9"><a href="evolutionary-biology.html#cb162-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb162-10"><a href="evolutionary-biology.html#cb162-10" aria-hidden="true" tabindex="-1"></a><span class="co"># create vector for storing results</span></span>
<span id="cb162-11"><a href="evolutionary-biology.html#cb162-11" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="cn">NA</span>, ngen)</span>
<span id="cb162-12"><a href="evolutionary-biology.html#cb162-12" aria-hidden="true" tabindex="-1"></a><span class="co"># set first element to initial p</span></span>
<span id="cb162-13"><a href="evolutionary-biology.html#cb162-13" aria-hidden="true" tabindex="-1"></a>p[<span class="dv">1</span>] <span class="ot">&lt;-</span> p_init</span>
<span id="cb162-14"><a href="evolutionary-biology.html#cb162-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb162-15"><a href="evolutionary-biology.html#cb162-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb162-16"><a href="evolutionary-biology.html#cb162-16" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">2</span><span class="sc">:</span>ngen){</span>
<span id="cb162-17"><a href="evolutionary-biology.html#cb162-17" aria-hidden="true" tabindex="-1"></a>  <span class="co"># sample number of A1 alleles based on p in previous generation</span></span>
<span id="cb162-18"><a href="evolutionary-biology.html#cb162-18" aria-hidden="true" tabindex="-1"></a>  nA1 <span class="ot">&lt;-</span> <span class="fu">rbinom</span>(<span class="dv">1</span>, <span class="dv">2</span><span class="sc">*</span>N, p[i<span class="dv">-1</span>])</span>
<span id="cb162-19"><a href="evolutionary-biology.html#cb162-19" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb162-20"><a href="evolutionary-biology.html#cb162-20" aria-hidden="true" tabindex="-1"></a>  <span class="co"># set frequency of A1 as p in the current generation</span></span>
<span id="cb162-21"><a href="evolutionary-biology.html#cb162-21" aria-hidden="true" tabindex="-1"></a>  p[i] <span class="ot">&lt;-</span> nA1 <span class="sc">/</span> (<span class="dv">2</span><span class="sc">*</span>N)</span>
<span id="cb162-22"><a href="evolutionary-biology.html#cb162-22" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb162-23"><a href="evolutionary-biology.html#cb162-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb162-24"><a href="evolutionary-biology.html#cb162-24" aria-hidden="true" tabindex="-1"></a>p</span>
<span id="cb162-25"><a href="evolutionary-biology.html#cb162-25" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   [1] 0.5000 0.5625 0.6250 0.6875 0.6875 0.7500 0.8125 1.0000 1.0000 1.0000</span></span>
<span id="cb162-26"><a href="evolutionary-biology.html#cb162-26" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  [11] 1.0000 1.0000 1.0000 1.0000 1.0000 1.0000 1.0000 1.0000 1.0000 1.0000</span></span>
<span id="cb162-27"><a href="evolutionary-biology.html#cb162-27" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  [21] 1.0000 1.0000 1.0000 1.0000 1.0000 1.0000 1.0000 1.0000 1.0000 1.0000</span></span>
<span id="cb162-28"><a href="evolutionary-biology.html#cb162-28" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  [31] 1.0000 1.0000 1.0000 1.0000 1.0000 1.0000 1.0000 1.0000 1.0000 1.0000</span></span>
<span id="cb162-29"><a href="evolutionary-biology.html#cb162-29" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  [41] 1.0000 1.0000 1.0000 1.0000 1.0000 1.0000 1.0000 1.0000 1.0000 1.0000</span></span>
<span id="cb162-30"><a href="evolutionary-biology.html#cb162-30" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  [51] 1.0000 1.0000 1.0000 1.0000 1.0000 1.0000 1.0000 1.0000 1.0000 1.0000</span></span>
<span id="cb162-31"><a href="evolutionary-biology.html#cb162-31" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  [61] 1.0000 1.0000 1.0000 1.0000 1.0000 1.0000 1.0000 1.0000 1.0000 1.0000</span></span>
<span id="cb162-32"><a href="evolutionary-biology.html#cb162-32" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  [71] 1.0000 1.0000 1.0000 1.0000 1.0000 1.0000 1.0000 1.0000 1.0000 1.0000</span></span>
<span id="cb162-33"><a href="evolutionary-biology.html#cb162-33" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  [81] 1.0000 1.0000 1.0000 1.0000 1.0000 1.0000 1.0000 1.0000 1.0000 1.0000</span></span>
<span id="cb162-34"><a href="evolutionary-biology.html#cb162-34" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  [91] 1.0000 1.0000 1.0000 1.0000 1.0000 1.0000 1.0000 1.0000 1.0000 1.0000</span></span></code></pre></div>
</div>
<p>Note that I assigned the population size <code>N</code>, the number of generations <code>ngen</code> and the initial frequency of <span class="math inline">\(A_1\)</span> <code>p_init</code> to named objects. This way, if you want to change either of these, you only need to change them in one place!</p>
</div>
<div id="visualizing-drift" class="section level4" number="3.2.3.3">
<h4><span class="header-section-number">3.2.3.3</span> Visualizing drift</h4>
<p>Now that we our for-loop to simulate data across generations, we can easily do this for different population sizes. Let’s simulate 1000 generations of drift for three populations of 10, 100 and 1000 individuals. We’ll keep the initial allele frequency, <code>p</code> at 0.5 in all cases. We will then combine them all into a <code>data.frame</code> alongside the number of generations. I will skip most of the comments in the code here so the code doesn’t take up too much space, but remember that you can look at the code in the previous section if you wonder what something does.</p>
<p>First, for a population size of 10, storing the results to an object called <code>n10</code></p>
<div class="sourceCode" id="cb163"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb163-1"><a href="evolutionary-biology.html#cb163-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Population size 10</span></span>
<span id="cb163-2"><a href="evolutionary-biology.html#cb163-2" aria-hidden="true" tabindex="-1"></a>N <span class="ot">&lt;-</span> <span class="dv">10</span></span>
<span id="cb163-3"><a href="evolutionary-biology.html#cb163-3" aria-hidden="true" tabindex="-1"></a>ngen <span class="ot">&lt;-</span> <span class="dv">1000</span></span>
<span id="cb163-4"><a href="evolutionary-biology.html#cb163-4" aria-hidden="true" tabindex="-1"></a>p_init <span class="ot">&lt;-</span> <span class="fl">0.5</span></span>
<span id="cb163-5"><a href="evolutionary-biology.html#cb163-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb163-6"><a href="evolutionary-biology.html#cb163-6" aria-hidden="true" tabindex="-1"></a>n10 <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="cn">NA</span>, ngen)</span>
<span id="cb163-7"><a href="evolutionary-biology.html#cb163-7" aria-hidden="true" tabindex="-1"></a>n10[<span class="dv">1</span>] <span class="ot">&lt;-</span> p_init</span>
<span id="cb163-8"><a href="evolutionary-biology.html#cb163-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb163-9"><a href="evolutionary-biology.html#cb163-9" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">2</span><span class="sc">:</span>ngen){</span>
<span id="cb163-10"><a href="evolutionary-biology.html#cb163-10" aria-hidden="true" tabindex="-1"></a>  nA1 <span class="ot">&lt;-</span> <span class="fu">rbinom</span>(<span class="dv">1</span>, <span class="dv">2</span><span class="sc">*</span>N, n10[i<span class="dv">-1</span>])</span>
<span id="cb163-11"><a href="evolutionary-biology.html#cb163-11" aria-hidden="true" tabindex="-1"></a>  n10[i] <span class="ot">&lt;-</span> nA1 <span class="sc">/</span> (<span class="dv">2</span><span class="sc">*</span>N)</span>
<span id="cb163-12"><a href="evolutionary-biology.html#cb163-12" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<p>Then we do the same for a population size of 100.</p>
<div class="sourceCode" id="cb164"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb164-1"><a href="evolutionary-biology.html#cb164-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Population size 100</span></span>
<span id="cb164-2"><a href="evolutionary-biology.html#cb164-2" aria-hidden="true" tabindex="-1"></a>N <span class="ot">&lt;-</span> <span class="dv">100</span></span>
<span id="cb164-3"><a href="evolutionary-biology.html#cb164-3" aria-hidden="true" tabindex="-1"></a>ngen <span class="ot">&lt;-</span> <span class="dv">1000</span></span>
<span id="cb164-4"><a href="evolutionary-biology.html#cb164-4" aria-hidden="true" tabindex="-1"></a>p_init <span class="ot">&lt;-</span> <span class="fl">0.5</span></span>
<span id="cb164-5"><a href="evolutionary-biology.html#cb164-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb164-6"><a href="evolutionary-biology.html#cb164-6" aria-hidden="true" tabindex="-1"></a>n100 <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="cn">NA</span>, ngen)</span>
<span id="cb164-7"><a href="evolutionary-biology.html#cb164-7" aria-hidden="true" tabindex="-1"></a>n100[<span class="dv">1</span>] <span class="ot">&lt;-</span> p_init</span>
<span id="cb164-8"><a href="evolutionary-biology.html#cb164-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb164-9"><a href="evolutionary-biology.html#cb164-9" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">2</span><span class="sc">:</span>ngen){</span>
<span id="cb164-10"><a href="evolutionary-biology.html#cb164-10" aria-hidden="true" tabindex="-1"></a>  nA1 <span class="ot">&lt;-</span> <span class="fu">rbinom</span>(<span class="dv">1</span>, <span class="dv">2</span><span class="sc">*</span>N, n100[i<span class="dv">-1</span>])</span>
<span id="cb164-11"><a href="evolutionary-biology.html#cb164-11" aria-hidden="true" tabindex="-1"></a>  n100[i] <span class="ot">&lt;-</span> nA1 <span class="sc">/</span> (<span class="dv">2</span><span class="sc">*</span>N)</span>
<span id="cb164-12"><a href="evolutionary-biology.html#cb164-12" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<p>And finally for population size 1000.</p>
<div class="sourceCode" id="cb165"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb165-1"><a href="evolutionary-biology.html#cb165-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Population size 1000</span></span>
<span id="cb165-2"><a href="evolutionary-biology.html#cb165-2" aria-hidden="true" tabindex="-1"></a>N <span class="ot">&lt;-</span> <span class="dv">1000</span></span>
<span id="cb165-3"><a href="evolutionary-biology.html#cb165-3" aria-hidden="true" tabindex="-1"></a>ngen <span class="ot">&lt;-</span> <span class="dv">1000</span></span>
<span id="cb165-4"><a href="evolutionary-biology.html#cb165-4" aria-hidden="true" tabindex="-1"></a>p_init <span class="ot">&lt;-</span> <span class="fl">0.5</span></span>
<span id="cb165-5"><a href="evolutionary-biology.html#cb165-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb165-6"><a href="evolutionary-biology.html#cb165-6" aria-hidden="true" tabindex="-1"></a>n1000 <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="cn">NA</span>, ngen)</span>
<span id="cb165-7"><a href="evolutionary-biology.html#cb165-7" aria-hidden="true" tabindex="-1"></a>n1000[<span class="dv">1</span>] <span class="ot">&lt;-</span> p_init</span>
<span id="cb165-8"><a href="evolutionary-biology.html#cb165-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb165-9"><a href="evolutionary-biology.html#cb165-9" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">2</span><span class="sc">:</span>ngen){</span>
<span id="cb165-10"><a href="evolutionary-biology.html#cb165-10" aria-hidden="true" tabindex="-1"></a>  nA1 <span class="ot">&lt;-</span> <span class="fu">rbinom</span>(<span class="dv">1</span>, <span class="dv">2</span><span class="sc">*</span>N, n1000[i<span class="dv">-1</span>])</span>
<span id="cb165-11"><a href="evolutionary-biology.html#cb165-11" aria-hidden="true" tabindex="-1"></a>  n1000[i] <span class="ot">&lt;-</span> nA1 <span class="sc">/</span> (<span class="dv">2</span><span class="sc">*</span>N)</span>
<span id="cb165-12"><a href="evolutionary-biology.html#cb165-12" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<p>As you may remember, we need to make it into a data frame in order to plot it with <code>ggplot2</code>. We also add a column called <code>g</code> that contains the number of generations.</p>
<div class="sourceCode" id="cb166"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb166-1"><a href="evolutionary-biology.html#cb166-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get number of generations</span></span>
<span id="cb166-2"><a href="evolutionary-biology.html#cb166-2" aria-hidden="true" tabindex="-1"></a>g <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="dv">1</span>, <span class="dv">1000</span>, <span class="dv">1</span>)</span>
<span id="cb166-3"><a href="evolutionary-biology.html#cb166-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb166-4"><a href="evolutionary-biology.html#cb166-4" aria-hidden="true" tabindex="-1"></a><span class="co"># combine into a data.frame</span></span>
<span id="cb166-5"><a href="evolutionary-biology.html#cb166-5" aria-hidden="true" tabindex="-1"></a>mydrift <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(g, n10, n100, n1000)</span></code></pre></div>
<p>Anyway, what we ultimately want to do is <em>visualise</em> drift over time. So we need to get our data in a position to plot it using <code>ggplot2</code> - once again, we turn to <code>pivot_longer()</code>. <code>pivot_longer()</code> in this case takes all our simulations (which at the moment are in different columns) and combines them into a single column <code>p</code>. A new column <code>pop_size</code> is made which indicates the name of the column where the data came from.</p>
<div class="sourceCode" id="cb167"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb167-1"><a href="evolutionary-biology.html#cb167-1" aria-hidden="true" tabindex="-1"></a><span class="co"># use pivot_longer to get data ready for plotting</span></span>
<span id="cb167-2"><a href="evolutionary-biology.html#cb167-2" aria-hidden="true" tabindex="-1"></a>mydrift_long <span class="ot">&lt;-</span> <span class="fu">pivot_longer</span>(mydrift, <span class="sc">-</span>g, <span class="at">names_to =</span> <span class="st">&quot;pop_size&quot;</span>, <span class="at">values_to =</span> <span class="st">&quot;p&quot;</span>)</span></code></pre></div>
<p>Now we will plot our data using the following code:</p>
<div class="sourceCode" id="cb168"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb168-1"><a href="evolutionary-biology.html#cb168-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot data</span></span>
<span id="cb168-2"><a href="evolutionary-biology.html#cb168-2" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(mydrift_long, <span class="fu">aes</span>(g, p, <span class="at">colour =</span> pop_size)) <span class="sc">+</span></span>
<span id="cb168-3"><a href="evolutionary-biology.html#cb168-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>() <span class="sc">+</span></span>
<span id="cb168-4"><a href="evolutionary-biology.html#cb168-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">&quot;No. generations&quot;</span>, <span class="at">y =</span> <span class="st">&quot;Allele freq (p)&quot;</span>)</span>
<span id="cb168-5"><a href="evolutionary-biology.html#cb168-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb168-6"><a href="evolutionary-biology.html#cb168-6" aria-hidden="true" tabindex="-1"></a><span class="co"># print nicely with some theme options</span></span>
<span id="cb168-7"><a href="evolutionary-biology.html#cb168-7" aria-hidden="true" tabindex="-1"></a>p <span class="sc">+</span> <span class="fu">theme_bw</span>() <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">&quot;bottom&quot;</span>) </span></code></pre></div>
<div class="figure"><span id="fig:unnamed-chunk-45"></span>
<img src="Exercise3_files/figure-html/unnamed-chunk-45-1.png" alt="A simulation of genetic drift over 1000 generations with different population sizes." width="672" />
<p class="caption">
Figure 3.2: A simulation of genetic drift over 1000 generations with different population sizes.
</p>
</div>
<p>And there we have it! A plot of our simulated data showing allele frequency changes over time as a result of genetic drift. Although the plot you get will differ from the one you see here (because of the random sampling), what you should be able to see is that with a low population size alleles are either lost or go to fixation very rapidly. This is because the effect of drift is much stronger here, whereas in larger populations it may take a long time for drift to fix or remove an allele - indeed in some cases it will not happen at all.</p>
</div>
</div>
</div>
<div class="footnotes">
<hr />
<ol start="18">
<li id="fn18"><p>Keep in mind that the value 0.05 is arbitrary, and that other (often lower) values are often used in other fields. Researchers simply decided that 5% is an acceptable degree of error. This also means that 5% of studies conducted (where the statistics are done correctly) will come to the wrong conclusion.<a href="evolutionary-biology.html#fnref18" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="r-programming-for-loops.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="study-questions-1.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
