<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>9.2 More on data handling: categorising and joining | R tutorials for the course BIOS1140 at the University of Oslo</title>
  <meta name="description" content="This is a set of R tutorials connected to the course BIOS1140 at the University of Oslo, and are meant as a companion to the book Evolutionary Genetics (Ravinet &amp; Sætre, 2018). They do not assume any prior R knowledge." />
  <meta name="generator" content="bookdown 0.39 and GitBook 2.6.7" />

  <meta property="og:title" content="9.2 More on data handling: categorising and joining | R tutorials for the course BIOS1140 at the University of Oslo" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="This is a set of R tutorials connected to the course BIOS1140 at the University of Oslo, and are meant as a companion to the book Evolutionary Genetics (Ravinet &amp; Sætre, 2018). They do not assume any prior R knowledge." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="9.2 More on data handling: categorising and joining | R tutorials for the course BIOS1140 at the University of Oslo" />
  
  <meta name="twitter:description" content="This is a set of R tutorials connected to the course BIOS1140 at the University of Oslo, and are meant as a companion to the book Evolutionary Genetics (Ravinet &amp; Sætre, 2018). They do not assume any prior R knowledge." />
  

<meta name="author" content="Original tutorials written by Mark Ravinet. Revamped in 2021 by Even Sletteng Garvang and Solveig Brochmann. Revamped again in 2024 by Emily Enevoldsen and Marius F. Maurstad" />


<meta name="date" content="2024-06-28" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="advanced-features-of-rstudio.html"/>
<link rel="next" href="lists.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">R tutorials for the course BIOS1140</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a>
<ul>
<li class="chapter" data-level="" data-path="how-to-use-these-tutorials.html"><a href="how-to-use-these-tutorials.html"><i class="fa fa-check"></i>How to use these tutorials</a></li>
</ul></li>
<li class="part"><span><b>I Introductory R course</b></span></li>
<li class="chapter" data-level="1" data-path="ch01.html"><a href="ch01.html"><i class="fa fa-check"></i><b>1</b> Introduction to R</a>
<ul>
<li class="chapter" data-level="" data-path="ch01.html"><a href="ch01.html#what-to-expect"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="1.1" data-path="start-using-r.html"><a href="start-using-r.html"><i class="fa fa-check"></i><b>1.1</b> Start using R!</a>
<ul>
<li class="chapter" data-level="1.1.1" data-path="start-using-r.html"><a href="start-using-r.html#install-r-and-rstudio"><i class="fa fa-check"></i><b>1.1.1</b> Install R and Rstudio</a></li>
<li class="chapter" data-level="1.1.2" data-path="start-using-r.html"><a href="start-using-r.html#getting-familiar-with-r-and-the-console"><i class="fa fa-check"></i><b>1.1.2</b> Getting familiar with R and the console</a></li>
<li class="chapter" data-level="1.1.3" data-path="start-using-r.html"><a href="start-using-r.html#scripts"><i class="fa fa-check"></i><b>1.1.3</b> Scripts</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="r-essentials.html"><a href="r-essentials.html"><i class="fa fa-check"></i><b>1.2</b> R essentials</a>
<ul>
<li class="chapter" data-level="1.2.1" data-path="r-essentials.html"><a href="r-essentials.html#assigning-to-objects"><i class="fa fa-check"></i><b>1.2.1</b> Assigning to objects</a></li>
<li class="chapter" data-level="1.2.2" data-path="r-essentials.html"><a href="r-essentials.html#vectors"><i class="fa fa-check"></i><b>1.2.2</b> Vectors</a></li>
<li class="chapter" data-level="1.2.3" data-path="r-essentials.html"><a href="r-essentials.html#strings"><i class="fa fa-check"></i><b>1.2.3</b> Strings</a></li>
<li class="chapter" data-level="1.2.4" data-path="r-essentials.html"><a href="r-essentials.html#logical"><i class="fa fa-check"></i><b>1.2.4</b> Logical values</a></li>
<li class="chapter" data-level="1.2.5" data-path="r-essentials.html"><a href="r-essentials.html#functions"><i class="fa fa-check"></i><b>1.2.5</b> Functions</a></li>
<li class="chapter" data-level="1.2.6" data-path="r-essentials.html"><a href="r-essentials.html#data-frames"><i class="fa fa-check"></i><b>1.2.6</b> Data frames</a></li>
</ul></li>
<li class="chapter" data-level="1.3" data-path="dataimport.html"><a href="dataimport.html"><i class="fa fa-check"></i><b>1.3</b> Data import</a>
<ul>
<li class="chapter" data-level="1.3.1" data-path="dataimport.html"><a href="dataimport.html#data-formats"><i class="fa fa-check"></i><b>1.3.1</b> Data formats</a></li>
<li class="chapter" data-level="1.3.2" data-path="dataimport.html"><a href="dataimport.html#working-directories"><i class="fa fa-check"></i><b>1.3.2</b> Working directories</a></li>
<li class="chapter" data-level="1.3.3" data-path="dataimport.html"><a href="dataimport.html#importing"><i class="fa fa-check"></i><b>1.3.3</b> Importing</a></li>
</ul></li>
<li class="chapter" data-level="1.4" data-path="plotting.html"><a href="plotting.html"><i class="fa fa-check"></i><b>1.4</b> Plotting</a>
<ul>
<li class="chapter" data-level="1.4.1" data-path="plotting.html"><a href="plotting.html#plotting-vectors"><i class="fa fa-check"></i><b>1.4.1</b> Plotting vectors</a></li>
<li class="chapter" data-level="1.4.2" data-path="plotting.html"><a href="plotting.html#customizing-your-plots"><i class="fa fa-check"></i><b>1.4.2</b> Customizing your plots</a></li>
</ul></li>
<li class="chapter" data-level="1.5" data-path="installing-and-loading-packages.html"><a href="installing-and-loading-packages.html"><i class="fa fa-check"></i><b>1.5</b> Installing and loading packages</a></li>
<li class="chapter" data-level="1.6" data-path="going-further.html"><a href="going-further.html"><i class="fa fa-check"></i><b>1.6</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="ch02.html"><a href="ch02.html"><i class="fa fa-check"></i><b>2</b> Building on your foundations: going further with R</a>
<ul>
<li class="chapter" data-level="" data-path="ch02.html"><a href="ch02.html#what-to-expect-1"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="2.1" data-path="intro-to-data-manipulation-with-tidyverse.html"><a href="intro-to-data-manipulation-with-tidyverse.html"><i class="fa fa-check"></i><b>2.1</b> Intro to data manipulation with tidyverse</a>
<ul>
<li class="chapter" data-level="2.1.1" data-path="intro-to-data-manipulation-with-tidyverse.html"><a href="intro-to-data-manipulation-with-tidyverse.html#what-is-the-tidyverse"><i class="fa fa-check"></i><b>2.1.1</b> What is the tidyverse?</a></li>
<li class="chapter" data-level="2.1.2" data-path="intro-to-data-manipulation-with-tidyverse.html"><a href="intro-to-data-manipulation-with-tidyverse.html#the-dplyr-package"><i class="fa fa-check"></i><b>2.1.2</b> The dplyr package</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="using-dplyr-to-work-with-your-data.html"><a href="using-dplyr-to-work-with-your-data.html"><i class="fa fa-check"></i><b>2.2</b> Using dplyr to work with your data</a>
<ul>
<li class="chapter" data-level="2.2.1" data-path="using-dplyr-to-work-with-your-data.html"><a href="using-dplyr-to-work-with-your-data.html#the-pipe"><i class="fa fa-check"></i><b>2.2.1</b> The pipe</a></li>
<li class="chapter" data-level="2.2.2" data-path="using-dplyr-to-work-with-your-data.html"><a href="using-dplyr-to-work-with-your-data.html#selecting-columns-with-select"><i class="fa fa-check"></i><b>2.2.2</b> Selecting columns with <code>select()</code></a></li>
<li class="chapter" data-level="2.2.3" data-path="using-dplyr-to-work-with-your-data.html"><a href="using-dplyr-to-work-with-your-data.html#filtering-colums-using-filter"><i class="fa fa-check"></i><b>2.2.3</b> Filtering colums using <code>filter()</code></a></li>
<li class="chapter" data-level="2.2.4" data-path="using-dplyr-to-work-with-your-data.html"><a href="using-dplyr-to-work-with-your-data.html#grouped-summaries-with-group_by-and-summarise"><i class="fa fa-check"></i><b>2.2.4</b> Grouped summaries with <code>group_by()</code> and <code>summarise()</code></a></li>
<li class="chapter" data-level="2.2.5" data-path="using-dplyr-to-work-with-your-data.html"><a href="using-dplyr-to-work-with-your-data.html#using-everything-weve-learned-in-a-single-pipe-and-a-dplyr-exercise"><i class="fa fa-check"></i><b>2.2.5</b> Using everything we’ve learned in a single pipe, and a <code>dplyr</code> exercise</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="plotting-your-data-with-ggplot2.html"><a href="plotting-your-data-with-ggplot2.html"><i class="fa fa-check"></i><b>2.3</b> Plotting your data with <code>ggplot2</code></a>
<ul>
<li class="chapter" data-level="2.3.1" data-path="plotting-your-data-with-ggplot2.html"><a href="plotting-your-data-with-ggplot2.html#the-three-things-you-need-in-a-ggplot"><i class="fa fa-check"></i><b>2.3.1</b> The three things you need in a ggplot</a></li>
<li class="chapter" data-level="2.3.2" data-path="plotting-your-data-with-ggplot2.html"><a href="plotting-your-data-with-ggplot2.html#storing-ggplots-in-objects"><i class="fa fa-check"></i><b>2.3.2</b> Storing ggplots in objects</a></li>
<li class="chapter" data-level="2.3.3" data-path="plotting-your-data-with-ggplot2.html"><a href="plotting-your-data-with-ggplot2.html#customizing-your-plots-1"><i class="fa fa-check"></i><b>2.3.3</b> Customizing your plots</a></li>
<li class="chapter" data-level="2.3.4" data-path="plotting-your-data-with-ggplot2.html"><a href="plotting-your-data-with-ggplot2.html#saving-your-plots"><i class="fa fa-check"></i><b>2.3.4</b> Saving your plots</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="pivot-longer.html"><a href="pivot-longer.html"><i class="fa fa-check"></i><b>2.4</b> Reshaping data with pivot_longer()</a>
<ul>
<li class="chapter" data-level="2.4.1" data-path="pivot-longer.html"><a href="pivot-longer.html#wide-and-long-format"><i class="fa fa-check"></i><b>2.4.1</b> Wide and long format</a></li>
<li class="chapter" data-level="2.4.2" data-path="pivot-longer.html"><a href="pivot-longer.html#import-example-data"><i class="fa fa-check"></i><b>2.4.2</b> Import example data</a></li>
<li class="chapter" data-level="2.4.3" data-path="pivot-longer.html"><a href="pivot-longer.html#reshape-the-data"><i class="fa fa-check"></i><b>2.4.3</b> Reshape the data</a></li>
<li class="chapter" data-level="2.4.4" data-path="pivot-longer.html"><a href="pivot-longer.html#plot-the-data"><i class="fa fa-check"></i><b>2.4.4</b> Plot the data</a></li>
</ul></li>
<li class="chapter" data-level="2.5" data-path="study-questions.html"><a href="study-questions.html"><i class="fa fa-check"></i><b>2.5</b> Study questions</a></li>
<li class="chapter" data-level="2.6" data-path="going-further-1.html"><a href="going-further-1.html"><i class="fa fa-check"></i><b>2.6</b> Going further</a></li>
</ul></li>
<li class="part"><span><b>II Evolutionary genetics with R</b></span></li>
<li class="chapter" data-level="3" data-path="ch03.html"><a href="ch03.html"><i class="fa fa-check"></i><b>3</b> Changes in Allele and Genotype Frequency</a>
<ul>
<li class="chapter" data-level="" data-path="ch03.html"><a href="ch03.html#what-to-expect-2"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="3.1" data-path="r-programming-for-loops.html"><a href="r-programming-for-loops.html"><i class="fa fa-check"></i><b>3.1</b> R programming: for-loops</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="r-programming-for-loops.html"><a href="r-programming-for-loops.html#motivation-why-loop"><i class="fa fa-check"></i><b>3.1.1</b> Motivation: why loop?</a></li>
<li class="chapter" data-level="3.1.2" data-path="r-programming-for-loops.html"><a href="r-programming-for-loops.html#how-a-for-loop-works"><i class="fa fa-check"></i><b>3.1.2</b> How a for-loop works</a></li>
<li class="chapter" data-level="3.1.3" data-path="r-programming-for-loops.html"><a href="r-programming-for-loops.html#indexing"><i class="fa fa-check"></i><b>3.1.3</b> Indexing with for-loops</a></li>
<li class="chapter" data-level="3.1.4" data-path="r-programming-for-loops.html"><a href="r-programming-for-loops.html#solving-our-problem"><i class="fa fa-check"></i><b>3.1.4</b> Solving our problem</a></li>
<li class="chapter" data-level="3.1.5" data-path="r-programming-for-loops.html"><a href="r-programming-for-loops.html#storing-values-from-a-for-loop"><i class="fa fa-check"></i><b>3.1.5</b> Storing values from a for-loop</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="evolutionary-biology.html"><a href="evolutionary-biology.html"><i class="fa fa-check"></i><b>3.2</b> Evolutionary biology</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="evolutionary-biology.html"><a href="evolutionary-biology.html#the-hardy-weinberg-model"><i class="fa fa-check"></i><b>3.2.1</b> The Hardy-Weinberg Model</a></li>
<li class="chapter" data-level="3.2.2" data-path="evolutionary-biology.html"><a href="evolutionary-biology.html#testing-for-deviations-from-the-hardy-weinberg-expectation"><i class="fa fa-check"></i><b>3.2.2</b> Testing for deviations from the Hardy Weinberg Expectation</a></li>
<li class="chapter" data-level="3.2.3" data-path="evolutionary-biology.html"><a href="evolutionary-biology.html#simulating-genetic-drift"><i class="fa fa-check"></i><b>3.2.3</b> Simulating genetic drift</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="study-questions-1.html"><a href="study-questions-1.html"><i class="fa fa-check"></i><b>3.3</b> Study questions</a></li>
<li class="chapter" data-level="3.4" data-path="going-further-2.html"><a href="going-further-2.html"><i class="fa fa-check"></i><b>3.4</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="ch04.html"><a href="ch04.html"><i class="fa fa-check"></i><b>4</b> The Theory of Natural Selection</a>
<ul>
<li class="chapter" data-level="" data-path="ch04.html"><a href="ch04.html#what-to-expect-3"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="" data-path="ch04.html"><a href="ch04.html#getting-started"><i class="fa fa-check"></i>Getting started</a></li>
<li class="chapter" data-level="4.1" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html"><i class="fa fa-check"></i><b>4.1</b> R programming: Making custom functions</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#motivation"><i class="fa fa-check"></i><b>4.1.1</b> Motivation</a></li>
<li class="chapter" data-level="4.1.2" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#function-basics"><i class="fa fa-check"></i><b>4.1.2</b> Function basics</a></li>
<li class="chapter" data-level="4.1.3" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#a-simple-example"><i class="fa fa-check"></i><b>4.1.3</b> A simple example</a></li>
<li class="chapter" data-level="4.1.4" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#some-important-function-properties"><i class="fa fa-check"></i><b>4.1.4</b> Some important function properties</a></li>
<li class="chapter" data-level="4.1.5" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#a-slightly-more-useful-example"><i class="fa fa-check"></i><b>4.1.5</b> A slightly more useful example</a></li>
<li class="chapter" data-level="4.1.6" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#example-calculating-genotype-frequencies"><i class="fa fa-check"></i><b>4.1.6</b> Example: calculating genotype frequencies</a></li>
<li class="chapter" data-level="4.1.7" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#creating-a-function-of-the-drift-simulation"><i class="fa fa-check"></i><b>4.1.7</b> Creating a function of the drift simulation</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="evolutionary-biology-fitness.html"><a href="evolutionary-biology-fitness.html"><i class="fa fa-check"></i><b>4.2</b> Evolutionary biology: fitness</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="evolutionary-biology-fitness.html"><a href="evolutionary-biology-fitness.html#understanding-fitness"><i class="fa fa-check"></i><b>4.2.1</b> Understanding fitness</a></li>
<li class="chapter" data-level="4.2.2" data-path="evolutionary-biology-fitness.html"><a href="evolutionary-biology-fitness.html#one-locus-model-of-viability-selection"><i class="fa fa-check"></i><b>4.2.2</b> One-locus model of viability selection</a></li>
<li class="chapter" data-level="4.2.3" data-path="evolutionary-biology-fitness.html"><a href="evolutionary-biology-fitness.html#over-and-underdominance"><i class="fa fa-check"></i><b>4.2.3</b> Over and underdominance</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="study-questions-2.html"><a href="study-questions-2.html"><i class="fa fa-check"></i><b>4.3</b> Study questions</a></li>
<li class="chapter" data-level="4.4" data-path="going-further-3.html"><a href="going-further-3.html"><i class="fa fa-check"></i><b>4.4</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="ch05.html"><a href="ch05.html"><i class="fa fa-check"></i><b>5</b> Detecting Natural Selection</a>
<ul>
<li class="chapter" data-level="" data-path="ch05.html"><a href="ch05.html#what-to-expect-4"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="" data-path="ch05.html"><a href="ch05.html#getting-started-1"><i class="fa fa-check"></i>Getting started</a></li>
<li class="chapter" data-level="5.1" data-path="vectorisation.html"><a href="vectorisation.html"><i class="fa fa-check"></i><b>5.1</b> More on functions: Vectorisation</a>
<ul>
<li class="chapter" data-level="5.1.1" data-path="vectorisation.html"><a href="vectorisation.html#the-apply-function"><i class="fa fa-check"></i><b>5.1.1</b> The <code>apply()</code> function</a></li>
<li class="chapter" data-level="5.1.2" data-path="vectorisation.html"><a href="vectorisation.html#ifelse"><i class="fa fa-check"></i><b>5.1.2</b> The <code>ifelse()</code> function</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="understanding-fst---the-fixation-index.html"><a href="understanding-fst---the-fixation-index.html"><i class="fa fa-check"></i><b>5.2</b> Understanding <em>F</em><sub>ST</sub> - the fixation index</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="understanding-fst---the-fixation-index.html"><a href="understanding-fst---the-fixation-index.html#a-worked-example-of-fst-in-humans"><i class="fa fa-check"></i><b>5.2.1</b> A worked example of <em>F</em><sub>ST</sub> in humans</a></li>
<li class="chapter" data-level="5.2.2" data-path="understanding-fst---the-fixation-index.html"><a href="understanding-fst---the-fixation-index.html#writing-a-set-of-fst-functions"><i class="fa fa-check"></i><b>5.2.2</b> Writing a set of <em>F</em><sub>ST</sub> functions</a></li>
<li class="chapter" data-level="5.2.3" data-path="understanding-fst---the-fixation-index.html"><a href="understanding-fst---the-fixation-index.html#applying-functions-to-matrices-and-data-frames"><i class="fa fa-check"></i><b>5.2.3</b> Applying functions to matrices and data frames</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="visualising-fst-along-a-chromosome.html"><a href="visualising-fst-along-a-chromosome.html"><i class="fa fa-check"></i><b>5.3</b> Visualising <em>F</em><sub>ST</sub> along a chromosome</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="visualising-fst-along-a-chromosome.html"><a href="visualising-fst-along-a-chromosome.html#identifying-outliers-in-our-fst-distribution"><i class="fa fa-check"></i><b>5.3.1</b> Identifying outliers in our <em>F</em><sub>ST</sub> distribution</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="study-questions-3.html"><a href="study-questions-3.html"><i class="fa fa-check"></i><b>5.4</b> Study questions</a></li>
<li class="chapter" data-level="5.5" data-path="going-further-4.html"><a href="going-further-4.html"><i class="fa fa-check"></i><b>5.5</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="ch07.html"><a href="ch07.html"><i class="fa fa-check"></i><b>6</b> Inferring Evolutionary Processes from Sequence Data</a>
<ul>
<li class="chapter" data-level="" data-path="ch07.html"><a href="ch07.html#what-to-expect-5"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="" data-path="ch07.html"><a href="ch07.html#getting-started-2"><i class="fa fa-check"></i>Getting started</a></li>
<li class="chapter" data-level="6.1" data-path="working-with-dna-sequence-data.html"><a href="working-with-dna-sequence-data.html"><i class="fa fa-check"></i><b>6.1</b> Working with DNA sequence data</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="working-with-dna-sequence-data.html"><a href="working-with-dna-sequence-data.html#reading-sequence-data-into-r."><i class="fa fa-check"></i><b>6.1.1</b> Reading sequence data into R.</a></li>
<li class="chapter" data-level="6.1.2" data-path="working-with-dna-sequence-data.html"><a href="working-with-dna-sequence-data.html#exploring-dna-sequence-data"><i class="fa fa-check"></i><b>6.1.2</b> Exploring DNA sequence data</a></li>
<li class="chapter" data-level="6.1.3" data-path="working-with-dna-sequence-data.html"><a href="working-with-dna-sequence-data.html#calculating-basic-sequence-statistics"><i class="fa fa-check"></i><b>6.1.3</b> Calculating basic sequence statistics</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="working-with-a-larger-dataset.html"><a href="working-with-a-larger-dataset.html"><i class="fa fa-check"></i><b>6.2</b> Working with a larger dataset</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="working-with-a-larger-dataset.html"><a href="working-with-a-larger-dataset.html#sample-size-and-sequence-statistics"><i class="fa fa-check"></i><b>6.2.1</b> Sample size and sequence statistics</a></li>
<li class="chapter" data-level="6.2.2" data-path="working-with-a-larger-dataset.html"><a href="working-with-a-larger-dataset.html#inferring-evolutionary-processes-using-tajimas-d"><i class="fa fa-check"></i><b>6.2.2</b> Inferring evolutionary processes using Tajima’s <em>D</em></a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="calculating-statistics-at-the-whole-genome-level.html"><a href="calculating-statistics-at-the-whole-genome-level.html"><i class="fa fa-check"></i><b>6.3</b> Calculating statistics at the whole genome level</a>
<ul>
<li class="chapter" data-level="6.3.1" data-path="calculating-statistics-at-the-whole-genome-level.html"><a href="calculating-statistics-at-the-whole-genome-level.html#reading-in-variant-data"><i class="fa fa-check"></i><b>6.3.1</b> Reading in variant data</a></li>
<li class="chapter" data-level="6.3.2" data-path="calculating-statistics-at-the-whole-genome-level.html"><a href="calculating-statistics-at-the-whole-genome-level.html#calculating-nucleotide-diversity-statistics"><i class="fa fa-check"></i><b>6.3.2</b> Calculating nucleotide diversity statistics</a></li>
<li class="chapter" data-level="6.3.3" data-path="calculating-statistics-at-the-whole-genome-level.html"><a href="calculating-statistics-at-the-whole-genome-level.html#visualising-nucleotide-diversity-along-the-chromosome"><i class="fa fa-check"></i><b>6.3.3</b> Visualising nucleotide diversity along the chromosome</a></li>
<li class="chapter" data-level="6.3.4" data-path="calculating-statistics-at-the-whole-genome-level.html"><a href="calculating-statistics-at-the-whole-genome-level.html#performing-a-sliding-window-analysis"><i class="fa fa-check"></i><b>6.3.4</b> Performing a sliding window analysis</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="study-questions-4.html"><a href="study-questions-4.html"><i class="fa fa-check"></i><b>6.4</b> Study questions</a></li>
<li class="chapter" data-level="6.5" data-path="going-further-5.html"><a href="going-further-5.html"><i class="fa fa-check"></i><b>6.5</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="ch08.html"><a href="ch08.html"><i class="fa fa-check"></i><b>7</b> Speciation Genomics</a>
<ul>
<li class="chapter" data-level="" data-path="ch08.html"><a href="ch08.html#what-to-expect-6"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="" data-path="ch08.html"><a href="ch08.html#getting-started-3"><i class="fa fa-check"></i>Getting started</a></li>
<li class="chapter" data-level="7.1" data-path="visualizing-complex-data.html"><a href="visualizing-complex-data.html"><i class="fa fa-check"></i><b>7.1</b> Visualizing complex data</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="visualizing-complex-data.html"><a href="visualizing-complex-data.html#faceting"><i class="fa fa-check"></i><b>7.1.1</b> Faceting</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="returning-to-the-sparrow-dataset.html"><a href="returning-to-the-sparrow-dataset.html"><i class="fa fa-check"></i><b>7.2</b> Returning to the sparrow dataset</a>
<ul>
<li class="chapter" data-level="7.2.1" data-path="returning-to-the-sparrow-dataset.html"><a href="returning-to-the-sparrow-dataset.html#reading-in-the-sparrow-vcf"><i class="fa fa-check"></i><b>7.2.1</b> Reading in the sparrow vcf</a></li>
<li class="chapter" data-level="7.2.2" data-path="returning-to-the-sparrow-dataset.html"><a href="returning-to-the-sparrow-dataset.html#examining-the-variant-data"><i class="fa fa-check"></i><b>7.2.2</b> Examining the variant data</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="setting-up-sliding-windows.html"><a href="setting-up-sliding-windows.html"><i class="fa fa-check"></i><b>7.3</b> Setting up sliding windows</a>
<ul>
<li class="chapter" data-level="7.3.1" data-path="setting-up-sliding-windows.html"><a href="setting-up-sliding-windows.html#calculating-sliding-window-estimates-of-nucleotide-diversity-and-differentiation"><i class="fa fa-check"></i><b>7.3.1</b> Calculating sliding window estimates of nucleotide diversity and differentiation</a></li>
<li class="chapter" data-level="7.3.2" data-path="setting-up-sliding-windows.html"><a href="setting-up-sliding-windows.html#sparrow-data"><i class="fa fa-check"></i><b>7.3.2</b> Extracting statistics for visualisation</a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="sparrow-viz.html"><a href="sparrow-viz.html"><i class="fa fa-check"></i><b>7.4</b> Visualising the data</a>
<ul>
<li class="chapter" data-level="7.4.1" data-path="sparrow-viz.html"><a href="sparrow-viz.html#visualising-patterns-along-the-chromosome"><i class="fa fa-check"></i><b>7.4.1</b> Visualising patterns along the chromosome</a></li>
<li class="chapter" data-level="7.4.2" data-path="sparrow-viz.html"><a href="sparrow-viz.html#interlude-relative-vs.-absolute-measures-of-nucleotide-diversity"><i class="fa fa-check"></i><b>7.4.2</b> Interlude: relative vs. absolute measures of nucleotide diversity</a></li>
<li class="chapter" data-level="7.4.3" data-path="sparrow-viz.html"><a href="sparrow-viz.html#investigating-recombination-rate-variation"><i class="fa fa-check"></i><b>7.4.3</b> Investigating recombination rate variation</a></li>
</ul></li>
<li class="chapter" data-level="7.5" data-path="study-questions-5.html"><a href="study-questions-5.html"><i class="fa fa-check"></i><b>7.5</b> Study questions</a></li>
<li class="chapter" data-level="7.6" data-path="going-further-6.html"><a href="going-further-6.html"><i class="fa fa-check"></i><b>7.6</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="ch09.html"><a href="ch09.html"><i class="fa fa-check"></i><b>8</b> Reconstructing the Past</a>
<ul>
<li class="chapter" data-level="" data-path="ch09.html"><a href="ch09.html#what-to-expect-7"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="" data-path="ch09.html"><a href="ch09.html#getting-started-4"><i class="fa fa-check"></i>Getting started</a></li>
<li class="chapter" data-level="8.1" data-path="phylogenetics-in-r.html"><a href="phylogenetics-in-r.html"><i class="fa fa-check"></i><b>8.1</b> Phylogenetics in R</a>
<ul>
<li class="chapter" data-level="8.1.1" data-path="phylogenetics-in-r.html"><a href="phylogenetics-in-r.html#storing-trees-in-r"><i class="fa fa-check"></i><b>8.1.1</b> Storing trees in R</a></li>
<li class="chapter" data-level="8.1.2" data-path="phylogenetics-in-r.html"><a href="phylogenetics-in-r.html#plotting-trees"><i class="fa fa-check"></i><b>8.1.2</b> Plotting trees</a></li>
<li class="chapter" data-level="8.1.3" data-path="phylogenetics-in-r.html"><a href="phylogenetics-in-r.html#a-simple-example-with-real-data---avian-phylogenetics"><i class="fa fa-check"></i><b>8.1.3</b> A simple example with real data - avian phylogenetics</a></li>
<li class="chapter" data-level="8.1.4" data-path="phylogenetics-in-r.html"><a href="phylogenetics-in-r.html#constructing-trees-with-r"><i class="fa fa-check"></i><b>8.1.4</b> Constructing trees with R</a></li>
</ul></li>
<li class="chapter" data-level="8.2" data-path="population-structure.html"><a href="population-structure.html"><i class="fa fa-check"></i><b>8.2</b> Population structure</a>
<ul>
<li class="chapter" data-level="8.2.1" data-path="population-structure.html"><a href="population-structure.html#village-dogs-as-an-insight-to-dog-domestication"><i class="fa fa-check"></i><b>8.2.1</b> Village dogs as an insight to dog domestication</a></li>
<li class="chapter" data-level="8.2.2" data-path="population-structure.html"><a href="population-structure.html#reading-the-data-into-r"><i class="fa fa-check"></i><b>8.2.2</b> Reading the data into R</a></li>
<li class="chapter" data-level="8.2.3" data-path="population-structure.html"><a href="population-structure.html#performing-a-pca"><i class="fa fa-check"></i><b>8.2.3</b> Performing a PCA</a></li>
<li class="chapter" data-level="8.2.4" data-path="population-structure.html"><a href="population-structure.html#visualising-the-pca"><i class="fa fa-check"></i><b>8.2.4</b> Visualising the PCA</a></li>
<li class="chapter" data-level="8.2.5" data-path="population-structure.html"><a href="population-structure.html#eigenvalues"><i class="fa fa-check"></i><b>8.2.5</b> Eigenvalues</a></li>
<li class="chapter" data-level="8.2.6" data-path="population-structure.html"><a href="population-structure.html#the-full-data-set"><i class="fa fa-check"></i><b>8.2.6</b> The full data set</a></li>
</ul></li>
<li class="chapter" data-level="8.3" data-path="study-questions-6.html"><a href="study-questions-6.html"><i class="fa fa-check"></i><b>8.3</b> Study questions</a></li>
<li class="chapter" data-level="8.4" data-path="going-further-7.html"><a href="going-further-7.html"><i class="fa fa-check"></i><b>8.4</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="ch10.html"><a href="ch10.html"><i class="fa fa-check"></i><b>9</b> Advancing Further in R</a>
<ul>
<li class="chapter" data-level="" data-path="ch10.html"><a href="ch10.html#what-to-expect-8"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="" data-path="ch10.html"><a href="ch10.html#getting-started-5"><i class="fa fa-check"></i>Getting started</a></li>
<li class="chapter" data-level="9.1" data-path="advanced-features-of-rstudio.html"><a href="advanced-features-of-rstudio.html"><i class="fa fa-check"></i><b>9.1</b> Advanced features of RStudio</a>
<ul>
<li class="chapter" data-level="9.1.1" data-path="advanced-features-of-rstudio.html"><a href="advanced-features-of-rstudio.html#projects-projects-projects"><i class="fa fa-check"></i><b>9.1.1</b> Projects, projects, projects</a></li>
<li class="chapter" data-level="9.1.2" data-path="advanced-features-of-rstudio.html"><a href="advanced-features-of-rstudio.html#everyone-has-a-history"><i class="fa fa-check"></i><b>9.1.2</b> Everyone has a history</a></li>
<li class="chapter" data-level="9.1.3" data-path="advanced-features-of-rstudio.html"><a href="advanced-features-of-rstudio.html#tab-complete-and-other-hotkeys"><i class="fa fa-check"></i><b>9.1.3</b> Tab complete and other hotkeys</a></li>
</ul></li>
<li class="chapter" data-level="9.2" data-path="more-on-data-handling-categorising-and-joining.html"><a href="more-on-data-handling-categorising-and-joining.html"><i class="fa fa-check"></i><b>9.2</b> More on data handling: categorising and joining</a>
<ul>
<li class="chapter" data-level="9.2.1" data-path="more-on-data-handling-categorising-and-joining.html"><a href="more-on-data-handling-categorising-and-joining.html#ifelse-for-making-categories"><i class="fa fa-check"></i><b>9.2.1</b> <code>ifelse()</code> for making categories</a></li>
<li class="chapter" data-level="9.2.2" data-path="more-on-data-handling-categorising-and-joining.html"><a href="more-on-data-handling-categorising-and-joining.html#joining"><i class="fa fa-check"></i><b>9.2.2</b> Joining</a></li>
</ul></li>
<li class="chapter" data-level="9.3" data-path="lists.html"><a href="lists.html"><i class="fa fa-check"></i><b>9.3</b> Lists</a></li>
<li class="chapter" data-level="9.4" data-path="vectorisation2.html"><a href="vectorisation2.html"><i class="fa fa-check"></i><b>9.4</b> Vectorisation</a>
<ul>
<li class="chapter" data-level="9.4.1" data-path="vectorisation2.html"><a href="vectorisation2.html#using-lapply-to-vectorise"><i class="fa fa-check"></i><b>9.4.1</b> Using <code>lapply()</code> to vectorise</a></li>
<li class="chapter" data-level="9.4.2" data-path="vectorisation2.html"><a href="vectorisation2.html#sapply"><i class="fa fa-check"></i><b>9.4.2</b> <code>sapply()</code></a></li>
<li class="chapter" data-level="9.4.3" data-path="vectorisation2.html"><a href="vectorisation2.html#anonymous-functions"><i class="fa fa-check"></i><b>9.4.3</b> Anonymous functions</a></li>
</ul></li>
<li class="chapter" data-level="9.5" data-path="concluding-remarks.html"><a href="concluding-remarks.html"><i class="fa fa-check"></i><b>9.5</b> Concluding remarks</a></li>
<li class="chapter" data-level="9.6" data-path="study-questions-7.html"><a href="study-questions-7.html"><i class="fa fa-check"></i><b>9.6</b> Study questions</a></li>
<li class="chapter" data-level="9.7" data-path="going-further-8.html"><a href="going-further-8.html"><i class="fa fa-check"></i><b>9.7</b> Going further</a></li>
</ul></li>
<li class="part"><span><b>III Assignments</b></span></li>
<li class="chapter" data-level="" data-path="w01.html"><a href="w01.html"><i class="fa fa-check"></i>Week 1-2 assignment 1</a></li>
<li class="chapter" data-level="" data-path="w03.html"><a href="w03.html"><i class="fa fa-check"></i>Week 3 Assignment 2</a></li>
<li class="chapter" data-level="" data-path="w04.html"><a href="w04.html"><i class="fa fa-check"></i>Week 4 assignment 3</a></li>
<li class="chapter" data-level="" data-path="w05.html"><a href="w05.html"><i class="fa fa-check"></i>Week 5 assignment 4</a></li>
<li class="chapter" data-level="" data-path="w07.html"><a href="w07.html"><i class="fa fa-check"></i>Week 6-7 assignment 5</a></li>
<li class="chapter" data-level="" data-path="w09.html"><a href="w09.html"><i class="fa fa-check"></i>Week 8 assignment 6</a></li>
<li class="chapter" data-level="" data-path="w10.html"><a href="w10.html"><i class="fa fa-check"></i>Week 9 assignment 7</a></li>
<li class="appendix"><span><b>Appendix</b></span></li>
<li class="chapter" data-level="A" data-path="rmarkdown.html"><a href="rmarkdown.html"><i class="fa fa-check"></i><b>A</b> About assignments and RMarkdown</a>
<ul>
<li class="chapter" data-level="" data-path="rmarkdown.html"><a href="rmarkdown.html#r-markdown"><i class="fa fa-check"></i>R Markdown</a></li>
<li class="chapter" data-level="" data-path="rmarkdown.html"><a href="rmarkdown.html#getting-started-6"><i class="fa fa-check"></i>Getting started</a></li>
</ul></li>
<li class="chapter" data-level="B" data-path="tech.html"><a href="tech.html"><i class="fa fa-check"></i><b>B</b> Technical information</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Made with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">R tutorials for the course BIOS1140 at the University of Oslo</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="more-on-data-handling-categorising-and-joining" class="section level2 hasAnchor" number="9.2">
<h2><span class="header-section-number">9.2</span> More on data handling: categorising and joining<a href="more-on-data-handling-categorising-and-joining.html#more-on-data-handling-categorising-and-joining" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<script src="js/hideOutput.js"></script>
<p>Handling large amounts of data with few lines of code is one of R’s strong points. In this section we will show how you can use <code>ifelse()</code> and the <code>left_join()</code> function from dplyr to make categories and add information to your data</p>
<div id="ifelse-for-making-categories" class="section level3 hasAnchor" number="9.2.1">
<h3><span class="header-section-number">9.2.1</span> <code>ifelse()</code> for making categories<a href="more-on-data-handling-categorising-and-joining.html#ifelse-for-making-categories" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Let’s first review what the <code>ifelse()</code> function does. As you learned back in <a href="vectorisation.html#ifelse">chapter 5</a>, <code>ifelse()</code> takes a logical statement, and returns something different depending on whether the condition is <code>TRUE</code> or <code>FALSE</code>:</p>
<div class="sourceCode" id="cb405"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb405-1"><a href="more-on-data-handling-categorising-and-joining.html#cb405-1" tabindex="-1"></a><span class="co"># make a small vector</span></span>
<span id="cb405-2"><a href="more-on-data-handling-categorising-and-joining.html#cb405-2" tabindex="-1"></a>y <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">20</span>, <span class="dv">30</span>, <span class="dv">50</span>)</span>
<span id="cb405-3"><a href="more-on-data-handling-categorising-and-joining.html#cb405-3" tabindex="-1"></a><span class="co"># use ifelse to evaluate it</span></span>
<span id="cb405-4"><a href="more-on-data-handling-categorising-and-joining.html#cb405-4" tabindex="-1"></a><span class="fu">ifelse</span>(y <span class="sc">&gt;</span> <span class="dv">25</span>, <span class="st">&quot;Greater than 25&quot;</span>, <span class="st">&quot;Not greater than 25&quot;</span>)</span></code></pre></div>
<p>This function can be extremely useful for creating new variables in datasets. Let’s return to the familiar <code>starwars</code> data from <code>dplyr</code> in order to use the function in this way.</p>
<div class="sourceCode" id="cb406"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb406-1"><a href="more-on-data-handling-categorising-and-joining.html#cb406-1" tabindex="-1"></a>starwars</span>
<span id="cb406-2"><a href="more-on-data-handling-categorising-and-joining.html#cb406-2" tabindex="-1"></a><span class="co">#&gt; # A tibble: 87 × 14</span></span>
<span id="cb406-3"><a href="more-on-data-handling-categorising-and-joining.html#cb406-3" tabindex="-1"></a><span class="co">#&gt;    name     height  mass hair_color skin_color eye_color birth_year sex   gender</span></span>
<span id="cb406-4"><a href="more-on-data-handling-categorising-and-joining.html#cb406-4" tabindex="-1"></a><span class="co">#&gt;    &lt;chr&gt;     &lt;int&gt; &lt;dbl&gt; &lt;chr&gt;      &lt;chr&gt;      &lt;chr&gt;          &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt; </span></span>
<span id="cb406-5"><a href="more-on-data-handling-categorising-and-joining.html#cb406-5" tabindex="-1"></a><span class="co">#&gt;  1 Luke Sk…    172    77 blond      fair       blue            19   male  mascu…</span></span>
<span id="cb406-6"><a href="more-on-data-handling-categorising-and-joining.html#cb406-6" tabindex="-1"></a><span class="co">#&gt;  2 C-3PO       167    75 &lt;NA&gt;       gold       yellow         112   none  mascu…</span></span>
<span id="cb406-7"><a href="more-on-data-handling-categorising-and-joining.html#cb406-7" tabindex="-1"></a><span class="co">#&gt;  3 R2-D2        96    32 &lt;NA&gt;       white, bl… red             33   none  mascu…</span></span>
<span id="cb406-8"><a href="more-on-data-handling-categorising-and-joining.html#cb406-8" tabindex="-1"></a><span class="co">#&gt;  4 Darth V…    202   136 none       white      yellow          41.9 male  mascu…</span></span>
<span id="cb406-9"><a href="more-on-data-handling-categorising-and-joining.html#cb406-9" tabindex="-1"></a><span class="co">#&gt;  5 Leia Or…    150    49 brown      light      brown           19   fema… femin…</span></span>
<span id="cb406-10"><a href="more-on-data-handling-categorising-and-joining.html#cb406-10" tabindex="-1"></a><span class="co">#&gt;  6 Owen La…    178   120 brown, gr… light      blue            52   male  mascu…</span></span>
<span id="cb406-11"><a href="more-on-data-handling-categorising-and-joining.html#cb406-11" tabindex="-1"></a><span class="co">#&gt;  7 Beru Wh…    165    75 brown      light      blue            47   fema… femin…</span></span>
<span id="cb406-12"><a href="more-on-data-handling-categorising-and-joining.html#cb406-12" tabindex="-1"></a><span class="co">#&gt;  8 R5-D4        97    32 &lt;NA&gt;       white, red red             NA   none  mascu…</span></span>
<span id="cb406-13"><a href="more-on-data-handling-categorising-and-joining.html#cb406-13" tabindex="-1"></a><span class="co">#&gt;  9 Biggs D…    183    84 black      light      brown           24   male  mascu…</span></span>
<span id="cb406-14"><a href="more-on-data-handling-categorising-and-joining.html#cb406-14" tabindex="-1"></a><span class="co">#&gt; 10 Obi-Wan…    182    77 auburn, w… fair       blue-gray       57   male  mascu…</span></span>
<span id="cb406-15"><a href="more-on-data-handling-categorising-and-joining.html#cb406-15" tabindex="-1"></a><span class="co">#&gt; # ℹ 77 more rows</span></span>
<span id="cb406-16"><a href="more-on-data-handling-categorising-and-joining.html#cb406-16" tabindex="-1"></a><span class="co">#&gt; # ℹ 5 more variables: homeworld &lt;chr&gt;, species &lt;chr&gt;, films &lt;list&gt;,</span></span>
<span id="cb406-17"><a href="more-on-data-handling-categorising-and-joining.html#cb406-17" tabindex="-1"></a><span class="co">#&gt; #   vehicles &lt;list&gt;, starships &lt;list&gt;</span></span></code></pre></div>
<p>Now we can take a look at the <code>starwars$species</code> vector. There are a lot of different species, so what if we wanted to create a vector that simply states whether an individual is a droid or not?</p>
<div class="sourceCode" id="cb407"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb407-1"><a href="more-on-data-handling-categorising-and-joining.html#cb407-1" tabindex="-1"></a><span class="co"># if else to identify droids and non-droids</span></span>
<span id="cb407-2"><a href="more-on-data-handling-categorising-and-joining.html#cb407-2" tabindex="-1"></a><span class="fu">ifelse</span>(starwars<span class="sc">$</span>species <span class="sc">==</span> <span class="st">&quot;Droid&quot;</span>, <span class="st">&quot;Droid&quot;</span>, <span class="st">&quot;Non-Droid&quot;</span>)</span></code></pre></div>
<p>This can be useful e.g. for comparing droids and non-droids in a plot or table. Say we want to label our species based on whether they are a droid, human or neither of the two. A useful thing with <code>ifelse()</code> is that the third argument can be <em>another</em> <code>ifelse()</code> function! So we can actually chain <code>ifelse()</code> commands like this:</p>
<div class="sourceCode" id="cb408"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb408-1"><a href="more-on-data-handling-categorising-and-joining.html#cb408-1" tabindex="-1"></a><span class="fu">ifelse</span>(starwars<span class="sc">$</span>species <span class="sc">==</span> <span class="st">&quot;Droid&quot;</span>, <span class="st">&quot;Droid&quot;</span>, <span class="fu">ifelse</span>(starwars<span class="sc">$</span>species <span class="sc">==</span> <span class="st">&quot;Human&quot;</span>, <span class="st">&quot;Human&quot;</span>, <span class="st">&quot;Neither human nor droid&quot;</span>))</span></code></pre></div>
<p>This is useful, but quickly becomes convoluted. Imagine how the code would look if we threw in a third and fourth category there! In cases like this, remember to use linebreaks to make the code more readable. You can have a linebreak anywhere after starting a function, and R will still understand that it’s part of the same function. A suggestion for better formatting than above:</p>
<div class="sourceCode" id="cb409"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb409-1"><a href="more-on-data-handling-categorising-and-joining.html#cb409-1" tabindex="-1"></a><span class="fu">ifelse</span>(</span>
<span id="cb409-2"><a href="more-on-data-handling-categorising-and-joining.html#cb409-2" tabindex="-1"></a>  starwars<span class="sc">$</span>species <span class="sc">==</span> <span class="st">&quot;Droid&quot;</span>, <span class="st">&quot;Droid&quot;</span>, </span>
<span id="cb409-3"><a href="more-on-data-handling-categorising-and-joining.html#cb409-3" tabindex="-1"></a>  <span class="fu">ifelse</span>(</span>
<span id="cb409-4"><a href="more-on-data-handling-categorising-and-joining.html#cb409-4" tabindex="-1"></a>    starwars<span class="sc">$</span>species <span class="sc">==</span> <span class="st">&quot;Human&quot;</span>, <span class="st">&quot;Human&quot;</span>, </span>
<span id="cb409-5"><a href="more-on-data-handling-categorising-and-joining.html#cb409-5" tabindex="-1"></a>    <span class="st">&quot;Neither human nor droid&quot;</span></span>
<span id="cb409-6"><a href="more-on-data-handling-categorising-and-joining.html#cb409-6" tabindex="-1"></a>    )</span>
<span id="cb409-7"><a href="more-on-data-handling-categorising-and-joining.html#cb409-7" tabindex="-1"></a>  )</span></code></pre></div>
<p>Still, if you have more than, say, four-five categories, this becomes difficult to read and time-consuming. For e.g. adding more information to a data frame, <em>joining</em> may be a better alternative, which we will go through next.</p>
</div>
<div id="joining" class="section level3 hasAnchor" number="9.2.2">
<h3><span class="header-section-number">9.2.2</span> Joining<a href="more-on-data-handling-categorising-and-joining.html#joining" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>For this section we will revisit the <a href="https://BIOS1140.github.io/data/copepods.txt">copepods.txt</a> data that we encountered way back in <a href="#w02">week 2</a>. Start by reading in this data. You should know enough by now to do this by yourself, so we won’t show you how.</p>
<div class="sourceCode" id="cb410"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb410-1"><a href="more-on-data-handling-categorising-and-joining.html#cb410-1" tabindex="-1"></a>copepods</span>
<span id="cb410-2"><a href="more-on-data-handling-categorising-and-joining.html#cb410-2" tabindex="-1"></a><span class="co">#&gt;   depth acartia calanus harpacticoida oithona oncaea temora</span></span>
<span id="cb410-3"><a href="more-on-data-handling-categorising-and-joining.html#cb410-3" tabindex="-1"></a><span class="co">#&gt; 1     0       0       3             0       2      0      0</span></span>
<span id="cb410-4"><a href="more-on-data-handling-categorising-and-joining.html#cb410-4" tabindex="-1"></a><span class="co">#&gt; 2     2       1       0             0       6      1      0</span></span>
<span id="cb410-5"><a href="more-on-data-handling-categorising-and-joining.html#cb410-5" tabindex="-1"></a><span class="co">#&gt; 3     4       1       0             0       7      0      1</span></span>
<span id="cb410-6"><a href="more-on-data-handling-categorising-and-joining.html#cb410-6" tabindex="-1"></a><span class="co">#&gt; 4     6      27       0             1       0      0      2</span></span>
<span id="cb410-7"><a href="more-on-data-handling-categorising-and-joining.html#cb410-7" tabindex="-1"></a><span class="co">#&gt; 5     8      11       0             2       6      0      3</span></span>
<span id="cb410-8"><a href="more-on-data-handling-categorising-and-joining.html#cb410-8" tabindex="-1"></a><span class="co">#&gt; 6    10      17       0             3       0      0      2</span></span>
<span id="cb410-9"><a href="more-on-data-handling-categorising-and-joining.html#cb410-9" tabindex="-1"></a><span class="co">#&gt; 7    12      13       0             1       0      0      1</span></span>
<span id="cb410-10"><a href="more-on-data-handling-categorising-and-joining.html#cb410-10" tabindex="-1"></a><span class="co">#&gt; 8    14       7       0            13       0      0      0</span></span>
<span id="cb410-11"><a href="more-on-data-handling-categorising-and-joining.html#cb410-11" tabindex="-1"></a><span class="co">#&gt; 9    16       6       0             6       0      0      1</span></span></code></pre></div>
<p>Next, we use <code>pivot_longer()</code> to get all taxa in a single column, i.e., convert to long format. See if you manage to do this yourself before looking at my code below.</p>
<div class="fold c">
<div class="sourceCode" id="cb411"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb411-1"><a href="more-on-data-handling-categorising-and-joining.html#cb411-1" tabindex="-1"></a>copepods_long <span class="ot">&lt;-</span> <span class="fu">pivot_longer</span>(copepods, </span>
<span id="cb411-2"><a href="more-on-data-handling-categorising-and-joining.html#cb411-2" tabindex="-1"></a>                              <span class="sc">-</span>depth, </span>
<span id="cb411-3"><a href="more-on-data-handling-categorising-and-joining.html#cb411-3" tabindex="-1"></a>                              <span class="at">names_to =</span> <span class="st">&quot;taxon&quot;</span>, </span>
<span id="cb411-4"><a href="more-on-data-handling-categorising-and-joining.html#cb411-4" tabindex="-1"></a>                              <span class="at">values_to =</span> <span class="st">&quot;count&quot;</span>)</span></code></pre></div>
</div>
<div class="sourceCode" id="cb412"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb412-1"><a href="more-on-data-handling-categorising-and-joining.html#cb412-1" tabindex="-1"></a>copepods_long</span>
<span id="cb412-2"><a href="more-on-data-handling-categorising-and-joining.html#cb412-2" tabindex="-1"></a><span class="co">#&gt; # A tibble: 54 × 3</span></span>
<span id="cb412-3"><a href="more-on-data-handling-categorising-and-joining.html#cb412-3" tabindex="-1"></a><span class="co">#&gt;    depth taxon         count</span></span>
<span id="cb412-4"><a href="more-on-data-handling-categorising-and-joining.html#cb412-4" tabindex="-1"></a><span class="co">#&gt;    &lt;int&gt; &lt;chr&gt;         &lt;int&gt;</span></span>
<span id="cb412-5"><a href="more-on-data-handling-categorising-and-joining.html#cb412-5" tabindex="-1"></a><span class="co">#&gt;  1     0 acartia           0</span></span>
<span id="cb412-6"><a href="more-on-data-handling-categorising-and-joining.html#cb412-6" tabindex="-1"></a><span class="co">#&gt;  2     0 calanus           3</span></span>
<span id="cb412-7"><a href="more-on-data-handling-categorising-and-joining.html#cb412-7" tabindex="-1"></a><span class="co">#&gt;  3     0 harpacticoida     0</span></span>
<span id="cb412-8"><a href="more-on-data-handling-categorising-and-joining.html#cb412-8" tabindex="-1"></a><span class="co">#&gt;  4     0 oithona           2</span></span>
<span id="cb412-9"><a href="more-on-data-handling-categorising-and-joining.html#cb412-9" tabindex="-1"></a><span class="co">#&gt;  5     0 oncaea            0</span></span>
<span id="cb412-10"><a href="more-on-data-handling-categorising-and-joining.html#cb412-10" tabindex="-1"></a><span class="co">#&gt;  6     0 temora            0</span></span>
<span id="cb412-11"><a href="more-on-data-handling-categorising-and-joining.html#cb412-11" tabindex="-1"></a><span class="co">#&gt;  7     2 acartia           1</span></span>
<span id="cb412-12"><a href="more-on-data-handling-categorising-and-joining.html#cb412-12" tabindex="-1"></a><span class="co">#&gt;  8     2 calanus           0</span></span>
<span id="cb412-13"><a href="more-on-data-handling-categorising-and-joining.html#cb412-13" tabindex="-1"></a><span class="co">#&gt;  9     2 harpacticoida     0</span></span>
<span id="cb412-14"><a href="more-on-data-handling-categorising-and-joining.html#cb412-14" tabindex="-1"></a><span class="co">#&gt; 10     2 oithona           6</span></span>
<span id="cb412-15"><a href="more-on-data-handling-categorising-and-joining.html#cb412-15" tabindex="-1"></a><span class="co">#&gt; # ℹ 44 more rows</span></span></code></pre></div>
<p>Now, say that you have recorded the temperature at each depth, and want to add that information to your copepod data. How would you go about doing that? First, here is the data in a data frame:</p>
<div class="sourceCode" id="cb413"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb413-1"><a href="more-on-data-handling-categorising-and-joining.html#cb413-1" tabindex="-1"></a>temps <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb413-2"><a href="more-on-data-handling-categorising-and-joining.html#cb413-2" tabindex="-1"></a>  <span class="at">depth =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">2</span>,<span class="dv">4</span>,<span class="dv">6</span>,<span class="dv">8</span>,<span class="dv">10</span>,<span class="dv">12</span>,<span class="dv">14</span>,<span class="dv">16</span>),</span>
<span id="cb413-3"><a href="more-on-data-handling-categorising-and-joining.html#cb413-3" tabindex="-1"></a>  <span class="at">temp =</span> <span class="fu">c</span>(<span class="fl">15.5</span>, <span class="fl">15.4</span>, <span class="fl">15.2</span>, <span class="fl">14.7</span>, <span class="fl">11.4</span>, <span class="fl">8.3</span>, <span class="fl">7.6</span>, <span class="fl">7.0</span>, <span class="fl">6.8</span>)</span>
<span id="cb413-4"><a href="more-on-data-handling-categorising-and-joining.html#cb413-4" tabindex="-1"></a>)</span>
<span id="cb413-5"><a href="more-on-data-handling-categorising-and-joining.html#cb413-5" tabindex="-1"></a>temps</span>
<span id="cb413-6"><a href="more-on-data-handling-categorising-and-joining.html#cb413-6" tabindex="-1"></a><span class="co">#&gt;   depth temp</span></span>
<span id="cb413-7"><a href="more-on-data-handling-categorising-and-joining.html#cb413-7" tabindex="-1"></a><span class="co">#&gt; 1     0 15.5</span></span>
<span id="cb413-8"><a href="more-on-data-handling-categorising-and-joining.html#cb413-8" tabindex="-1"></a><span class="co">#&gt; 2     2 15.4</span></span>
<span id="cb413-9"><a href="more-on-data-handling-categorising-and-joining.html#cb413-9" tabindex="-1"></a><span class="co">#&gt; 3     4 15.2</span></span>
<span id="cb413-10"><a href="more-on-data-handling-categorising-and-joining.html#cb413-10" tabindex="-1"></a><span class="co">#&gt; 4     6 14.7</span></span>
<span id="cb413-11"><a href="more-on-data-handling-categorising-and-joining.html#cb413-11" tabindex="-1"></a><span class="co">#&gt; 5     8 11.4</span></span>
<span id="cb413-12"><a href="more-on-data-handling-categorising-and-joining.html#cb413-12" tabindex="-1"></a><span class="co">#&gt; 6    10  8.3</span></span>
<span id="cb413-13"><a href="more-on-data-handling-categorising-and-joining.html#cb413-13" tabindex="-1"></a><span class="co">#&gt; 7    12  7.6</span></span>
<span id="cb413-14"><a href="more-on-data-handling-categorising-and-joining.html#cb413-14" tabindex="-1"></a><span class="co">#&gt; 8    14  7.0</span></span>
<span id="cb413-15"><a href="more-on-data-handling-categorising-and-joining.html#cb413-15" tabindex="-1"></a><span class="co">#&gt; 9    16  6.8</span></span></code></pre></div>
<p>One way would be using nested <code>ifelse()</code> functions, like we learned in the previous section. This is a lot of work and doesn’t look good, but it’s written it out below just to show you it’s possible:</p>
<div class="fold c">
<div class="sourceCode" id="cb414"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb414-1"><a href="more-on-data-handling-categorising-and-joining.html#cb414-1" tabindex="-1"></a>copepods_long<span class="sc">$</span>depthtemp <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(</span>
<span id="cb414-2"><a href="more-on-data-handling-categorising-and-joining.html#cb414-2" tabindex="-1"></a>  copepods_long<span class="sc">$</span>depth <span class="sc">==</span> <span class="dv">0</span>, <span class="fl">15.5</span>,</span>
<span id="cb414-3"><a href="more-on-data-handling-categorising-and-joining.html#cb414-3" tabindex="-1"></a>  <span class="fu">ifelse</span>(</span>
<span id="cb414-4"><a href="more-on-data-handling-categorising-and-joining.html#cb414-4" tabindex="-1"></a>    copepods_long<span class="sc">$</span>depth <span class="sc">==</span> <span class="dv">2</span>, <span class="fl">15.4</span>,</span>
<span id="cb414-5"><a href="more-on-data-handling-categorising-and-joining.html#cb414-5" tabindex="-1"></a>    <span class="fu">ifelse</span>(</span>
<span id="cb414-6"><a href="more-on-data-handling-categorising-and-joining.html#cb414-6" tabindex="-1"></a>      copepods_long<span class="sc">$</span>depth <span class="sc">==</span> <span class="dv">4</span>, <span class="fl">15.2</span>,</span>
<span id="cb414-7"><a href="more-on-data-handling-categorising-and-joining.html#cb414-7" tabindex="-1"></a>      <span class="fu">ifelse</span>(</span>
<span id="cb414-8"><a href="more-on-data-handling-categorising-and-joining.html#cb414-8" tabindex="-1"></a>        copepods_long<span class="sc">$</span>depth <span class="sc">==</span> <span class="dv">6</span>, <span class="fl">14.7</span>,</span>
<span id="cb414-9"><a href="more-on-data-handling-categorising-and-joining.html#cb414-9" tabindex="-1"></a>        <span class="fu">ifelse</span>(</span>
<span id="cb414-10"><a href="more-on-data-handling-categorising-and-joining.html#cb414-10" tabindex="-1"></a>          copepods_long<span class="sc">$</span>depth <span class="sc">==</span> <span class="dv">8</span>, <span class="fl">11.4</span>,</span>
<span id="cb414-11"><a href="more-on-data-handling-categorising-and-joining.html#cb414-11" tabindex="-1"></a>          <span class="fu">ifelse</span>(</span>
<span id="cb414-12"><a href="more-on-data-handling-categorising-and-joining.html#cb414-12" tabindex="-1"></a>            copepods_long<span class="sc">$</span>depth <span class="sc">==</span> <span class="dv">10</span>, <span class="fl">8.3</span>,</span>
<span id="cb414-13"><a href="more-on-data-handling-categorising-and-joining.html#cb414-13" tabindex="-1"></a>            <span class="fu">ifelse</span>(</span>
<span id="cb414-14"><a href="more-on-data-handling-categorising-and-joining.html#cb414-14" tabindex="-1"></a>              copepods_long<span class="sc">$</span>depth <span class="sc">==</span> <span class="dv">12</span>, <span class="fl">7.6</span>,</span>
<span id="cb414-15"><a href="more-on-data-handling-categorising-and-joining.html#cb414-15" tabindex="-1"></a>              <span class="fu">ifelse</span>(</span>
<span id="cb414-16"><a href="more-on-data-handling-categorising-and-joining.html#cb414-16" tabindex="-1"></a>                copepods_long<span class="sc">$</span>depth <span class="sc">==</span> <span class="dv">14</span>, <span class="fl">7.0</span>,</span>
<span id="cb414-17"><a href="more-on-data-handling-categorising-and-joining.html#cb414-17" tabindex="-1"></a>                <span class="fu">ifelse</span>(</span>
<span id="cb414-18"><a href="more-on-data-handling-categorising-and-joining.html#cb414-18" tabindex="-1"></a>                  copepods_long<span class="sc">$</span>depth <span class="sc">==</span> <span class="dv">16</span>, <span class="fl">6.8</span>,</span>
<span id="cb414-19"><a href="more-on-data-handling-categorising-and-joining.html#cb414-19" tabindex="-1"></a>                  <span class="cn">NA</span></span>
<span id="cb414-20"><a href="more-on-data-handling-categorising-and-joining.html#cb414-20" tabindex="-1"></a>                )</span>
<span id="cb414-21"><a href="more-on-data-handling-categorising-and-joining.html#cb414-21" tabindex="-1"></a>              )</span>
<span id="cb414-22"><a href="more-on-data-handling-categorising-and-joining.html#cb414-22" tabindex="-1"></a>            )</span>
<span id="cb414-23"><a href="more-on-data-handling-categorising-and-joining.html#cb414-23" tabindex="-1"></a>          )</span>
<span id="cb414-24"><a href="more-on-data-handling-categorising-and-joining.html#cb414-24" tabindex="-1"></a>        )</span>
<span id="cb414-25"><a href="more-on-data-handling-categorising-and-joining.html#cb414-25" tabindex="-1"></a>      )</span>
<span id="cb414-26"><a href="more-on-data-handling-categorising-and-joining.html#cb414-26" tabindex="-1"></a>    )</span>
<span id="cb414-27"><a href="more-on-data-handling-categorising-and-joining.html#cb414-27" tabindex="-1"></a>  )</span>
<span id="cb414-28"><a href="more-on-data-handling-categorising-and-joining.html#cb414-28" tabindex="-1"></a>)</span></code></pre></div>
</div>
<p>Instead, you can use the <code>left_join()</code> function from <code>dplyr</code>. You have to supply it the original data, the data you want to join with, and a vector of column names to join by (here “depth”).</p>
<div class="sourceCode" id="cb415"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb415-1"><a href="more-on-data-handling-categorising-and-joining.html#cb415-1" tabindex="-1"></a>copepods_temp <span class="ot">&lt;-</span> <span class="fu">left_join</span>(copepods_long, temps, <span class="at">by =</span> <span class="st">&quot;depth&quot;</span>)</span>
<span id="cb415-2"><a href="more-on-data-handling-categorising-and-joining.html#cb415-2" tabindex="-1"></a>copepods_temp</span>
<span id="cb415-3"><a href="more-on-data-handling-categorising-and-joining.html#cb415-3" tabindex="-1"></a><span class="co">#&gt; # A tibble: 54 × 5</span></span>
<span id="cb415-4"><a href="more-on-data-handling-categorising-and-joining.html#cb415-4" tabindex="-1"></a><span class="co">#&gt;    depth taxon         count depthtemp  temp</span></span>
<span id="cb415-5"><a href="more-on-data-handling-categorising-and-joining.html#cb415-5" tabindex="-1"></a><span class="co">#&gt;    &lt;dbl&gt; &lt;chr&gt;         &lt;int&gt;     &lt;dbl&gt; &lt;dbl&gt;</span></span>
<span id="cb415-6"><a href="more-on-data-handling-categorising-and-joining.html#cb415-6" tabindex="-1"></a><span class="co">#&gt;  1     0 acartia           0      15.5  15.5</span></span>
<span id="cb415-7"><a href="more-on-data-handling-categorising-and-joining.html#cb415-7" tabindex="-1"></a><span class="co">#&gt;  2     0 calanus           3      15.5  15.5</span></span>
<span id="cb415-8"><a href="more-on-data-handling-categorising-and-joining.html#cb415-8" tabindex="-1"></a><span class="co">#&gt;  3     0 harpacticoida     0      15.5  15.5</span></span>
<span id="cb415-9"><a href="more-on-data-handling-categorising-and-joining.html#cb415-9" tabindex="-1"></a><span class="co">#&gt;  4     0 oithona           2      15.5  15.5</span></span>
<span id="cb415-10"><a href="more-on-data-handling-categorising-and-joining.html#cb415-10" tabindex="-1"></a><span class="co">#&gt;  5     0 oncaea            0      15.5  15.5</span></span>
<span id="cb415-11"><a href="more-on-data-handling-categorising-and-joining.html#cb415-11" tabindex="-1"></a><span class="co">#&gt;  6     0 temora            0      15.5  15.5</span></span>
<span id="cb415-12"><a href="more-on-data-handling-categorising-and-joining.html#cb415-12" tabindex="-1"></a><span class="co">#&gt;  7     2 acartia           1      15.4  15.4</span></span>
<span id="cb415-13"><a href="more-on-data-handling-categorising-and-joining.html#cb415-13" tabindex="-1"></a><span class="co">#&gt;  8     2 calanus           0      15.4  15.4</span></span>
<span id="cb415-14"><a href="more-on-data-handling-categorising-and-joining.html#cb415-14" tabindex="-1"></a><span class="co">#&gt;  9     2 harpacticoida     0      15.4  15.4</span></span>
<span id="cb415-15"><a href="more-on-data-handling-categorising-and-joining.html#cb415-15" tabindex="-1"></a><span class="co">#&gt; 10     2 oithona           6      15.4  15.4</span></span>
<span id="cb415-16"><a href="more-on-data-handling-categorising-and-joining.html#cb415-16" tabindex="-1"></a><span class="co">#&gt; # ℹ 44 more rows</span></span></code></pre></div>
<p><code>left_join()</code> matches one or more columns in your two data sets, and add rows from the second data set into the first data set in the correct place. You see that the <code>temp</code> column is equal to the <code>depthtemp</code> we created earlier, but it’s so much easier to work with! Keep in mind that it is this simple in our case because <code>depth</code> has the exact same name in both data frames. Remember this when recording data in the future!</p>
<div class="green">
<p><strong>Important concept:</strong><br />
If you have your data spread out over multiple files, remember to name columns appropriately. All columns that contain the same kind of data should have the <em>exact same name</em> across all data sets. Similarly, the data should be entered in the same way in both data sets (e.g., don’t record depth as “2” in one data set and “2m” in the other). If you do this, you can easily join data sets with the <code>left_join()</code> function.</p>
</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="advanced-features-of-rstudio.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="lists.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
