<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>9.1 Phylogenetics in R | R tutorials for the course BIOS1140 at the University of Oslo</title>
  <meta name="description" content="This is a set of R tutorials connected to the course BIOS1140 at the University of Oslo, and are meant as a companion to the book Evolutionary Genetics (Ravinet &amp; Sætre, 2018). They do not assume any prior R knowledge." />
  <meta name="generator" content="bookdown 0.25 and GitBook 2.6.7" />

  <meta property="og:title" content="9.1 Phylogenetics in R | R tutorials for the course BIOS1140 at the University of Oslo" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="This is a set of R tutorials connected to the course BIOS1140 at the University of Oslo, and are meant as a companion to the book Evolutionary Genetics (Ravinet &amp; Sætre, 2018). They do not assume any prior R knowledge." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="9.1 Phylogenetics in R | R tutorials for the course BIOS1140 at the University of Oslo" />
  
  <meta name="twitter:description" content="This is a set of R tutorials connected to the course BIOS1140 at the University of Oslo, and are meant as a companion to the book Evolutionary Genetics (Ravinet &amp; Sætre, 2018). They do not assume any prior R knowledge." />
  

<meta name="author" content="Original tutorials written by Mark Ravinet. Revamped in 2021 by Even Sletteng Garvang and Solveig Brochmann." />


<meta name="date" content="2022-10-07" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="ch09.html"/>
<link rel="next" href="population-structure.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">R tutorials for the course BIOS1140</a></li>

<li class="divider"></li>
<li><a href="index.html#preface">Preface<span></span></a>
<ul>
<li><a href="how-to-use-these-tutorials.html#how-to-use-these-tutorials">How to use these tutorials<span></span></a></li>
</ul></li>
<li class="part"><span><b>I Introductory R course<span></span></b></span></li>
<li class="chapter" data-level="1" data-path="ch01.html"><a href="ch01.html"><i class="fa fa-check"></i><b>1</b> Introduction to R<span></span></a>
<ul>
<li><a href="ch01.html#what-to-expect">What to expect<span></span></a></li>
<li class="chapter" data-level="1.1" data-path="start-using-r.html"><a href="start-using-r.html"><i class="fa fa-check"></i><b>1.1</b> Start using R!<span></span></a>
<ul>
<li class="chapter" data-level="1.1.1" data-path="start-using-r.html"><a href="start-using-r.html#install-r-and-rstudio"><i class="fa fa-check"></i><b>1.1.1</b> Install R and Rstudio<span></span></a></li>
<li class="chapter" data-level="1.1.2" data-path="start-using-r.html"><a href="start-using-r.html#getting-familiar-with-r-and-the-console"><i class="fa fa-check"></i><b>1.1.2</b> Getting familiar with R and the console<span></span></a></li>
<li class="chapter" data-level="1.1.3" data-path="start-using-r.html"><a href="start-using-r.html#scripts"><i class="fa fa-check"></i><b>1.1.3</b> Scripts<span></span></a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="r-essentials.html"><a href="r-essentials.html"><i class="fa fa-check"></i><b>1.2</b> R essentials<span></span></a>
<ul>
<li class="chapter" data-level="1.2.1" data-path="r-essentials.html"><a href="r-essentials.html#assigning-to-objects"><i class="fa fa-check"></i><b>1.2.1</b> Assigning to objects<span></span></a></li>
<li class="chapter" data-level="1.2.2" data-path="r-essentials.html"><a href="r-essentials.html#vectors"><i class="fa fa-check"></i><b>1.2.2</b> Vectors<span></span></a></li>
<li class="chapter" data-level="1.2.3" data-path="r-essentials.html"><a href="r-essentials.html#strings"><i class="fa fa-check"></i><b>1.2.3</b> Strings<span></span></a></li>
<li class="chapter" data-level="1.2.4" data-path="r-essentials.html"><a href="r-essentials.html#logical"><i class="fa fa-check"></i><b>1.2.4</b> Logical values<span></span></a></li>
<li class="chapter" data-level="1.2.5" data-path="r-essentials.html"><a href="r-essentials.html#functions"><i class="fa fa-check"></i><b>1.2.5</b> Functions<span></span></a></li>
<li class="chapter" data-level="1.2.6" data-path="r-essentials.html"><a href="r-essentials.html#data-frames"><i class="fa fa-check"></i><b>1.2.6</b> Data frames<span></span></a></li>
</ul></li>
<li class="chapter" data-level="1.3" data-path="dataimport.html"><a href="dataimport.html"><i class="fa fa-check"></i><b>1.3</b> Data import<span></span></a>
<ul>
<li class="chapter" data-level="1.3.1" data-path="dataimport.html"><a href="dataimport.html#data-formats"><i class="fa fa-check"></i><b>1.3.1</b> Data formats<span></span></a></li>
<li class="chapter" data-level="1.3.2" data-path="dataimport.html"><a href="dataimport.html#working-directories"><i class="fa fa-check"></i><b>1.3.2</b> Working directories<span></span></a></li>
<li class="chapter" data-level="1.3.3" data-path="dataimport.html"><a href="dataimport.html#importing"><i class="fa fa-check"></i><b>1.3.3</b> Importing<span></span></a></li>
</ul></li>
<li class="chapter" data-level="1.4" data-path="plotting.html"><a href="plotting.html"><i class="fa fa-check"></i><b>1.4</b> Plotting<span></span></a>
<ul>
<li class="chapter" data-level="1.4.1" data-path="plotting.html"><a href="plotting.html#plotting-vectors"><i class="fa fa-check"></i><b>1.4.1</b> Plotting vectors<span></span></a></li>
<li class="chapter" data-level="1.4.2" data-path="plotting.html"><a href="plotting.html#customizing-your-plots"><i class="fa fa-check"></i><b>1.4.2</b> Customizing your plots<span></span></a></li>
</ul></li>
<li class="chapter" data-level="1.5" data-path="installing-and-loading-packages.html"><a href="installing-and-loading-packages.html"><i class="fa fa-check"></i><b>1.5</b> Installing and loading packages<span></span></a></li>
<li class="chapter" data-level="1.6" data-path="going-further.html"><a href="going-further.html"><i class="fa fa-check"></i><b>1.6</b> Going further<span></span></a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="ch02.html"><a href="ch02.html"><i class="fa fa-check"></i><b>2</b> Building on your foundations: going further with R<span></span></a>
<ul>
<li><a href="ch02.html#what-to-expect-1">What to expect<span></span></a></li>
<li class="chapter" data-level="2.1" data-path="intro-to-data-manipulation-with-tidyverse.html"><a href="intro-to-data-manipulation-with-tidyverse.html"><i class="fa fa-check"></i><b>2.1</b> Intro to data manipulation with tidyverse<span></span></a>
<ul>
<li class="chapter" data-level="2.1.1" data-path="intro-to-data-manipulation-with-tidyverse.html"><a href="intro-to-data-manipulation-with-tidyverse.html#what-is-the-tidyverse"><i class="fa fa-check"></i><b>2.1.1</b> What is the tidyverse?<span></span></a></li>
<li class="chapter" data-level="2.1.2" data-path="intro-to-data-manipulation-with-tidyverse.html"><a href="intro-to-data-manipulation-with-tidyverse.html#the-dplyr-package"><i class="fa fa-check"></i><b>2.1.2</b> The dplyr package<span></span></a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="using-dplyr-to-work-with-your-data.html"><a href="using-dplyr-to-work-with-your-data.html"><i class="fa fa-check"></i><b>2.2</b> Using dplyr to work with your data<span></span></a>
<ul>
<li class="chapter" data-level="2.2.1" data-path="using-dplyr-to-work-with-your-data.html"><a href="using-dplyr-to-work-with-your-data.html#the-pipe"><i class="fa fa-check"></i><b>2.2.1</b> The pipe<span></span></a></li>
<li class="chapter" data-level="2.2.2" data-path="using-dplyr-to-work-with-your-data.html"><a href="using-dplyr-to-work-with-your-data.html#selecting-columns-with-select"><i class="fa fa-check"></i><b>2.2.2</b> Selecting columns with <code>select()</code><span></span></a></li>
<li class="chapter" data-level="2.2.3" data-path="using-dplyr-to-work-with-your-data.html"><a href="using-dplyr-to-work-with-your-data.html#filtering-colums-using-filter"><i class="fa fa-check"></i><b>2.2.3</b> Filtering colums using <code>filter()</code><span></span></a></li>
<li class="chapter" data-level="2.2.4" data-path="using-dplyr-to-work-with-your-data.html"><a href="using-dplyr-to-work-with-your-data.html#grouped-summaries-with-group_by-and-summarise"><i class="fa fa-check"></i><b>2.2.4</b> Grouped summaries with <code>group_by()</code> and <code>summarise()</code><span></span></a></li>
<li class="chapter" data-level="2.2.5" data-path="using-dplyr-to-work-with-your-data.html"><a href="using-dplyr-to-work-with-your-data.html#using-everything-weve-learned-in-a-single-pipe-and-a-dplyr-exercise"><i class="fa fa-check"></i><b>2.2.5</b> Using everything we’ve learned in a single pipe, and a <code>dplyr</code> exercise<span></span></a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="plotting-your-data-with-ggplot2.html"><a href="plotting-your-data-with-ggplot2.html"><i class="fa fa-check"></i><b>2.3</b> Plotting your data with <code>ggplot2</code><span></span></a>
<ul>
<li class="chapter" data-level="2.3.1" data-path="plotting-your-data-with-ggplot2.html"><a href="plotting-your-data-with-ggplot2.html#the-three-things-you-need-in-a-ggplot"><i class="fa fa-check"></i><b>2.3.1</b> The three things you need in a ggplot<span></span></a></li>
<li class="chapter" data-level="2.3.2" data-path="plotting-your-data-with-ggplot2.html"><a href="plotting-your-data-with-ggplot2.html#storing-ggplots-in-objects"><i class="fa fa-check"></i><b>2.3.2</b> Storing ggplots in objects<span></span></a></li>
<li class="chapter" data-level="2.3.3" data-path="plotting-your-data-with-ggplot2.html"><a href="plotting-your-data-with-ggplot2.html#customizing-your-plots-1"><i class="fa fa-check"></i><b>2.3.3</b> Customizing your plots<span></span></a></li>
<li class="chapter" data-level="2.3.4" data-path="plotting-your-data-with-ggplot2.html"><a href="plotting-your-data-with-ggplot2.html#saving-your-plots"><i class="fa fa-check"></i><b>2.3.4</b> Saving your plots<span></span></a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="pivot-longer.html"><a href="pivot-longer.html"><i class="fa fa-check"></i><b>2.4</b> Reshaping data with pivot_longer()<span></span></a>
<ul>
<li class="chapter" data-level="2.4.1" data-path="pivot-longer.html"><a href="pivot-longer.html#wide-and-long-format"><i class="fa fa-check"></i><b>2.4.1</b> Wide and long format<span></span></a></li>
<li class="chapter" data-level="2.4.2" data-path="pivot-longer.html"><a href="pivot-longer.html#import-example-data"><i class="fa fa-check"></i><b>2.4.2</b> Import example data<span></span></a></li>
<li class="chapter" data-level="2.4.3" data-path="pivot-longer.html"><a href="pivot-longer.html#reshape-the-data"><i class="fa fa-check"></i><b>2.4.3</b> Reshape the data<span></span></a></li>
<li class="chapter" data-level="2.4.4" data-path="pivot-longer.html"><a href="pivot-longer.html#plot-the-data"><i class="fa fa-check"></i><b>2.4.4</b> Plot the data<span></span></a></li>
</ul></li>
<li class="chapter" data-level="2.5" data-path="study-questions.html"><a href="study-questions.html"><i class="fa fa-check"></i><b>2.5</b> Study questions<span></span></a></li>
<li class="chapter" data-level="2.6" data-path="going-further-1.html"><a href="going-further-1.html"><i class="fa fa-check"></i><b>2.6</b> Going further<span></span></a></li>
</ul></li>
<li class="part"><span><b>II Evolutionary genetics with R<span></span></b></span></li>
<li class="chapter" data-level="3" data-path="ch03.html"><a href="ch03.html"><i class="fa fa-check"></i><b>3</b> Changes in Allele and Genotype Frequency<span></span></a>
<ul>
<li><a href="ch03.html#what-to-expect-2">What to expect<span></span></a></li>
<li class="chapter" data-level="3.1" data-path="r-programming-for-loops.html"><a href="r-programming-for-loops.html"><i class="fa fa-check"></i><b>3.1</b> R programming: for-loops<span></span></a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="r-programming-for-loops.html"><a href="r-programming-for-loops.html#motivation-why-loop"><i class="fa fa-check"></i><b>3.1.1</b> Motivation: why loop?<span></span></a></li>
<li class="chapter" data-level="3.1.2" data-path="r-programming-for-loops.html"><a href="r-programming-for-loops.html#how-a-for-loop-works"><i class="fa fa-check"></i><b>3.1.2</b> How a for-loop works<span></span></a></li>
<li class="chapter" data-level="3.1.3" data-path="r-programming-for-loops.html"><a href="r-programming-for-loops.html#indexing"><i class="fa fa-check"></i><b>3.1.3</b> Indexing with for-loops<span></span></a></li>
<li class="chapter" data-level="3.1.4" data-path="r-programming-for-loops.html"><a href="r-programming-for-loops.html#solving-our-problem"><i class="fa fa-check"></i><b>3.1.4</b> Solving our problem<span></span></a></li>
<li class="chapter" data-level="3.1.5" data-path="r-programming-for-loops.html"><a href="r-programming-for-loops.html#storing-values-from-a-for-loop"><i class="fa fa-check"></i><b>3.1.5</b> Storing values from a for-loop<span></span></a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="evolutionary-biology.html"><a href="evolutionary-biology.html"><i class="fa fa-check"></i><b>3.2</b> Evolutionary biology<span></span></a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="evolutionary-biology.html"><a href="evolutionary-biology.html#the-hardy-weinberg-model"><i class="fa fa-check"></i><b>3.2.1</b> The Hardy-Weinberg Model<span></span></a></li>
<li class="chapter" data-level="3.2.2" data-path="evolutionary-biology.html"><a href="evolutionary-biology.html#testing-for-deviations-from-the-hardy-weinberg-expectation"><i class="fa fa-check"></i><b>3.2.2</b> Testing for deviations from the Hardy Weinberg Expectation<span></span></a></li>
<li class="chapter" data-level="3.2.3" data-path="evolutionary-biology.html"><a href="evolutionary-biology.html#simulating-genetic-drift"><i class="fa fa-check"></i><b>3.2.3</b> Simulating genetic drift<span></span></a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="study-questions-1.html"><a href="study-questions-1.html"><i class="fa fa-check"></i><b>3.3</b> Study questions<span></span></a></li>
<li class="chapter" data-level="3.4" data-path="going-further-2.html"><a href="going-further-2.html"><i class="fa fa-check"></i><b>3.4</b> Going further<span></span></a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="ch04.html"><a href="ch04.html"><i class="fa fa-check"></i><b>4</b> The Theory of Natural Selection<span></span></a>
<ul>
<li><a href="ch04.html#what-to-expect-3">What to expect<span></span></a></li>
<li><a href="ch04.html#getting-started">Getting started<span></span></a></li>
<li class="chapter" data-level="4.1" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html"><i class="fa fa-check"></i><b>4.1</b> R programming: Making custom functions<span></span></a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#motivation"><i class="fa fa-check"></i><b>4.1.1</b> Motivation<span></span></a></li>
<li class="chapter" data-level="4.1.2" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#function-basics"><i class="fa fa-check"></i><b>4.1.2</b> Function basics<span></span></a></li>
<li class="chapter" data-level="4.1.3" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#a-simple-example"><i class="fa fa-check"></i><b>4.1.3</b> A simple example<span></span></a></li>
<li class="chapter" data-level="4.1.4" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#some-important-function-properties"><i class="fa fa-check"></i><b>4.1.4</b> Some important function properties<span></span></a></li>
<li class="chapter" data-level="4.1.5" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#a-slightly-more-useful-example"><i class="fa fa-check"></i><b>4.1.5</b> A slightly more useful example<span></span></a></li>
<li class="chapter" data-level="4.1.6" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#example-calculating-genotype-frequencies"><i class="fa fa-check"></i><b>4.1.6</b> Example: calculating genotype frequencies<span></span></a></li>
<li class="chapter" data-level="4.1.7" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#creating-a-function-of-the-drift-simulation"><i class="fa fa-check"></i><b>4.1.7</b> Creating a function of the drift simulation<span></span></a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="evolutionary-biology-fitness.html"><a href="evolutionary-biology-fitness.html"><i class="fa fa-check"></i><b>4.2</b> Evolutionary biology: fitness<span></span></a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="evolutionary-biology-fitness.html"><a href="evolutionary-biology-fitness.html#understanding-fitness"><i class="fa fa-check"></i><b>4.2.1</b> Understanding fitness<span></span></a></li>
<li class="chapter" data-level="4.2.2" data-path="evolutionary-biology-fitness.html"><a href="evolutionary-biology-fitness.html#one-locus-model-of-viability-selection"><i class="fa fa-check"></i><b>4.2.2</b> One-locus model of viability selection<span></span></a></li>
<li class="chapter" data-level="4.2.3" data-path="evolutionary-biology-fitness.html"><a href="evolutionary-biology-fitness.html#over-and-underdominance"><i class="fa fa-check"></i><b>4.2.3</b> Over and underdominance<span></span></a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="study-questions-2.html"><a href="study-questions-2.html"><i class="fa fa-check"></i><b>4.3</b> Study questions<span></span></a></li>
<li class="chapter" data-level="4.4" data-path="going-further-3.html"><a href="going-further-3.html"><i class="fa fa-check"></i><b>4.4</b> Going further<span></span></a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="ch05.html"><a href="ch05.html"><i class="fa fa-check"></i><b>5</b> Detecting Natural Selection<span></span></a>
<ul>
<li><a href="ch05.html#what-to-expect-4">What to expect<span></span></a></li>
<li><a href="ch05.html#getting-started-1">Getting started<span></span></a></li>
<li class="chapter" data-level="5.1" data-path="vectorisation.html"><a href="vectorisation.html"><i class="fa fa-check"></i><b>5.1</b> More on functions: Vectorisation<span></span></a>
<ul>
<li class="chapter" data-level="5.1.1" data-path="vectorisation.html"><a href="vectorisation.html#the-apply-function"><i class="fa fa-check"></i><b>5.1.1</b> The <code>apply()</code> function<span></span></a></li>
<li class="chapter" data-level="5.1.2" data-path="vectorisation.html"><a href="vectorisation.html#ifelse"><i class="fa fa-check"></i><b>5.1.2</b> The <code>ifelse()</code> function<span></span></a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="understanding-fst---the-fixation-index.html"><a href="understanding-fst---the-fixation-index.html"><i class="fa fa-check"></i><b>5.2</b> Understanding <em>F</em><sub>ST</sub> - the fixation index<span></span></a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="understanding-fst---the-fixation-index.html"><a href="understanding-fst---the-fixation-index.html#a-worked-example-of-fst-in-humans"><i class="fa fa-check"></i><b>5.2.1</b> A worked example of <em>F</em><sub>ST</sub> in humans<span></span></a></li>
<li class="chapter" data-level="5.2.2" data-path="understanding-fst---the-fixation-index.html"><a href="understanding-fst---the-fixation-index.html#writing-a-set-of-fst-functions"><i class="fa fa-check"></i><b>5.2.2</b> Writing a set of <em>F</em><sub>ST</sub> functions<span></span></a></li>
<li class="chapter" data-level="5.2.3" data-path="understanding-fst---the-fixation-index.html"><a href="understanding-fst---the-fixation-index.html#applying-functions-to-matrices-and-data-frames"><i class="fa fa-check"></i><b>5.2.3</b> Applying functions to matrices and data frames<span></span></a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="visualising-fst-along-a-chromosome.html"><a href="visualising-fst-along-a-chromosome.html"><i class="fa fa-check"></i><b>5.3</b> Visualising <em>F</em><sub>ST</sub> along a chromosome<span></span></a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="visualising-fst-along-a-chromosome.html"><a href="visualising-fst-along-a-chromosome.html#identifying-outliers-in-our-fst-distribution"><i class="fa fa-check"></i><b>5.3.1</b> Identifying outliers in our <em>F</em><sub>ST</sub> distribution<span></span></a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="study-questions-3.html"><a href="study-questions-3.html"><i class="fa fa-check"></i><b>5.4</b> Study questions<span></span></a></li>
<li class="chapter" data-level="5.5" data-path="going-further-4.html"><a href="going-further-4.html"><i class="fa fa-check"></i><b>5.5</b> Going further<span></span></a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="ch06.html"><a href="ch06.html"><i class="fa fa-check"></i><b>6</b> Multilocus Evolution<span></span></a>
<ul>
<li><a href="ch06.html#what-to-expect-5">What to expect<span></span></a></li>
<li><a href="ch06.html#getting-started-2">Getting started<span></span></a></li>
<li class="chapter" data-level="6.1" data-path="tools-for-investigating-differences-between-groups.html"><a href="tools-for-investigating-differences-between-groups.html"><i class="fa fa-check"></i><b>6.1</b> Tools for investigating differences between groups<span></span></a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="tools-for-investigating-differences-between-groups.html"><a href="tools-for-investigating-differences-between-groups.html#exploring-group-differences"><i class="fa fa-check"></i><b>6.1.1</b> Exploring group differences<span></span></a></li>
<li class="chapter" data-level="6.1.2" data-path="tools-for-investigating-differences-between-groups.html"><a href="tools-for-investigating-differences-between-groups.html#analysis-of-variance-anova"><i class="fa fa-check"></i><b>6.1.2</b> Analysis of variance (ANOVA)<span></span></a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="performing-a-qtl-analysis-in-bedbugs.html"><a href="performing-a-qtl-analysis-in-bedbugs.html"><i class="fa fa-check"></i><b>6.2</b> Performing a QTL analysis in bedbugs<span></span></a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="performing-a-qtl-analysis-in-bedbugs.html"><a href="performing-a-qtl-analysis-in-bedbugs.html#bedbugs-pesticide-resistance-and-study-design"><i class="fa fa-check"></i><b>6.2.1</b> Bedbugs, pesticide resistance and study design<span></span></a></li>
<li class="chapter" data-level="6.2.2" data-path="performing-a-qtl-analysis-in-bedbugs.html"><a href="performing-a-qtl-analysis-in-bedbugs.html#linking-anova-to-our-qtl-analysis"><i class="fa fa-check"></i><b>6.2.2</b> Linking ANOVA to our QTL analysis<span></span></a></li>
<li class="chapter" data-level="6.2.3" data-path="performing-a-qtl-analysis-in-bedbugs.html"><a href="performing-a-qtl-analysis-in-bedbugs.html#testing-the-same-genotype-phenotype-association-with-anova"><i class="fa fa-check"></i><b>6.2.3</b> Testing the same genotype phenotype association with ANOVA<span></span></a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="study-questions-4.html"><a href="study-questions-4.html"><i class="fa fa-check"></i><b>6.3</b> Study questions<span></span></a></li>
<li class="chapter" data-level="6.4" data-path="going-further-5.html"><a href="going-further-5.html"><i class="fa fa-check"></i><b>6.4</b> Going further<span></span></a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="ch07.html"><a href="ch07.html"><i class="fa fa-check"></i><b>7</b> Inferring Evolutionary Processes from Sequence Data<span></span></a>
<ul>
<li><a href="ch07.html#what-to-expect-6">What to expect<span></span></a></li>
<li><a href="ch07.html#getting-started-3">Getting started<span></span></a></li>
<li class="chapter" data-level="7.1" data-path="working-with-dna-sequence-data.html"><a href="working-with-dna-sequence-data.html"><i class="fa fa-check"></i><b>7.1</b> Working with DNA sequence data<span></span></a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="working-with-dna-sequence-data.html"><a href="working-with-dna-sequence-data.html#reading-sequence-data-into-r."><i class="fa fa-check"></i><b>7.1.1</b> Reading sequence data into R.<span></span></a></li>
<li class="chapter" data-level="7.1.2" data-path="working-with-dna-sequence-data.html"><a href="working-with-dna-sequence-data.html#exploring-dna-sequence-data"><i class="fa fa-check"></i><b>7.1.2</b> Exploring DNA sequence data<span></span></a></li>
<li class="chapter" data-level="7.1.3" data-path="working-with-dna-sequence-data.html"><a href="working-with-dna-sequence-data.html#calculating-basic-sequence-statistics"><i class="fa fa-check"></i><b>7.1.3</b> Calculating basic sequence statistics<span></span></a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="working-with-a-larger-dataset.html"><a href="working-with-a-larger-dataset.html"><i class="fa fa-check"></i><b>7.2</b> Working with a larger dataset<span></span></a>
<ul>
<li class="chapter" data-level="7.2.1" data-path="working-with-a-larger-dataset.html"><a href="working-with-a-larger-dataset.html#sample-size-and-sequence-statistics"><i class="fa fa-check"></i><b>7.2.1</b> Sample size and sequence statistics<span></span></a></li>
<li class="chapter" data-level="7.2.2" data-path="working-with-a-larger-dataset.html"><a href="working-with-a-larger-dataset.html#inferring-evolutionary-processes-using-tajimas-d"><i class="fa fa-check"></i><b>7.2.2</b> Inferring evolutionary processes using Tajima’s <em>D</em><span></span></a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="calculating-statistics-at-the-whole-genome-level.html"><a href="calculating-statistics-at-the-whole-genome-level.html"><i class="fa fa-check"></i><b>7.3</b> Calculating statistics at the whole genome level<span></span></a>
<ul>
<li class="chapter" data-level="7.3.1" data-path="calculating-statistics-at-the-whole-genome-level.html"><a href="calculating-statistics-at-the-whole-genome-level.html#reading-in-variant-data"><i class="fa fa-check"></i><b>7.3.1</b> Reading in variant data<span></span></a></li>
<li class="chapter" data-level="7.3.2" data-path="calculating-statistics-at-the-whole-genome-level.html"><a href="calculating-statistics-at-the-whole-genome-level.html#calculating-nucleotide-diversity-statistics"><i class="fa fa-check"></i><b>7.3.2</b> Calculating nucleotide diversity statistics<span></span></a></li>
<li class="chapter" data-level="7.3.3" data-path="calculating-statistics-at-the-whole-genome-level.html"><a href="calculating-statistics-at-the-whole-genome-level.html#visualising-nucleotide-diversity-along-the-chromosome"><i class="fa fa-check"></i><b>7.3.3</b> Visualising nucleotide diversity along the chromosome<span></span></a></li>
<li class="chapter" data-level="7.3.4" data-path="calculating-statistics-at-the-whole-genome-level.html"><a href="calculating-statistics-at-the-whole-genome-level.html#performing-a-sliding-window-analysis"><i class="fa fa-check"></i><b>7.3.4</b> Performing a sliding window analysis<span></span></a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="study-questions-5.html"><a href="study-questions-5.html"><i class="fa fa-check"></i><b>7.4</b> Study questions<span></span></a></li>
<li class="chapter" data-level="7.5" data-path="going-further-6.html"><a href="going-further-6.html"><i class="fa fa-check"></i><b>7.5</b> Going further<span></span></a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="ch08.html"><a href="ch08.html"><i class="fa fa-check"></i><b>8</b> Speciation Genomics<span></span></a>
<ul>
<li><a href="ch08.html#what-to-expect-7">What to expect<span></span></a></li>
<li><a href="ch08.html#getting-started-4">Getting started<span></span></a></li>
<li class="chapter" data-level="8.1" data-path="visualizing-complex-data.html"><a href="visualizing-complex-data.html"><i class="fa fa-check"></i><b>8.1</b> Visualizing complex data<span></span></a>
<ul>
<li class="chapter" data-level="8.1.1" data-path="visualizing-complex-data.html"><a href="visualizing-complex-data.html#faceting"><i class="fa fa-check"></i><b>8.1.1</b> Faceting<span></span></a></li>
</ul></li>
<li class="chapter" data-level="8.2" data-path="returning-to-the-sparrow-dataset.html"><a href="returning-to-the-sparrow-dataset.html"><i class="fa fa-check"></i><b>8.2</b> Returning to the sparrow dataset<span></span></a>
<ul>
<li class="chapter" data-level="8.2.1" data-path="returning-to-the-sparrow-dataset.html"><a href="returning-to-the-sparrow-dataset.html#reading-in-the-sparrow-vcf"><i class="fa fa-check"></i><b>8.2.1</b> Reading in the sparrow vcf<span></span></a></li>
<li class="chapter" data-level="8.2.2" data-path="returning-to-the-sparrow-dataset.html"><a href="returning-to-the-sparrow-dataset.html#examining-the-variant-data"><i class="fa fa-check"></i><b>8.2.2</b> Examining the variant data<span></span></a></li>
</ul></li>
<li class="chapter" data-level="8.3" data-path="setting-up-sliding-windows.html"><a href="setting-up-sliding-windows.html"><i class="fa fa-check"></i><b>8.3</b> Setting up sliding windows<span></span></a>
<ul>
<li class="chapter" data-level="8.3.1" data-path="setting-up-sliding-windows.html"><a href="setting-up-sliding-windows.html#calculating-sliding-window-estimates-of-nucleotide-diversity-and-differentiation"><i class="fa fa-check"></i><b>8.3.1</b> Calculating sliding window estimates of nucleotide diversity and differentiation<span></span></a></li>
<li class="chapter" data-level="8.3.2" data-path="setting-up-sliding-windows.html"><a href="setting-up-sliding-windows.html#sparrow-data"><i class="fa fa-check"></i><b>8.3.2</b> Extracting statistics for visualisation<span></span></a></li>
</ul></li>
<li class="chapter" data-level="8.4" data-path="sparrow-viz.html"><a href="sparrow-viz.html"><i class="fa fa-check"></i><b>8.4</b> Visualising the data<span></span></a>
<ul>
<li class="chapter" data-level="8.4.1" data-path="sparrow-viz.html"><a href="sparrow-viz.html#visualising-patterns-along-the-chromosome"><i class="fa fa-check"></i><b>8.4.1</b> Visualising patterns along the chromosome<span></span></a></li>
<li class="chapter" data-level="8.4.2" data-path="sparrow-viz.html"><a href="sparrow-viz.html#interlude-relative-vs.-absolute-measures-of-nucleotide-diversity"><i class="fa fa-check"></i><b>8.4.2</b> Interlude: relative vs. absolute measures of nucleotide diversity<span></span></a></li>
<li class="chapter" data-level="8.4.3" data-path="sparrow-viz.html"><a href="sparrow-viz.html#investigating-recombination-rate-variation"><i class="fa fa-check"></i><b>8.4.3</b> Investigating recombination rate variation<span></span></a></li>
</ul></li>
<li class="chapter" data-level="8.5" data-path="study-questions-6.html"><a href="study-questions-6.html"><i class="fa fa-check"></i><b>8.5</b> Study questions<span></span></a></li>
<li class="chapter" data-level="8.6" data-path="going-further-7.html"><a href="going-further-7.html"><i class="fa fa-check"></i><b>8.6</b> Going further<span></span></a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="ch09.html"><a href="ch09.html"><i class="fa fa-check"></i><b>9</b> Reconstructing the Past<span></span></a>
<ul>
<li><a href="ch09.html#what-to-expect-8">What to expect<span></span></a></li>
<li><a href="ch09.html#getting-started-5">Getting started<span></span></a></li>
<li class="chapter" data-level="9.1" data-path="phylogenetics-in-r.html"><a href="phylogenetics-in-r.html"><i class="fa fa-check"></i><b>9.1</b> Phylogenetics in R<span></span></a>
<ul>
<li class="chapter" data-level="9.1.1" data-path="phylogenetics-in-r.html"><a href="phylogenetics-in-r.html#storing-trees-in-r"><i class="fa fa-check"></i><b>9.1.1</b> Storing trees in R<span></span></a></li>
<li class="chapter" data-level="9.1.2" data-path="phylogenetics-in-r.html"><a href="phylogenetics-in-r.html#plotting-trees"><i class="fa fa-check"></i><b>9.1.2</b> Plotting trees<span></span></a></li>
<li class="chapter" data-level="9.1.3" data-path="phylogenetics-in-r.html"><a href="phylogenetics-in-r.html#a-simple-example-with-real-data---avian-phylogenetics"><i class="fa fa-check"></i><b>9.1.3</b> A simple example with real data - avian phylogenetics<span></span></a></li>
<li class="chapter" data-level="9.1.4" data-path="phylogenetics-in-r.html"><a href="phylogenetics-in-r.html#constructing-trees-with-r"><i class="fa fa-check"></i><b>9.1.4</b> Constructing trees with R<span></span></a></li>
</ul></li>
<li class="chapter" data-level="9.2" data-path="population-structure.html"><a href="population-structure.html"><i class="fa fa-check"></i><b>9.2</b> Population structure<span></span></a>
<ul>
<li class="chapter" data-level="9.2.1" data-path="population-structure.html"><a href="population-structure.html#village-dogs-as-an-insight-to-dog-domestication"><i class="fa fa-check"></i><b>9.2.1</b> Village dogs as an insight to dog domestication<span></span></a></li>
<li class="chapter" data-level="9.2.2" data-path="population-structure.html"><a href="population-structure.html#reading-the-data-into-r"><i class="fa fa-check"></i><b>9.2.2</b> Reading the data into R<span></span></a></li>
<li class="chapter" data-level="9.2.3" data-path="population-structure.html"><a href="population-structure.html#performing-a-pca"><i class="fa fa-check"></i><b>9.2.3</b> Performing a PCA<span></span></a></li>
<li class="chapter" data-level="9.2.4" data-path="population-structure.html"><a href="population-structure.html#visualising-the-pca"><i class="fa fa-check"></i><b>9.2.4</b> Visualising the PCA<span></span></a></li>
<li class="chapter" data-level="9.2.5" data-path="population-structure.html"><a href="population-structure.html#eigenvalues"><i class="fa fa-check"></i><b>9.2.5</b> Eigenvalues<span></span></a></li>
<li class="chapter" data-level="9.2.6" data-path="population-structure.html"><a href="population-structure.html#the-full-data-set"><i class="fa fa-check"></i><b>9.2.6</b> The full data set<span></span></a></li>
</ul></li>
<li class="chapter" data-level="9.3" data-path="study-questions-7.html"><a href="study-questions-7.html"><i class="fa fa-check"></i><b>9.3</b> Study questions<span></span></a></li>
<li class="chapter" data-level="9.4" data-path="going-further-8.html"><a href="going-further-8.html"><i class="fa fa-check"></i><b>9.4</b> Going further<span></span></a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="ch10.html"><a href="ch10.html"><i class="fa fa-check"></i><b>10</b> Advancing Further in R<span></span></a>
<ul>
<li><a href="ch10.html#what-to-expect-9">What to expect<span></span></a></li>
<li><a href="ch10.html#getting-started-6">Getting started<span></span></a></li>
<li class="chapter" data-level="10.1" data-path="advanced-features-of-rstudio.html"><a href="advanced-features-of-rstudio.html"><i class="fa fa-check"></i><b>10.1</b> Advanced features of RStudio<span></span></a>
<ul>
<li class="chapter" data-level="10.1.1" data-path="advanced-features-of-rstudio.html"><a href="advanced-features-of-rstudio.html#projects-projects-projects"><i class="fa fa-check"></i><b>10.1.1</b> Projects, projects, projects<span></span></a></li>
<li class="chapter" data-level="10.1.2" data-path="advanced-features-of-rstudio.html"><a href="advanced-features-of-rstudio.html#everyone-has-a-history"><i class="fa fa-check"></i><b>10.1.2</b> Everyone has a history<span></span></a></li>
<li class="chapter" data-level="10.1.3" data-path="advanced-features-of-rstudio.html"><a href="advanced-features-of-rstudio.html#tab-complete-and-other-hotkeys"><i class="fa fa-check"></i><b>10.1.3</b> Tab complete and other hotkeys<span></span></a></li>
</ul></li>
<li class="chapter" data-level="10.2" data-path="more-on-data-handling-categorising-and-joining.html"><a href="more-on-data-handling-categorising-and-joining.html"><i class="fa fa-check"></i><b>10.2</b> More on data handling: categorising and joining<span></span></a>
<ul>
<li class="chapter" data-level="10.2.1" data-path="more-on-data-handling-categorising-and-joining.html"><a href="more-on-data-handling-categorising-and-joining.html#ifelse-for-making-categories"><i class="fa fa-check"></i><b>10.2.1</b> <code>ifelse()</code> for making categories<span></span></a></li>
<li class="chapter" data-level="10.2.2" data-path="more-on-data-handling-categorising-and-joining.html"><a href="more-on-data-handling-categorising-and-joining.html#joining"><i class="fa fa-check"></i><b>10.2.2</b> Joining<span></span></a></li>
</ul></li>
<li class="chapter" data-level="10.3" data-path="lists.html"><a href="lists.html"><i class="fa fa-check"></i><b>10.3</b> Lists<span></span></a></li>
<li class="chapter" data-level="10.4" data-path="vectorisation2.html"><a href="vectorisation2.html"><i class="fa fa-check"></i><b>10.4</b> Vectorisation<span></span></a>
<ul>
<li class="chapter" data-level="10.4.1" data-path="vectorisation2.html"><a href="vectorisation2.html#using-lapply-to-vectorise"><i class="fa fa-check"></i><b>10.4.1</b> Using <code>lapply()</code> to vectorise<span></span></a></li>
<li class="chapter" data-level="10.4.2" data-path="vectorisation2.html"><a href="vectorisation2.html#sapply"><i class="fa fa-check"></i><b>10.4.2</b> <code>sapply()</code><span></span></a></li>
<li class="chapter" data-level="10.4.3" data-path="vectorisation2.html"><a href="vectorisation2.html#anonymous-functions"><i class="fa fa-check"></i><b>10.4.3</b> Anonymous functions<span></span></a></li>
</ul></li>
<li class="chapter" data-level="10.5" data-path="concluding-remarks.html"><a href="concluding-remarks.html"><i class="fa fa-check"></i><b>10.5</b> Concluding remarks<span></span></a></li>
<li class="chapter" data-level="10.6" data-path="study-questions-8.html"><a href="study-questions-8.html"><i class="fa fa-check"></i><b>10.6</b> Study questions<span></span></a></li>
<li class="chapter" data-level="10.7" data-path="going-further-9.html"><a href="going-further-9.html"><i class="fa fa-check"></i><b>10.7</b> Going further<span></span></a></li>
</ul></li>
<li class="part"><span><b>III Assignments<span></span></b></span></li>
<li><a href="w01.html#w01">Week 1-2 assignment<span></span></a></li>
<li><a href="w03.html#w03">Week 3 Assignment<span></span></a></li>
<li><a href="w04.html#w04">Week 4 assignment<span></span></a></li>
<li><a href="w05.html#w05">Week 5 assignment<span></span></a></li>
<li><a href="w06.html#w06">Week 6 assignment<span></span></a></li>
<li><a href="w07.html#w07">Week 7 assignment<span></span></a></li>
<li><a href="w08.html#w08">Week 8 assignment<span></span></a></li>
<li><a href="w09.html#w09">Week 9 assignment<span></span></a></li>
<li><a href="w10.html#w10">Week 10 assignment<span></span></a></li>
<li class="appendix"><span><b>Appendix<span></span></b></span></li>
<li class="chapter" data-level="A" data-path="rmarkdown.html"><a href="rmarkdown.html"><i class="fa fa-check"></i><b>A</b> About assignments and RMarkdown<span></span></a>
<ul>
<li><a href="rmarkdown.html#r-markdown">R Markdown<span></span></a></li>
<li><a href="rmarkdown.html#getting-started-7">Getting started<span></span></a></li>
</ul></li>
<li class="chapter" data-level="B" data-path="tech.html"><a href="tech.html"><i class="fa fa-check"></i><b>B</b> Technical information<span></span></a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Made with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">R tutorials for the course BIOS1140 at the University of Oslo</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="phylogenetics-in-r" class="section level2 hasAnchor" number="9.1">
<h2><span class="header-section-number">9.1</span> Phylogenetics in R<a href="phylogenetics-in-r.html#phylogenetics-in-r" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<script src="js/hideOutput.js"></script>
<p>R has a number of extremely powerful packages for performing phylogenetic analysis, from plotting trees to testing comparative models of evolution. You can see <a href="https://cran.r-project.org/web/views/Phylogenetics.html">here</a> for more information if you are interested in learning about what sort of things are possible. For today’s session, we will learn how to handle and visualise phylogenetic trees in R. We will also construct a series of trees from a sequence alignment. First, let’s familiarise ourselves with how R handles phylogenetic data.</p>
<div id="storing-trees-in-r" class="section level3 hasAnchor" number="9.1.1">
<h3><span class="header-section-number">9.1.1</span> Storing trees in R<a href="phylogenetics-in-r.html#storing-trees-in-r" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The backbone of most phylogenetic analysis in R comes from the functions that are part of the <code>ape</code> package. <code>ape</code> stores trees as <code>phylo</code> objects, which are easy to access and manipulate. The easiest way to understand this is to have a look at a simple phylogeny, so we’ll create a random tree now.</p>
<div class="sourceCode" id="cb354"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb354-1"><a href="phylogenetics-in-r.html#cb354-1" aria-hidden="true" tabindex="-1"></a><span class="co"># set seed to ensure the same tree is produced</span></span>
<span id="cb354-2"><a href="phylogenetics-in-r.html#cb354-2" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">32</span>)</span>
<span id="cb354-3"><a href="phylogenetics-in-r.html#cb354-3" aria-hidden="true" tabindex="-1"></a><span class="co"># generate a tree</span></span>
<span id="cb354-4"><a href="phylogenetics-in-r.html#cb354-4" aria-hidden="true" tabindex="-1"></a>tree <span class="ot">&lt;-</span> <span class="fu">rtree</span>(<span class="at">n =</span> <span class="dv">4</span>, <span class="at">tip.label =</span> <span class="fu">c</span>(<span class="st">&quot;a&quot;</span>, <span class="st">&quot;b&quot;</span>, <span class="st">&quot;c&quot;</span>, <span class="st">&quot;d&quot;</span>))</span></code></pre></div>
<p>What have we done here? First, the <code>set.seed</code> function just sets a seed for our random simulation of a tree. You won’t need to worry about this for the majority of the time, here we are using it to make sure that when we randomly create a tree, we all create the same one.</p>
<p>What you need to focus on is the second line of code that uses the <code>rtree</code> function. This is simply a means to generate a random tree. With the <code>n = 4</code> argument, we are simply stating our tree will have four taxa and we are already specifying what they should be called with the <code>tip.label</code> argument.</p>
<p>Let’s take a closer look at our <code>tree</code> object. It is a <code>phylo</code> object - you can demonstrate this to yourself with <code>class(tree)</code>.</p>
<div class="sourceCode" id="cb355"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb355-1"><a href="phylogenetics-in-r.html#cb355-1" aria-hidden="true" tabindex="-1"></a>tree</span>
<span id="cb355-2"><a href="phylogenetics-in-r.html#cb355-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb355-3"><a href="phylogenetics-in-r.html#cb355-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Phylogenetic tree with 4 tips and 3 internal nodes.</span></span>
<span id="cb355-4"><a href="phylogenetics-in-r.html#cb355-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb355-5"><a href="phylogenetics-in-r.html#cb355-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Tip labels:</span></span>
<span id="cb355-6"><a href="phylogenetics-in-r.html#cb355-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   c, a, d, b</span></span>
<span id="cb355-7"><a href="phylogenetics-in-r.html#cb355-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb355-8"><a href="phylogenetics-in-r.html#cb355-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Rooted; includes branch lengths.</span></span></code></pre></div>
<p>By printing <code>tree</code> to the console, we see it is a tree with 4 tips and 3 internal nodes, a set of tip labels. We also see it is rooted and that the branch lengths are stored in this object too.</p>
<p>You can actually look more deeply into the data stored within the <code>tree</code> object if you want to. Try the following code and see what is inside.</p>
<div class="sourceCode" id="cb356"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb356-1"><a href="phylogenetics-in-r.html#cb356-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(tree)</span>
<span id="cb356-2"><a href="phylogenetics-in-r.html#cb356-2" aria-hidden="true" tabindex="-1"></a><span class="fu">objects</span>(tree)</span>
<span id="cb356-3"><a href="phylogenetics-in-r.html#cb356-3" aria-hidden="true" tabindex="-1"></a>tree<span class="sc">$</span>edge</span>
<span id="cb356-4"><a href="phylogenetics-in-r.html#cb356-4" aria-hidden="true" tabindex="-1"></a>tree<span class="sc">$</span>edge.length</span></code></pre></div>
<p>It is of course, much easier to understand a tree when we visualise it. Luckily this is easy in R.</p>
<div class="sourceCode" id="cb357"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb357-1"><a href="phylogenetics-in-r.html#cb357-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(tree)</span></code></pre></div>
<p><img src="Exercise9_files/figure-html/unnamed-chunk-6-1.png" width="672" /></p>
<p>In the next section, we will learn more about how to plot trees.</p>
</div>
<div id="plotting-trees" class="section level3 hasAnchor" number="9.1.2">
<h3><span class="header-section-number">9.1.2</span> Plotting trees<a href="phylogenetics-in-r.html#plotting-trees" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>We can do a lot with our trees in R using a few simple plot commands. We will use some of these later in the tutorial and assignment, so here’s a quick introduction of some of the options you have.</p>
<p>First, let’s generate another random tree, this time with 5 taxa.</p>
<div class="sourceCode" id="cb358"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb358-1"><a href="phylogenetics-in-r.html#cb358-1" aria-hidden="true" tabindex="-1"></a><span class="co"># set seed to ensure the same tree is produced</span></span>
<span id="cb358-2"><a href="phylogenetics-in-r.html#cb358-2" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">32</span>)</span>
<span id="cb358-3"><a href="phylogenetics-in-r.html#cb358-3" aria-hidden="true" tabindex="-1"></a><span class="co"># generate a tree</span></span>
<span id="cb358-4"><a href="phylogenetics-in-r.html#cb358-4" aria-hidden="true" tabindex="-1"></a>tree <span class="ot">&lt;-</span> <span class="fu">rtree</span>(<span class="at">n =</span> <span class="dv">5</span>, <span class="at">tip.label =</span> <span class="fu">c</span>(<span class="st">&quot;a&quot;</span>, <span class="st">&quot;b&quot;</span>, <span class="st">&quot;c&quot;</span>, <span class="st">&quot;d&quot;</span>, <span class="st">&quot;e&quot;</span>))</span></code></pre></div>
<p>Now, try modifying the appearance of the tree using some of these arguments to <code>plot()</code>:</p>
<ul>
<li><code>use.edge.length</code> (<code>TRUE</code> (default) or <code>FALSE</code>): should branch length be used to represent evolutionary distance?</li>
<li><code>type</code>: the type of tree to plot. Options include “phylogram” (default), “cladogram”, “unrooted” and “fan”.</li>
<li><code>edge.width</code>: sets the thickness of the branches</li>
<li><code>edge.color</code>: sets the color of the branches</li>
</ul>
<p>See <code>?plot.phylo</code> for a comprehensive list of arguments.</p>
<p>You can also manipulate the contents of your tree:</p>
<ul>
<li><code>drop.tip()</code> removes a tip from the tree</li>
<li><code>rotate()</code> switches places of two tips in the visualisation of the tree (without altering the evolutionary relationship among taxa)</li>
<li><code>extract.clade()</code> subsets the tree to a given clade</li>
</ul>
<p>See the help pages for the functions to find out more about how they work. Now, let’s use some of the options we’ve learned here for looking at some real data.</p>
</div>
<div id="a-simple-example-with-real-data---avian-phylogenetics" class="section level3 hasAnchor" number="9.1.3">
<h3><span class="header-section-number">9.1.3</span> A simple example with real data - avian phylogenetics<a href="phylogenetics-in-r.html#a-simple-example-with-real-data---avian-phylogenetics" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>So far, we have only looked at randomly generated trees. Let’s have a look at some data stored within <code>ape</code>—a phylogeny of birds at the order level.</p>
<div class="sourceCode" id="cb359"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb359-1"><a href="phylogenetics-in-r.html#cb359-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get bird order data</span></span>
<span id="cb359-2"><a href="phylogenetics-in-r.html#cb359-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">&quot;bird.orders&quot;</span>)</span></code></pre></div>
<p>Let’s plot the phylogeny to have a look at it. We will also add some annotation to make sense of the phylogeny.</p>
<div class="sourceCode" id="cb360"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb360-1"><a href="phylogenetics-in-r.html#cb360-1" aria-hidden="true" tabindex="-1"></a><span class="co"># no.margin = TRUE gives prettier plots</span></span>
<span id="cb360-2"><a href="phylogenetics-in-r.html#cb360-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(bird.orders, <span class="at">no.margin =</span> <span class="cn">TRUE</span>)</span>
<span id="cb360-3"><a href="phylogenetics-in-r.html#cb360-3" aria-hidden="true" tabindex="-1"></a><span class="fu">segments</span>(<span class="dv">38</span>, <span class="dv">1</span>, <span class="dv">38</span>, <span class="dv">5</span>, <span class="at">lwd =</span> <span class="dv">2</span>)</span>
<span id="cb360-4"><a href="phylogenetics-in-r.html#cb360-4" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="dv">39</span>, <span class="dv">3</span>, <span class="st">&quot;Proaves&quot;</span>, <span class="at">srt =</span> <span class="dv">270</span>)</span>
<span id="cb360-5"><a href="phylogenetics-in-r.html#cb360-5" aria-hidden="true" tabindex="-1"></a><span class="fu">segments</span>(<span class="dv">38</span>, <span class="dv">6</span>, <span class="dv">38</span>, <span class="dv">23</span>, <span class="at">lwd =</span> <span class="dv">2</span>)</span>
<span id="cb360-6"><a href="phylogenetics-in-r.html#cb360-6" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="dv">39</span>, <span class="fl">14.5</span>, <span class="st">&quot;Neoaves&quot;</span>, <span class="at">srt =</span> <span class="dv">270</span>)</span></code></pre></div>
<p><img src="Exercise9_files/figure-html/unnamed-chunk-9-1.png" width="672" /></p>
<p>Here, the <code>segments</code> and <code>text</code> functions specify the bars and names of the two major groups in our avian phylogeny. We are just using them for display purposes here, but if you’d like to know more about them, you can look at the R help with <code>?segments</code> and <code>?text</code> commands.</p>
<p>Let’s focus on the Neoaves clade for now. Perhaps we want to test whether certain families within Neoaves form a monophyletic group? We can do this with the <code>is.monophyletic</code> function.</p>
<div class="sourceCode" id="cb361"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb361-1"><a href="phylogenetics-in-r.html#cb361-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Parrots and Passerines?</span></span>
<span id="cb361-2"><a href="phylogenetics-in-r.html#cb361-2" aria-hidden="true" tabindex="-1"></a><span class="fu">is.monophyletic</span>(bird.orders, <span class="fu">c</span>(<span class="st">&quot;Passeriformes&quot;</span>, <span class="st">&quot;Psittaciformes&quot;</span>))</span>
<span id="cb361-3"><a href="phylogenetics-in-r.html#cb361-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] FALSE</span></span>
<span id="cb361-4"><a href="phylogenetics-in-r.html#cb361-4" aria-hidden="true" tabindex="-1"></a><span class="co"># hummingbirds and swifts?</span></span>
<span id="cb361-5"><a href="phylogenetics-in-r.html#cb361-5" aria-hidden="true" tabindex="-1"></a><span class="fu">is.monophyletic</span>(bird.orders, <span class="fu">c</span>(<span class="st">&quot;Trochiliformes&quot;</span>, <span class="st">&quot;Apodiformes&quot;</span>))</span>
<span id="cb361-6"><a href="phylogenetics-in-r.html#cb361-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] TRUE</span></span></code></pre></div>
<p>If we want to look at just the Neoaves, we can subset our tree using <code>extract.clade()</code>. We need to supply a node from our tree to <code>extract.clade</code>, so let’s find the correct node first. The nodes in the tree can be found by running the <code>nodelabels()</code> function after using <code>plot()</code>:</p>
<div class="sourceCode" id="cb362"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb362-1"><a href="phylogenetics-in-r.html#cb362-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(bird.orders, <span class="at">no.margin =</span> <span class="cn">TRUE</span>)</span>
<span id="cb362-2"><a href="phylogenetics-in-r.html#cb362-2" aria-hidden="true" tabindex="-1"></a><span class="fu">segments</span>(<span class="dv">38</span>, <span class="dv">1</span>, <span class="dv">38</span>, <span class="dv">5</span>, <span class="at">lwd =</span> <span class="dv">2</span>)</span>
<span id="cb362-3"><a href="phylogenetics-in-r.html#cb362-3" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="dv">39</span>, <span class="dv">3</span>, <span class="st">&quot;Proaves&quot;</span>, <span class="at">srt =</span> <span class="dv">270</span>)</span>
<span id="cb362-4"><a href="phylogenetics-in-r.html#cb362-4" aria-hidden="true" tabindex="-1"></a><span class="fu">segments</span>(<span class="dv">38</span>, <span class="dv">6</span>, <span class="dv">38</span>, <span class="dv">23</span>, <span class="at">lwd =</span> <span class="dv">2</span>)</span>
<span id="cb362-5"><a href="phylogenetics-in-r.html#cb362-5" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="dv">39</span>, <span class="fl">14.5</span>, <span class="st">&quot;Neoaves&quot;</span>, <span class="at">srt =</span> <span class="dv">270</span>)</span>
<span id="cb362-6"><a href="phylogenetics-in-r.html#cb362-6" aria-hidden="true" tabindex="-1"></a><span class="fu">nodelabels</span>()</span></code></pre></div>
<p><img src="Exercise9_files/figure-html/unnamed-chunk-11-1.png" width="672" /></p>
<p>We can see that the Neoaves start at node 29, so let’s extract that one.</p>
<div class="sourceCode" id="cb363"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb363-1"><a href="phylogenetics-in-r.html#cb363-1" aria-hidden="true" tabindex="-1"></a><span class="co"># extract clade</span></span>
<span id="cb363-2"><a href="phylogenetics-in-r.html#cb363-2" aria-hidden="true" tabindex="-1"></a>neoaves <span class="ot">&lt;-</span> <span class="fu">extract.clade</span>(bird.orders, <span class="dv">29</span>)</span>
<span id="cb363-3"><a href="phylogenetics-in-r.html#cb363-3" aria-hidden="true" tabindex="-1"></a><span class="co"># plot</span></span>
<span id="cb363-4"><a href="phylogenetics-in-r.html#cb363-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(neoaves, <span class="at">no.margin =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p><img src="Exercise9_files/figure-html/unnamed-chunk-12-1.png" width="672" /></p>
<p>The functions provided by <code>ape</code> make it quite easy to handle phylogenies in R, feel free to experiment further to find out what you can do!</p>
</div>
<div id="constructing-trees-with-r" class="section level3 hasAnchor" number="9.1.4">
<h3><span class="header-section-number">9.1.4</span> Constructing trees with R<a href="phylogenetics-in-r.html#constructing-trees-with-r" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>So far, we have only looked at examples of trees that are already constructed in some way. However, if you are working with your own data, this is not the case - you need to actually make the tree yourself. Luckily, <code>phangorn</code> is ideally suited for this. We will use some data, bundled with the package, for the next steps. The following code loads the data:</p>
<div class="yellow">
<div class="sourceCode" id="cb364"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb364-1"><a href="phylogenetics-in-r.html#cb364-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get phangorn primates data</span></span>
<span id="cb364-2"><a href="phylogenetics-in-r.html#cb364-2" aria-hidden="true" tabindex="-1"></a>fdir <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&quot;extdata/trees&quot;</span>, <span class="at">package =</span> <span class="st">&quot;phangorn&quot;</span>)</span>
<span id="cb364-3"><a href="phylogenetics-in-r.html#cb364-3" aria-hidden="true" tabindex="-1"></a>primates <span class="ot">&lt;-</span> <span class="fu">read.dna</span>(<span class="fu">file.path</span>(fdir, <span class="st">&quot;primates.dna&quot;</span>), <span class="at">format =</span> <span class="st">&quot;interleaved&quot;</span>)</span></code></pre></div>
</div>
<p>This is a set of 14 mitochondrial DNA sequences from 12 primate species and 2 outgroups - a mouse and a cow. The sequences are 232 basepairs long. The data is originally from <a href="https://academic.oup.com/mbe/article/5/6/626/1044336">this paper</a> and is a well-known example dataset in phylogenetics.</p>
<p>We have seen the structure this data is stored in before - it is a <code>DNA.bin</code> object like we worked with in <a href="ch07.html#ch07">Chapter 7</a>.</p>
<p>Print <code>primates</code> to your screen and have a look at it. For the next section, we will use just four species - the hominidae (i.e. Orangutan, Gorilla, Chimpanzee and Human). Let’s subset our data in order to do that.</p>
<div class="sourceCode" id="cb365"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb365-1"><a href="phylogenetics-in-r.html#cb365-1" aria-hidden="true" tabindex="-1"></a><span class="co"># subset data to get hominidae</span></span>
<span id="cb365-2"><a href="phylogenetics-in-r.html#cb365-2" aria-hidden="true" tabindex="-1"></a>hominidae <span class="ot">&lt;-</span> primates[<span class="dv">11</span><span class="sc">:</span><span class="dv">14</span>, ]</span></code></pre></div>
<p>We also need to convert our dataset so that <code>phangorn</code> is able to use it properly. The package uses a data structure called <code>phyDAT</code>. Luckily conversion is very easy indeed:</p>
<div class="sourceCode" id="cb366"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb366-1"><a href="phylogenetics-in-r.html#cb366-1" aria-hidden="true" tabindex="-1"></a><span class="co"># convert data</span></span>
<span id="cb366-2"><a href="phylogenetics-in-r.html#cb366-2" aria-hidden="true" tabindex="-1"></a>hominidae <span class="ot">&lt;-</span> <span class="fu">as.phyDat</span>(hominidae)</span></code></pre></div>
<p>We are going to create two types of trees - UPGMA and Neighbour Joining. These are distance based measures and so we must first make a distance matrix among our taxa, which requires a substitution model. The default substitution model is the Jukes &amp; Cantor model, but we can also use Felsenstein’s 1981 model. Which is the best to apply here? To find that out, we should first test the different models using <code>modelTest</code>:</p>
<div class="sourceCode" id="cb367"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb367-1"><a href="phylogenetics-in-r.html#cb367-1" aria-hidden="true" tabindex="-1"></a><span class="co"># perform model selection</span></span>
<span id="cb367-2"><a href="phylogenetics-in-r.html#cb367-2" aria-hidden="true" tabindex="-1"></a>hominidae_mt <span class="ot">&lt;-</span> <span class="fu">modelTest</span>(hominidae, <span class="at">model =</span> <span class="fu">c</span>(<span class="st">&quot;JC&quot;</span>, <span class="st">&quot;F81&quot;</span>), <span class="at">G =</span> <span class="cn">FALSE</span>, <span class="at">I =</span> <span class="cn">FALSE</span>)</span>
<span id="cb367-3"><a href="phylogenetics-in-r.html#cb367-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] &quot;JC&quot;</span></span>
<span id="cb367-4"><a href="phylogenetics-in-r.html#cb367-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] &quot;F81&quot;</span></span></code></pre></div>
<p>Take a look at the <code>hominidae_mt</code> table. What we have done here is performed a maximum likelihood analysis and a form of model selection to determine which of the two models we tested - JC69 and F81 (specified by <code>model = c("JC", "F81")</code>) best fits our data. We also set <code>G</code> and <code>I</code> to false in order to simplify the output. Don’t worry too much about what these are for now, but feel free to use <code>?modelTest</code> if you wish to learn more.</p>
<p>Anyway, how can we interpret this table? Well, we are looking for the model with the <strong>log likelihood</strong> closest to zero and also the lowest value of AIC (Akaike information criterion - <a href="https://en.wikipedia.org/wiki/Akaike_information_criterion">see here for more information</a>). In this case, it is clear that F81 is a better fit for the data than the JC model, so we will calculate our distance matrix with this model instead.</p>
<p>We can now calculate evolutionary distance using <code>dist.ml</code> - a function that compares pairwise distances among sequences the substitution model we chose.</p>
<div class="sourceCode" id="cb368"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb368-1"><a href="phylogenetics-in-r.html#cb368-1" aria-hidden="true" tabindex="-1"></a><span class="co"># first generate a distance matrix</span></span>
<span id="cb368-2"><a href="phylogenetics-in-r.html#cb368-2" aria-hidden="true" tabindex="-1"></a>hominidae_dist <span class="ot">&lt;-</span> <span class="fu">dist.ml</span>(hominidae, <span class="at">model =</span> <span class="st">&quot;F81&quot;</span>)</span></code></pre></div>
<p>Take a look at <code>hominidae_dist</code>. You will see it is a matrix of the distance or difference between the sequences. The distances are based on the number of nucleotide substitutions, and the actual values depend on the model we use—here, the F81. It is not straightforward to interpret how much the groups differ from the numbers directly, but in general, the larger the number the greater the genetic distance.</p>
<p>Next we can create our trees. For an UPGMA tree, we use the <code>upgma</code> function:</p>
<div class="sourceCode" id="cb369"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb369-1"><a href="phylogenetics-in-r.html#cb369-1" aria-hidden="true" tabindex="-1"></a><span class="co"># upgma tree</span></span>
<span id="cb369-2"><a href="phylogenetics-in-r.html#cb369-2" aria-hidden="true" tabindex="-1"></a>hom_upgma <span class="ot">&lt;-</span> <span class="fu">upgma</span>(hominidae_dist)</span></code></pre></div>
<p>Next we will make a neighbour joining tree. This is easily done with the <code>NJ</code> function.</p>
<div class="sourceCode" id="cb370"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb370-1"><a href="phylogenetics-in-r.html#cb370-1" aria-hidden="true" tabindex="-1"></a><span class="co"># NJ tree</span></span>
<span id="cb370-2"><a href="phylogenetics-in-r.html#cb370-2" aria-hidden="true" tabindex="-1"></a>hom_nj <span class="ot">&lt;-</span> <span class="fu">NJ</span>(hominidae_dist)</span></code></pre></div>
<p>Now that we have created both of our trees, we should plot them to have a look at them.</p>
<div class="sourceCode" id="cb371"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb371-1"><a href="phylogenetics-in-r.html#cb371-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot them both</span></span>
<span id="cb371-2"><a href="phylogenetics-in-r.html#cb371-2" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">1</span>)) <span class="co"># 2 plots in same window</span></span>
<span id="cb371-3"><a href="phylogenetics-in-r.html#cb371-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(hom_upgma, <span class="at">no.margin =</span> <span class="cn">TRUE</span>)</span>
<span id="cb371-4"><a href="phylogenetics-in-r.html#cb371-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(hom_nj, <span class="at">type =</span> <span class="st">&quot;unrooted&quot;</span>, <span class="at">no.margin =</span> <span class="cn">TRUE</span>)</span>
<span id="cb371-5"><a href="phylogenetics-in-r.html#cb371-5" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>)) <span class="co"># reset mfrow</span></span></code></pre></div>
<p><img src="Exercise9_files/figure-html/unnamed-chunk-20-1.png" width="672" /></p>
<p>Note that when we plot the NJ tree, we add an extra argument to get an unrooted tree. The default in R is to plot rooted trees, but since the neighbour joining algorithm produces an unrooted phylogeny, the correct way to plot it is unrooted.</p>
<p>We can verify that the tree is unrooted (compared to the UPGMA tree) using the <code>is.rooted()</code> function.</p>
<div class="sourceCode" id="cb372"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb372-1"><a href="phylogenetics-in-r.html#cb372-1" aria-hidden="true" tabindex="-1"></a><span class="co"># check whether the tree is rooted</span></span>
<span id="cb372-2"><a href="phylogenetics-in-r.html#cb372-2" aria-hidden="true" tabindex="-1"></a><span class="fu">is.rooted</span>(hom_nj)</span>
<span id="cb372-3"><a href="phylogenetics-in-r.html#cb372-3" aria-hidden="true" tabindex="-1"></a><span class="fu">is.rooted</span>(hom_upgma)</span></code></pre></div>
<p>We can also set a root on our tree, if we know what we should set the outgroup to. In this case, we can set our outgroup to Orangutan, because we know it is the most divergent from the clade that consists of humans, chimps and gorillas.</p>
<p>We will set the root of our neighbour joining tree below using the <code>root</code> function and we’ll then plot it to see how it looks.</p>
<div class="sourceCode" id="cb373"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb373-1"><a href="phylogenetics-in-r.html#cb373-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot nj rooted</span></span>
<span id="cb373-2"><a href="phylogenetics-in-r.html#cb373-2" aria-hidden="true" tabindex="-1"></a>hom_nj_r <span class="ot">&lt;-</span> <span class="fu">root</span>(hom_nj, <span class="st">&quot;Orang&quot;</span>)</span>
<span id="cb373-3"><a href="phylogenetics-in-r.html#cb373-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(hom_nj_r, <span class="at">no.margin =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p><img src="Exercise9_files/figure-html/unnamed-chunk-22-1.png" width="672" /></p>
<p>In this case, it hasn’t actually made a huge difference to our tree topology, but with a larger dataset, it might do.</p>
<p>As a final point here, we might want to try and compare our two trees and see which we should accept as the best model for the evolutionary relationships among our taxa. One way to do this is to use the <strong>parsimony score</strong> of a phylogeny. Essentially, the lower the parsimony score is for a tree, the more parsimonious explanation of the data it might be. This is very easy to achieve with the <code>parsimony</code> function.</p>
<div class="sourceCode" id="cb374"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb374-1"><a href="phylogenetics-in-r.html#cb374-1" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate parsimony</span></span>
<span id="cb374-2"><a href="phylogenetics-in-r.html#cb374-2" aria-hidden="true" tabindex="-1"></a><span class="fu">parsimony</span>(hom_upgma, hominidae)</span>
<span id="cb374-3"><a href="phylogenetics-in-r.html#cb374-3" aria-hidden="true" tabindex="-1"></a><span class="fu">parsimony</span>(hom_nj_r, hominidae)</span></code></pre></div>
<p>For the <code>parsimony</code> function, the first argument is the tree, the second is the data. Here we can see that both parsimony scores are equal for the two trees, suggesting that they are both equivalent models of the evolutionary relationships among the taxa we are studying here.</p>
<p>If you test the parsimony score for the rooted and the unrooted NJ tree, you will see that they are the same. It is important to note that this is not usually the case! Choosing an outgroup will normally change the tree length, and therefore the parsimony score.</p>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="ch09.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="population-structure.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
