<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>9.1 Phylogenetics in R | R tutorials in BIOS1140</title>
  <meta name="description" content="This is a set of R tutorials connected to the course BIOS1140 at the University of Oslo, and are meant as a companion to the book Evolutionary Genetics (Ravinet &amp; Sætre, 2018). They do not assume any prior R knowledge." />
  <meta name="generator" content="bookdown 0.21 and GitBook 2.6.7" />

  <meta property="og:title" content="9.1 Phylogenetics in R | R tutorials in BIOS1140" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="This is a set of R tutorials connected to the course BIOS1140 at the University of Oslo, and are meant as a companion to the book Evolutionary Genetics (Ravinet &amp; Sætre, 2018). They do not assume any prior R knowledge." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="9.1 Phylogenetics in R | R tutorials in BIOS1140" />
  
  <meta name="twitter:description" content="This is a set of R tutorials connected to the course BIOS1140 at the University of Oslo, and are meant as a companion to the book Evolutionary Genetics (Ravinet &amp; Sætre, 2018). They do not assume any prior R knowledge." />
  



<meta name="date" content="2021-06-05" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="reconstructing-the-past.html"/>
<link rel="next" href="population-structure.html"/>
<script src="libs/header-attrs-2.6/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">R tutorials for the course BIOS1140</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a></li>
<li class="chapter" data-level="1" data-path="introduction-to-r.html"><a href="introduction-to-r.html"><i class="fa fa-check"></i><b>1</b> Introduction to R</a>
<ul>
<li class="chapter" data-level="" data-path="introduction-to-r.html"><a href="introduction-to-r.html#what-to-expect"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="1.1" data-path="start-using-r.html"><a href="start-using-r.html"><i class="fa fa-check"></i><b>1.1</b> Start using R!</a>
<ul>
<li class="chapter" data-level="1.1.1" data-path="start-using-r.html"><a href="start-using-r.html#install-r-and-rstudio"><i class="fa fa-check"></i><b>1.1.1</b> Install R and Rstudio</a></li>
<li class="chapter" data-level="1.1.2" data-path="start-using-r.html"><a href="start-using-r.html#getting-familiar-with-r-and-the-console"><i class="fa fa-check"></i><b>1.1.2</b> Getting familiar with R and the console</a></li>
<li class="chapter" data-level="1.1.3" data-path="start-using-r.html"><a href="start-using-r.html#scripts"><i class="fa fa-check"></i><b>1.1.3</b> Scripts</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="r-essentials.html"><a href="r-essentials.html"><i class="fa fa-check"></i><b>1.2</b> R essentials</a>
<ul>
<li class="chapter" data-level="1.2.1" data-path="r-essentials.html"><a href="r-essentials.html#assigning-to-objects"><i class="fa fa-check"></i><b>1.2.1</b> Assigning to objects</a></li>
<li class="chapter" data-level="1.2.2" data-path="r-essentials.html"><a href="r-essentials.html#vectors"><i class="fa fa-check"></i><b>1.2.2</b> Vectors</a></li>
<li class="chapter" data-level="1.2.3" data-path="r-essentials.html"><a href="r-essentials.html#strings"><i class="fa fa-check"></i><b>1.2.3</b> Strings</a></li>
<li class="chapter" data-level="1.2.4" data-path="r-essentials.html"><a href="r-essentials.html#logical"><i class="fa fa-check"></i><b>1.2.4</b> Logical values</a></li>
<li class="chapter" data-level="1.2.5" data-path="r-essentials.html"><a href="r-essentials.html#functions"><i class="fa fa-check"></i><b>1.2.5</b> Functions</a></li>
<li class="chapter" data-level="1.2.6" data-path="r-essentials.html"><a href="r-essentials.html#data-frames"><i class="fa fa-check"></i><b>1.2.6</b> Data frames</a></li>
</ul></li>
<li class="chapter" data-level="1.3" data-path="dataimport.html"><a href="dataimport.html"><i class="fa fa-check"></i><b>1.3</b> Data import</a>
<ul>
<li class="chapter" data-level="1.3.1" data-path="dataimport.html"><a href="dataimport.html#data-formats"><i class="fa fa-check"></i><b>1.3.1</b> Data formats</a></li>
<li class="chapter" data-level="1.3.2" data-path="dataimport.html"><a href="dataimport.html#working-directories"><i class="fa fa-check"></i><b>1.3.2</b> Working directories</a></li>
<li class="chapter" data-level="1.3.3" data-path="dataimport.html"><a href="dataimport.html#importing"><i class="fa fa-check"></i><b>1.3.3</b> Importing</a></li>
</ul></li>
<li class="chapter" data-level="1.4" data-path="plotting.html"><a href="plotting.html"><i class="fa fa-check"></i><b>1.4</b> Plotting</a>
<ul>
<li class="chapter" data-level="1.4.1" data-path="plotting.html"><a href="plotting.html#plotting-vectors"><i class="fa fa-check"></i><b>1.4.1</b> Plotting vectors</a></li>
<li class="chapter" data-level="1.4.2" data-path="plotting.html"><a href="plotting.html#customizing-your-plots"><i class="fa fa-check"></i><b>1.4.2</b> Customizing your plots</a></li>
</ul></li>
<li class="chapter" data-level="1.5" data-path="installing-and-loading-packages.html"><a href="installing-and-loading-packages.html"><i class="fa fa-check"></i><b>1.5</b> Installing and loading packages</a></li>
<li class="chapter" data-level="1.6" data-path="going-further.html"><a href="going-further.html"><i class="fa fa-check"></i><b>1.6</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="building-on-your-foundations-going-further-with-r.html"><a href="building-on-your-foundations-going-further-with-r.html"><i class="fa fa-check"></i><b>2</b> Building on your foundations: going further with R</a>
<ul>
<li class="chapter" data-level="" data-path="building-on-your-foundations-going-further-with-r.html"><a href="building-on-your-foundations-going-further-with-r.html#what-to-expect-1"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="2.1" data-path="intro-to-data-manipulation-with-tidyverse.html"><a href="intro-to-data-manipulation-with-tidyverse.html"><i class="fa fa-check"></i><b>2.1</b> Intro to data manipulation with tidyverse</a>
<ul>
<li class="chapter" data-level="2.1.1" data-path="intro-to-data-manipulation-with-tidyverse.html"><a href="intro-to-data-manipulation-with-tidyverse.html#what-is-the-tidyverse"><i class="fa fa-check"></i><b>2.1.1</b> What is the tidyverse?</a></li>
<li class="chapter" data-level="2.1.2" data-path="intro-to-data-manipulation-with-tidyverse.html"><a href="intro-to-data-manipulation-with-tidyverse.html#the-dplyr-package"><i class="fa fa-check"></i><b>2.1.2</b> The dplyr package</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="using-dplyr-to-work-with-your-data.html"><a href="using-dplyr-to-work-with-your-data.html"><i class="fa fa-check"></i><b>2.2</b> Using dplyr to work with your data</a>
<ul>
<li class="chapter" data-level="2.2.1" data-path="using-dplyr-to-work-with-your-data.html"><a href="using-dplyr-to-work-with-your-data.html#the-pipe"><i class="fa fa-check"></i><b>2.2.1</b> The pipe</a></li>
<li class="chapter" data-level="2.2.2" data-path="using-dplyr-to-work-with-your-data.html"><a href="using-dplyr-to-work-with-your-data.html#selecting-colums-with-select"><i class="fa fa-check"></i><b>2.2.2</b> Selecting colums with <code>select()</code></a></li>
<li class="chapter" data-level="2.2.3" data-path="using-dplyr-to-work-with-your-data.html"><a href="using-dplyr-to-work-with-your-data.html#filtering-colums-using-filter"><i class="fa fa-check"></i><b>2.2.3</b> Filtering colums using <code>filter()</code></a></li>
<li class="chapter" data-level="2.2.4" data-path="using-dplyr-to-work-with-your-data.html"><a href="using-dplyr-to-work-with-your-data.html#creating-new-columns-with-mutate"><i class="fa fa-check"></i><b>2.2.4</b> Creating new columns with <code>mutate()</code></a></li>
<li class="chapter" data-level="2.2.5" data-path="using-dplyr-to-work-with-your-data.html"><a href="using-dplyr-to-work-with-your-data.html#grouped-summaries-with-group_by-and-summarise"><i class="fa fa-check"></i><b>2.2.5</b> Grouped summaries with <code>group_by()</code> and <code>summarise()</code></a></li>
<li class="chapter" data-level="2.2.6" data-path="using-dplyr-to-work-with-your-data.html"><a href="using-dplyr-to-work-with-your-data.html#using-everything-weve-learned-in-a-single-pipe-and-a-dplyr-exercise"><i class="fa fa-check"></i><b>2.2.6</b> Using everything we’ve learned in a single pipe, and a <code>dplyr</code> exercise</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="plotting-your-data-with-ggplot2.html"><a href="plotting-your-data-with-ggplot2.html"><i class="fa fa-check"></i><b>2.3</b> Plotting your data with <code>ggplot2</code></a>
<ul>
<li class="chapter" data-level="2.3.1" data-path="plotting-your-data-with-ggplot2.html"><a href="plotting-your-data-with-ggplot2.html#the-three-things-you-need-in-a-ggplot"><i class="fa fa-check"></i><b>2.3.1</b> The three things you need in a ggplot</a></li>
<li class="chapter" data-level="2.3.2" data-path="plotting-your-data-with-ggplot2.html"><a href="plotting-your-data-with-ggplot2.html#storing-ggplots-in-objects"><i class="fa fa-check"></i><b>2.3.2</b> Storing ggplots in objects</a></li>
<li class="chapter" data-level="2.3.3" data-path="plotting-your-data-with-ggplot2.html"><a href="plotting-your-data-with-ggplot2.html#customizing-your-plots-1"><i class="fa fa-check"></i><b>2.3.3</b> Customizing your plots</a></li>
<li class="chapter" data-level="2.3.4" data-path="plotting-your-data-with-ggplot2.html"><a href="plotting-your-data-with-ggplot2.html#more-ggplot-examples"><i class="fa fa-check"></i><b>2.3.4</b> More ggplot examples</a></li>
<li class="chapter" data-level="2.3.5" data-path="plotting-your-data-with-ggplot2.html"><a href="plotting-your-data-with-ggplot2.html#saving-your-plots"><i class="fa fa-check"></i><b>2.3.5</b> Saving your plots</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="reshaping-data-with-pivot-longer.html"><a href="reshaping-data-with-pivot-longer.html"><i class="fa fa-check"></i><b>2.4</b> Reshaping data with pivot_longer()</a>
<ul>
<li class="chapter" data-level="2.4.1" data-path="reshaping-data-with-pivot-longer.html"><a href="reshaping-data-with-pivot-longer.html#wide-and-long-format"><i class="fa fa-check"></i><b>2.4.1</b> Wide and long format</a></li>
<li class="chapter" data-level="2.4.2" data-path="reshaping-data-with-pivot-longer.html"><a href="reshaping-data-with-pivot-longer.html#import-example-data"><i class="fa fa-check"></i><b>2.4.2</b> Import example data</a></li>
<li class="chapter" data-level="2.4.3" data-path="reshaping-data-with-pivot-longer.html"><a href="reshaping-data-with-pivot-longer.html#reshape-the-data"><i class="fa fa-check"></i><b>2.4.3</b> Reshape the data</a></li>
<li class="chapter" data-level="2.4.4" data-path="reshaping-data-with-pivot-longer.html"><a href="reshaping-data-with-pivot-longer.html#plot-the-data"><i class="fa fa-check"></i><b>2.4.4</b> Plot the data</a></li>
</ul></li>
<li class="chapter" data-level="2.5" data-path="going-further-1.html"><a href="going-further-1.html"><i class="fa fa-check"></i><b>2.5</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="changes-in-allele-and-genotype-frequency.html"><a href="changes-in-allele-and-genotype-frequency.html"><i class="fa fa-check"></i><b>3</b> Changes in Allele and Genotype Frequency</a>
<ul>
<li class="chapter" data-level="" data-path="changes-in-allele-and-genotype-frequency.html"><a href="changes-in-allele-and-genotype-frequency.html#what-to-expect-2"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="3.1" data-path="r-programming-for-loops.html"><a href="r-programming-for-loops.html"><i class="fa fa-check"></i><b>3.1</b> R programming: for-loops</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="r-programming-for-loops.html"><a href="r-programming-for-loops.html#motivation-why-loop"><i class="fa fa-check"></i><b>3.1.1</b> Motivation: why loop?</a></li>
<li class="chapter" data-level="3.1.2" data-path="r-programming-for-loops.html"><a href="r-programming-for-loops.html#how-a-for-loop-works"><i class="fa fa-check"></i><b>3.1.2</b> How a for-loop works</a></li>
<li class="chapter" data-level="3.1.3" data-path="r-programming-for-loops.html"><a href="r-programming-for-loops.html#indexing"><i class="fa fa-check"></i><b>3.1.3</b> Indexing with for-loops</a></li>
<li class="chapter" data-level="3.1.4" data-path="r-programming-for-loops.html"><a href="r-programming-for-loops.html#solving-our-problem"><i class="fa fa-check"></i><b>3.1.4</b> Solving our problem</a></li>
<li class="chapter" data-level="3.1.5" data-path="r-programming-for-loops.html"><a href="r-programming-for-loops.html#storing-values-from-a-for-loop"><i class="fa fa-check"></i><b>3.1.5</b> Storing values from a for-loop</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="evolutionary-biology.html"><a href="evolutionary-biology.html"><i class="fa fa-check"></i><b>3.2</b> Evolutionary biology</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="evolutionary-biology.html"><a href="evolutionary-biology.html#the-hardy-weinberg-model"><i class="fa fa-check"></i><b>3.2.1</b> The Hardy-Weinberg Model</a></li>
<li class="chapter" data-level="3.2.2" data-path="evolutionary-biology.html"><a href="evolutionary-biology.html#testing-for-deviations-from-the-hardy-weinberg-expectation"><i class="fa fa-check"></i><b>3.2.2</b> Testing for deviations from the Hardy Weinberg Expectation</a></li>
<li class="chapter" data-level="3.2.3" data-path="evolutionary-biology.html"><a href="evolutionary-biology.html#simulating-genetic-drift"><i class="fa fa-check"></i><b>3.2.3</b> Simulating genetic drift</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="study-questions.html"><a href="study-questions.html"><i class="fa fa-check"></i><b>3.3</b> Study questions</a></li>
<li class="chapter" data-level="3.4" data-path="going-further-2.html"><a href="going-further-2.html"><i class="fa fa-check"></i><b>3.4</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="the-theory-of-natural-selection.html"><a href="the-theory-of-natural-selection.html"><i class="fa fa-check"></i><b>4</b> The Theory of Natural Selection</a>
<ul>
<li class="chapter" data-level="" data-path="the-theory-of-natural-selection.html"><a href="the-theory-of-natural-selection.html#what-to-expect-3"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="" data-path="the-theory-of-natural-selection.html"><a href="the-theory-of-natural-selection.html#getting-started"><i class="fa fa-check"></i>Getting started</a></li>
<li class="chapter" data-level="4.1" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html"><i class="fa fa-check"></i><b>4.1</b> R programming: Making custom functions</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#motivation"><i class="fa fa-check"></i><b>4.1.1</b> Motivation</a></li>
<li class="chapter" data-level="4.1.2" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#function-basics"><i class="fa fa-check"></i><b>4.1.2</b> Function basics</a></li>
<li class="chapter" data-level="4.1.3" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#a-simple-example"><i class="fa fa-check"></i><b>4.1.3</b> A simple example</a></li>
<li class="chapter" data-level="4.1.4" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#some-important-function-properties"><i class="fa fa-check"></i><b>4.1.4</b> Some important function properties</a></li>
<li class="chapter" data-level="4.1.5" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#a-slightly-more-useful-example"><i class="fa fa-check"></i><b>4.1.5</b> A slightly more useful example</a></li>
<li class="chapter" data-level="4.1.6" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#example-calculating-genotype-frequencies"><i class="fa fa-check"></i><b>4.1.6</b> Example: calculating genotype frequencies</a></li>
<li class="chapter" data-level="4.1.7" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#creating-a-function-of-the-drift-simulation"><i class="fa fa-check"></i><b>4.1.7</b> Creating a function of the drift simulation</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="evolutionary-biology-fitness.html"><a href="evolutionary-biology-fitness.html"><i class="fa fa-check"></i><b>4.2</b> Evolutionary biology: fitness</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="evolutionary-biology-fitness.html"><a href="evolutionary-biology-fitness.html#understanding-fitness"><i class="fa fa-check"></i><b>4.2.1</b> Understanding fitness</a></li>
<li class="chapter" data-level="4.2.2" data-path="evolutionary-biology-fitness.html"><a href="evolutionary-biology-fitness.html#one-locus-model-of-viability-selection"><i class="fa fa-check"></i><b>4.2.2</b> One-locus model of viability selection</a></li>
<li class="chapter" data-level="4.2.3" data-path="evolutionary-biology-fitness.html"><a href="evolutionary-biology-fitness.html#directional-selection"><i class="fa fa-check"></i><b>4.2.3</b> Directional selection</a></li>
<li class="chapter" data-level="4.2.4" data-path="evolutionary-biology-fitness.html"><a href="evolutionary-biology-fitness.html#over-and-underdominance"><i class="fa fa-check"></i><b>4.2.4</b> Over and underdominance</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="study-questions-1.html"><a href="study-questions-1.html"><i class="fa fa-check"></i><b>4.3</b> Study questions</a></li>
<li class="chapter" data-level="4.4" data-path="going-further-3.html"><a href="going-further-3.html"><i class="fa fa-check"></i><b>4.4</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="detecting-natural-selection.html"><a href="detecting-natural-selection.html"><i class="fa fa-check"></i><b>5</b> Detecting Natural Selection</a>
<ul>
<li class="chapter" data-level="" data-path="detecting-natural-selection.html"><a href="detecting-natural-selection.html#what-to-expect-4"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="" data-path="detecting-natural-selection.html"><a href="detecting-natural-selection.html#getting-started-1"><i class="fa fa-check"></i>Getting started</a></li>
<li class="chapter" data-level="5.1" data-path="r-programming-vectorisation.html"><a href="r-programming-vectorisation.html"><i class="fa fa-check"></i><b>5.1</b> R-programming: Vectorisation</a>
<ul>
<li class="chapter" data-level="5.1.1" data-path="r-programming-vectorisation.html"><a href="r-programming-vectorisation.html#more-on-functions-vectorisation"><i class="fa fa-check"></i><b>5.1.1</b> More on functions: vectorisation</a></li>
<li class="chapter" data-level="5.1.2" data-path="r-programming-vectorisation.html"><a href="r-programming-vectorisation.html#the-apply-function"><i class="fa fa-check"></i><b>5.1.2</b> The <code>apply()</code> function</a></li>
<li class="chapter" data-level="5.1.3" data-path="r-programming-vectorisation.html"><a href="r-programming-vectorisation.html#ifelse"><i class="fa fa-check"></i><b>5.1.3</b> The <code>ifelse()</code> function</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="evolutionary-biology-fst.html"><a href="evolutionary-biology-fst.html"><i class="fa fa-check"></i><b>5.2</b> Evolutionary biology: F<sub>ST</sub></a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="evolutionary-biology-fst.html"><a href="evolutionary-biology-fst.html#understanding-fst---the-fixation-index"><i class="fa fa-check"></i><b>5.2.1</b> Understanding <em>F</em><sub>ST</sub> - the fixation index</a></li>
<li class="chapter" data-level="5.2.2" data-path="evolutionary-biology-fst.html"><a href="evolutionary-biology-fst.html#visualising-fst-along-a-chromosome"><i class="fa fa-check"></i><b>5.2.2</b> Visualising <em>F</em><sub>ST</sub> along a chromosome</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="study-questions-2.html"><a href="study-questions-2.html"><i class="fa fa-check"></i><b>5.3</b> Study questions</a></li>
<li class="chapter" data-level="5.4" data-path="going-further-4.html"><a href="going-further-4.html"><i class="fa fa-check"></i><b>5.4</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="multilocus-evolution.html"><a href="multilocus-evolution.html"><i class="fa fa-check"></i><b>6</b> Multilocus Evolution</a>
<ul>
<li class="chapter" data-level="" data-path="multilocus-evolution.html"><a href="multilocus-evolution.html#what-to-expect-5"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="" data-path="multilocus-evolution.html"><a href="multilocus-evolution.html#getting-started-2"><i class="fa fa-check"></i>Getting started</a></li>
<li class="chapter" data-level="6.1" data-path="tools-for-investigating-differences-between-groups.html"><a href="tools-for-investigating-differences-between-groups.html"><i class="fa fa-check"></i><b>6.1</b> Tools for investigating differences between groups</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="tools-for-investigating-differences-between-groups.html"><a href="tools-for-investigating-differences-between-groups.html#exploring-group-differences"><i class="fa fa-check"></i><b>6.1.1</b> Exploring group differences</a></li>
<li class="chapter" data-level="6.1.2" data-path="tools-for-investigating-differences-between-groups.html"><a href="tools-for-investigating-differences-between-groups.html#analysis-of-variance-anova"><i class="fa fa-check"></i><b>6.1.2</b> Analysis of variance (ANOVA)</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="performing-a-qtl-analysis-in-bedbugs.html"><a href="performing-a-qtl-analysis-in-bedbugs.html"><i class="fa fa-check"></i><b>6.2</b> Performing a QTL analysis in bedbugs</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="performing-a-qtl-analysis-in-bedbugs.html"><a href="performing-a-qtl-analysis-in-bedbugs.html#bedbugs-pesticide-resistance-and-study-design"><i class="fa fa-check"></i><b>6.2.1</b> Bedbugs, pesticide resistance and study design</a></li>
<li class="chapter" data-level="6.2.2" data-path="performing-a-qtl-analysis-in-bedbugs.html"><a href="performing-a-qtl-analysis-in-bedbugs.html#linking-anova-to-our-qtl-analysis"><i class="fa fa-check"></i><b>6.2.2</b> Linking ANOVA to our QTL analysis</a></li>
<li class="chapter" data-level="6.2.3" data-path="performing-a-qtl-analysis-in-bedbugs.html"><a href="performing-a-qtl-analysis-in-bedbugs.html#testing-the-same-genotype-phenotype-association-with-anova"><i class="fa fa-check"></i><b>6.2.3</b> Testing the same genotype phenotype association with ANOVA</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="study-questions-3.html"><a href="study-questions-3.html"><i class="fa fa-check"></i><b>6.3</b> Study questions</a></li>
<li class="chapter" data-level="6.4" data-path="going-further-5.html"><a href="going-further-5.html"><i class="fa fa-check"></i><b>6.4</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="inferring-evolutionary-processes-from-sequence-data.html"><a href="inferring-evolutionary-processes-from-sequence-data.html"><i class="fa fa-check"></i><b>7</b> Inferring Evolutionary Processes from Sequence Data</a>
<ul>
<li class="chapter" data-level="" data-path="inferring-evolutionary-processes-from-sequence-data.html"><a href="inferring-evolutionary-processes-from-sequence-data.html#what-to-expect-6"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="" data-path="inferring-evolutionary-processes-from-sequence-data.html"><a href="inferring-evolutionary-processes-from-sequence-data.html#getting-started-3"><i class="fa fa-check"></i>Getting started</a></li>
<li class="chapter" data-level="7.1" data-path="working-with-dna-sequence-data.html"><a href="working-with-dna-sequence-data.html"><i class="fa fa-check"></i><b>7.1</b> Working with DNA sequence data</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="working-with-dna-sequence-data.html"><a href="working-with-dna-sequence-data.html#reading-sequence-data-into-r."><i class="fa fa-check"></i><b>7.1.1</b> Reading sequence data into R.</a></li>
<li class="chapter" data-level="7.1.2" data-path="working-with-dna-sequence-data.html"><a href="working-with-dna-sequence-data.html#exploring-dna-sequence-data"><i class="fa fa-check"></i><b>7.1.2</b> Exploring DNA sequence data</a></li>
<li class="chapter" data-level="7.1.3" data-path="working-with-dna-sequence-data.html"><a href="working-with-dna-sequence-data.html#calculating-basic-sequence-statistics"><i class="fa fa-check"></i><b>7.1.3</b> Calculating basic sequence statistics</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="working-with-a-larger-dataset.html"><a href="working-with-a-larger-dataset.html"><i class="fa fa-check"></i><b>7.2</b> Working with a larger dataset</a>
<ul>
<li class="chapter" data-level="7.2.1" data-path="working-with-a-larger-dataset.html"><a href="working-with-a-larger-dataset.html#sample-size-and-sequence-statistics"><i class="fa fa-check"></i><b>7.2.1</b> Sample size and sequence statistics</a></li>
<li class="chapter" data-level="7.2.2" data-path="working-with-a-larger-dataset.html"><a href="working-with-a-larger-dataset.html#inferring-evolutionary-processes-using-tajimas-d"><i class="fa fa-check"></i><b>7.2.2</b> Inferring evolutionary processes using Tajima’s <em>D</em></a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="calculating-statistics-at-the-whole-genome-level.html"><a href="calculating-statistics-at-the-whole-genome-level.html"><i class="fa fa-check"></i><b>7.3</b> Calculating statistics at the whole genome level</a>
<ul>
<li class="chapter" data-level="7.3.1" data-path="calculating-statistics-at-the-whole-genome-level.html"><a href="calculating-statistics-at-the-whole-genome-level.html#reading-in-variant-data"><i class="fa fa-check"></i><b>7.3.1</b> Reading in variant data</a></li>
<li class="chapter" data-level="7.3.2" data-path="calculating-statistics-at-the-whole-genome-level.html"><a href="calculating-statistics-at-the-whole-genome-level.html#calculating-nucleotide-diversity-statistics"><i class="fa fa-check"></i><b>7.3.2</b> Calculating nucleotide diversity statistics</a></li>
<li class="chapter" data-level="7.3.3" data-path="calculating-statistics-at-the-whole-genome-level.html"><a href="calculating-statistics-at-the-whole-genome-level.html#visualising-nucleotide-diversity-along-the-chromosome"><i class="fa fa-check"></i><b>7.3.3</b> Visualising nucleotide diversity along the chromosome</a></li>
<li class="chapter" data-level="7.3.4" data-path="calculating-statistics-at-the-whole-genome-level.html"><a href="calculating-statistics-at-the-whole-genome-level.html#performing-a-sliding-window-analysis"><i class="fa fa-check"></i><b>7.3.4</b> Performing a sliding window analysis</a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="study-questions-4.html"><a href="study-questions-4.html"><i class="fa fa-check"></i><b>7.4</b> Study questions</a></li>
<li class="chapter" data-level="7.5" data-path="going-further-6.html"><a href="going-further-6.html"><i class="fa fa-check"></i><b>7.5</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="speciation-genomics.html"><a href="speciation-genomics.html"><i class="fa fa-check"></i><b>8</b> Speciation Genomics</a>
<ul>
<li class="chapter" data-level="" data-path="speciation-genomics.html"><a href="speciation-genomics.html#what-to-expect-7"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="" data-path="speciation-genomics.html"><a href="speciation-genomics.html#getting-started-4"><i class="fa fa-check"></i>Getting started</a></li>
<li class="chapter" data-level="8.1" data-path="r-programming-visualizing-complex-data.html"><a href="r-programming-visualizing-complex-data.html"><i class="fa fa-check"></i><b>8.1</b> R programming: Visualizing complex data</a>
<ul>
<li class="chapter" data-level="8.1.1" data-path="r-programming-visualizing-complex-data.html"><a href="r-programming-visualizing-complex-data.html#faceting"><i class="fa fa-check"></i><b>8.1.1</b> Faceting</a></li>
</ul></li>
<li class="chapter" data-level="8.2" data-path="evolutionary-genetics-returning-to-the-sparrow-dataset.html"><a href="evolutionary-genetics-returning-to-the-sparrow-dataset.html"><i class="fa fa-check"></i><b>8.2</b> Evolutionary genetics: Returning to the sparrow dataset</a>
<ul>
<li class="chapter" data-level="8.2.1" data-path="evolutionary-genetics-returning-to-the-sparrow-dataset.html"><a href="evolutionary-genetics-returning-to-the-sparrow-dataset.html#reading-in-the-sparrow-vcf"><i class="fa fa-check"></i><b>8.2.1</b> Reading in the sparrow vcf</a></li>
<li class="chapter" data-level="8.2.2" data-path="evolutionary-genetics-returning-to-the-sparrow-dataset.html"><a href="evolutionary-genetics-returning-to-the-sparrow-dataset.html#examining-the-variant-data"><i class="fa fa-check"></i><b>8.2.2</b> Examining the variant data</a></li>
<li class="chapter" data-level="8.2.3" data-path="evolutionary-genetics-returning-to-the-sparrow-dataset.html"><a href="evolutionary-genetics-returning-to-the-sparrow-dataset.html#setting-up-sliding-windows"><i class="fa fa-check"></i><b>8.2.3</b> Setting up sliding windows</a></li>
<li class="chapter" data-level="8.2.4" data-path="evolutionary-genetics-returning-to-the-sparrow-dataset.html"><a href="evolutionary-genetics-returning-to-the-sparrow-dataset.html#calculating-sliding-window-estimates-of-nucleotide-diversity-and-differentiation"><i class="fa fa-check"></i><b>8.2.4</b> Calculating sliding window estimates of nucleotide diversity and differentiation</a></li>
<li class="chapter" data-level="8.2.5" data-path="evolutionary-genetics-returning-to-the-sparrow-dataset.html"><a href="evolutionary-genetics-returning-to-the-sparrow-dataset.html#visualising-the-data"><i class="fa fa-check"></i><b>8.2.5</b> Visualising the data</a></li>
<li class="chapter" data-level="8.2.6" data-path="evolutionary-genetics-returning-to-the-sparrow-dataset.html"><a href="evolutionary-genetics-returning-to-the-sparrow-dataset.html#interlude-relative-vs.-absolute-measures-of-nucleotide-diversity"><i class="fa fa-check"></i><b>8.2.6</b> Interlude: relative vs. absolute measures of nucleotide diversity</a></li>
<li class="chapter" data-level="8.2.7" data-path="evolutionary-genetics-returning-to-the-sparrow-dataset.html"><a href="evolutionary-genetics-returning-to-the-sparrow-dataset.html#investigating-recombination-rate-variation"><i class="fa fa-check"></i><b>8.2.7</b> Investigating recombination rate variation</a></li>
</ul></li>
<li class="chapter" data-level="8.3" data-path="study-questions-5.html"><a href="study-questions-5.html"><i class="fa fa-check"></i><b>8.3</b> Study questions</a></li>
<li class="chapter" data-level="8.4" data-path="going-further-7.html"><a href="going-further-7.html"><i class="fa fa-check"></i><b>8.4</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="reconstructing-the-past.html"><a href="reconstructing-the-past.html"><i class="fa fa-check"></i><b>9</b> Reconstructing the Past</a>
<ul>
<li class="chapter" data-level="" data-path="reconstructing-the-past.html"><a href="reconstructing-the-past.html#what-to-expect-8"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="" data-path="reconstructing-the-past.html"><a href="reconstructing-the-past.html#getting-started-5"><i class="fa fa-check"></i>Getting started</a></li>
<li class="chapter" data-level="9.1" data-path="phylogenetics-in-r.html"><a href="phylogenetics-in-r.html"><i class="fa fa-check"></i><b>9.1</b> Phylogenetics in R</a>
<ul>
<li class="chapter" data-level="9.1.1" data-path="phylogenetics-in-r.html"><a href="phylogenetics-in-r.html#storing-trees-in-r"><i class="fa fa-check"></i><b>9.1.1</b> Storing trees in R</a></li>
<li class="chapter" data-level="9.1.2" data-path="phylogenetics-in-r.html"><a href="phylogenetics-in-r.html#plotting-trees"><i class="fa fa-check"></i><b>9.1.2</b> Plotting trees</a></li>
<li class="chapter" data-level="9.1.3" data-path="phylogenetics-in-r.html"><a href="phylogenetics-in-r.html#manipulating-trees"><i class="fa fa-check"></i><b>9.1.3</b> Manipulating trees</a></li>
<li class="chapter" data-level="9.1.4" data-path="phylogenetics-in-r.html"><a href="phylogenetics-in-r.html#a-simple-example-with-real-data---avian-phylogenetics"><i class="fa fa-check"></i><b>9.1.4</b> A simple example with real data - avian phylogenetics</a></li>
<li class="chapter" data-level="9.1.5" data-path="phylogenetics-in-r.html"><a href="phylogenetics-in-r.html#constructing-trees-with-r"><i class="fa fa-check"></i><b>9.1.5</b> Constructing trees with R</a></li>
</ul></li>
<li class="chapter" data-level="9.2" data-path="population-structure.html"><a href="population-structure.html"><i class="fa fa-check"></i><b>9.2</b> Population structure</a>
<ul>
<li class="chapter" data-level="9.2.1" data-path="population-structure.html"><a href="population-structure.html#village-dogs-as-an-insight-to-dog-domestication"><i class="fa fa-check"></i><b>9.2.1</b> Village dogs as an insight to dog domestication</a></li>
<li class="chapter" data-level="9.2.2" data-path="population-structure.html"><a href="population-structure.html#reading-the-data-into-r"><i class="fa fa-check"></i><b>9.2.2</b> Reading the data into R</a></li>
<li class="chapter" data-level="9.2.3" data-path="population-structure.html"><a href="population-structure.html#performing-a-pca"><i class="fa fa-check"></i><b>9.2.3</b> Performing a PCA</a></li>
<li class="chapter" data-level="9.2.4" data-path="population-structure.html"><a href="population-structure.html#visualising-the-pca"><i class="fa fa-check"></i><b>9.2.4</b> Visualising the PCA</a></li>
</ul></li>
<li class="chapter" data-level="9.3" data-path="study-questions-6.html"><a href="study-questions-6.html"><i class="fa fa-check"></i><b>9.3</b> Study questions</a></li>
<li class="chapter" data-level="9.4" data-path="going-further-8.html"><a href="going-further-8.html"><i class="fa fa-check"></i><b>9.4</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="advancing-further-in-r.html"><a href="advancing-further-in-r.html"><i class="fa fa-check"></i><b>10</b> Advancing Further in R</a>
<ul>
<li class="chapter" data-level="" data-path="advancing-further-in-r.html"><a href="advancing-further-in-r.html#what-to-expect-9"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="" data-path="advancing-further-in-r.html"><a href="advancing-further-in-r.html#getting-started-6"><i class="fa fa-check"></i>Getting started</a></li>
<li class="chapter" data-level="10.1" data-path="advanced-features-of-rstudio.html"><a href="advanced-features-of-rstudio.html"><i class="fa fa-check"></i><b>10.1</b> Advanced features of RStudio</a>
<ul>
<li class="chapter" data-level="10.1.1" data-path="advanced-features-of-rstudio.html"><a href="advanced-features-of-rstudio.html#projects-projects-projects"><i class="fa fa-check"></i><b>10.1.1</b> Projects, projects, projects</a></li>
<li class="chapter" data-level="10.1.2" data-path="advanced-features-of-rstudio.html"><a href="advanced-features-of-rstudio.html#everyone-has-a-history"><i class="fa fa-check"></i><b>10.1.2</b> Everyone has a history</a></li>
<li class="chapter" data-level="10.1.3" data-path="advanced-features-of-rstudio.html"><a href="advanced-features-of-rstudio.html#tab-complete-and-other-hotkeys"><i class="fa fa-check"></i><b>10.1.3</b> Tab complete and other hotkeys</a></li>
<li class="chapter" data-level="10.1.4" data-path="advanced-features-of-rstudio.html"><a href="advanced-features-of-rstudio.html#code-checking"><i class="fa fa-check"></i><b>10.1.4</b> Code checking</a></li>
</ul></li>
<li class="chapter" data-level="10.2" data-path="control-flow-for-else-if-and-ifelse.html"><a href="control-flow-for-else-if-and-ifelse.html"><i class="fa fa-check"></i><b>10.2</b> Control flow - <code>for</code>, <code>else</code>, <code>if</code> and <code>ifelse</code></a>
<ul>
<li class="chapter" data-level="10.2.1" data-path="control-flow-for-else-if-and-ifelse.html"><a href="control-flow-for-else-if-and-ifelse.html#if-and-else"><i class="fa fa-check"></i><b>10.2.1</b> <code>if</code> and <code>else</code></a></li>
<li class="chapter" data-level="10.2.2" data-path="control-flow-for-else-if-and-ifelse.html"><a href="control-flow-for-else-if-and-ifelse.html#ifelse---a-single-line-alternative"><i class="fa fa-check"></i><b>10.2.2</b> <code>ifelse</code> - a single line alternative</a></li>
<li class="chapter" data-level="10.2.3" data-path="control-flow-for-else-if-and-ifelse.html"><a href="control-flow-for-else-if-and-ifelse.html#for-loops"><i class="fa fa-check"></i><b>10.2.3</b> <code>for</code> loops</a></li>
</ul></li>
<li class="chapter" data-level="10.3" data-path="vectorisation.html"><a href="vectorisation.html"><i class="fa fa-check"></i><b>10.3</b> Vectorisation</a>
<ul>
<li class="chapter" data-level="10.3.1" data-path="vectorisation.html"><a href="vectorisation.html#why-is-vectorisation-preferable-in-r"><i class="fa fa-check"></i><b>10.3.1</b> Why is vectorisation preferable in R?</a></li>
<li class="chapter" data-level="10.3.2" data-path="vectorisation.html"><a href="vectorisation.html#using-sapply-instead-of-for-on-vectors"><i class="fa fa-check"></i><b>10.3.2</b> Using <code>sapply</code> instead of <code>for</code> on vectors</a></li>
<li class="chapter" data-level="10.3.3" data-path="vectorisation.html"><a href="vectorisation.html#using-apply-instead-of-for-on-matrices-and-dataframes"><i class="fa fa-check"></i><b>10.3.3</b> Using apply instead of <code>for</code> on matrices and dataframes</a></li>
</ul></li>
<li class="chapter" data-level="10.4" data-path="returning-to-writing-your-own-functions.html"><a href="returning-to-writing-your-own-functions.html"><i class="fa fa-check"></i><b>10.4</b> Returning to writing your own functions</a></li>
<li class="chapter" data-level="10.5" data-path="concluding-remarks.html"><a href="concluding-remarks.html"><i class="fa fa-check"></i><b>10.5</b> Concluding remarks</a></li>
<li class="chapter" data-level="10.6" data-path="study-questions-7.html"><a href="study-questions-7.html"><i class="fa fa-check"></i><b>10.6</b> Study questions</a></li>
<li class="chapter" data-level="10.7" data-path="going-further-9.html"><a href="going-further-9.html"><i class="fa fa-check"></i><b>10.7</b> Going further</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Made with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">R tutorials in BIOS1140</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="phylogenetics-in-r" class="section level2" number="9.1">
<h2><span class="header-section-number">9.1</span> Phylogenetics in R</h2>
<script src="js/hideOutput.js"></script>
<p>R has a number of extremely powerful packages for performing phylogenetic analysis, from plotting trees to testing comparative models of evolution. You can see <a href="https://cran.r-project.org/web/views/Phylogenetics.html">here</a> for more information if you are interested in learning about what sort of things are possible. For today’s session, we will learn how to handle and visualise phylogenetic trees in R. We will also construct a series of trees from a sequence alignment. First, let’s familiarise ourselves with how R handles phylogenetic data.</p>
<div id="storing-trees-in-r" class="section level3" number="9.1.1">
<h3><span class="header-section-number">9.1.1</span> Storing trees in R</h3>
<p>The backbone of most phylogenetic analysis in R comes from the functions that are part of the <code>ape</code> package. <code>ape</code> stores trees as <code>phylo</code> objects, which are easy to access and manipulate. The easiest way to understand this is to have a look at a simple phylogeny, so we’ll create a random tree now.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="phylogenetics-in-r.html#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># set seed to ensure the same tree is produced</span></span>
<span id="cb2-2"><a href="phylogenetics-in-r.html#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">32</span>)</span>
<span id="cb2-3"><a href="phylogenetics-in-r.html#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="co"># generate a tree</span></span>
<span id="cb2-4"><a href="phylogenetics-in-r.html#cb2-4" aria-hidden="true" tabindex="-1"></a>tree <span class="ot">&lt;-</span> <span class="fu">rtree</span>(<span class="at">n =</span> <span class="dv">4</span>, <span class="at">tip.label =</span> <span class="fu">c</span>(<span class="st">&quot;a&quot;</span>, <span class="st">&quot;b&quot;</span>, <span class="st">&quot;c&quot;</span>, <span class="st">&quot;d&quot;</span>))</span></code></pre></div>
<p>What have we done here? First, the <code>set.seed</code> function just sets a seed for our random simulation of a tree. You won’t need to worry about this for the majority of the time, here we are using it to make sure that when we randomly create a tree, we all create the same one.</p>
<p>What you need to focus on is the second line of code that uses the <code>rtree</code> function. This is simply a means to generate a random tree. With the <code>n = 4</code> argument, we are simply stating our tree will have four taxa and we are already specifying what they should be called with the <code>tip.label</code> argument.</p>
<p>Let’s take a closer look at our <code>tree</code> object. It is a <code>phylo</code> object - you can demonstrate this to yourself with <code>class(tree)</code>.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="phylogenetics-in-r.html#cb3-1" aria-hidden="true" tabindex="-1"></a>tree</span>
<span id="cb3-2"><a href="phylogenetics-in-r.html#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb3-3"><a href="phylogenetics-in-r.html#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Phylogenetic tree with 4 tips and 3 internal nodes.</span></span>
<span id="cb3-4"><a href="phylogenetics-in-r.html#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb3-5"><a href="phylogenetics-in-r.html#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Tip labels:</span></span>
<span id="cb3-6"><a href="phylogenetics-in-r.html#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   c, a, d, b</span></span>
<span id="cb3-7"><a href="phylogenetics-in-r.html#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb3-8"><a href="phylogenetics-in-r.html#cb3-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Rooted; includes branch lengths.</span></span></code></pre></div>
<p>By printing <code>tree</code> to the console, we see it is a tree with 4 tips and 3 internal nodes, a set of tip labels. We also see it is rooted and that the branch lengths are stored in this object too.</p>
<p>You can actually look more deeply into the data stored within the <code>tree</code> object if you want to. Try the following code and see what is inside.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="phylogenetics-in-r.html#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(tree)</span>
<span id="cb4-2"><a href="phylogenetics-in-r.html#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="fu">objects</span>(tree)</span>
<span id="cb4-3"><a href="phylogenetics-in-r.html#cb4-3" aria-hidden="true" tabindex="-1"></a>tree<span class="sc">$</span>edge</span>
<span id="cb4-4"><a href="phylogenetics-in-r.html#cb4-4" aria-hidden="true" tabindex="-1"></a>tree<span class="sc">$</span>edge.length</span></code></pre></div>
<p>It is of course, much easier to understand a tree when we visualise it. Luckily this is easy in R.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="phylogenetics-in-r.html#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(tree)</span></code></pre></div>
<p><img src="Exercise9_files/figure-html/unnamed-chunk-6-1.png" width="672" /></p>
<p>In the next section, we will learn more about how to plot trees.</p>
</div>
<div id="plotting-trees" class="section level3" number="9.1.2">
<h3><span class="header-section-number">9.1.2</span> Plotting trees</h3>
<p><em>In this section, we show various ways to control the appearance of phylogenetic trees. The plots are hidden by default to avoid extensive scrolling, click “show plot” to see them! Or better, run the code yourself, try to change some arguments and see what happens.</em></p>
<script src="js/hideOutput.js"></script>
<p>We can actually do a lot with our trees in R using a few simple plot commands. Let’s generate another random tree, this time with 5 taxa.</p>
<div class="fold o">
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="phylogenetics-in-r.html#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># set seed to ensure the same tree is produced</span></span>
<span id="cb6-2"><a href="phylogenetics-in-r.html#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">32</span>)</span>
<span id="cb6-3"><a href="phylogenetics-in-r.html#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="co"># generate a tree</span></span>
<span id="cb6-4"><a href="phylogenetics-in-r.html#cb6-4" aria-hidden="true" tabindex="-1"></a>tree <span class="ot">&lt;-</span> <span class="fu">rtree</span>(<span class="at">n =</span> <span class="dv">5</span>, <span class="at">tip.label =</span> <span class="fu">c</span>(<span class="st">&quot;a&quot;</span>, <span class="st">&quot;b&quot;</span>, <span class="st">&quot;c&quot;</span>, <span class="st">&quot;d&quot;</span>, <span class="st">&quot;e&quot;</span>))</span></code></pre></div>
</div>
<p>First of all, let’s plot our new tree using the default plot settings. Except unlike last time, we will suppress the branch lenghts.</p>
<div class="fold o">
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="phylogenetics-in-r.html#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot tree and suppress branch lengths</span></span>
<span id="cb7-2"><a href="phylogenetics-in-r.html#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(tree, <span class="at">use.edge.length =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p><img src="Exercise9_files/figure-html/unnamed-chunk-8-1.png" width="672" /></p>
</div>
<p>All we did here was set the <code>use.edge.length</code> argument to false, suppressing the branch lengths. The standar tree plot that <code>ape</code> creates is a <strong>phylogram</strong>. However, there are a number of other different plotting methods we can use.</p>
<p>We can plot a <strong>cladogram</strong> for example.</p>
<div class="fold o">
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="phylogenetics-in-r.html#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot cladogram</span></span>
<span id="cb8-2"><a href="phylogenetics-in-r.html#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(tree, <span class="at">type =</span> <span class="st">&quot;cladogram&quot;</span>, <span class="at">use.edge.length =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p><img src="Exercise9_files/figure-html/unnamed-chunk-9-1.png" width="672" /></p>
</div>
<p>We can plot our tree as if it had no root.</p>
<div class="fold o">
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="phylogenetics-in-r.html#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot unrooted</span></span>
<span id="cb9-2"><a href="phylogenetics-in-r.html#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(tree, <span class="at">type =</span> <span class="st">&quot;unrooted&quot;</span>, <span class="at">use.edge.length =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p><img src="Exercise9_files/figure-html/unnamed-chunk-10-1.png" width="672" /></p>
</div>
<p>We can also make a radial tree, which is a convenient way to plot many taxa at once.</p>
<div class="fold o">
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="phylogenetics-in-r.html#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot fan/radial treee</span></span>
<span id="cb10-2"><a href="phylogenetics-in-r.html#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(tree, <span class="at">type =</span> <span class="st">&quot;fan&quot;</span>, <span class="at">use.edge.length =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p><img src="Exercise9_files/figure-html/unnamed-chunk-11-1.png" width="672" /></p>
</div>
<p>We can also easily set many different graphical parameters to make our tree look a specific way. So for example</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="phylogenetics-in-r.html#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot red tree (with suppressed branch lengths)</span></span>
<span id="cb11-2"><a href="phylogenetics-in-r.html#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(tree, <span class="at">use.edge.length =</span> <span class="cn">FALSE</span>, <span class="at">edge.width =</span> <span class="dv">2</span>, <span class="at">edge.color =</span> <span class="st">&quot;red&quot;</span>)</span></code></pre></div>
<p><img src="Exercise9_files/figure-html/unnamed-chunk-12-1.png" width="672" /></p>
</div>
<div id="manipulating-trees" class="section level3" number="9.1.3">
<h3><span class="header-section-number">9.1.3</span> Manipulating trees</h3>
<p>As well as simply plotting trees, it is possible to manipulate them to make them look a certain way or to focus on a specific group of taxa. <code>ape</code> has a number of powerful functions that allow us to do this quite easily. To start, let’s make a new random tree, this time with 5 taxa.</p>
<div class="fold o">
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="phylogenetics-in-r.html#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># set seed (to ensure the same tree is drawn for all)</span></span>
<span id="cb12-2"><a href="phylogenetics-in-r.html#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">32</span>)</span>
<span id="cb12-3"><a href="phylogenetics-in-r.html#cb12-3" aria-hidden="true" tabindex="-1"></a><span class="co"># make a new tree with 5 tips - one is root</span></span>
<span id="cb12-4"><a href="phylogenetics-in-r.html#cb12-4" aria-hidden="true" tabindex="-1"></a>tree <span class="ot">&lt;-</span> <span class="fu">rtree</span>(<span class="at">n =</span> <span class="dv">5</span>, <span class="at">tip.label =</span> <span class="fu">c</span>(<span class="st">&quot;a&quot;</span>, <span class="st">&quot;b&quot;</span>, <span class="st">&quot;c&quot;</span>, <span class="st">&quot;d&quot;</span>, <span class="st">&quot;e&quot;</span>))</span>
<span id="cb12-5"><a href="phylogenetics-in-r.html#cb12-5" aria-hidden="true" tabindex="-1"></a><span class="co"># plot the tree</span></span>
<span id="cb12-6"><a href="phylogenetics-in-r.html#cb12-6" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(tree)</span></code></pre></div>
<p><img src="Exercise9_files/figure-html/unnamed-chunk-13-1.png" width="672" /></p>
</div>
<p>Let’s say we want to remove a tip from our tree - we can easily achieve this with the <code>drop.tip</code> function, and supplying the name of the tip.</p>
<div class="fold o">
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="phylogenetics-in-r.html#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># drop a tip </span></span>
<span id="cb13-2"><a href="phylogenetics-in-r.html#cb13-2" aria-hidden="true" tabindex="-1"></a>prune_tree <span class="ot">&lt;-</span> <span class="fu">drop.tip</span>(tree, <span class="st">&quot;d&quot;</span>)</span>
<span id="cb13-3"><a href="phylogenetics-in-r.html#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(prune_tree)</span></code></pre></div>
<p><img src="Exercise9_files/figure-html/unnamed-chunk-14-1.png" width="672" /></p>
</div>
<p>Perhaps we want to rotate our tree, so that the tips are orientated slightly differently? This is also very easily achieved with the <code>rotate</code> function.</p>
<div class="fold o">
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="phylogenetics-in-r.html#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># rotate tree</span></span>
<span id="cb14-2"><a href="phylogenetics-in-r.html#cb14-2" aria-hidden="true" tabindex="-1"></a>prune_tree1 <span class="ot">&lt;-</span> <span class="fu">rotate</span>(prune_tree, <span class="fu">c</span>(<span class="st">&quot;b&quot;</span>, <span class="st">&quot;a&quot;</span>))</span>
<span id="cb14-3"><a href="phylogenetics-in-r.html#cb14-3" aria-hidden="true" tabindex="-1"></a><span class="co"># plot newly rotate tree</span></span>
<span id="cb14-4"><a href="phylogenetics-in-r.html#cb14-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(prune_tree1)</span></code></pre></div>
<p><img src="Exercise9_files/figure-html/unnamed-chunk-15-1.png" width="672" /></p>
</div>
<p>Here, we specified we want the tree rotated on the internal node that joins <code>a</code> and <code>b</code>, which in this case is the root of the tree. What if we want to rotate another set of taxa, such as <code>c</code> and <code>e</code>? Well it is actually much more straightforward to specify the <strong>node</strong> rather than the tip names. But how can you know the names of each of the nodes? Luckily there is an easy way to display them.</p>
<div class="fold o">
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="phylogenetics-in-r.html#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># show nodes on tree</span></span>
<span id="cb15-2"><a href="phylogenetics-in-r.html#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(prune_tree1)</span>
<span id="cb15-3"><a href="phylogenetics-in-r.html#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="fu">nodelabels</span>()</span></code></pre></div>
<p><img src="Exercise9_files/figure-html/unnamed-chunk-16-1.png" width="672" /></p>
</div>
<p>So we want to rotate our tree on node 7. Thus we can rerun the code using <code>rotate</code> but this time with the node specified, not the tips:</p>
<div class="fold o">
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="phylogenetics-in-r.html#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># rotate tree specifying node</span></span>
<span id="cb16-2"><a href="phylogenetics-in-r.html#cb16-2" aria-hidden="true" tabindex="-1"></a>prune_tree2 <span class="ot">&lt;-</span> <span class="fu">rotate</span>(prune_tree1, <span class="dv">7</span>)</span>
<span id="cb16-3"><a href="phylogenetics-in-r.html#cb16-3" aria-hidden="true" tabindex="-1"></a><span class="co"># plot newly rotate tree</span></span>
<span id="cb16-4"><a href="phylogenetics-in-r.html#cb16-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(prune_tree2)</span></code></pre></div>
<p><img src="Exercise9_files/figure-html/unnamed-chunk-17-1.png" width="672" /></p>
</div>
<p>Note that rotating trees just alters the way we plot them, it does not make any difference to the evolutionary relationship among the taxa shown on them.</p>
<div class="fold o">
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="phylogenetics-in-r.html#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># set seed</span></span>
<span id="cb17-2"><a href="phylogenetics-in-r.html#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">32</span>)</span>
<span id="cb17-3"><a href="phylogenetics-in-r.html#cb17-3" aria-hidden="true" tabindex="-1"></a><span class="co"># make a random tree of three taxa</span></span>
<span id="cb17-4"><a href="phylogenetics-in-r.html#cb17-4" aria-hidden="true" tabindex="-1"></a>tree <span class="ot">&lt;-</span> <span class="fu">rtree</span>(<span class="at">n =</span> <span class="dv">3</span>, <span class="at">tip.label =</span> <span class="fu">c</span>(<span class="st">&quot;a&quot;</span>, <span class="st">&quot;b&quot;</span>, <span class="st">&quot;c&quot;</span>))</span>
<span id="cb17-5"><a href="phylogenetics-in-r.html#cb17-5" aria-hidden="true" tabindex="-1"></a><span class="co"># set up plot layout</span></span>
<span id="cb17-6"><a href="phylogenetics-in-r.html#cb17-6" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">1</span>))</span>
<span id="cb17-7"><a href="phylogenetics-in-r.html#cb17-7" aria-hidden="true" tabindex="-1"></a><span class="co"># plot the tree</span></span>
<span id="cb17-8"><a href="phylogenetics-in-r.html#cb17-8" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(tree, <span class="at">use.edge.length =</span> <span class="cn">FALSE</span>, <span class="at">cex =</span> <span class="fl">1.5</span>, <span class="at">no.margin =</span> <span class="cn">TRUE</span>)</span>
<span id="cb17-9"><a href="phylogenetics-in-r.html#cb17-9" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">rotate</span>(tree, <span class="dv">4</span>), <span class="at">use.edge.length =</span> <span class="cn">FALSE</span>, <span class="at">cex =</span> <span class="fl">1.5</span>, <span class="at">no.margin =</span> <span class="cn">TRUE</span>)</span>
<span id="cb17-10"><a href="phylogenetics-in-r.html#cb17-10" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">rotate</span>(tree, <span class="dv">5</span>), <span class="at">use.edge.length =</span> <span class="cn">FALSE</span>, <span class="at">cex =</span> <span class="fl">1.5</span>, <span class="at">no.margin =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p><img src="Exercise9_files/figure-html/unnamed-chunk-18-1.png" width="672" /></p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="phylogenetics-in-r.html#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># set layout back to normal</span></span>
<span id="cb18-2"><a href="phylogenetics-in-r.html#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">1</span>))</span></code></pre></div>
</div>
<p>Note that <code>cex</code> just alters the size of the tip labels and <code>no.margin = TRUE</code> suppresses the use of margins for visualisation. All three of these trees are the same - they are just rotated differently.</p>
</div>
<div id="a-simple-example-with-real-data---avian-phylogenetics" class="section level3" number="9.1.4">
<h3><span class="header-section-number">9.1.4</span> A simple example with real data - avian phylogenetics</h3>
<p>So far, we have only looked at randomly generated trees. Let’s have a look at some data stored within <code>ape</code> - a phylogeny of birds at the order level.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="phylogenetics-in-r.html#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get bird order data</span></span>
<span id="cb19-2"><a href="phylogenetics-in-r.html#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">&quot;bird.orders&quot;</span>)</span></code></pre></div>
<p>Let’s plot the phylogeny to have a look at it. We will also add some annotation to make sense of the phylogeny.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="phylogenetics-in-r.html#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(bird.orders, <span class="at">no.margin =</span> <span class="cn">TRUE</span>)</span>
<span id="cb20-2"><a href="phylogenetics-in-r.html#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="fu">segments</span>(<span class="dv">38</span>, <span class="dv">1</span>, <span class="dv">38</span>, <span class="dv">5</span>, <span class="at">lwd =</span> <span class="dv">2</span>)</span>
<span id="cb20-3"><a href="phylogenetics-in-r.html#cb20-3" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="dv">39</span>, <span class="dv">3</span>, <span class="st">&quot;Proaves&quot;</span>, <span class="at">srt =</span> <span class="dv">270</span>)</span>
<span id="cb20-4"><a href="phylogenetics-in-r.html#cb20-4" aria-hidden="true" tabindex="-1"></a><span class="fu">segments</span>(<span class="dv">38</span>, <span class="dv">6</span>, <span class="dv">38</span>, <span class="dv">23</span>, <span class="at">lwd =</span> <span class="dv">2</span>)</span>
<span id="cb20-5"><a href="phylogenetics-in-r.html#cb20-5" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="dv">39</span>, <span class="fl">14.5</span>, <span class="st">&quot;Neoaves&quot;</span>, <span class="at">srt =</span> <span class="dv">270</span>)</span></code></pre></div>
<p><img src="Exercise9_files/figure-html/unnamed-chunk-20-1.png" width="672" /></p>
<p>Here, the <code>segments</code> and <code>text</code> functions specify the bars and names of the two major groups in our avian phylogeny. We are just using them for display purposes here, but if you’d like to know more about them, you can look at the R help with <code>?segments</code> and <code>?text</code> commands.</p>
<p>Let’s focus on the Neoaves clade for now. Perhaps we want to test whether certain families within Neoaves form a monophyletic group? We can do this with the <code>is.monophyletic</code> function.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="phylogenetics-in-r.html#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Parrots and Passerines?</span></span>
<span id="cb21-2"><a href="phylogenetics-in-r.html#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="fu">is.monophyletic</span>(bird.orders, <span class="fu">c</span>(<span class="st">&quot;Passeriformes&quot;</span>, <span class="st">&quot;Psittaciformes&quot;</span>))</span>
<span id="cb21-3"><a href="phylogenetics-in-r.html#cb21-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] FALSE</span></span>
<span id="cb21-4"><a href="phylogenetics-in-r.html#cb21-4" aria-hidden="true" tabindex="-1"></a><span class="co"># hummingbirds and swifts?</span></span>
<span id="cb21-5"><a href="phylogenetics-in-r.html#cb21-5" aria-hidden="true" tabindex="-1"></a><span class="fu">is.monophyletic</span>(bird.orders, <span class="fu">c</span>(<span class="st">&quot;Trochiliformes&quot;</span>, <span class="st">&quot;Apodiformes&quot;</span>))</span>
<span id="cb21-6"><a href="phylogenetics-in-r.html#cb21-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] TRUE</span></span></code></pre></div>
<p>We can also easily subset our tree using <code>extract.clade</code>. Let’s extract the Neoaves and then plot them. We need to supply the correct node to <code>extract.clade</code>, so let’s first find the correct node by plotting nodelabels on top of the previous tree.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="phylogenetics-in-r.html#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(bird.orders, <span class="at">no.margin =</span> <span class="cn">TRUE</span>)</span>
<span id="cb22-2"><a href="phylogenetics-in-r.html#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="fu">segments</span>(<span class="dv">38</span>, <span class="dv">1</span>, <span class="dv">38</span>, <span class="dv">5</span>, <span class="at">lwd =</span> <span class="dv">2</span>)</span>
<span id="cb22-3"><a href="phylogenetics-in-r.html#cb22-3" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="dv">39</span>, <span class="dv">3</span>, <span class="st">&quot;Proaves&quot;</span>, <span class="at">srt =</span> <span class="dv">270</span>)</span>
<span id="cb22-4"><a href="phylogenetics-in-r.html#cb22-4" aria-hidden="true" tabindex="-1"></a><span class="fu">segments</span>(<span class="dv">38</span>, <span class="dv">6</span>, <span class="dv">38</span>, <span class="dv">23</span>, <span class="at">lwd =</span> <span class="dv">2</span>)</span>
<span id="cb22-5"><a href="phylogenetics-in-r.html#cb22-5" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="dv">39</span>, <span class="fl">14.5</span>, <span class="st">&quot;Neoaves&quot;</span>, <span class="at">srt =</span> <span class="dv">270</span>)</span>
<span id="cb22-6"><a href="phylogenetics-in-r.html#cb22-6" aria-hidden="true" tabindex="-1"></a><span class="fu">nodelabels</span>()</span></code></pre></div>
<p><img src="Exercise9_files/figure-html/unnamed-chunk-22-1.png" width="672" /></p>
<p>We can see that the Neoaves start at node 29, so let’s extract that one.</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="phylogenetics-in-r.html#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co"># extract clade</span></span>
<span id="cb23-2"><a href="phylogenetics-in-r.html#cb23-2" aria-hidden="true" tabindex="-1"></a>neoaves <span class="ot">&lt;-</span> <span class="fu">extract.clade</span>(bird.orders, <span class="dv">29</span>)</span>
<span id="cb23-3"><a href="phylogenetics-in-r.html#cb23-3" aria-hidden="true" tabindex="-1"></a><span class="co"># plot</span></span>
<span id="cb23-4"><a href="phylogenetics-in-r.html#cb23-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(neoaves)</span></code></pre></div>
<p><img src="Exercise9_files/figure-html/unnamed-chunk-23-1.png" width="672" /></p>
<p>The functions provided by <code>ape</code> make it quite easy to handle phylogenies in R, feel free to experiment further to find out what you can do!</p>
</div>
<div id="constructing-trees-with-r" class="section level3" number="9.1.5">
<h3><span class="header-section-number">9.1.5</span> Constructing trees with R</h3>
<p>So far, we have only looked at examples of trees that are already constructed in some way. However, if you are working with your own data, this is not the case - you need to actually make the tree yourself. Luckily, <code>phangorn</code> is ideally suited for this. We will use some data, bundled with the package, for the next steps. The following code loads the data:</p>
<div class="yellow">
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="phylogenetics-in-r.html#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get phangorn primates data</span></span>
<span id="cb24-2"><a href="phylogenetics-in-r.html#cb24-2" aria-hidden="true" tabindex="-1"></a>fdir <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&quot;extdata/trees&quot;</span>, <span class="at">package =</span> <span class="st">&quot;phangorn&quot;</span>)</span>
<span id="cb24-3"><a href="phylogenetics-in-r.html#cb24-3" aria-hidden="true" tabindex="-1"></a>primates <span class="ot">&lt;-</span> <span class="fu">read.dna</span>(<span class="fu">file.path</span>(fdir, <span class="st">&quot;primates.dna&quot;</span>), <span class="at">format =</span> <span class="st">&quot;interleaved&quot;</span>)</span></code></pre></div>
</div>
<p>This is a set of 14 mitochondrial DNA sequences from 12 primate species and 2 outgroups - a mouse and a cow. The sequences are 232 basepairs long. The data is originally from <a href="https://academic.oup.com/mbe/article/5/6/626/1044336">this paper</a> and is a well-known example dataset in phylogenetics.</p>
<p>We have seen the structure this data is stored in before - it is a <code>DNA.bin</code> object like we worked with in <a href="https://evolutionarygenetics.github.io/Chapter7.html">Chapter 7</a>.</p>
<p>Print <code>primates</code> to your screen and have a look at it. For the next section, we will use just four species - the hominidae (i.e. Orangutan, Gorilla, Chimpanzee and Human). Let’s subset our data in order to do that.</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="phylogenetics-in-r.html#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co"># subset data to get hominidae</span></span>
<span id="cb25-2"><a href="phylogenetics-in-r.html#cb25-2" aria-hidden="true" tabindex="-1"></a>hominidae <span class="ot">&lt;-</span> primates[<span class="dv">11</span><span class="sc">:</span><span class="dv">14</span>, ]</span></code></pre></div>
<p>We also need to convert our dataset so that <code>phangorn</code> is able to use it properly. The package uses a data structure called <code>phyDAT</code>. Luckily conversion is very easy indeed:</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="phylogenetics-in-r.html#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co"># convert data</span></span>
<span id="cb26-2"><a href="phylogenetics-in-r.html#cb26-2" aria-hidden="true" tabindex="-1"></a>hominidae <span class="ot">&lt;-</span> <span class="fu">as.phyDat</span>(hominidae)</span></code></pre></div>
<div class="blue">
<p><strong>Droppe?</strong><br />
We are going to create two types of trees - UPGMA and Neighbour Joining. These are distance based measures and so we must first make a distance matrix among our taxa, which requires a substitution model. The default substitution model is the Jukes &amp; Cantor model, but we can also use Felsenstein’s 1981 model. Which is the best to apply here? To find that out, we should first test the different models using <code>modelTest</code>:</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="phylogenetics-in-r.html#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="co"># perform model selection</span></span>
<span id="cb27-2"><a href="phylogenetics-in-r.html#cb27-2" aria-hidden="true" tabindex="-1"></a>hominidae_mt <span class="ot">&lt;-</span> <span class="fu">modelTest</span>(hominidae, <span class="at">model =</span> <span class="fu">c</span>(<span class="st">&quot;JC&quot;</span>, <span class="st">&quot;F81&quot;</span>), <span class="at">G =</span> <span class="cn">FALSE</span>, <span class="at">I =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p>Take a look at the <code>hominidae_mt</code> table. What we have done here is performed a maximum likelihood analysis and a form of model selection to determine which of the two models we tested - JC69 and F81 (specified by <code>model = c("JC", "F81")</code>) best fits our data. We also set <code>G</code> and <code>I</code> to false in order to simplify the output. Don’t worry too much about what these are for now, but feel free to use <code>?modelTest</code> if you wish to learn more.</p>
<p>Anyway, how can we interpret this table? Well, we are looking for the model with the ** log likelihood** closest to zero and also the lowest value of AIC (Akaike information criterion - <a href="https://en.wikipedia.org/wiki/Akaike_information_criterion">see here for more information</a>). In this case, it is clear that F81 is a better fit for the data than the JC model, so we will calculate our distance matrix with this model instead.</p>
</div>
<p>We can now calculate evolutionary distance using <code>dist.ml</code> - a function that compares pairwise distances among sequences the substitution model we chose.</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="phylogenetics-in-r.html#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="co"># first generate a distance matrix</span></span>
<span id="cb28-2"><a href="phylogenetics-in-r.html#cb28-2" aria-hidden="true" tabindex="-1"></a>hominidae_dist <span class="ot">&lt;-</span> <span class="fu">dist.ml</span>(hominidae, <span class="at">model =</span> <span class="st">&quot;F81&quot;</span>)</span></code></pre></div>
<p>Take a look at <code>hominidae_mt</code> - you will see it is a matrix of the distance between the sequences - i.e. in terms of the number of nucleotide substitions. Next we can create our trees. For an UPGMA tree, we use the <code>upgma</code> function:</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="phylogenetics-in-r.html#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="co"># upgma tree</span></span>
<span id="cb29-2"><a href="phylogenetics-in-r.html#cb29-2" aria-hidden="true" tabindex="-1"></a>hom_upgma <span class="ot">&lt;-</span> <span class="fu">upgma</span>(hominidae_dist)</span></code></pre></div>
<p>Next we will make a neighbour joining tree. This is easily done with the <code>NJ</code> function.</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="phylogenetics-in-r.html#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="co"># NJ tree</span></span>
<span id="cb30-2"><a href="phylogenetics-in-r.html#cb30-2" aria-hidden="true" tabindex="-1"></a>hom_nj <span class="ot">&lt;-</span> <span class="fu">NJ</span>(hominidae_dist)</span></code></pre></div>
<p>Now that we have created both of our trees, we should plot them to have a look at them.</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="phylogenetics-in-r.html#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot them both</span></span>
<span id="cb31-2"><a href="phylogenetics-in-r.html#cb31-2" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">1</span>))</span>
<span id="cb31-3"><a href="phylogenetics-in-r.html#cb31-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(hom_upgma, <span class="at">no.margin =</span> <span class="cn">TRUE</span>)</span>
<span id="cb31-4"><a href="phylogenetics-in-r.html#cb31-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(hom_nj, <span class="at">no.margin =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p><img src="Exercise9_files/figure-html/unnamed-chunk-31-1.png" width="672" /></p>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="phylogenetics-in-r.html#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))</span></code></pre></div>
<p>However, you will remember that the neighbour joining algorithm produces an unrooted phylogeny. This means the way we plotted it above is incorrect. We can verify that the tree is unrooted (compared to the UPGMA tree) using the <code>is.rooted</code> function.</p>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb33-1"><a href="phylogenetics-in-r.html#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co"># check whether the tree is rooted</span></span>
<span id="cb33-2"><a href="phylogenetics-in-r.html#cb33-2" aria-hidden="true" tabindex="-1"></a><span class="fu">is.rooted</span>(hom_nj)</span>
<span id="cb33-3"><a href="phylogenetics-in-r.html#cb33-3" aria-hidden="true" tabindex="-1"></a><span class="fu">is.rooted</span>(hom_upgma)</span></code></pre></div>
<p>So it makes more sense to plot our tree as unrooted. We do this below.</p>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb34-1"><a href="phylogenetics-in-r.html#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot nj unrooted</span></span>
<span id="cb34-2"><a href="phylogenetics-in-r.html#cb34-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(hom_nj, <span class="at">type =</span> <span class="st">&quot;unrooted&quot;</span>)</span></code></pre></div>
<p><img src="Exercise9_files/figure-html/unnamed-chunk-33-1.png" width="672" /></p>
<p>We can also set a root on our tree, if we know what we should set the outgroup to. In this case, we can set our outgroup to Orangutan, because we know it is the most divergent from the clade that consists of humans, chimps and gorillas.</p>
<p>We will set the root of our neighbour joining tree below using the <code>root</code> function and we’ll then plot it to see how it looks.</p>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb35-1"><a href="phylogenetics-in-r.html#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot nj rooted</span></span>
<span id="cb35-2"><a href="phylogenetics-in-r.html#cb35-2" aria-hidden="true" tabindex="-1"></a>hom_nj_r <span class="ot">&lt;-</span> <span class="fu">root</span>(hom_nj, <span class="st">&quot;Orang&quot;</span>)</span>
<span id="cb35-3"><a href="phylogenetics-in-r.html#cb35-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(hom_nj_r)</span></code></pre></div>
<p><img src="Exercise9_files/figure-html/unnamed-chunk-34-1.png" width="672" /></p>
<p>In this case, it hasn’t actually made a huge difference to our tree topology, but with a larger dataset, it might do.</p>
<p>As a final point here, we might want to try and compare our two trees and see which we should accept as the best model for the evolutionary relationships among our taxa. One way to do this is to use the <strong>parsimony score</strong> of a phylogeny. Essentially, the lower the parsimony score is for a tree, the more parsimonious explanation of the data it might be. This is very easy to achieve with the <code>parsimony</code> function.</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="phylogenetics-in-r.html#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate parsimony</span></span>
<span id="cb36-2"><a href="phylogenetics-in-r.html#cb36-2" aria-hidden="true" tabindex="-1"></a><span class="fu">parsimony</span>(hom_upgma, hominidae)</span>
<span id="cb36-3"><a href="phylogenetics-in-r.html#cb36-3" aria-hidden="true" tabindex="-1"></a><span class="fu">parsimony</span>(hom_nj, hominidae)</span></code></pre></div>
<p>For the <code>parsimony</code> function, the first argument is the tree, the second is the data. Here we can that both parsimony scores are equal for the two trees, suggesting that they are both equivalent models of the evolutionary relationships among the taxa we are studying here.</p>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="reconstructing-the-past.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="population-structure.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
