<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>A About assignments and RMarkdown | R tutorials for the course BIOS1140 at the University of Oslo</title>
  <meta name="description" content="This is a set of R tutorials connected to the course BIOS1140 at the University of Oslo, and are meant as a companion to the book Evolutionary Genetics (Ravinet &amp; Sætre, 2018). They do not assume any prior R knowledge." />
  <meta name="generator" content="bookdown 0.21 and GitBook 2.6.7" />

  <meta property="og:title" content="A About assignments and RMarkdown | R tutorials for the course BIOS1140 at the University of Oslo" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="This is a set of R tutorials connected to the course BIOS1140 at the University of Oslo, and are meant as a companion to the book Evolutionary Genetics (Ravinet &amp; Sætre, 2018). They do not assume any prior R knowledge." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="A About assignments and RMarkdown | R tutorials for the course BIOS1140 at the University of Oslo" />
  
  <meta name="twitter:description" content="This is a set of R tutorials connected to the course BIOS1140 at the University of Oslo, and are meant as a companion to the book Evolutionary Genetics (Ravinet &amp; Sætre, 2018). They do not assume any prior R knowledge." />
  

<meta name="author" content="Original tutorials written by Mark Ravinet. Revamped in 2021 by Even Sletteng Garvang and Solveig Brochmann." />


<meta name="date" content="2021-06-23" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="w10.html"/>

<script src="libs/header-attrs-2.5/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<link href="libs/anchor-sections-1.0/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">R tutorials for the course BIOS1140</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a>
<ul>
<li class="chapter" data-level="" data-path="how-to-use-these-tutorials.html"><a href="how-to-use-these-tutorials.html"><i class="fa fa-check"></i>How to use these tutorials</a></li>
</ul></li>
<li class="part"><span><b>I Introductory R course</b></span></li>
<li class="chapter" data-level="1" data-path="ch01.html"><a href="ch01.html"><i class="fa fa-check"></i><b>1</b> Introduction to R</a>
<ul>
<li class="chapter" data-level="" data-path="ch01.html"><a href="ch01.html#what-to-expect"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="1.1" data-path="start-using-r.html"><a href="start-using-r.html"><i class="fa fa-check"></i><b>1.1</b> Start using R!</a>
<ul>
<li class="chapter" data-level="1.1.1" data-path="start-using-r.html"><a href="start-using-r.html#install-r-and-rstudio"><i class="fa fa-check"></i><b>1.1.1</b> Install R and Rstudio</a></li>
<li class="chapter" data-level="1.1.2" data-path="start-using-r.html"><a href="start-using-r.html#getting-familiar-with-r-and-the-console"><i class="fa fa-check"></i><b>1.1.2</b> Getting familiar with R and the console</a></li>
<li class="chapter" data-level="1.1.3" data-path="start-using-r.html"><a href="start-using-r.html#scripts"><i class="fa fa-check"></i><b>1.1.3</b> Scripts</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="r-essentials.html"><a href="r-essentials.html"><i class="fa fa-check"></i><b>1.2</b> R essentials</a>
<ul>
<li class="chapter" data-level="1.2.1" data-path="r-essentials.html"><a href="r-essentials.html#assigning-to-objects"><i class="fa fa-check"></i><b>1.2.1</b> Assigning to objects</a></li>
<li class="chapter" data-level="1.2.2" data-path="r-essentials.html"><a href="r-essentials.html#vectors"><i class="fa fa-check"></i><b>1.2.2</b> Vectors</a></li>
<li class="chapter" data-level="1.2.3" data-path="r-essentials.html"><a href="r-essentials.html#strings"><i class="fa fa-check"></i><b>1.2.3</b> Strings</a></li>
<li class="chapter" data-level="1.2.4" data-path="r-essentials.html"><a href="r-essentials.html#logical"><i class="fa fa-check"></i><b>1.2.4</b> Logical values</a></li>
<li class="chapter" data-level="1.2.5" data-path="r-essentials.html"><a href="r-essentials.html#functions"><i class="fa fa-check"></i><b>1.2.5</b> Functions</a></li>
<li class="chapter" data-level="1.2.6" data-path="r-essentials.html"><a href="r-essentials.html#data-frames"><i class="fa fa-check"></i><b>1.2.6</b> Data frames</a></li>
</ul></li>
<li class="chapter" data-level="1.3" data-path="dataimport.html"><a href="dataimport.html"><i class="fa fa-check"></i><b>1.3</b> Data import</a>
<ul>
<li class="chapter" data-level="1.3.1" data-path="dataimport.html"><a href="dataimport.html#data-formats"><i class="fa fa-check"></i><b>1.3.1</b> Data formats</a></li>
<li class="chapter" data-level="1.3.2" data-path="dataimport.html"><a href="dataimport.html#working-directories"><i class="fa fa-check"></i><b>1.3.2</b> Working directories</a></li>
<li class="chapter" data-level="1.3.3" data-path="dataimport.html"><a href="dataimport.html#importing"><i class="fa fa-check"></i><b>1.3.3</b> Importing</a></li>
</ul></li>
<li class="chapter" data-level="1.4" data-path="plotting.html"><a href="plotting.html"><i class="fa fa-check"></i><b>1.4</b> Plotting</a>
<ul>
<li class="chapter" data-level="1.4.1" data-path="plotting.html"><a href="plotting.html#plotting-vectors"><i class="fa fa-check"></i><b>1.4.1</b> Plotting vectors</a></li>
<li class="chapter" data-level="1.4.2" data-path="plotting.html"><a href="plotting.html#customizing-your-plots"><i class="fa fa-check"></i><b>1.4.2</b> Customizing your plots</a></li>
</ul></li>
<li class="chapter" data-level="1.5" data-path="installing-and-loading-packages.html"><a href="installing-and-loading-packages.html"><i class="fa fa-check"></i><b>1.5</b> Installing and loading packages</a></li>
<li class="chapter" data-level="1.6" data-path="going-further.html"><a href="going-further.html"><i class="fa fa-check"></i><b>1.6</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="ch02.html"><a href="ch02.html"><i class="fa fa-check"></i><b>2</b> Building on your foundations: going further with R</a>
<ul>
<li class="chapter" data-level="" data-path="ch02.html"><a href="ch02.html#what-to-expect-1"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="2.1" data-path="intro-to-data-manipulation-with-tidyverse.html"><a href="intro-to-data-manipulation-with-tidyverse.html"><i class="fa fa-check"></i><b>2.1</b> Intro to data manipulation with tidyverse</a>
<ul>
<li class="chapter" data-level="2.1.1" data-path="intro-to-data-manipulation-with-tidyverse.html"><a href="intro-to-data-manipulation-with-tidyverse.html#what-is-the-tidyverse"><i class="fa fa-check"></i><b>2.1.1</b> What is the tidyverse?</a></li>
<li class="chapter" data-level="2.1.2" data-path="intro-to-data-manipulation-with-tidyverse.html"><a href="intro-to-data-manipulation-with-tidyverse.html#the-dplyr-package"><i class="fa fa-check"></i><b>2.1.2</b> The dplyr package</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="using-dplyr-to-work-with-your-data.html"><a href="using-dplyr-to-work-with-your-data.html"><i class="fa fa-check"></i><b>2.2</b> Using dplyr to work with your data</a>
<ul>
<li class="chapter" data-level="2.2.1" data-path="using-dplyr-to-work-with-your-data.html"><a href="using-dplyr-to-work-with-your-data.html#the-pipe"><i class="fa fa-check"></i><b>2.2.1</b> The pipe</a></li>
<li class="chapter" data-level="2.2.2" data-path="using-dplyr-to-work-with-your-data.html"><a href="using-dplyr-to-work-with-your-data.html#selecting-colums-with-select"><i class="fa fa-check"></i><b>2.2.2</b> Selecting colums with <code>select()</code></a></li>
<li class="chapter" data-level="2.2.3" data-path="using-dplyr-to-work-with-your-data.html"><a href="using-dplyr-to-work-with-your-data.html#filtering-colums-using-filter"><i class="fa fa-check"></i><b>2.2.3</b> Filtering colums using <code>filter()</code></a></li>
<li class="chapter" data-level="2.2.4" data-path="using-dplyr-to-work-with-your-data.html"><a href="using-dplyr-to-work-with-your-data.html#grouped-summaries-with-group_by-and-summarise"><i class="fa fa-check"></i><b>2.2.4</b> Grouped summaries with <code>group_by()</code> and <code>summarise()</code></a></li>
<li class="chapter" data-level="2.2.5" data-path="using-dplyr-to-work-with-your-data.html"><a href="using-dplyr-to-work-with-your-data.html#using-everything-weve-learned-in-a-single-pipe-and-a-dplyr-exercise"><i class="fa fa-check"></i><b>2.2.5</b> Using everything we’ve learned in a single pipe, and a <code>dplyr</code> exercise</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="plotting-your-data-with-ggplot2.html"><a href="plotting-your-data-with-ggplot2.html"><i class="fa fa-check"></i><b>2.3</b> Plotting your data with <code>ggplot2</code></a>
<ul>
<li class="chapter" data-level="2.3.1" data-path="plotting-your-data-with-ggplot2.html"><a href="plotting-your-data-with-ggplot2.html#the-three-things-you-need-in-a-ggplot"><i class="fa fa-check"></i><b>2.3.1</b> The three things you need in a ggplot</a></li>
<li class="chapter" data-level="2.3.2" data-path="plotting-your-data-with-ggplot2.html"><a href="plotting-your-data-with-ggplot2.html#storing-ggplots-in-objects"><i class="fa fa-check"></i><b>2.3.2</b> Storing ggplots in objects</a></li>
<li class="chapter" data-level="2.3.3" data-path="plotting-your-data-with-ggplot2.html"><a href="plotting-your-data-with-ggplot2.html#customizing-your-plots-1"><i class="fa fa-check"></i><b>2.3.3</b> Customizing your plots</a></li>
<li class="chapter" data-level="2.3.4" data-path="plotting-your-data-with-ggplot2.html"><a href="plotting-your-data-with-ggplot2.html#saving-your-plots"><i class="fa fa-check"></i><b>2.3.4</b> Saving your plots</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="reshaping-data-with-pivot-longer.html"><a href="reshaping-data-with-pivot-longer.html"><i class="fa fa-check"></i><b>2.4</b> Reshaping data with pivot_longer()</a>
<ul>
<li class="chapter" data-level="2.4.1" data-path="reshaping-data-with-pivot-longer.html"><a href="reshaping-data-with-pivot-longer.html#wide-and-long-format"><i class="fa fa-check"></i><b>2.4.1</b> Wide and long format</a></li>
<li class="chapter" data-level="2.4.2" data-path="reshaping-data-with-pivot-longer.html"><a href="reshaping-data-with-pivot-longer.html#import-example-data"><i class="fa fa-check"></i><b>2.4.2</b> Import example data</a></li>
<li class="chapter" data-level="2.4.3" data-path="reshaping-data-with-pivot-longer.html"><a href="reshaping-data-with-pivot-longer.html#reshape-the-data"><i class="fa fa-check"></i><b>2.4.3</b> Reshape the data</a></li>
<li class="chapter" data-level="2.4.4" data-path="reshaping-data-with-pivot-longer.html"><a href="reshaping-data-with-pivot-longer.html#plot-the-data"><i class="fa fa-check"></i><b>2.4.4</b> Plot the data</a></li>
</ul></li>
<li class="chapter" data-level="2.5" data-path="study-questions.html"><a href="study-questions.html"><i class="fa fa-check"></i><b>2.5</b> Study questions</a></li>
<li class="chapter" data-level="2.6" data-path="going-further-1.html"><a href="going-further-1.html"><i class="fa fa-check"></i><b>2.6</b> Going further</a></li>
</ul></li>
<li class="part"><span><b>II Evolutionary genetics with R</b></span></li>
<li class="chapter" data-level="3" data-path="ch03.html"><a href="ch03.html"><i class="fa fa-check"></i><b>3</b> Changes in Allele and Genotype Frequency</a>
<ul>
<li class="chapter" data-level="" data-path="ch03.html"><a href="ch03.html#what-to-expect-2"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="3.1" data-path="r-programming-for-loops.html"><a href="r-programming-for-loops.html"><i class="fa fa-check"></i><b>3.1</b> R programming: for-loops</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="r-programming-for-loops.html"><a href="r-programming-for-loops.html#motivation-why-loop"><i class="fa fa-check"></i><b>3.1.1</b> Motivation: why loop?</a></li>
<li class="chapter" data-level="3.1.2" data-path="r-programming-for-loops.html"><a href="r-programming-for-loops.html#how-a-for-loop-works"><i class="fa fa-check"></i><b>3.1.2</b> How a for-loop works</a></li>
<li class="chapter" data-level="3.1.3" data-path="r-programming-for-loops.html"><a href="r-programming-for-loops.html#indexing"><i class="fa fa-check"></i><b>3.1.3</b> Indexing with for-loops</a></li>
<li class="chapter" data-level="3.1.4" data-path="r-programming-for-loops.html"><a href="r-programming-for-loops.html#solving-our-problem"><i class="fa fa-check"></i><b>3.1.4</b> Solving our problem</a></li>
<li class="chapter" data-level="3.1.5" data-path="r-programming-for-loops.html"><a href="r-programming-for-loops.html#storing-values-from-a-for-loop"><i class="fa fa-check"></i><b>3.1.5</b> Storing values from a for-loop</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="evolutionary-biology.html"><a href="evolutionary-biology.html"><i class="fa fa-check"></i><b>3.2</b> Evolutionary biology</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="evolutionary-biology.html"><a href="evolutionary-biology.html#the-hardy-weinberg-model"><i class="fa fa-check"></i><b>3.2.1</b> The Hardy-Weinberg Model</a></li>
<li class="chapter" data-level="3.2.2" data-path="evolutionary-biology.html"><a href="evolutionary-biology.html#testing-for-deviations-from-the-hardy-weinberg-expectation"><i class="fa fa-check"></i><b>3.2.2</b> Testing for deviations from the Hardy Weinberg Expectation</a></li>
<li class="chapter" data-level="3.2.3" data-path="evolutionary-biology.html"><a href="evolutionary-biology.html#simulating-genetic-drift"><i class="fa fa-check"></i><b>3.2.3</b> Simulating genetic drift</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="study-questions-1.html"><a href="study-questions-1.html"><i class="fa fa-check"></i><b>3.3</b> Study questions</a></li>
<li class="chapter" data-level="3.4" data-path="going-further-2.html"><a href="going-further-2.html"><i class="fa fa-check"></i><b>3.4</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="ch04.html"><a href="ch04.html"><i class="fa fa-check"></i><b>4</b> The Theory of Natural Selection</a>
<ul>
<li class="chapter" data-level="" data-path="ch04.html"><a href="ch04.html#what-to-expect-3"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="" data-path="ch04.html"><a href="ch04.html#getting-started"><i class="fa fa-check"></i>Getting started</a></li>
<li class="chapter" data-level="4.1" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html"><i class="fa fa-check"></i><b>4.1</b> R programming: Making custom functions</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#motivation"><i class="fa fa-check"></i><b>4.1.1</b> Motivation</a></li>
<li class="chapter" data-level="4.1.2" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#function-basics"><i class="fa fa-check"></i><b>4.1.2</b> Function basics</a></li>
<li class="chapter" data-level="4.1.3" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#a-simple-example"><i class="fa fa-check"></i><b>4.1.3</b> A simple example</a></li>
<li class="chapter" data-level="4.1.4" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#some-important-function-properties"><i class="fa fa-check"></i><b>4.1.4</b> Some important function properties</a></li>
<li class="chapter" data-level="4.1.5" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#a-slightly-more-useful-example"><i class="fa fa-check"></i><b>4.1.5</b> A slightly more useful example</a></li>
<li class="chapter" data-level="4.1.6" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#example-calculating-genotype-frequencies"><i class="fa fa-check"></i><b>4.1.6</b> Example: calculating genotype frequencies</a></li>
<li class="chapter" data-level="4.1.7" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#creating-a-function-of-the-drift-simulation"><i class="fa fa-check"></i><b>4.1.7</b> Creating a function of the drift simulation</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="evolutionary-biology-fitness.html"><a href="evolutionary-biology-fitness.html"><i class="fa fa-check"></i><b>4.2</b> Evolutionary biology: fitness</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="evolutionary-biology-fitness.html"><a href="evolutionary-biology-fitness.html#understanding-fitness"><i class="fa fa-check"></i><b>4.2.1</b> Understanding fitness</a></li>
<li class="chapter" data-level="4.2.2" data-path="evolutionary-biology-fitness.html"><a href="evolutionary-biology-fitness.html#one-locus-model-of-viability-selection"><i class="fa fa-check"></i><b>4.2.2</b> One-locus model of viability selection</a></li>
<li class="chapter" data-level="4.2.3" data-path="evolutionary-biology-fitness.html"><a href="evolutionary-biology-fitness.html#over-and-underdominance"><i class="fa fa-check"></i><b>4.2.3</b> Over and underdominance</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="study-questions-2.html"><a href="study-questions-2.html"><i class="fa fa-check"></i><b>4.3</b> Study questions</a></li>
<li class="chapter" data-level="4.4" data-path="going-further-3.html"><a href="going-further-3.html"><i class="fa fa-check"></i><b>4.4</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="ch05.html"><a href="ch05.html"><i class="fa fa-check"></i><b>5</b> Detecting Natural Selection</a>
<ul>
<li class="chapter" data-level="" data-path="ch05.html"><a href="ch05.html#what-to-expect-4"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="" data-path="ch05.html"><a href="ch05.html#getting-started-1"><i class="fa fa-check"></i>Getting started</a></li>
<li class="chapter" data-level="5.1" data-path="vectorisation.html"><a href="vectorisation.html"><i class="fa fa-check"></i><b>5.1</b> More on functions: Vectorisation</a>
<ul>
<li class="chapter" data-level="5.1.1" data-path="vectorisation.html"><a href="vectorisation.html#the-apply-function"><i class="fa fa-check"></i><b>5.1.1</b> The <code>apply()</code> function</a></li>
<li class="chapter" data-level="5.1.2" data-path="vectorisation.html"><a href="vectorisation.html#ifelse"><i class="fa fa-check"></i><b>5.1.2</b> The <code>ifelse()</code> function</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="understanding-fst-the-fixation-index.html"><a href="understanding-fst-the-fixation-index.html"><i class="fa fa-check"></i><b>5.2</b> Understanding <em>F</em><sub>ST</sub> - the fixation index</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="understanding-fst-the-fixation-index.html"><a href="understanding-fst-the-fixation-index.html#a-worked-example-of-fst-in-humans"><i class="fa fa-check"></i><b>5.2.1</b> A worked example of <em>F</em><sub>ST</sub> in humans</a></li>
<li class="chapter" data-level="5.2.2" data-path="understanding-fst-the-fixation-index.html"><a href="understanding-fst-the-fixation-index.html#writing-a-set-of-fst-functions"><i class="fa fa-check"></i><b>5.2.2</b> Writing a set of <em>F</em><sub>ST</sub> functions</a></li>
<li class="chapter" data-level="5.2.3" data-path="understanding-fst-the-fixation-index.html"><a href="understanding-fst-the-fixation-index.html#applying-functions-to-matrices-and-data-frames"><i class="fa fa-check"></i><b>5.2.3</b> Applying functions to matrices and data frames</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="visualising-fst-along-a-chromosome.html"><a href="visualising-fst-along-a-chromosome.html"><i class="fa fa-check"></i><b>5.3</b> Visualising <em>F</em><sub>ST</sub> along a chromosome</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="visualising-fst-along-a-chromosome.html"><a href="visualising-fst-along-a-chromosome.html#identifying-outliers-in-our-fst-distribution"><i class="fa fa-check"></i><b>5.3.1</b> Identifying outliers in our <em>F</em><sub>ST</sub> distribution</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="study-questions-3.html"><a href="study-questions-3.html"><i class="fa fa-check"></i><b>5.4</b> Study questions</a></li>
<li class="chapter" data-level="5.5" data-path="going-further-4.html"><a href="going-further-4.html"><i class="fa fa-check"></i><b>5.5</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="ch06.html"><a href="ch06.html"><i class="fa fa-check"></i><b>6</b> Multilocus Evolution</a>
<ul>
<li class="chapter" data-level="" data-path="ch06.html"><a href="ch06.html#what-to-expect-5"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="" data-path="ch06.html"><a href="ch06.html#getting-started-2"><i class="fa fa-check"></i>Getting started</a></li>
<li class="chapter" data-level="6.1" data-path="tools-for-investigating-differences-between-groups.html"><a href="tools-for-investigating-differences-between-groups.html"><i class="fa fa-check"></i><b>6.1</b> Tools for investigating differences between groups</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="tools-for-investigating-differences-between-groups.html"><a href="tools-for-investigating-differences-between-groups.html#exploring-group-differences"><i class="fa fa-check"></i><b>6.1.1</b> Exploring group differences</a></li>
<li class="chapter" data-level="6.1.2" data-path="tools-for-investigating-differences-between-groups.html"><a href="tools-for-investigating-differences-between-groups.html#analysis-of-variance-anova"><i class="fa fa-check"></i><b>6.1.2</b> Analysis of variance (ANOVA)</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="performing-a-qtl-analysis-in-bedbugs.html"><a href="performing-a-qtl-analysis-in-bedbugs.html"><i class="fa fa-check"></i><b>6.2</b> Performing a QTL analysis in bedbugs</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="performing-a-qtl-analysis-in-bedbugs.html"><a href="performing-a-qtl-analysis-in-bedbugs.html#bedbugs-pesticide-resistance-and-study-design"><i class="fa fa-check"></i><b>6.2.1</b> Bedbugs, pesticide resistance and study design</a></li>
<li class="chapter" data-level="6.2.2" data-path="performing-a-qtl-analysis-in-bedbugs.html"><a href="performing-a-qtl-analysis-in-bedbugs.html#linking-anova-to-our-qtl-analysis"><i class="fa fa-check"></i><b>6.2.2</b> Linking ANOVA to our QTL analysis</a></li>
<li class="chapter" data-level="6.2.3" data-path="performing-a-qtl-analysis-in-bedbugs.html"><a href="performing-a-qtl-analysis-in-bedbugs.html#testing-the-same-genotype-phenotype-association-with-anova"><i class="fa fa-check"></i><b>6.2.3</b> Testing the same genotype phenotype association with ANOVA</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="study-questions-4.html"><a href="study-questions-4.html"><i class="fa fa-check"></i><b>6.3</b> Study questions</a></li>
<li class="chapter" data-level="6.4" data-path="going-further-5.html"><a href="going-further-5.html"><i class="fa fa-check"></i><b>6.4</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="ch07.html"><a href="ch07.html"><i class="fa fa-check"></i><b>7</b> Inferring Evolutionary Processes from Sequence Data</a>
<ul>
<li class="chapter" data-level="" data-path="ch07.html"><a href="ch07.html#what-to-expect-6"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="" data-path="ch07.html"><a href="ch07.html#getting-started-3"><i class="fa fa-check"></i>Getting started</a></li>
<li class="chapter" data-level="7.1" data-path="working-with-dna-sequence-data.html"><a href="working-with-dna-sequence-data.html"><i class="fa fa-check"></i><b>7.1</b> Working with DNA sequence data</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="working-with-dna-sequence-data.html"><a href="working-with-dna-sequence-data.html#reading-sequence-data-into-r."><i class="fa fa-check"></i><b>7.1.1</b> Reading sequence data into R.</a></li>
<li class="chapter" data-level="7.1.2" data-path="working-with-dna-sequence-data.html"><a href="working-with-dna-sequence-data.html#exploring-dna-sequence-data"><i class="fa fa-check"></i><b>7.1.2</b> Exploring DNA sequence data</a></li>
<li class="chapter" data-level="7.1.3" data-path="working-with-dna-sequence-data.html"><a href="working-with-dna-sequence-data.html#calculating-basic-sequence-statistics"><i class="fa fa-check"></i><b>7.1.3</b> Calculating basic sequence statistics</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="working-with-a-larger-dataset.html"><a href="working-with-a-larger-dataset.html"><i class="fa fa-check"></i><b>7.2</b> Working with a larger dataset</a>
<ul>
<li class="chapter" data-level="7.2.1" data-path="working-with-a-larger-dataset.html"><a href="working-with-a-larger-dataset.html#sample-size-and-sequence-statistics"><i class="fa fa-check"></i><b>7.2.1</b> Sample size and sequence statistics</a></li>
<li class="chapter" data-level="7.2.2" data-path="working-with-a-larger-dataset.html"><a href="working-with-a-larger-dataset.html#inferring-evolutionary-processes-using-tajimas-d"><i class="fa fa-check"></i><b>7.2.2</b> Inferring evolutionary processes using Tajima’s <em>D</em></a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="calculating-statistics-at-the-whole-genome-level.html"><a href="calculating-statistics-at-the-whole-genome-level.html"><i class="fa fa-check"></i><b>7.3</b> Calculating statistics at the whole genome level</a>
<ul>
<li class="chapter" data-level="7.3.1" data-path="calculating-statistics-at-the-whole-genome-level.html"><a href="calculating-statistics-at-the-whole-genome-level.html#reading-in-variant-data"><i class="fa fa-check"></i><b>7.3.1</b> Reading in variant data</a></li>
<li class="chapter" data-level="7.3.2" data-path="calculating-statistics-at-the-whole-genome-level.html"><a href="calculating-statistics-at-the-whole-genome-level.html#calculating-nucleotide-diversity-statistics"><i class="fa fa-check"></i><b>7.3.2</b> Calculating nucleotide diversity statistics</a></li>
<li class="chapter" data-level="7.3.3" data-path="calculating-statistics-at-the-whole-genome-level.html"><a href="calculating-statistics-at-the-whole-genome-level.html#visualising-nucleotide-diversity-along-the-chromosome"><i class="fa fa-check"></i><b>7.3.3</b> Visualising nucleotide diversity along the chromosome</a></li>
<li class="chapter" data-level="7.3.4" data-path="calculating-statistics-at-the-whole-genome-level.html"><a href="calculating-statistics-at-the-whole-genome-level.html#performing-a-sliding-window-analysis"><i class="fa fa-check"></i><b>7.3.4</b> Performing a sliding window analysis</a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="study-questions-5.html"><a href="study-questions-5.html"><i class="fa fa-check"></i><b>7.4</b> Study questions</a></li>
<li class="chapter" data-level="7.5" data-path="going-further-6.html"><a href="going-further-6.html"><i class="fa fa-check"></i><b>7.5</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="ch08.html"><a href="ch08.html"><i class="fa fa-check"></i><b>8</b> Speciation Genomics</a>
<ul>
<li class="chapter" data-level="" data-path="ch08.html"><a href="ch08.html#what-to-expect-7"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="" data-path="ch08.html"><a href="ch08.html#getting-started-4"><i class="fa fa-check"></i>Getting started</a></li>
<li class="chapter" data-level="8.1" data-path="visualizing-complex-data.html"><a href="visualizing-complex-data.html"><i class="fa fa-check"></i><b>8.1</b> Visualizing complex data</a>
<ul>
<li class="chapter" data-level="8.1.1" data-path="visualizing-complex-data.html"><a href="visualizing-complex-data.html#faceting"><i class="fa fa-check"></i><b>8.1.1</b> Faceting</a></li>
</ul></li>
<li class="chapter" data-level="8.2" data-path="returning-to-the-sparrow-dataset.html"><a href="returning-to-the-sparrow-dataset.html"><i class="fa fa-check"></i><b>8.2</b> Returning to the sparrow dataset</a>
<ul>
<li class="chapter" data-level="8.2.1" data-path="returning-to-the-sparrow-dataset.html"><a href="returning-to-the-sparrow-dataset.html#reading-in-the-sparrow-vcf"><i class="fa fa-check"></i><b>8.2.1</b> Reading in the sparrow vcf</a></li>
<li class="chapter" data-level="8.2.2" data-path="returning-to-the-sparrow-dataset.html"><a href="returning-to-the-sparrow-dataset.html#examining-the-variant-data"><i class="fa fa-check"></i><b>8.2.2</b> Examining the variant data</a></li>
</ul></li>
<li class="chapter" data-level="8.3" data-path="setting-up-sliding-windows.html"><a href="setting-up-sliding-windows.html"><i class="fa fa-check"></i><b>8.3</b> Setting up sliding windows</a>
<ul>
<li class="chapter" data-level="8.3.1" data-path="setting-up-sliding-windows.html"><a href="setting-up-sliding-windows.html#calculating-sliding-window-estimates-of-nucleotide-diversity-and-differentiation"><i class="fa fa-check"></i><b>8.3.1</b> Calculating sliding window estimates of nucleotide diversity and differentiation</a></li>
<li class="chapter" data-level="8.3.2" data-path="setting-up-sliding-windows.html"><a href="setting-up-sliding-windows.html#sparrow-data"><i class="fa fa-check"></i><b>8.3.2</b> Extracting statistics for visualisation</a></li>
</ul></li>
<li class="chapter" data-level="8.4" data-path="sparrow-viz.html"><a href="sparrow-viz.html"><i class="fa fa-check"></i><b>8.4</b> Visualising the data</a>
<ul>
<li class="chapter" data-level="8.4.1" data-path="sparrow-viz.html"><a href="sparrow-viz.html#visualising-patterns-along-the-chromosome"><i class="fa fa-check"></i><b>8.4.1</b> Visualising patterns along the chromosome</a></li>
<li class="chapter" data-level="8.4.2" data-path="sparrow-viz.html"><a href="sparrow-viz.html#interlude-relative-vs.-absolute-measures-of-nucleotide-diversity"><i class="fa fa-check"></i><b>8.4.2</b> Interlude: relative vs. absolute measures of nucleotide diversity</a></li>
<li class="chapter" data-level="8.4.3" data-path="sparrow-viz.html"><a href="sparrow-viz.html#investigating-recombination-rate-variation"><i class="fa fa-check"></i><b>8.4.3</b> Investigating recombination rate variation</a></li>
</ul></li>
<li class="chapter" data-level="8.5" data-path="study-questions-6.html"><a href="study-questions-6.html"><i class="fa fa-check"></i><b>8.5</b> Study questions</a></li>
<li class="chapter" data-level="8.6" data-path="going-further-7.html"><a href="going-further-7.html"><i class="fa fa-check"></i><b>8.6</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="ch09.html"><a href="ch09.html"><i class="fa fa-check"></i><b>9</b> Reconstructing the Past</a>
<ul>
<li class="chapter" data-level="" data-path="ch09.html"><a href="ch09.html#what-to-expect-8"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="" data-path="ch09.html"><a href="ch09.html#getting-started-5"><i class="fa fa-check"></i>Getting started</a></li>
<li class="chapter" data-level="9.1" data-path="phylogenetics-in-r.html"><a href="phylogenetics-in-r.html"><i class="fa fa-check"></i><b>9.1</b> Phylogenetics in R</a>
<ul>
<li class="chapter" data-level="9.1.1" data-path="phylogenetics-in-r.html"><a href="phylogenetics-in-r.html#storing-trees-in-r"><i class="fa fa-check"></i><b>9.1.1</b> Storing trees in R</a></li>
<li class="chapter" data-level="9.1.2" data-path="phylogenetics-in-r.html"><a href="phylogenetics-in-r.html#plotting-trees"><i class="fa fa-check"></i><b>9.1.2</b> Plotting trees</a></li>
<li class="chapter" data-level="9.1.3" data-path="phylogenetics-in-r.html"><a href="phylogenetics-in-r.html#a-simple-example-with-real-data---avian-phylogenetics"><i class="fa fa-check"></i><b>9.1.3</b> A simple example with real data - avian phylogenetics</a></li>
<li class="chapter" data-level="9.1.4" data-path="phylogenetics-in-r.html"><a href="phylogenetics-in-r.html#constructing-trees-with-r"><i class="fa fa-check"></i><b>9.1.4</b> Constructing trees with R</a></li>
</ul></li>
<li class="chapter" data-level="9.2" data-path="population-structure.html"><a href="population-structure.html"><i class="fa fa-check"></i><b>9.2</b> Population structure</a>
<ul>
<li class="chapter" data-level="9.2.1" data-path="population-structure.html"><a href="population-structure.html#village-dogs-as-an-insight-to-dog-domestication"><i class="fa fa-check"></i><b>9.2.1</b> Village dogs as an insight to dog domestication</a></li>
<li class="chapter" data-level="9.2.2" data-path="population-structure.html"><a href="population-structure.html#reading-the-data-into-r"><i class="fa fa-check"></i><b>9.2.2</b> Reading the data into R</a></li>
<li class="chapter" data-level="9.2.3" data-path="population-structure.html"><a href="population-structure.html#performing-a-pca"><i class="fa fa-check"></i><b>9.2.3</b> Performing a PCA</a></li>
<li class="chapter" data-level="9.2.4" data-path="population-structure.html"><a href="population-structure.html#visualising-the-pca"><i class="fa fa-check"></i><b>9.2.4</b> Visualising the PCA</a></li>
<li class="chapter" data-level="9.2.5" data-path="population-structure.html"><a href="population-structure.html#eigenvalues"><i class="fa fa-check"></i><b>9.2.5</b> Eigenvalues</a></li>
<li class="chapter" data-level="9.2.6" data-path="population-structure.html"><a href="population-structure.html#the-full-data-set"><i class="fa fa-check"></i><b>9.2.6</b> The full data set</a></li>
</ul></li>
<li class="chapter" data-level="9.3" data-path="study-questions-7.html"><a href="study-questions-7.html"><i class="fa fa-check"></i><b>9.3</b> Study questions</a></li>
<li class="chapter" data-level="9.4" data-path="going-further-8.html"><a href="going-further-8.html"><i class="fa fa-check"></i><b>9.4</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="ch10.html"><a href="ch10.html"><i class="fa fa-check"></i><b>10</b> Advancing Further in R</a>
<ul>
<li class="chapter" data-level="" data-path="ch10.html"><a href="ch10.html#what-to-expect-9"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="" data-path="ch10.html"><a href="ch10.html#getting-started-6"><i class="fa fa-check"></i>Getting started</a></li>
<li class="chapter" data-level="10.1" data-path="advanced-features-of-rstudio.html"><a href="advanced-features-of-rstudio.html"><i class="fa fa-check"></i><b>10.1</b> Advanced features of RStudio</a>
<ul>
<li class="chapter" data-level="10.1.1" data-path="advanced-features-of-rstudio.html"><a href="advanced-features-of-rstudio.html#projects-projects-projects"><i class="fa fa-check"></i><b>10.1.1</b> Projects, projects, projects</a></li>
<li class="chapter" data-level="10.1.2" data-path="advanced-features-of-rstudio.html"><a href="advanced-features-of-rstudio.html#everyone-has-a-history"><i class="fa fa-check"></i><b>10.1.2</b> Everyone has a history</a></li>
<li class="chapter" data-level="10.1.3" data-path="advanced-features-of-rstudio.html"><a href="advanced-features-of-rstudio.html#tab-complete-and-other-hotkeys"><i class="fa fa-check"></i><b>10.1.3</b> Tab complete and other hotkeys</a></li>
<li class="chapter" data-level="10.1.4" data-path="advanced-features-of-rstudio.html"><a href="advanced-features-of-rstudio.html#code-checking"><i class="fa fa-check"></i><b>10.1.4</b> Code checking</a></li>
</ul></li>
<li class="chapter" data-level="10.2" data-path="more-on-data-handlingcategorising-and-relating.html"><a href="more-on-data-handlingcategorising-and-relating.html"><i class="fa fa-check"></i><b>10.2</b> More on data handling—categorising and relating</a>
<ul>
<li class="chapter" data-level="10.2.1" data-path="more-on-data-handlingcategorising-and-relating.html"><a href="more-on-data-handlingcategorising-and-relating.html#ifelse-for-making-categories"><i class="fa fa-check"></i><b>10.2.1</b> <code>ifelse()</code> for making categories</a></li>
<li class="chapter" data-level="10.2.2" data-path="more-on-data-handlingcategorising-and-relating.html"><a href="more-on-data-handlingcategorising-and-relating.html#joining"><i class="fa fa-check"></i><b>10.2.2</b> Joining</a></li>
</ul></li>
<li class="chapter" data-level="10.3" data-path="vectorisation-1.html"><a href="vectorisation-1.html"><i class="fa fa-check"></i><b>10.3</b> Vectorisation</a>
<ul>
<li class="chapter" data-level="10.3.1" data-path="vectorisation-1.html"><a href="vectorisation-1.html#why-is-vectorisation-preferable-in-r"><i class="fa fa-check"></i><b>10.3.1</b> Why is vectorisation preferable in R?</a></li>
<li class="chapter" data-level="10.3.2" data-path="vectorisation-1.html"><a href="vectorisation-1.html#using-sapply-instead-of-for-on-vectors"><i class="fa fa-check"></i><b>10.3.2</b> Using <code>sapply</code> instead of <code>for</code> on vectors</a></li>
<li class="chapter" data-level="10.3.3" data-path="vectorisation-1.html"><a href="vectorisation-1.html#using-apply-instead-of-for-on-matrices-and-dataframes"><i class="fa fa-check"></i><b>10.3.3</b> Using apply instead of <code>for</code> on matrices and dataframes</a></li>
</ul></li>
<li class="chapter" data-level="10.4" data-path="returning-to-writing-your-own-functions.html"><a href="returning-to-writing-your-own-functions.html"><i class="fa fa-check"></i><b>10.4</b> Returning to writing your own functions</a></li>
<li class="chapter" data-level="10.5" data-path="concluding-remarks.html"><a href="concluding-remarks.html"><i class="fa fa-check"></i><b>10.5</b> Concluding remarks</a></li>
<li class="chapter" data-level="10.6" data-path="study-questions-8.html"><a href="study-questions-8.html"><i class="fa fa-check"></i><b>10.6</b> Study questions</a></li>
<li class="chapter" data-level="10.7" data-path="going-further-9.html"><a href="going-further-9.html"><i class="fa fa-check"></i><b>10.7</b> Going further</a></li>
</ul></li>
<li class="part"><span><b>III Assignments</b></span></li>
<li class="chapter" data-level="" data-path="w01.html"><a href="w01.html"><i class="fa fa-check"></i>Week 1-2 assignment</a></li>
<li class="chapter" data-level="" data-path="w03.html"><a href="w03.html"><i class="fa fa-check"></i>Week 3 Assignment</a></li>
<li class="chapter" data-level="" data-path="w04.html"><a href="w04.html"><i class="fa fa-check"></i>Week 4 assignment</a></li>
<li class="chapter" data-level="" data-path="w05.html"><a href="w05.html"><i class="fa fa-check"></i>Week 5 assignment</a></li>
<li class="chapter" data-level="" data-path="w06.html"><a href="w06.html"><i class="fa fa-check"></i>Week 6 assignment</a></li>
<li class="chapter" data-level="" data-path="w07.html"><a href="w07.html"><i class="fa fa-check"></i>Week 7 assignment</a></li>
<li class="chapter" data-level="" data-path="w08.html"><a href="w08.html"><i class="fa fa-check"></i>Week 8 assignment</a></li>
<li class="chapter" data-level="" data-path="w09.html"><a href="w09.html"><i class="fa fa-check"></i>Week 9 assignment</a></li>
<li class="chapter" data-level="" data-path="w10.html"><a href="w10.html"><i class="fa fa-check"></i>Week 10 assignment</a></li>
<li class="appendix"><span><b>Appendix</b></span></li>
<li class="chapter" data-level="A" data-path="rmarkdown.html"><a href="rmarkdown.html"><i class="fa fa-check"></i><b>A</b> About assignments and RMarkdown</a>
<ul>
<li class="chapter" data-level="" data-path="rmarkdown.html"><a href="rmarkdown.html#r-markdown"><i class="fa fa-check"></i>R Markdown</a></li>
<li class="chapter" data-level="" data-path="rmarkdown.html"><a href="rmarkdown.html#getting-started-7"><i class="fa fa-check"></i>Getting started</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Made with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">R tutorials for the course BIOS1140 at the University of Oslo</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="rmarkdown" class="section level1" number="11">
<h1><span class="header-section-number">A</span> About assignments and RMarkdown</h1>
<p>There are some important things to consider when handing in the assignments:</p>
<ol style="list-style-type: decimal">
<li>All answers, plots and code should be contained in <em>a single document</em>. This needs to be either a word or a pdf document.</li>
<li>The document should contain <strong>all the code</strong> you used to complete the assignment. It should also contain <strong>all relevant output</strong> from the code. This should be in a font that’s suitable for programming (i.e., where all characters have the same width), for example Source code pro.</li>
<li>You are allowed to work together, but <strong>identical hand-ins are not allowed</strong>. Similarly, all the text should be your own, <strong>copying from the book, tutorials or any other sources is not allowed</strong>. If you really want to quote someone, you should make it clear that it’s a quote, and cite it properly. It is OK to use the <em>code</em> in the tutorials without citing that.</li>
</ol>
<p>One simple way of combining code, plots and text is to write in a word document where you also paste your code and figures that you’ve exported from R. Another way is to use RMarkdown, which will be introduced in the next section.</p>
<p>A simple template for handing in assignments in R Markdown can be found <a href="https://BIOS1140.github.io/data/assignment_template.Rmd">here</a></p>
<div id="r-markdown" class="section level3 unnumbered">
<h3>R Markdown</h3>
<p>R Markdown is a way to work where you can write, code and show output in the same document. It can be a very convenient way to communicate science, and to write comprehensive data journals detailing how you have analysed your data. In fact, these tutorials are all written in R Markdown! Here we will briefly introduce R Markdown so you will be able to use it for the assignments if you want to. For a more comprehensive introduction, see <a href="https://rmarkdown.rstudio.com/lesson-1.html">the introduction tutorials</a>.</p>
<p>Before using R Markdown, you need to install it. You install it like you would any normal R package.</p>
<div class="sourceCode" id="cb448"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb448-1"><a href="rmarkdown.html#cb448-1" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;rmarkdown&quot;</span>)</span></code></pre></div>
<p>When you write R Markdown documents, you write within RStudio, before converting your text and code to either a HTML, PDF or Word document (known as “knitting”). For this course you need to deliver assignments either in PDF or Word format. PDF requires <span class="math inline">\(\latex\)</span>, and can thus be a bit tricky sometimes. The simplest way to download Latex is by installing <a href="https://yihui.org/tinytex/">TinyTex</a> with the following code:</p>
<div class="sourceCode" id="cb449"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb449-1"><a href="rmarkdown.html#cb449-1" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&#39;tinytex&#39;</span>)</span>
<span id="cb449-2"><a href="rmarkdown.html#cb449-2" aria-hidden="true" tabindex="-1"></a>tinytex<span class="sc">::</span><span class="fu">install_tinytex</span>()</span></code></pre></div>
<p>Now you have all you need for creating PDF documents with R Markdown!</p>
</div>
<div id="getting-started-7" class="section level3 unnumbered">
<h3>Getting started</h3>
<p>Create a new R Markdown document by navigating to <code>File &gt; New File &gt; R Markdown ...</code>. You will then be prompted to provide a title, an author and an output format (you can change these later). Choose PDF as the format and proceed. You have now created an R Markdown file! Notice that it already contains some usage examples. You don’t need these for the assignments, so you can safely delete everything from <code>## R Markdown</code> and to the end of the document.</p>
<p>Here, we will only provide the very basics needed to hand in the assignments. If you’re interested, you can learn more about the basics <a href="https://bookdown.org/yihui/rmarkdown-cookbook/basics.html">here</a>.</p>
<ol style="list-style-type: decimal">
<li><p>Anything you write in the document will be rendered as normal text.</p></li>
<li><p>You can write R code by inserting a <em>code chunk</em>. A code chunk looks like this (hotkey <code>ctrl</code>/<code>cmd</code>+<code>alt</code>+<code>i</code>):</p>
<div class="sourceCode" id="cb450"><pre class="sourceCode markdown"><code class="sourceCode markdown"><span id="cb450-1"><a href="rmarkdown.html#cb450-1" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb450-2"><a href="rmarkdown.html#cb450-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb450-3"><a href="rmarkdown.html#cb450-3" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span></code></pre></div>
<p>Inside the code chunk, you can write regular R code that you can execute normally with <code>ctrl</code>/<code>cmd</code>+<code>enter</code>.</p></li>
<li><p>You can get a top-level header by starting a line with <code>#</code> (outside of a code chunk). You get a second level header with <code>##</code>, third level with <code>###</code> and so on.</p></li>
<li><p>You get <strong>bold</strong> text by flanking the text with two asterisks like this: <code>**bold text**</code> and <em>italic</em> text with a single asterisk: <code>*italic text*</code></p></li>
</ol>
<p>When you’re done with the document, you can press the tiny yarnball at the top of the document that says “knit” to render your document (or press <code>ctrl</code>/<code>cmd</code>+<code>shift</code>+<code>k</code>). This will execute all code that is in the code chunks from a clean environment, and include any plots that are produced by your code. If everything worked, you now have a beautiful document containing code, text and plots!</p>
<div class="blue">
<p><strong>Some notes about R Markdown:</strong></p>
<ul>
<li>All your code is run from a clean slate. This means that all objects must be defined within the document to make it work.</li>
<li>If you even have a single error in the code, the document will not knit. On the plus side, this can be a good way of checking that all your code works!</li>
<li>When working with data files, the data file has to be contained in the same folder as the R Markdown document. Setting working directory within the document will not work most of the time.</li>
<li>Plots will appear inline instead of in the plot window. Disable this by navigating to the cog wheel at the top, and select “chunk output in console”.</li>
</ul>
</div>

</div>
</div>







































































































            </section>

          </div>
        </div>
      </div>
<a href="w10.html" class="navigation navigation-prev navigation-unique" aria-label="Previous page"><i class="fa fa-angle-left"></i></a>

    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
