<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>2.2 Using dplyr to work with your data | R tutorials for the course BIOS1140 at the University of Oslo</title>
  <meta name="description" content="This is a set of R tutorials connected to the course BIOS1140 at the University of Oslo, and are meant as a companion to the book Evolutionary Genetics (Ravinet &amp; Sætre, 2018). They do not assume any prior R knowledge." />
  <meta name="generator" content="bookdown 0.21 and GitBook 2.6.7" />

  <meta property="og:title" content="2.2 Using dplyr to work with your data | R tutorials for the course BIOS1140 at the University of Oslo" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="This is a set of R tutorials connected to the course BIOS1140 at the University of Oslo, and are meant as a companion to the book Evolutionary Genetics (Ravinet &amp; Sætre, 2018). They do not assume any prior R knowledge." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="2.2 Using dplyr to work with your data | R tutorials for the course BIOS1140 at the University of Oslo" />
  
  <meta name="twitter:description" content="This is a set of R tutorials connected to the course BIOS1140 at the University of Oslo, and are meant as a companion to the book Evolutionary Genetics (Ravinet &amp; Sætre, 2018). They do not assume any prior R knowledge." />
  

<meta name="author" content="Original tutorials written by Mark Ravinet. Revamped in 2021 by Even Sletteng Garvang and Solveig Brochmann." />


<meta name="date" content="2021-07-01" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="intro-to-data-manipulation-with-tidyverse.html"/>
<link rel="next" href="plotting-your-data-with-ggplot2.html"/>
<script src="libs/header-attrs-2.6/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">R tutorials for the course BIOS1140</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a>
<ul>
<li class="chapter" data-level="" data-path="how-to-use-these-tutorials.html"><a href="how-to-use-these-tutorials.html"><i class="fa fa-check"></i>How to use these tutorials</a></li>
</ul></li>
<li class="part"><span><b>I Introductory R course</b></span></li>
<li class="chapter" data-level="1" data-path="ch01.html"><a href="ch01.html"><i class="fa fa-check"></i><b>1</b> Introduction to R</a>
<ul>
<li class="chapter" data-level="" data-path="ch01.html"><a href="ch01.html#what-to-expect"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="1.1" data-path="start-using-r.html"><a href="start-using-r.html"><i class="fa fa-check"></i><b>1.1</b> Start using R!</a>
<ul>
<li class="chapter" data-level="1.1.1" data-path="start-using-r.html"><a href="start-using-r.html#install-r-and-rstudio"><i class="fa fa-check"></i><b>1.1.1</b> Install R and Rstudio</a></li>
<li class="chapter" data-level="1.1.2" data-path="start-using-r.html"><a href="start-using-r.html#getting-familiar-with-r-and-the-console"><i class="fa fa-check"></i><b>1.1.2</b> Getting familiar with R and the console</a></li>
<li class="chapter" data-level="1.1.3" data-path="start-using-r.html"><a href="start-using-r.html#scripts"><i class="fa fa-check"></i><b>1.1.3</b> Scripts</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="r-essentials.html"><a href="r-essentials.html"><i class="fa fa-check"></i><b>1.2</b> R essentials</a>
<ul>
<li class="chapter" data-level="1.2.1" data-path="r-essentials.html"><a href="r-essentials.html#assigning-to-objects"><i class="fa fa-check"></i><b>1.2.1</b> Assigning to objects</a></li>
<li class="chapter" data-level="1.2.2" data-path="r-essentials.html"><a href="r-essentials.html#vectors"><i class="fa fa-check"></i><b>1.2.2</b> Vectors</a></li>
<li class="chapter" data-level="1.2.3" data-path="r-essentials.html"><a href="r-essentials.html#strings"><i class="fa fa-check"></i><b>1.2.3</b> Strings</a></li>
<li class="chapter" data-level="1.2.4" data-path="r-essentials.html"><a href="r-essentials.html#logical"><i class="fa fa-check"></i><b>1.2.4</b> Logical values</a></li>
<li class="chapter" data-level="1.2.5" data-path="r-essentials.html"><a href="r-essentials.html#functions"><i class="fa fa-check"></i><b>1.2.5</b> Functions</a></li>
<li class="chapter" data-level="1.2.6" data-path="r-essentials.html"><a href="r-essentials.html#data-frames"><i class="fa fa-check"></i><b>1.2.6</b> Data frames</a></li>
</ul></li>
<li class="chapter" data-level="1.3" data-path="dataimport.html"><a href="dataimport.html"><i class="fa fa-check"></i><b>1.3</b> Data import</a>
<ul>
<li class="chapter" data-level="1.3.1" data-path="dataimport.html"><a href="dataimport.html#data-formats"><i class="fa fa-check"></i><b>1.3.1</b> Data formats</a></li>
<li class="chapter" data-level="1.3.2" data-path="dataimport.html"><a href="dataimport.html#working-directories"><i class="fa fa-check"></i><b>1.3.2</b> Working directories</a></li>
<li class="chapter" data-level="1.3.3" data-path="dataimport.html"><a href="dataimport.html#importing"><i class="fa fa-check"></i><b>1.3.3</b> Importing</a></li>
</ul></li>
<li class="chapter" data-level="1.4" data-path="plotting.html"><a href="plotting.html"><i class="fa fa-check"></i><b>1.4</b> Plotting</a>
<ul>
<li class="chapter" data-level="1.4.1" data-path="plotting.html"><a href="plotting.html#plotting-vectors"><i class="fa fa-check"></i><b>1.4.1</b> Plotting vectors</a></li>
<li class="chapter" data-level="1.4.2" data-path="plotting.html"><a href="plotting.html#customizing-your-plots"><i class="fa fa-check"></i><b>1.4.2</b> Customizing your plots</a></li>
</ul></li>
<li class="chapter" data-level="1.5" data-path="installing-and-loading-packages.html"><a href="installing-and-loading-packages.html"><i class="fa fa-check"></i><b>1.5</b> Installing and loading packages</a></li>
<li class="chapter" data-level="1.6" data-path="going-further.html"><a href="going-further.html"><i class="fa fa-check"></i><b>1.6</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="ch02.html"><a href="ch02.html"><i class="fa fa-check"></i><b>2</b> Building on your foundations: going further with R</a>
<ul>
<li class="chapter" data-level="" data-path="ch02.html"><a href="ch02.html#what-to-expect-1"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="2.1" data-path="intro-to-data-manipulation-with-tidyverse.html"><a href="intro-to-data-manipulation-with-tidyverse.html"><i class="fa fa-check"></i><b>2.1</b> Intro to data manipulation with tidyverse</a>
<ul>
<li class="chapter" data-level="2.1.1" data-path="intro-to-data-manipulation-with-tidyverse.html"><a href="intro-to-data-manipulation-with-tidyverse.html#what-is-the-tidyverse"><i class="fa fa-check"></i><b>2.1.1</b> What is the tidyverse?</a></li>
<li class="chapter" data-level="2.1.2" data-path="intro-to-data-manipulation-with-tidyverse.html"><a href="intro-to-data-manipulation-with-tidyverse.html#the-dplyr-package"><i class="fa fa-check"></i><b>2.1.2</b> The dplyr package</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="using-dplyr-to-work-with-your-data.html"><a href="using-dplyr-to-work-with-your-data.html"><i class="fa fa-check"></i><b>2.2</b> Using dplyr to work with your data</a>
<ul>
<li class="chapter" data-level="2.2.1" data-path="using-dplyr-to-work-with-your-data.html"><a href="using-dplyr-to-work-with-your-data.html#the-pipe"><i class="fa fa-check"></i><b>2.2.1</b> The pipe</a></li>
<li class="chapter" data-level="2.2.2" data-path="using-dplyr-to-work-with-your-data.html"><a href="using-dplyr-to-work-with-your-data.html#selecting-colums-with-select"><i class="fa fa-check"></i><b>2.2.2</b> Selecting colums with <code>select()</code></a></li>
<li class="chapter" data-level="2.2.3" data-path="using-dplyr-to-work-with-your-data.html"><a href="using-dplyr-to-work-with-your-data.html#filtering-colums-using-filter"><i class="fa fa-check"></i><b>2.2.3</b> Filtering colums using <code>filter()</code></a></li>
<li class="chapter" data-level="2.2.4" data-path="using-dplyr-to-work-with-your-data.html"><a href="using-dplyr-to-work-with-your-data.html#grouped-summaries-with-group_by-and-summarise"><i class="fa fa-check"></i><b>2.2.4</b> Grouped summaries with <code>group_by()</code> and <code>summarise()</code></a></li>
<li class="chapter" data-level="2.2.5" data-path="using-dplyr-to-work-with-your-data.html"><a href="using-dplyr-to-work-with-your-data.html#using-everything-weve-learned-in-a-single-pipe-and-a-dplyr-exercise"><i class="fa fa-check"></i><b>2.2.5</b> Using everything we’ve learned in a single pipe, and a <code>dplyr</code> exercise</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="plotting-your-data-with-ggplot2.html"><a href="plotting-your-data-with-ggplot2.html"><i class="fa fa-check"></i><b>2.3</b> Plotting your data with <code>ggplot2</code></a>
<ul>
<li class="chapter" data-level="2.3.1" data-path="plotting-your-data-with-ggplot2.html"><a href="plotting-your-data-with-ggplot2.html#the-three-things-you-need-in-a-ggplot"><i class="fa fa-check"></i><b>2.3.1</b> The three things you need in a ggplot</a></li>
<li class="chapter" data-level="2.3.2" data-path="plotting-your-data-with-ggplot2.html"><a href="plotting-your-data-with-ggplot2.html#storing-ggplots-in-objects"><i class="fa fa-check"></i><b>2.3.2</b> Storing ggplots in objects</a></li>
<li class="chapter" data-level="2.3.3" data-path="plotting-your-data-with-ggplot2.html"><a href="plotting-your-data-with-ggplot2.html#customizing-your-plots-1"><i class="fa fa-check"></i><b>2.3.3</b> Customizing your plots</a></li>
<li class="chapter" data-level="2.3.4" data-path="plotting-your-data-with-ggplot2.html"><a href="plotting-your-data-with-ggplot2.html#saving-your-plots"><i class="fa fa-check"></i><b>2.3.4</b> Saving your plots</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="pivot-longer.html"><a href="pivot-longer.html"><i class="fa fa-check"></i><b>2.4</b> Reshaping data with pivot_longer()</a>
<ul>
<li class="chapter" data-level="2.4.1" data-path="pivot-longer.html"><a href="pivot-longer.html#wide-and-long-format"><i class="fa fa-check"></i><b>2.4.1</b> Wide and long format</a></li>
<li class="chapter" data-level="2.4.2" data-path="pivot-longer.html"><a href="pivot-longer.html#import-example-data"><i class="fa fa-check"></i><b>2.4.2</b> Import example data</a></li>
<li class="chapter" data-level="2.4.3" data-path="pivot-longer.html"><a href="pivot-longer.html#reshape-the-data"><i class="fa fa-check"></i><b>2.4.3</b> Reshape the data</a></li>
<li class="chapter" data-level="2.4.4" data-path="pivot-longer.html"><a href="pivot-longer.html#plot-the-data"><i class="fa fa-check"></i><b>2.4.4</b> Plot the data</a></li>
</ul></li>
<li class="chapter" data-level="2.5" data-path="study-questions.html"><a href="study-questions.html"><i class="fa fa-check"></i><b>2.5</b> Study questions</a></li>
<li class="chapter" data-level="2.6" data-path="going-further-1.html"><a href="going-further-1.html"><i class="fa fa-check"></i><b>2.6</b> Going further</a></li>
</ul></li>
<li class="part"><span><b>II Evolutionary genetics with R</b></span></li>
<li class="chapter" data-level="3" data-path="ch03.html"><a href="ch03.html"><i class="fa fa-check"></i><b>3</b> Changes in Allele and Genotype Frequency</a>
<ul>
<li class="chapter" data-level="" data-path="ch03.html"><a href="ch03.html#what-to-expect-2"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="3.1" data-path="r-programming-for-loops.html"><a href="r-programming-for-loops.html"><i class="fa fa-check"></i><b>3.1</b> R programming: for-loops</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="r-programming-for-loops.html"><a href="r-programming-for-loops.html#motivation-why-loop"><i class="fa fa-check"></i><b>3.1.1</b> Motivation: why loop?</a></li>
<li class="chapter" data-level="3.1.2" data-path="r-programming-for-loops.html"><a href="r-programming-for-loops.html#how-a-for-loop-works"><i class="fa fa-check"></i><b>3.1.2</b> How a for-loop works</a></li>
<li class="chapter" data-level="3.1.3" data-path="r-programming-for-loops.html"><a href="r-programming-for-loops.html#indexing"><i class="fa fa-check"></i><b>3.1.3</b> Indexing with for-loops</a></li>
<li class="chapter" data-level="3.1.4" data-path="r-programming-for-loops.html"><a href="r-programming-for-loops.html#solving-our-problem"><i class="fa fa-check"></i><b>3.1.4</b> Solving our problem</a></li>
<li class="chapter" data-level="3.1.5" data-path="r-programming-for-loops.html"><a href="r-programming-for-loops.html#storing-values-from-a-for-loop"><i class="fa fa-check"></i><b>3.1.5</b> Storing values from a for-loop</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="evolutionary-biology.html"><a href="evolutionary-biology.html"><i class="fa fa-check"></i><b>3.2</b> Evolutionary biology</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="evolutionary-biology.html"><a href="evolutionary-biology.html#the-hardy-weinberg-model"><i class="fa fa-check"></i><b>3.2.1</b> The Hardy-Weinberg Model</a></li>
<li class="chapter" data-level="3.2.2" data-path="evolutionary-biology.html"><a href="evolutionary-biology.html#testing-for-deviations-from-the-hardy-weinberg-expectation"><i class="fa fa-check"></i><b>3.2.2</b> Testing for deviations from the Hardy Weinberg Expectation</a></li>
<li class="chapter" data-level="3.2.3" data-path="evolutionary-biology.html"><a href="evolutionary-biology.html#simulating-genetic-drift"><i class="fa fa-check"></i><b>3.2.3</b> Simulating genetic drift</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="study-questions-1.html"><a href="study-questions-1.html"><i class="fa fa-check"></i><b>3.3</b> Study questions</a></li>
<li class="chapter" data-level="3.4" data-path="going-further-2.html"><a href="going-further-2.html"><i class="fa fa-check"></i><b>3.4</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="ch04.html"><a href="ch04.html"><i class="fa fa-check"></i><b>4</b> The Theory of Natural Selection</a>
<ul>
<li class="chapter" data-level="" data-path="ch04.html"><a href="ch04.html#what-to-expect-3"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="" data-path="ch04.html"><a href="ch04.html#getting-started"><i class="fa fa-check"></i>Getting started</a></li>
<li class="chapter" data-level="4.1" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html"><i class="fa fa-check"></i><b>4.1</b> R programming: Making custom functions</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#motivation"><i class="fa fa-check"></i><b>4.1.1</b> Motivation</a></li>
<li class="chapter" data-level="4.1.2" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#function-basics"><i class="fa fa-check"></i><b>4.1.2</b> Function basics</a></li>
<li class="chapter" data-level="4.1.3" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#a-simple-example"><i class="fa fa-check"></i><b>4.1.3</b> A simple example</a></li>
<li class="chapter" data-level="4.1.4" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#some-important-function-properties"><i class="fa fa-check"></i><b>4.1.4</b> Some important function properties</a></li>
<li class="chapter" data-level="4.1.5" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#a-slightly-more-useful-example"><i class="fa fa-check"></i><b>4.1.5</b> A slightly more useful example</a></li>
<li class="chapter" data-level="4.1.6" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#example-calculating-genotype-frequencies"><i class="fa fa-check"></i><b>4.1.6</b> Example: calculating genotype frequencies</a></li>
<li class="chapter" data-level="4.1.7" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#creating-a-function-of-the-drift-simulation"><i class="fa fa-check"></i><b>4.1.7</b> Creating a function of the drift simulation</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="evolutionary-biology-fitness.html"><a href="evolutionary-biology-fitness.html"><i class="fa fa-check"></i><b>4.2</b> Evolutionary biology: fitness</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="evolutionary-biology-fitness.html"><a href="evolutionary-biology-fitness.html#understanding-fitness"><i class="fa fa-check"></i><b>4.2.1</b> Understanding fitness</a></li>
<li class="chapter" data-level="4.2.2" data-path="evolutionary-biology-fitness.html"><a href="evolutionary-biology-fitness.html#one-locus-model-of-viability-selection"><i class="fa fa-check"></i><b>4.2.2</b> One-locus model of viability selection</a></li>
<li class="chapter" data-level="4.2.3" data-path="evolutionary-biology-fitness.html"><a href="evolutionary-biology-fitness.html#over-and-underdominance"><i class="fa fa-check"></i><b>4.2.3</b> Over and underdominance</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="study-questions-2.html"><a href="study-questions-2.html"><i class="fa fa-check"></i><b>4.3</b> Study questions</a></li>
<li class="chapter" data-level="4.4" data-path="going-further-3.html"><a href="going-further-3.html"><i class="fa fa-check"></i><b>4.4</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="ch05.html"><a href="ch05.html"><i class="fa fa-check"></i><b>5</b> Detecting Natural Selection</a>
<ul>
<li class="chapter" data-level="" data-path="ch05.html"><a href="ch05.html#what-to-expect-4"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="" data-path="ch05.html"><a href="ch05.html#getting-started-1"><i class="fa fa-check"></i>Getting started</a></li>
<li class="chapter" data-level="5.1" data-path="vectorisation.html"><a href="vectorisation.html"><i class="fa fa-check"></i><b>5.1</b> More on functions: Vectorisation</a>
<ul>
<li class="chapter" data-level="5.1.1" data-path="vectorisation.html"><a href="vectorisation.html#the-apply-function"><i class="fa fa-check"></i><b>5.1.1</b> The <code>apply()</code> function</a></li>
<li class="chapter" data-level="5.1.2" data-path="vectorisation.html"><a href="vectorisation.html#ifelse"><i class="fa fa-check"></i><b>5.1.2</b> The <code>ifelse()</code> function</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="understanding-fst-the-fixation-index.html"><a href="understanding-fst-the-fixation-index.html"><i class="fa fa-check"></i><b>5.2</b> Understanding <em>F</em><sub>ST</sub> - the fixation index</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="understanding-fst-the-fixation-index.html"><a href="understanding-fst-the-fixation-index.html#a-worked-example-of-fst-in-humans"><i class="fa fa-check"></i><b>5.2.1</b> A worked example of <em>F</em><sub>ST</sub> in humans</a></li>
<li class="chapter" data-level="5.2.2" data-path="understanding-fst-the-fixation-index.html"><a href="understanding-fst-the-fixation-index.html#writing-a-set-of-fst-functions"><i class="fa fa-check"></i><b>5.2.2</b> Writing a set of <em>F</em><sub>ST</sub> functions</a></li>
<li class="chapter" data-level="5.2.3" data-path="understanding-fst-the-fixation-index.html"><a href="understanding-fst-the-fixation-index.html#applying-functions-to-matrices-and-data-frames"><i class="fa fa-check"></i><b>5.2.3</b> Applying functions to matrices and data frames</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="visualising-fst-along-a-chromosome.html"><a href="visualising-fst-along-a-chromosome.html"><i class="fa fa-check"></i><b>5.3</b> Visualising <em>F</em><sub>ST</sub> along a chromosome</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="visualising-fst-along-a-chromosome.html"><a href="visualising-fst-along-a-chromosome.html#identifying-outliers-in-our-fst-distribution"><i class="fa fa-check"></i><b>5.3.1</b> Identifying outliers in our <em>F</em><sub>ST</sub> distribution</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="study-questions-3.html"><a href="study-questions-3.html"><i class="fa fa-check"></i><b>5.4</b> Study questions</a></li>
<li class="chapter" data-level="5.5" data-path="going-further-4.html"><a href="going-further-4.html"><i class="fa fa-check"></i><b>5.5</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="ch06.html"><a href="ch06.html"><i class="fa fa-check"></i><b>6</b> Multilocus Evolution</a>
<ul>
<li class="chapter" data-level="" data-path="ch06.html"><a href="ch06.html#what-to-expect-5"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="" data-path="ch06.html"><a href="ch06.html#getting-started-2"><i class="fa fa-check"></i>Getting started</a></li>
<li class="chapter" data-level="6.1" data-path="tools-for-investigating-differences-between-groups.html"><a href="tools-for-investigating-differences-between-groups.html"><i class="fa fa-check"></i><b>6.1</b> Tools for investigating differences between groups</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="tools-for-investigating-differences-between-groups.html"><a href="tools-for-investigating-differences-between-groups.html#exploring-group-differences"><i class="fa fa-check"></i><b>6.1.1</b> Exploring group differences</a></li>
<li class="chapter" data-level="6.1.2" data-path="tools-for-investigating-differences-between-groups.html"><a href="tools-for-investigating-differences-between-groups.html#analysis-of-variance-anova"><i class="fa fa-check"></i><b>6.1.2</b> Analysis of variance (ANOVA)</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="performing-a-qtl-analysis-in-bedbugs.html"><a href="performing-a-qtl-analysis-in-bedbugs.html"><i class="fa fa-check"></i><b>6.2</b> Performing a QTL analysis in bedbugs</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="performing-a-qtl-analysis-in-bedbugs.html"><a href="performing-a-qtl-analysis-in-bedbugs.html#bedbugs-pesticide-resistance-and-study-design"><i class="fa fa-check"></i><b>6.2.1</b> Bedbugs, pesticide resistance and study design</a></li>
<li class="chapter" data-level="6.2.2" data-path="performing-a-qtl-analysis-in-bedbugs.html"><a href="performing-a-qtl-analysis-in-bedbugs.html#linking-anova-to-our-qtl-analysis"><i class="fa fa-check"></i><b>6.2.2</b> Linking ANOVA to our QTL analysis</a></li>
<li class="chapter" data-level="6.2.3" data-path="performing-a-qtl-analysis-in-bedbugs.html"><a href="performing-a-qtl-analysis-in-bedbugs.html#testing-the-same-genotype-phenotype-association-with-anova"><i class="fa fa-check"></i><b>6.2.3</b> Testing the same genotype phenotype association with ANOVA</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="study-questions-4.html"><a href="study-questions-4.html"><i class="fa fa-check"></i><b>6.3</b> Study questions</a></li>
<li class="chapter" data-level="6.4" data-path="going-further-5.html"><a href="going-further-5.html"><i class="fa fa-check"></i><b>6.4</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="ch07.html"><a href="ch07.html"><i class="fa fa-check"></i><b>7</b> Inferring Evolutionary Processes from Sequence Data</a>
<ul>
<li class="chapter" data-level="" data-path="ch07.html"><a href="ch07.html#what-to-expect-6"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="" data-path="ch07.html"><a href="ch07.html#getting-started-3"><i class="fa fa-check"></i>Getting started</a></li>
<li class="chapter" data-level="7.1" data-path="working-with-dna-sequence-data.html"><a href="working-with-dna-sequence-data.html"><i class="fa fa-check"></i><b>7.1</b> Working with DNA sequence data</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="working-with-dna-sequence-data.html"><a href="working-with-dna-sequence-data.html#reading-sequence-data-into-r."><i class="fa fa-check"></i><b>7.1.1</b> Reading sequence data into R.</a></li>
<li class="chapter" data-level="7.1.2" data-path="working-with-dna-sequence-data.html"><a href="working-with-dna-sequence-data.html#exploring-dna-sequence-data"><i class="fa fa-check"></i><b>7.1.2</b> Exploring DNA sequence data</a></li>
<li class="chapter" data-level="7.1.3" data-path="working-with-dna-sequence-data.html"><a href="working-with-dna-sequence-data.html#calculating-basic-sequence-statistics"><i class="fa fa-check"></i><b>7.1.3</b> Calculating basic sequence statistics</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="working-with-a-larger-dataset.html"><a href="working-with-a-larger-dataset.html"><i class="fa fa-check"></i><b>7.2</b> Working with a larger dataset</a>
<ul>
<li class="chapter" data-level="7.2.1" data-path="working-with-a-larger-dataset.html"><a href="working-with-a-larger-dataset.html#sample-size-and-sequence-statistics"><i class="fa fa-check"></i><b>7.2.1</b> Sample size and sequence statistics</a></li>
<li class="chapter" data-level="7.2.2" data-path="working-with-a-larger-dataset.html"><a href="working-with-a-larger-dataset.html#inferring-evolutionary-processes-using-tajimas-d"><i class="fa fa-check"></i><b>7.2.2</b> Inferring evolutionary processes using Tajima’s <em>D</em></a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="calculating-statistics-at-the-whole-genome-level.html"><a href="calculating-statistics-at-the-whole-genome-level.html"><i class="fa fa-check"></i><b>7.3</b> Calculating statistics at the whole genome level</a>
<ul>
<li class="chapter" data-level="7.3.1" data-path="calculating-statistics-at-the-whole-genome-level.html"><a href="calculating-statistics-at-the-whole-genome-level.html#reading-in-variant-data"><i class="fa fa-check"></i><b>7.3.1</b> Reading in variant data</a></li>
<li class="chapter" data-level="7.3.2" data-path="calculating-statistics-at-the-whole-genome-level.html"><a href="calculating-statistics-at-the-whole-genome-level.html#calculating-nucleotide-diversity-statistics"><i class="fa fa-check"></i><b>7.3.2</b> Calculating nucleotide diversity statistics</a></li>
<li class="chapter" data-level="7.3.3" data-path="calculating-statistics-at-the-whole-genome-level.html"><a href="calculating-statistics-at-the-whole-genome-level.html#visualising-nucleotide-diversity-along-the-chromosome"><i class="fa fa-check"></i><b>7.3.3</b> Visualising nucleotide diversity along the chromosome</a></li>
<li class="chapter" data-level="7.3.4" data-path="calculating-statistics-at-the-whole-genome-level.html"><a href="calculating-statistics-at-the-whole-genome-level.html#performing-a-sliding-window-analysis"><i class="fa fa-check"></i><b>7.3.4</b> Performing a sliding window analysis</a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="study-questions-5.html"><a href="study-questions-5.html"><i class="fa fa-check"></i><b>7.4</b> Study questions</a></li>
<li class="chapter" data-level="7.5" data-path="going-further-6.html"><a href="going-further-6.html"><i class="fa fa-check"></i><b>7.5</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="ch08.html"><a href="ch08.html"><i class="fa fa-check"></i><b>8</b> Speciation Genomics</a>
<ul>
<li class="chapter" data-level="" data-path="ch08.html"><a href="ch08.html#what-to-expect-7"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="" data-path="ch08.html"><a href="ch08.html#getting-started-4"><i class="fa fa-check"></i>Getting started</a></li>
<li class="chapter" data-level="8.1" data-path="visualizing-complex-data.html"><a href="visualizing-complex-data.html"><i class="fa fa-check"></i><b>8.1</b> Visualizing complex data</a>
<ul>
<li class="chapter" data-level="8.1.1" data-path="visualizing-complex-data.html"><a href="visualizing-complex-data.html#faceting"><i class="fa fa-check"></i><b>8.1.1</b> Faceting</a></li>
</ul></li>
<li class="chapter" data-level="8.2" data-path="returning-to-the-sparrow-dataset.html"><a href="returning-to-the-sparrow-dataset.html"><i class="fa fa-check"></i><b>8.2</b> Returning to the sparrow dataset</a>
<ul>
<li class="chapter" data-level="8.2.1" data-path="returning-to-the-sparrow-dataset.html"><a href="returning-to-the-sparrow-dataset.html#reading-in-the-sparrow-vcf"><i class="fa fa-check"></i><b>8.2.1</b> Reading in the sparrow vcf</a></li>
<li class="chapter" data-level="8.2.2" data-path="returning-to-the-sparrow-dataset.html"><a href="returning-to-the-sparrow-dataset.html#examining-the-variant-data"><i class="fa fa-check"></i><b>8.2.2</b> Examining the variant data</a></li>
</ul></li>
<li class="chapter" data-level="8.3" data-path="setting-up-sliding-windows.html"><a href="setting-up-sliding-windows.html"><i class="fa fa-check"></i><b>8.3</b> Setting up sliding windows</a>
<ul>
<li class="chapter" data-level="8.3.1" data-path="setting-up-sliding-windows.html"><a href="setting-up-sliding-windows.html#calculating-sliding-window-estimates-of-nucleotide-diversity-and-differentiation"><i class="fa fa-check"></i><b>8.3.1</b> Calculating sliding window estimates of nucleotide diversity and differentiation</a></li>
<li class="chapter" data-level="8.3.2" data-path="setting-up-sliding-windows.html"><a href="setting-up-sliding-windows.html#sparrow-data"><i class="fa fa-check"></i><b>8.3.2</b> Extracting statistics for visualisation</a></li>
</ul></li>
<li class="chapter" data-level="8.4" data-path="sparrow-viz.html"><a href="sparrow-viz.html"><i class="fa fa-check"></i><b>8.4</b> Visualising the data</a>
<ul>
<li class="chapter" data-level="8.4.1" data-path="sparrow-viz.html"><a href="sparrow-viz.html#visualising-patterns-along-the-chromosome"><i class="fa fa-check"></i><b>8.4.1</b> Visualising patterns along the chromosome</a></li>
<li class="chapter" data-level="8.4.2" data-path="sparrow-viz.html"><a href="sparrow-viz.html#interlude-relative-vs.-absolute-measures-of-nucleotide-diversity"><i class="fa fa-check"></i><b>8.4.2</b> Interlude: relative vs. absolute measures of nucleotide diversity</a></li>
<li class="chapter" data-level="8.4.3" data-path="sparrow-viz.html"><a href="sparrow-viz.html#investigating-recombination-rate-variation"><i class="fa fa-check"></i><b>8.4.3</b> Investigating recombination rate variation</a></li>
</ul></li>
<li class="chapter" data-level="8.5" data-path="study-questions-6.html"><a href="study-questions-6.html"><i class="fa fa-check"></i><b>8.5</b> Study questions</a></li>
<li class="chapter" data-level="8.6" data-path="going-further-7.html"><a href="going-further-7.html"><i class="fa fa-check"></i><b>8.6</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="ch09.html"><a href="ch09.html"><i class="fa fa-check"></i><b>9</b> Reconstructing the Past</a>
<ul>
<li class="chapter" data-level="" data-path="ch09.html"><a href="ch09.html#what-to-expect-8"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="" data-path="ch09.html"><a href="ch09.html#getting-started-5"><i class="fa fa-check"></i>Getting started</a></li>
<li class="chapter" data-level="9.1" data-path="phylogenetics-in-r.html"><a href="phylogenetics-in-r.html"><i class="fa fa-check"></i><b>9.1</b> Phylogenetics in R</a>
<ul>
<li class="chapter" data-level="9.1.1" data-path="phylogenetics-in-r.html"><a href="phylogenetics-in-r.html#storing-trees-in-r"><i class="fa fa-check"></i><b>9.1.1</b> Storing trees in R</a></li>
<li class="chapter" data-level="9.1.2" data-path="phylogenetics-in-r.html"><a href="phylogenetics-in-r.html#plotting-trees"><i class="fa fa-check"></i><b>9.1.2</b> Plotting trees</a></li>
<li class="chapter" data-level="9.1.3" data-path="phylogenetics-in-r.html"><a href="phylogenetics-in-r.html#a-simple-example-with-real-data---avian-phylogenetics"><i class="fa fa-check"></i><b>9.1.3</b> A simple example with real data - avian phylogenetics</a></li>
<li class="chapter" data-level="9.1.4" data-path="phylogenetics-in-r.html"><a href="phylogenetics-in-r.html#constructing-trees-with-r"><i class="fa fa-check"></i><b>9.1.4</b> Constructing trees with R</a></li>
</ul></li>
<li class="chapter" data-level="9.2" data-path="population-structure.html"><a href="population-structure.html"><i class="fa fa-check"></i><b>9.2</b> Population structure</a>
<ul>
<li class="chapter" data-level="9.2.1" data-path="population-structure.html"><a href="population-structure.html#village-dogs-as-an-insight-to-dog-domestication"><i class="fa fa-check"></i><b>9.2.1</b> Village dogs as an insight to dog domestication</a></li>
<li class="chapter" data-level="9.2.2" data-path="population-structure.html"><a href="population-structure.html#reading-the-data-into-r"><i class="fa fa-check"></i><b>9.2.2</b> Reading the data into R</a></li>
<li class="chapter" data-level="9.2.3" data-path="population-structure.html"><a href="population-structure.html#performing-a-pca"><i class="fa fa-check"></i><b>9.2.3</b> Performing a PCA</a></li>
<li class="chapter" data-level="9.2.4" data-path="population-structure.html"><a href="population-structure.html#visualising-the-pca"><i class="fa fa-check"></i><b>9.2.4</b> Visualising the PCA</a></li>
<li class="chapter" data-level="9.2.5" data-path="population-structure.html"><a href="population-structure.html#eigenvalues"><i class="fa fa-check"></i><b>9.2.5</b> Eigenvalues</a></li>
<li class="chapter" data-level="9.2.6" data-path="population-structure.html"><a href="population-structure.html#the-full-data-set"><i class="fa fa-check"></i><b>9.2.6</b> The full data set</a></li>
</ul></li>
<li class="chapter" data-level="9.3" data-path="study-questions-7.html"><a href="study-questions-7.html"><i class="fa fa-check"></i><b>9.3</b> Study questions</a></li>
<li class="chapter" data-level="9.4" data-path="going-further-8.html"><a href="going-further-8.html"><i class="fa fa-check"></i><b>9.4</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="ch10.html"><a href="ch10.html"><i class="fa fa-check"></i><b>10</b> Advancing Further in R</a>
<ul>
<li class="chapter" data-level="" data-path="ch10.html"><a href="ch10.html#what-to-expect-9"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="" data-path="ch10.html"><a href="ch10.html#getting-started-6"><i class="fa fa-check"></i>Getting started</a></li>
<li class="chapter" data-level="10.1" data-path="advanced-features-of-rstudio.html"><a href="advanced-features-of-rstudio.html"><i class="fa fa-check"></i><b>10.1</b> Advanced features of RStudio</a>
<ul>
<li class="chapter" data-level="10.1.1" data-path="advanced-features-of-rstudio.html"><a href="advanced-features-of-rstudio.html#projects-projects-projects"><i class="fa fa-check"></i><b>10.1.1</b> Projects, projects, projects</a></li>
<li class="chapter" data-level="10.1.2" data-path="advanced-features-of-rstudio.html"><a href="advanced-features-of-rstudio.html#everyone-has-a-history"><i class="fa fa-check"></i><b>10.1.2</b> Everyone has a history</a></li>
<li class="chapter" data-level="10.1.3" data-path="advanced-features-of-rstudio.html"><a href="advanced-features-of-rstudio.html#tab-complete-and-other-hotkeys"><i class="fa fa-check"></i><b>10.1.3</b> Tab complete and other hotkeys</a></li>
<li class="chapter" data-level="10.1.4" data-path="advanced-features-of-rstudio.html"><a href="advanced-features-of-rstudio.html#code-checking"><i class="fa fa-check"></i><b>10.1.4</b> Code checking</a></li>
</ul></li>
<li class="chapter" data-level="10.2" data-path="more-on-data-handling-categorising-and-joining.html"><a href="more-on-data-handling-categorising-and-joining.html"><i class="fa fa-check"></i><b>10.2</b> More on data handling: categorising and joining</a>
<ul>
<li class="chapter" data-level="10.2.1" data-path="more-on-data-handling-categorising-and-joining.html"><a href="more-on-data-handling-categorising-and-joining.html#ifelse-for-making-categories"><i class="fa fa-check"></i><b>10.2.1</b> <code>ifelse()</code> for making categories</a></li>
<li class="chapter" data-level="10.2.2" data-path="more-on-data-handling-categorising-and-joining.html"><a href="more-on-data-handling-categorising-and-joining.html#joining"><i class="fa fa-check"></i><b>10.2.2</b> Joining</a></li>
</ul></li>
<li class="chapter" data-level="10.3" data-path="lists.html"><a href="lists.html"><i class="fa fa-check"></i><b>10.3</b> Lists</a></li>
<li class="chapter" data-level="10.4" data-path="vectorisation2.html"><a href="vectorisation2.html"><i class="fa fa-check"></i><b>10.4</b> Vectorisation</a>
<ul>
<li class="chapter" data-level="10.4.1" data-path="vectorisation2.html"><a href="vectorisation2.html#using-lapply-to-vectorise"><i class="fa fa-check"></i><b>10.4.1</b> Using <code>lapply()</code> to vectorise</a></li>
<li class="chapter" data-level="10.4.2" data-path="vectorisation2.html"><a href="vectorisation2.html#sapply"><i class="fa fa-check"></i><b>10.4.2</b> <code>sapply()</code></a></li>
<li class="chapter" data-level="10.4.3" data-path="vectorisation2.html"><a href="vectorisation2.html#anonymous-functions"><i class="fa fa-check"></i><b>10.4.3</b> Anonymous functions</a></li>
</ul></li>
<li class="chapter" data-level="10.5" data-path="concluding-remarks.html"><a href="concluding-remarks.html"><i class="fa fa-check"></i><b>10.5</b> Concluding remarks</a></li>
<li class="chapter" data-level="10.6" data-path="study-questions-8.html"><a href="study-questions-8.html"><i class="fa fa-check"></i><b>10.6</b> Study questions</a></li>
<li class="chapter" data-level="10.7" data-path="going-further-9.html"><a href="going-further-9.html"><i class="fa fa-check"></i><b>10.7</b> Going further</a></li>
</ul></li>
<li class="part"><span><b>III Assignments</b></span></li>
<li class="chapter" data-level="" data-path="w01.html"><a href="w01.html"><i class="fa fa-check"></i>Week 1-2 assignment</a></li>
<li class="chapter" data-level="" data-path="w03.html"><a href="w03.html"><i class="fa fa-check"></i>Week 3 Assignment</a></li>
<li class="chapter" data-level="" data-path="w04.html"><a href="w04.html"><i class="fa fa-check"></i>Week 4 assignment</a></li>
<li class="chapter" data-level="" data-path="w05.html"><a href="w05.html"><i class="fa fa-check"></i>Week 5 assignment</a></li>
<li class="chapter" data-level="" data-path="w06.html"><a href="w06.html"><i class="fa fa-check"></i>Week 6 assignment</a></li>
<li class="chapter" data-level="" data-path="w07.html"><a href="w07.html"><i class="fa fa-check"></i>Week 7 assignment</a></li>
<li class="chapter" data-level="" data-path="w08.html"><a href="w08.html"><i class="fa fa-check"></i>Week 8 assignment</a></li>
<li class="chapter" data-level="" data-path="w09.html"><a href="w09.html"><i class="fa fa-check"></i>Week 9 assignment</a></li>
<li class="chapter" data-level="" data-path="w10.html"><a href="w10.html"><i class="fa fa-check"></i>Week 10 assignment</a></li>
<li class="appendix"><span><b>Appendix</b></span></li>
<li class="chapter" data-level="A" data-path="rmarkdown.html"><a href="rmarkdown.html"><i class="fa fa-check"></i><b>A</b> About assignments and RMarkdown</a>
<ul>
<li class="chapter" data-level="" data-path="rmarkdown.html"><a href="rmarkdown.html#r-markdown"><i class="fa fa-check"></i>R Markdown</a></li>
<li class="chapter" data-level="" data-path="rmarkdown.html"><a href="rmarkdown.html#getting-started-7"><i class="fa fa-check"></i>Getting started</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Made with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">R tutorials for the course BIOS1140 at the University of Oslo</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="using-dplyr-to-work-with-your-data" class="section level2" number="2.2">
<h2><span class="header-section-number">2.2</span> Using dplyr to work with your data</h2>
<script src="js/hideOutput.js"></script>
<p>First, we have to install and load the tidyverse.<a href="#fn7" class="footnote-ref" id="fnref7"><sup>7</sup></a></p>
<div class="sourceCode" id="cb63"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb63-1"><a href="using-dplyr-to-work-with-your-data.html#cb63-1" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;tidyverse&quot;</span>)</span>
<span id="cb63-2"><a href="using-dplyr-to-work-with-your-data.html#cb63-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code></pre></div>
<p>Together with the tidyverse, you get a built in data set called <code>starwars</code>, containing information about the characters from the Star Wars films. We will use this data set for most of this tutorial. Since it’s already loaded, all you have to do to access it is run:</p>
<div class="sourceCode" id="cb64"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb64-1"><a href="using-dplyr-to-work-with-your-data.html#cb64-1" aria-hidden="true" tabindex="-1"></a>starwars</span>
<span id="cb64-2"><a href="using-dplyr-to-work-with-your-data.html#cb64-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; # A tibble: 87 x 14</span></span>
<span id="cb64-3"><a href="using-dplyr-to-work-with-your-data.html#cb64-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    name  height  mass hair_color skin_color eye_color birth_year sex   gender</span></span>
<span id="cb64-4"><a href="using-dplyr-to-work-with-your-data.html#cb64-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    &lt;chr&gt;  &lt;int&gt; &lt;dbl&gt; &lt;chr&gt;      &lt;chr&gt;      &lt;chr&gt;          &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt; </span></span>
<span id="cb64-5"><a href="using-dplyr-to-work-with-your-data.html#cb64-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  1 Luke~    172    77 blond      fair       blue            19   male  mascu~</span></span>
<span id="cb64-6"><a href="using-dplyr-to-work-with-your-data.html#cb64-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  2 C-3PO    167    75 &lt;NA&gt;       gold       yellow         112   none  mascu~</span></span>
<span id="cb64-7"><a href="using-dplyr-to-work-with-your-data.html#cb64-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  3 R2-D2     96    32 &lt;NA&gt;       white, bl~ red             33   none  mascu~</span></span>
<span id="cb64-8"><a href="using-dplyr-to-work-with-your-data.html#cb64-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  4 Dart~    202   136 none       white      yellow          41.9 male  mascu~</span></span>
<span id="cb64-9"><a href="using-dplyr-to-work-with-your-data.html#cb64-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  5 Leia~    150    49 brown      light      brown           19   fema~ femin~</span></span>
<span id="cb64-10"><a href="using-dplyr-to-work-with-your-data.html#cb64-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  6 Owen~    178   120 brown, gr~ light      blue            52   male  mascu~</span></span>
<span id="cb64-11"><a href="using-dplyr-to-work-with-your-data.html#cb64-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  7 Beru~    165    75 brown      light      blue            47   fema~ femin~</span></span>
<span id="cb64-12"><a href="using-dplyr-to-work-with-your-data.html#cb64-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  8 R5-D4     97    32 &lt;NA&gt;       white, red red             NA   none  mascu~</span></span>
<span id="cb64-13"><a href="using-dplyr-to-work-with-your-data.html#cb64-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  9 Bigg~    183    84 black      light      brown           24   male  mascu~</span></span>
<span id="cb64-14"><a href="using-dplyr-to-work-with-your-data.html#cb64-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 10 Obi-~    182    77 auburn, w~ fair       blue-gray       57   male  mascu~</span></span>
<span id="cb64-15"><a href="using-dplyr-to-work-with-your-data.html#cb64-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; # ... with 77 more rows, and 5 more variables: homeworld &lt;chr&gt;, species &lt;chr&gt;,</span></span>
<span id="cb64-16"><a href="using-dplyr-to-work-with-your-data.html#cb64-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; #   films &lt;list&gt;, vehicles &lt;list&gt;, starships &lt;list&gt;</span></span></code></pre></div>
<div class="blue">
<p><strong>Additional info:</strong> <code>starwars</code> is a data frame like you learned about last week, but you might notice that it’s printing a bit differently (e.g. with red color for <code>NA</code>, condensing the output). This is because <code>starwars</code> is a special kind of data frame introduced in the tidyverse, called a <code>tibble</code>. For all practical purposes, a tibble and a data frame is the same, so throughout this course we won’t care much whether our data is contained in a tibble or a regular data frame.</p>
</div>
<div id="the-pipe" class="section level3" number="2.2.1">
<h3><span class="header-section-number">2.2.1</span> The pipe</h3>
<p>The tidyverse introduces a new operator called the pipe, which looks like this <code>%&gt;%</code>. Conceptually, the pipe allows you to do something with your data, and then send the result to a new function which does more work, sends it to the next function and so on until you’re satisfied. You can for instance use the pipe like this:</p>
<div class="sourceCode" id="cb65"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb65-1"><a href="using-dplyr-to-work-with-your-data.html#cb65-1" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">100</span></span>
<span id="cb65-2"><a href="using-dplyr-to-work-with-your-data.html#cb65-2" aria-hidden="true" tabindex="-1"></a>x<span class="sc">^</span><span class="dv">2</span> <span class="sc">%&gt;%</span> <span class="fu">mean</span>()</span>
<span id="cb65-3"><a href="using-dplyr-to-work-with-your-data.html#cb65-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 3383.5</span></span></code></pre></div>
<p>This is equivalent to writing:</p>
<div class="sourceCode" id="cb66"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb66-1"><a href="using-dplyr-to-work-with-your-data.html#cb66-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mean</span>(x<span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb66-2"><a href="using-dplyr-to-work-with-your-data.html#cb66-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 3383.5</span></span></code></pre></div>
<p>Basically, you are putting the left hand side of the pipe into the parentheses in the function on the right hand side. This may not seem useful right now, but as we will see later, this can make code much easier to read.</p>
<p>You can also assign the results of your pipe to an object like any ordinary calculation:</p>
<div class="sourceCode" id="cb67"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb67-1"><a href="using-dplyr-to-work-with-your-data.html#cb67-1" aria-hidden="true" tabindex="-1"></a>x2_mean <span class="ot">&lt;-</span> x<span class="sc">^</span><span class="dv">2</span> <span class="sc">%&gt;%</span> <span class="fu">mean</span>()</span></code></pre></div>
<p>There will be a lot of examples of using the pipe throughout this tutorial, showing how it can make quite complex code readable.</p>
<div class="green">
<p><strong>Important concept:</strong><br />
The pipe operator <code>%&gt;%</code> allows you to send an object from the left side of the pipe to a function on the right side.</p>
</div>
</div>
<div id="selecting-colums-with-select" class="section level3" number="2.2.2">
<h3><span class="header-section-number">2.2.2</span> Selecting colums with <code>select()</code></h3>
<p>Lets say we want to choose the <code>name</code> and <code>homeworld</code> columns from our <code>starwars</code> data, how can we do that? With standard R, we might do something like this.</p>
<div class="sourceCode" id="cb68"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb68-1"><a href="using-dplyr-to-work-with-your-data.html#cb68-1" aria-hidden="true" tabindex="-1"></a><span class="co"># with names</span></span>
<span id="cb68-2"><a href="using-dplyr-to-work-with-your-data.html#cb68-2" aria-hidden="true" tabindex="-1"></a>starwars[, <span class="fu">c</span>(<span class="st">&#39;name&#39;</span>, <span class="st">&#39;homeworld&#39;</span>)]</span>
<span id="cb68-3"><a href="using-dplyr-to-work-with-your-data.html#cb68-3" aria-hidden="true" tabindex="-1"></a><span class="co"># with indices</span></span>
<span id="cb68-4"><a href="using-dplyr-to-work-with-your-data.html#cb68-4" aria-hidden="true" tabindex="-1"></a>starwars[, <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">9</span>)]</span></code></pre></div>
<p>With <code>dplyr</code> we can do the following:</p>
<div class="sourceCode" id="cb69"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb69-1"><a href="using-dplyr-to-work-with-your-data.html#cb69-1" aria-hidden="true" tabindex="-1"></a><span class="fu">select</span>(starwars, name, homeworld)</span>
<span id="cb69-2"><a href="using-dplyr-to-work-with-your-data.html#cb69-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; # A tibble: 87 x 2</span></span>
<span id="cb69-3"><a href="using-dplyr-to-work-with-your-data.html#cb69-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    name               homeworld</span></span>
<span id="cb69-4"><a href="using-dplyr-to-work-with-your-data.html#cb69-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    &lt;chr&gt;              &lt;chr&gt;    </span></span>
<span id="cb69-5"><a href="using-dplyr-to-work-with-your-data.html#cb69-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  1 Luke Skywalker     Tatooine </span></span>
<span id="cb69-6"><a href="using-dplyr-to-work-with-your-data.html#cb69-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  2 C-3PO              Tatooine </span></span>
<span id="cb69-7"><a href="using-dplyr-to-work-with-your-data.html#cb69-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  3 R2-D2              Naboo    </span></span>
<span id="cb69-8"><a href="using-dplyr-to-work-with-your-data.html#cb69-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  4 Darth Vader        Tatooine </span></span>
<span id="cb69-9"><a href="using-dplyr-to-work-with-your-data.html#cb69-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  5 Leia Organa        Alderaan </span></span>
<span id="cb69-10"><a href="using-dplyr-to-work-with-your-data.html#cb69-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  6 Owen Lars          Tatooine </span></span>
<span id="cb69-11"><a href="using-dplyr-to-work-with-your-data.html#cb69-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  7 Beru Whitesun lars Tatooine </span></span>
<span id="cb69-12"><a href="using-dplyr-to-work-with-your-data.html#cb69-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  8 R5-D4              Tatooine </span></span>
<span id="cb69-13"><a href="using-dplyr-to-work-with-your-data.html#cb69-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  9 Biggs Darklighter  Tatooine </span></span>
<span id="cb69-14"><a href="using-dplyr-to-work-with-your-data.html#cb69-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 10 Obi-Wan Kenobi     Stewjon  </span></span>
<span id="cb69-15"><a href="using-dplyr-to-work-with-your-data.html#cb69-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; # ... with 77 more rows</span></span></code></pre></div>
<p>The first argument here is your data, while the others are the columns you want to select. Note that you don’t need to use the quotes <code>"</code> here, you generally don’t need those for the <code>dplyr</code> functions.</p>
<p><code>select()</code> becomes even more intuitive when using the pipe:</p>
<div class="sourceCode" id="cb70"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb70-1"><a href="using-dplyr-to-work-with-your-data.html#cb70-1" aria-hidden="true" tabindex="-1"></a>starwars <span class="sc">%&gt;%</span> <span class="fu">select</span>(name, homeworld)</span></code></pre></div>
<p>This style of code is closer to how we would write in English: “Take the <code>starwars</code> data, and select the <code>name</code> and <code>homeworld</code> columns”. From now on we will write all our <code>dplyr</code> code using the pipe.</p>
<p>If you want to omit a column, you can use <code>-</code> in front of its name:</p>
<div class="sourceCode" id="cb71"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb71-1"><a href="using-dplyr-to-work-with-your-data.html#cb71-1" aria-hidden="true" tabindex="-1"></a><span class="co"># choose all columns BUT name</span></span>
<span id="cb71-2"><a href="using-dplyr-to-work-with-your-data.html#cb71-2" aria-hidden="true" tabindex="-1"></a>starwars <span class="sc">%&gt;%</span> <span class="fu">select</span>(<span class="sc">-</span>name)</span></code></pre></div>
<p><code>select</code> also has additional ways of selecting columns, some examples of this is shown below:</p>
<div class="sourceCode" id="cb72"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb72-1"><a href="using-dplyr-to-work-with-your-data.html#cb72-1" aria-hidden="true" tabindex="-1"></a><span class="co"># choose only columns containing an underscore</span></span>
<span id="cb72-2"><a href="using-dplyr-to-work-with-your-data.html#cb72-2" aria-hidden="true" tabindex="-1"></a>starwars <span class="sc">%&gt;%</span> <span class="fu">select</span>(<span class="fu">contains</span>(<span class="st">&quot;_&quot;</span>))</span>
<span id="cb72-3"><a href="using-dplyr-to-work-with-your-data.html#cb72-3" aria-hidden="true" tabindex="-1"></a><span class="co"># choose only columns beginning with &quot;s&quot;</span></span>
<span id="cb72-4"><a href="using-dplyr-to-work-with-your-data.html#cb72-4" aria-hidden="true" tabindex="-1"></a>starwars <span class="sc">%&gt;%</span> <span class="fu">select</span>(<span class="fu">starts_with</span>(<span class="st">&quot;s&quot;</span>))</span>
<span id="cb72-5"><a href="using-dplyr-to-work-with-your-data.html#cb72-5" aria-hidden="true" tabindex="-1"></a><span class="co"># choose only columns ending with &quot;color&quot;</span></span>
<span id="cb72-6"><a href="using-dplyr-to-work-with-your-data.html#cb72-6" aria-hidden="true" tabindex="-1"></a>starwars <span class="sc">%&gt;%</span> <span class="fu">select</span>(<span class="fu">ends_with</span>(<span class="st">&quot;color&quot;</span>))</span></code></pre></div>
<p><strong>Exercise: Use <code>select()</code> to select numeric columns, i.e. all the columns that contain numbers, and save it to an object with a meaningful name. You can do this manually by looking at the columns. If you want a challenge, see if you can figure out how to automatically select numeric columns from your data.</strong></p>
<div class="fold s">
<div class="sourceCode" id="cb73"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb73-1"><a href="using-dplyr-to-work-with-your-data.html#cb73-1" aria-hidden="true" tabindex="-1"></a><span class="co"># manual method:</span></span>
<span id="cb73-2"><a href="using-dplyr-to-work-with-your-data.html#cb73-2" aria-hidden="true" tabindex="-1"></a>sw_hmb <span class="ot">&lt;-</span> starwars <span class="sc">%&gt;%</span> <span class="fu">select</span>(height, mass, birth_year)</span>
<span id="cb73-3"><a href="using-dplyr-to-work-with-your-data.html#cb73-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb73-4"><a href="using-dplyr-to-work-with-your-data.html#cb73-4" aria-hidden="true" tabindex="-1"></a><span class="co"># challenge:</span></span>
<span id="cb73-5"><a href="using-dplyr-to-work-with-your-data.html#cb73-5" aria-hidden="true" tabindex="-1"></a>sw_hmb <span class="ot">&lt;-</span> starwars <span class="sc">%&gt;%</span> <span class="fu">select</span>(<span class="fu">where</span>(is.numeric))</span></code></pre></div>
</div>
</div>
<div id="filtering-colums-using-filter" class="section level3" number="2.2.3">
<h3><span class="header-section-number">2.2.3</span> Filtering colums using <code>filter()</code></h3>
<p>Last week, you learned to filter a data set based on some criterion using the square brackets <code>[]</code>. To filter out only the humans from the <code>starwars</code> data set, you could write:</p>
<div class="sourceCode" id="cb74"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb74-1"><a href="using-dplyr-to-work-with-your-data.html#cb74-1" aria-hidden="true" tabindex="-1"></a>starwars[starwars<span class="sc">$</span>species <span class="sc">==</span> <span class="st">&quot;Human&quot;</span>, ]</span></code></pre></div>
<p><code>dplyr</code>’s <code>filter()</code> function does the same thing, but with a slightly different syntax:</p>
<div class="sourceCode" id="cb75"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb75-1"><a href="using-dplyr-to-work-with-your-data.html#cb75-1" aria-hidden="true" tabindex="-1"></a>starwars <span class="sc">%&gt;%</span> <span class="fu">filter</span>(species <span class="sc">==</span> <span class="st">&quot;Human&quot;</span>)</span></code></pre></div>
<p>This looks quite similar to using the square brackets, one notable difference being that you don’t need to use <code>starwars$</code> within <code>filter()</code>, the function already understands that we are working with <code>starwars</code>.</p>
<p>Like with base R, you can use this to filter with the other logical operators as well, like <code>&gt;</code> and <code>!=</code>:</p>
<div class="sourceCode" id="cb76"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb76-1"><a href="using-dplyr-to-work-with-your-data.html#cb76-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get people lower than 1m tall</span></span>
<span id="cb76-2"><a href="using-dplyr-to-work-with-your-data.html#cb76-2" aria-hidden="true" tabindex="-1"></a>starwars <span class="sc">%&gt;%</span> <span class="fu">filter</span>(height <span class="sc">&lt;</span> <span class="dv">100</span>)</span>
<span id="cb76-3"><a href="using-dplyr-to-work-with-your-data.html#cb76-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-4"><a href="using-dplyr-to-work-with-your-data.html#cb76-4" aria-hidden="true" tabindex="-1"></a><span class="co"># get all non-humans</span></span>
<span id="cb76-5"><a href="using-dplyr-to-work-with-your-data.html#cb76-5" aria-hidden="true" tabindex="-1"></a>starwars <span class="sc">%&gt;%</span> <span class="fu">filter</span>(species <span class="sc">!=</span> <span class="st">&quot;Human&quot;</span>)</span></code></pre></div>
<p>You can also filter using several criteria at once, simply separate the logical statements with a comma:</p>
<div class="sourceCode" id="cb77"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb77-1"><a href="using-dplyr-to-work-with-your-data.html#cb77-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get all non-humans shorter than 1m</span></span>
<span id="cb77-2"><a href="using-dplyr-to-work-with-your-data.html#cb77-2" aria-hidden="true" tabindex="-1"></a>starwars <span class="sc">%&gt;%</span> <span class="fu">filter</span>(height <span class="sc">&lt;</span> <span class="dv">100</span>, species <span class="sc">!=</span> <span class="st">&quot;Human&quot;</span>)</span></code></pre></div>
<div id="combining-filter-and-select" class="section level4" number="2.2.3.1">
<h4><span class="header-section-number">2.2.3.1</span> Combining <code>filter()</code> and <code>select()</code></h4>
<p>The real power of the pipe shows when you chain several operations together. To both filter and select from your data, simply first do the filtering and then pipe the result to <code>select</code>:<a href="#fn8" class="footnote-ref" id="fnref8"><sup>8</sup></a></p>
<div class="sourceCode" id="cb78"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb78-1"><a href="using-dplyr-to-work-with-your-data.html#cb78-1" aria-hidden="true" tabindex="-1"></a>starwars <span class="sc">%&gt;%</span> <span class="fu">filter</span>(height <span class="sc">&lt;</span> <span class="dv">100</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb78-2"><a href="using-dplyr-to-work-with-your-data.html#cb78-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(name, height, birth_year)</span>
<span id="cb78-3"><a href="using-dplyr-to-work-with-your-data.html#cb78-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; # A tibble: 7 x 3</span></span>
<span id="cb78-4"><a href="using-dplyr-to-work-with-your-data.html#cb78-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   name                  height birth_year</span></span>
<span id="cb78-5"><a href="using-dplyr-to-work-with-your-data.html#cb78-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   &lt;chr&gt;                  &lt;int&gt;      &lt;dbl&gt;</span></span>
<span id="cb78-6"><a href="using-dplyr-to-work-with-your-data.html#cb78-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 1 R2-D2                     96         33</span></span>
<span id="cb78-7"><a href="using-dplyr-to-work-with-your-data.html#cb78-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 2 R5-D4                     97         NA</span></span>
<span id="cb78-8"><a href="using-dplyr-to-work-with-your-data.html#cb78-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 3 Yoda                      66        896</span></span>
<span id="cb78-9"><a href="using-dplyr-to-work-with-your-data.html#cb78-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 4 Wicket Systri Warrick     88          8</span></span>
<span id="cb78-10"><a href="using-dplyr-to-work-with-your-data.html#cb78-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 5 Dud Bolt                  94         NA</span></span>
<span id="cb78-11"><a href="using-dplyr-to-work-with-your-data.html#cb78-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 6 Ratts Tyerell             79         NA</span></span>
<span id="cb78-12"><a href="using-dplyr-to-work-with-your-data.html#cb78-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 7 R4-P17                    96         NA</span></span></code></pre></div>
<p>Again, the code looks like how you would explain what you’re doing: “take the starwars data, filter based on height, and select the <code>name</code>, <code>height</code> and <code>birth_year</code> columns”. Note that you can have a line break after a pipe like you can inside parentheses, your code will still continue running.</p>
<p><strong>Exercise: use <code>filter()</code> to choose all the people that has “Naboo” as <code>homeworld</code>. Select name, skin_color and eye_color and save the result to an object.</strong></p>
<div class="fold s">
<div class="sourceCode" id="cb79"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb79-1"><a href="using-dplyr-to-work-with-your-data.html#cb79-1" aria-hidden="true" tabindex="-1"></a>sw_naboo_color <span class="ot">&lt;-</span> starwars <span class="sc">%&gt;%</span> <span class="fu">filter</span>(homeworld <span class="sc">==</span> <span class="st">&quot;Naboo&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="cb79-2"><a href="using-dplyr-to-work-with-your-data.html#cb79-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(name, skin_color, eye_color)</span></code></pre></div>
</div>
<div class="green">
<p><strong>Important concept:</strong><br />
- <code>filter()</code> is used to select specific rows. Example: <code>filter(height &lt; 100)</code> - <code>select()</code> is used to select specific columns. Example: <code>select(name, height)</code></p>
</div>
</div>
</div>
<div id="grouped-summaries-with-group_by-and-summarise" class="section level3" number="2.2.4">
<h3><span class="header-section-number">2.2.4</span> Grouped summaries with <code>group_by()</code> and <code>summarise()</code></h3>
<p>Imagine that you want to calculate the mean height of the people (and droids) in the <code>starwars</code> data set. You could use <code>mean()</code> on the column to achieve this (note the use of <code>na.rm</code> since the height column contains <code>NA</code>s):</p>
<div class="sourceCode" id="cb80"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb80-1"><a href="using-dplyr-to-work-with-your-data.html#cb80-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mean</span>(starwars<span class="sc">$</span>height, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb80-2"><a href="using-dplyr-to-work-with-your-data.html#cb80-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 174.358</span></span></code></pre></div>
<p>But what if you want to calculate the mean height separately for e.g. the different species? One way to do this is to do a <em>grouped summary</em>. Your group is the <code>species</code> column, and your summary statistic is <code>mean</code>. We create groups using the <code>group_by()</code> function:</p>
<div class="sourceCode" id="cb81"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb81-1"><a href="using-dplyr-to-work-with-your-data.html#cb81-1" aria-hidden="true" tabindex="-1"></a>starwars <span class="sc">%&gt;%</span> <span class="fu">group_by</span>(species)</span>
<span id="cb81-2"><a href="using-dplyr-to-work-with-your-data.html#cb81-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; # A tibble: 87 x 14</span></span>
<span id="cb81-3"><a href="using-dplyr-to-work-with-your-data.html#cb81-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; # Groups:   species [38]</span></span>
<span id="cb81-4"><a href="using-dplyr-to-work-with-your-data.html#cb81-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    name  height  mass hair_color skin_color eye_color birth_year sex   gender</span></span>
<span id="cb81-5"><a href="using-dplyr-to-work-with-your-data.html#cb81-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    &lt;chr&gt;  &lt;int&gt; &lt;dbl&gt; &lt;chr&gt;      &lt;chr&gt;      &lt;chr&gt;          &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt; </span></span>
<span id="cb81-6"><a href="using-dplyr-to-work-with-your-data.html#cb81-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  1 Luke~    172    77 blond      fair       blue            19   male  mascu~</span></span>
<span id="cb81-7"><a href="using-dplyr-to-work-with-your-data.html#cb81-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  2 C-3PO    167    75 &lt;NA&gt;       gold       yellow         112   none  mascu~</span></span>
<span id="cb81-8"><a href="using-dplyr-to-work-with-your-data.html#cb81-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  3 R2-D2     96    32 &lt;NA&gt;       white, bl~ red             33   none  mascu~</span></span>
<span id="cb81-9"><a href="using-dplyr-to-work-with-your-data.html#cb81-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  4 Dart~    202   136 none       white      yellow          41.9 male  mascu~</span></span>
<span id="cb81-10"><a href="using-dplyr-to-work-with-your-data.html#cb81-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  5 Leia~    150    49 brown      light      brown           19   fema~ femin~</span></span>
<span id="cb81-11"><a href="using-dplyr-to-work-with-your-data.html#cb81-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  6 Owen~    178   120 brown, gr~ light      blue            52   male  mascu~</span></span>
<span id="cb81-12"><a href="using-dplyr-to-work-with-your-data.html#cb81-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  7 Beru~    165    75 brown      light      blue            47   fema~ femin~</span></span>
<span id="cb81-13"><a href="using-dplyr-to-work-with-your-data.html#cb81-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  8 R5-D4     97    32 &lt;NA&gt;       white, red red             NA   none  mascu~</span></span>
<span id="cb81-14"><a href="using-dplyr-to-work-with-your-data.html#cb81-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  9 Bigg~    183    84 black      light      brown           24   male  mascu~</span></span>
<span id="cb81-15"><a href="using-dplyr-to-work-with-your-data.html#cb81-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 10 Obi-~    182    77 auburn, w~ fair       blue-gray       57   male  mascu~</span></span>
<span id="cb81-16"><a href="using-dplyr-to-work-with-your-data.html#cb81-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; # ... with 77 more rows, and 5 more variables: homeworld &lt;chr&gt;, species &lt;chr&gt;,</span></span>
<span id="cb81-17"><a href="using-dplyr-to-work-with-your-data.html#cb81-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; #   films &lt;list&gt;, vehicles &lt;list&gt;, starships &lt;list&gt;</span></span></code></pre></div>
<p>Notice that nothing has changed in the data, but at the top you can see the text <code># Groups:   species [38]</code>, showing that you indeed have created a group, and that you have 38 different species in your data. The main use of <code>group_by()</code> is together with <code>summarise()</code>, which does a summary based on the groups you’ve created:</p>
<div class="sourceCode" id="cb82"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb82-1"><a href="using-dplyr-to-work-with-your-data.html#cb82-1" aria-hidden="true" tabindex="-1"></a>starwars <span class="sc">%&gt;%</span> </span>
<span id="cb82-2"><a href="using-dplyr-to-work-with-your-data.html#cb82-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(species) <span class="sc">%&gt;%</span> <span class="co">#create group</span></span>
<span id="cb82-3"><a href="using-dplyr-to-work-with-your-data.html#cb82-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">mean_height =</span> <span class="fu">mean</span>(height, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)) <span class="co"># calculate summary statistic</span></span>
<span id="cb82-4"><a href="using-dplyr-to-work-with-your-data.html#cb82-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; # A tibble: 38 x 2</span></span>
<span id="cb82-5"><a href="using-dplyr-to-work-with-your-data.html#cb82-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    species   mean_height</span></span>
<span id="cb82-6"><a href="using-dplyr-to-work-with-your-data.html#cb82-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    &lt;chr&gt;           &lt;dbl&gt;</span></span>
<span id="cb82-7"><a href="using-dplyr-to-work-with-your-data.html#cb82-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  1 Aleena            79 </span></span>
<span id="cb82-8"><a href="using-dplyr-to-work-with-your-data.html#cb82-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  2 Besalisk         198 </span></span>
<span id="cb82-9"><a href="using-dplyr-to-work-with-your-data.html#cb82-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  3 Cerean           198 </span></span>
<span id="cb82-10"><a href="using-dplyr-to-work-with-your-data.html#cb82-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  4 Chagrian         196 </span></span>
<span id="cb82-11"><a href="using-dplyr-to-work-with-your-data.html#cb82-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  5 Clawdite         168 </span></span>
<span id="cb82-12"><a href="using-dplyr-to-work-with-your-data.html#cb82-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  6 Droid            131.</span></span>
<span id="cb82-13"><a href="using-dplyr-to-work-with-your-data.html#cb82-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  7 Dug              112 </span></span>
<span id="cb82-14"><a href="using-dplyr-to-work-with-your-data.html#cb82-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  8 Ewok              88 </span></span>
<span id="cb82-15"><a href="using-dplyr-to-work-with-your-data.html#cb82-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  9 Geonosian        183 </span></span>
<span id="cb82-16"><a href="using-dplyr-to-work-with-your-data.html#cb82-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 10 Gungan           209.</span></span>
<span id="cb82-17"><a href="using-dplyr-to-work-with-your-data.html#cb82-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; # ... with 28 more rows</span></span></code></pre></div>
<p>Note how again, like in <code>data.frame</code>, the argument name to <code>summarise</code> becomes the column name in your new data frame. You can use several summary functions inside <code>summarise()</code>, like <code>median()</code>, <code>sd()</code>, <code>sum()</code> and <code>max()</code> to name some. You can also do several summaries within a single <code>summarise()</code> function:</p>
<div class="sourceCode" id="cb83"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb83-1"><a href="using-dplyr-to-work-with-your-data.html#cb83-1" aria-hidden="true" tabindex="-1"></a>starwars <span class="sc">%&gt;%</span> </span>
<span id="cb83-2"><a href="using-dplyr-to-work-with-your-data.html#cb83-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(species) <span class="sc">%&gt;%</span></span>
<span id="cb83-3"><a href="using-dplyr-to-work-with-your-data.html#cb83-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">mean_height =</span> <span class="fu">mean</span>(height, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb83-4"><a href="using-dplyr-to-work-with-your-data.html#cb83-4" aria-hidden="true" tabindex="-1"></a>            <span class="at">median_height =</span> <span class="fu">median</span>(height, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb83-5"><a href="using-dplyr-to-work-with-your-data.html#cb83-5" aria-hidden="true" tabindex="-1"></a>            <span class="at">sd_height =</span> <span class="fu">sd</span>(height, <span class="at">na.rm =</span> <span class="cn">TRUE</span>))</span>
<span id="cb83-6"><a href="using-dplyr-to-work-with-your-data.html#cb83-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; # A tibble: 38 x 4</span></span>
<span id="cb83-7"><a href="using-dplyr-to-work-with-your-data.html#cb83-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    species   mean_height median_height sd_height</span></span>
<span id="cb83-8"><a href="using-dplyr-to-work-with-your-data.html#cb83-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    &lt;chr&gt;           &lt;dbl&gt;         &lt;dbl&gt;     &lt;dbl&gt;</span></span>
<span id="cb83-9"><a href="using-dplyr-to-work-with-your-data.html#cb83-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  1 Aleena            79             79      NA  </span></span>
<span id="cb83-10"><a href="using-dplyr-to-work-with-your-data.html#cb83-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  2 Besalisk         198            198      NA  </span></span>
<span id="cb83-11"><a href="using-dplyr-to-work-with-your-data.html#cb83-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  3 Cerean           198            198      NA  </span></span>
<span id="cb83-12"><a href="using-dplyr-to-work-with-your-data.html#cb83-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  4 Chagrian         196            196      NA  </span></span>
<span id="cb83-13"><a href="using-dplyr-to-work-with-your-data.html#cb83-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  5 Clawdite         168            168      NA  </span></span>
<span id="cb83-14"><a href="using-dplyr-to-work-with-your-data.html#cb83-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  6 Droid            131.            97      49.1</span></span>
<span id="cb83-15"><a href="using-dplyr-to-work-with-your-data.html#cb83-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  7 Dug              112            112      NA  </span></span>
<span id="cb83-16"><a href="using-dplyr-to-work-with-your-data.html#cb83-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  8 Ewok              88             88      NA  </span></span>
<span id="cb83-17"><a href="using-dplyr-to-work-with-your-data.html#cb83-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  9 Geonosian        183            183      NA  </span></span>
<span id="cb83-18"><a href="using-dplyr-to-work-with-your-data.html#cb83-18" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 10 Gungan           209.           206      14.2</span></span>
<span id="cb83-19"><a href="using-dplyr-to-work-with-your-data.html#cb83-19" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; # ... with 28 more rows</span></span></code></pre></div>
<p>We can even group by several variables, creating more detailed summaries:</p>
<div class="sourceCode" id="cb84"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb84-1"><a href="using-dplyr-to-work-with-your-data.html#cb84-1" aria-hidden="true" tabindex="-1"></a>starwars <span class="sc">%&gt;%</span></span>
<span id="cb84-2"><a href="using-dplyr-to-work-with-your-data.html#cb84-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(homeworld, sex) <span class="sc">%&gt;%</span></span>
<span id="cb84-3"><a href="using-dplyr-to-work-with-your-data.html#cb84-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">mean_height =</span> <span class="fu">mean</span>(height, <span class="at">na.rm =</span> <span class="cn">TRUE</span>))</span>
<span id="cb84-4"><a href="using-dplyr-to-work-with-your-data.html#cb84-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; # A tibble: 61 x 3</span></span>
<span id="cb84-5"><a href="using-dplyr-to-work-with-your-data.html#cb84-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; # Groups:   homeworld [49]</span></span>
<span id="cb84-6"><a href="using-dplyr-to-work-with-your-data.html#cb84-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    homeworld      sex    mean_height</span></span>
<span id="cb84-7"><a href="using-dplyr-to-work-with-your-data.html#cb84-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    &lt;chr&gt;          &lt;chr&gt;        &lt;dbl&gt;</span></span>
<span id="cb84-8"><a href="using-dplyr-to-work-with-your-data.html#cb84-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  1 Alderaan       female        150 </span></span>
<span id="cb84-9"><a href="using-dplyr-to-work-with-your-data.html#cb84-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  2 Alderaan       male          190.</span></span>
<span id="cb84-10"><a href="using-dplyr-to-work-with-your-data.html#cb84-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  3 Aleen Minor    male           79 </span></span>
<span id="cb84-11"><a href="using-dplyr-to-work-with-your-data.html#cb84-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  4 Bespin         male          175 </span></span>
<span id="cb84-12"><a href="using-dplyr-to-work-with-your-data.html#cb84-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  5 Bestine IV     male          180 </span></span>
<span id="cb84-13"><a href="using-dplyr-to-work-with-your-data.html#cb84-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  6 Cato Neimoidia male          191 </span></span>
<span id="cb84-14"><a href="using-dplyr-to-work-with-your-data.html#cb84-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  7 Cerea          male          198 </span></span>
<span id="cb84-15"><a href="using-dplyr-to-work-with-your-data.html#cb84-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  8 Champala       male          196 </span></span>
<span id="cb84-16"><a href="using-dplyr-to-work-with-your-data.html#cb84-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  9 Chandrila      female        150 </span></span>
<span id="cb84-17"><a href="using-dplyr-to-work-with-your-data.html#cb84-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 10 Concord Dawn   male          183 </span></span>
<span id="cb84-18"><a href="using-dplyr-to-work-with-your-data.html#cb84-18" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; # ... with 51 more rows</span></span></code></pre></div>
<p>Now you get two groups for homeworld Alderaan, one with males and one with females. For the following homeworld groups there are only males, so you just get one group for each (giving a pretty accurate picture of the gender balance in Star Wars).</p>
<div id="counting-how-many-observations-we-have-in-our-groups-with-tally" class="section level4" number="2.2.4.1">
<h4><span class="header-section-number">2.2.4.1</span> Counting how many observations we have in our groups with <code>tally()</code></h4>
<p>When we have created our groups, we can also use the <code>tally()</code> function to count the number of observations we have in the groups:</p>
<div class="sourceCode" id="cb85"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb85-1"><a href="using-dplyr-to-work-with-your-data.html#cb85-1" aria-hidden="true" tabindex="-1"></a>starwars <span class="sc">%&gt;%</span> </span>
<span id="cb85-2"><a href="using-dplyr-to-work-with-your-data.html#cb85-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(species) <span class="sc">%&gt;%</span></span>
<span id="cb85-3"><a href="using-dplyr-to-work-with-your-data.html#cb85-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tally</span>()</span>
<span id="cb85-4"><a href="using-dplyr-to-work-with-your-data.html#cb85-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; # A tibble: 38 x 2</span></span>
<span id="cb85-5"><a href="using-dplyr-to-work-with-your-data.html#cb85-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    species       n</span></span>
<span id="cb85-6"><a href="using-dplyr-to-work-with-your-data.html#cb85-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    &lt;chr&gt;     &lt;int&gt;</span></span>
<span id="cb85-7"><a href="using-dplyr-to-work-with-your-data.html#cb85-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  1 Aleena        1</span></span>
<span id="cb85-8"><a href="using-dplyr-to-work-with-your-data.html#cb85-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  2 Besalisk      1</span></span>
<span id="cb85-9"><a href="using-dplyr-to-work-with-your-data.html#cb85-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  3 Cerean        1</span></span>
<span id="cb85-10"><a href="using-dplyr-to-work-with-your-data.html#cb85-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  4 Chagrian      1</span></span>
<span id="cb85-11"><a href="using-dplyr-to-work-with-your-data.html#cb85-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  5 Clawdite      1</span></span>
<span id="cb85-12"><a href="using-dplyr-to-work-with-your-data.html#cb85-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  6 Droid         6</span></span>
<span id="cb85-13"><a href="using-dplyr-to-work-with-your-data.html#cb85-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  7 Dug           1</span></span>
<span id="cb85-14"><a href="using-dplyr-to-work-with-your-data.html#cb85-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  8 Ewok          1</span></span>
<span id="cb85-15"><a href="using-dplyr-to-work-with-your-data.html#cb85-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  9 Geonosian     1</span></span>
<span id="cb85-16"><a href="using-dplyr-to-work-with-your-data.html#cb85-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 10 Gungan        3</span></span>
<span id="cb85-17"><a href="using-dplyr-to-work-with-your-data.html#cb85-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; # ... with 28 more rows</span></span></code></pre></div>
<p>This can be useful to get an overview of your data<a href="#fn9" class="footnote-ref" id="fnref9"><sup>9</sup></a></p>
<div class="green">
<p><strong>Important concept:</strong><br />
<code>group_by()</code> can be combined with different functions to give an overview of your data.</p>
<ul>
<li><code>group_by() %&gt;% summarise()</code> does some calculation in each group. Example: <code>group_by(homeworld, sex) %&gt;% summarise(mean_height = mean(height))</code></li>
<li><code>group_by() %&gt;% tally()</code> counts the number of observations in the groups. Example: <code>group_by(species) %&gt;% tally()</code></li>
</ul>
</div>
</div>
</div>
<div id="using-everything-weve-learned-in-a-single-pipe-and-a-dplyr-exercise" class="section level3" number="2.2.5">
<h3><span class="header-section-number">2.2.5</span> Using everything we’ve learned in a single pipe, and a <code>dplyr</code> exercise</h3>
<p>One advantage of pipes is that you can do everything you want in a single operation<a href="#fn10" class="footnote-ref" id="fnref10"><sup>10</sup></a>. Below is an example using everything we’ve learned so far in a single pipe.</p>
<div class="sourceCode" id="cb86"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb86-1"><a href="using-dplyr-to-work-with-your-data.html#cb86-1" aria-hidden="true" tabindex="-1"></a>starwars <span class="sc">%&gt;%</span></span>
<span id="cb86-2"><a href="using-dplyr-to-work-with-your-data.html#cb86-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="sc">-</span>films, <span class="sc">-</span>vehicles, <span class="sc">-</span>starships) <span class="sc">%&gt;%</span></span>
<span id="cb86-3"><a href="using-dplyr-to-work-with-your-data.html#cb86-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(species <span class="sc">==</span> <span class="st">&quot;Human&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="cb86-4"><a href="using-dplyr-to-work-with-your-data.html#cb86-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(sex) <span class="sc">%&gt;%</span></span>
<span id="cb86-5"><a href="using-dplyr-to-work-with-your-data.html#cb86-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">mean_height =</span> <span class="fu">mean</span>(height, <span class="at">na.rm =</span> <span class="cn">TRUE</span>))</span>
<span id="cb86-6"><a href="using-dplyr-to-work-with-your-data.html#cb86-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; # A tibble: 2 x 2</span></span>
<span id="cb86-7"><a href="using-dplyr-to-work-with-your-data.html#cb86-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   sex    mean_height</span></span>
<span id="cb86-8"><a href="using-dplyr-to-work-with-your-data.html#cb86-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   &lt;chr&gt;        &lt;dbl&gt;</span></span>
<span id="cb86-9"><a href="using-dplyr-to-work-with-your-data.html#cb86-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 1 female        160.</span></span>
<span id="cb86-10"><a href="using-dplyr-to-work-with-your-data.html#cb86-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 2 male          182.</span></span></code></pre></div>
<p><strong>Exercise: Explain to another student what happens in the code above (or write it down if you’re doing this on your own).</strong></p>
<p><strong>Exercise: Take the starwars data set, filter so you keep all that are <em>below the mean height</em>. Then, calculate the mean height of these short individuals, grouped by homeworld.</strong></p>
<details>
<summary>
<p>Show hint</p>
</summary>
<p>You can supply the mean height to the logical statement inside <code>filter()</code>. Your filtering step should then look like this: <code>filter(height &lt; mean(height, na.rm = TRUE))</code></p>
</details>
<div class="fold s">
<div class="sourceCode" id="cb87"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb87-1"><a href="using-dplyr-to-work-with-your-data.html#cb87-1" aria-hidden="true" tabindex="-1"></a>starwars <span class="sc">%&gt;%</span></span>
<span id="cb87-2"><a href="using-dplyr-to-work-with-your-data.html#cb87-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(height <span class="sc">&lt;</span> <span class="fu">mean</span>(height, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb87-3"><a href="using-dplyr-to-work-with-your-data.html#cb87-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(homeworld) <span class="sc">%&gt;%</span></span>
<span id="cb87-4"><a href="using-dplyr-to-work-with-your-data.html#cb87-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">mean_height =</span> <span class="fu">mean</span>(height, <span class="at">na.rm =</span> <span class="cn">TRUE</span>))</span>
<span id="cb87-5"><a href="using-dplyr-to-work-with-your-data.html#cb87-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; # A tibble: 19 x 2</span></span>
<span id="cb87-6"><a href="using-dplyr-to-work-with-your-data.html#cb87-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    homeworld   mean_height</span></span>
<span id="cb87-7"><a href="using-dplyr-to-work-with-your-data.html#cb87-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    &lt;chr&gt;             &lt;dbl&gt;</span></span>
<span id="cb87-8"><a href="using-dplyr-to-work-with-your-data.html#cb87-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  1 Alderaan           150 </span></span>
<span id="cb87-9"><a href="using-dplyr-to-work-with-your-data.html#cb87-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  2 Aleen Minor         79 </span></span>
<span id="cb87-10"><a href="using-dplyr-to-work-with-your-data.html#cb87-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  3 Chandrila          150 </span></span>
<span id="cb87-11"><a href="using-dplyr-to-work-with-your-data.html#cb87-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  4 Corellia           170 </span></span>
<span id="cb87-12"><a href="using-dplyr-to-work-with-your-data.html#cb87-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  5 Coruscant          168.</span></span>
<span id="cb87-13"><a href="using-dplyr-to-work-with-your-data.html#cb87-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  6 Endor               88 </span></span>
<span id="cb87-14"><a href="using-dplyr-to-work-with-your-data.html#cb87-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  7 Iridonia           171 </span></span>
<span id="cb87-15"><a href="using-dplyr-to-work-with-your-data.html#cb87-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  8 Malastare          112 </span></span>
<span id="cb87-16"><a href="using-dplyr-to-work-with-your-data.html#cb87-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  9 Mirial             168 </span></span>
<span id="cb87-17"><a href="using-dplyr-to-work-with-your-data.html#cb87-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 10 Naboo              151.</span></span>
<span id="cb87-18"><a href="using-dplyr-to-work-with-your-data.html#cb87-18" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 11 Rodia              173 </span></span>
<span id="cb87-19"><a href="using-dplyr-to-work-with-your-data.html#cb87-19" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 12 Sullust            160 </span></span>
<span id="cb87-20"><a href="using-dplyr-to-work-with-your-data.html#cb87-20" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 13 Tatooine           153.</span></span>
<span id="cb87-21"><a href="using-dplyr-to-work-with-your-data.html#cb87-21" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 14 Toydaria           137 </span></span>
<span id="cb87-22"><a href="using-dplyr-to-work-with-your-data.html#cb87-22" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 15 Troiken            122 </span></span>
<span id="cb87-23"><a href="using-dplyr-to-work-with-your-data.html#cb87-23" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 16 Tund               163 </span></span>
<span id="cb87-24"><a href="using-dplyr-to-work-with-your-data.html#cb87-24" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 17 Vulpter             94 </span></span>
<span id="cb87-25"><a href="using-dplyr-to-work-with-your-data.html#cb87-25" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 18 Zolan              168 </span></span>
<span id="cb87-26"><a href="using-dplyr-to-work-with-your-data.html#cb87-26" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 19 &lt;NA&gt;                81</span></span></code></pre></div>
</div>
</div>
</div>
<div class="footnotes">
<hr />
<ol start="7">
<li id="fn7"><p>Remember that you only need to install a package once, but that it needs to be loaded with <code>library()</code> every time you want to use it.<a href="using-dplyr-to-work-with-your-data.html#fnref7" class="footnote-back">↩︎</a></p></li>
<li id="fn8"><p>Remember that what the pipe basically does is to put the left hand side of the pipe into the function on the right hand side. Without the pipe, filtering and selecting looks like this: <code>select(filter(starwars, height &lt; 100), name, height, birth_year)</code><a href="using-dplyr-to-work-with-your-data.html#fnref8" class="footnote-back">↩︎</a></p></li>
<li id="fn9"><p>for example, you could realise that it doesn’t make sense to calculate mean and standard deviation when you only have a single value, like we’ve done quite a bit<a href="using-dplyr-to-work-with-your-data.html#fnref9" class="footnote-back">↩︎</a></p></li>
<li id="fn10"><p>In practice, it’s probably smart to make an object of your intermediary results every now and then.<a href="using-dplyr-to-work-with-your-data.html#fnref10" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="intro-to-data-manipulation-with-tidyverse.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="plotting-your-data-with-ggplot2.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
