<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Week 1-2 assignment | R tutorials in BIOS1140</title>
  <meta name="description" content="This is a set of R tutorials connected to the course BIOS1140 at the University of Oslo, and are meant as a companion to the book Evolutionary Genetics (Ravinet &amp; Sætre, 2018). They do not assume any prior R knowledge." />
  <meta name="generator" content="bookdown 0.21 and GitBook 2.6.7" />

  <meta property="og:title" content="Week 1-2 assignment | R tutorials in BIOS1140" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="This is a set of R tutorials connected to the course BIOS1140 at the University of Oslo, and are meant as a companion to the book Evolutionary Genetics (Ravinet &amp; Sætre, 2018). They do not assume any prior R knowledge." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Week 1-2 assignment | R tutorials in BIOS1140" />
  
  <meta name="twitter:description" content="This is a set of R tutorials connected to the course BIOS1140 at the University of Oslo, and are meant as a companion to the book Evolutionary Genetics (Ravinet &amp; Sætre, 2018). They do not assume any prior R knowledge." />
  



<meta name="date" content="2021-06-20" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="going-further-9.html"/>
<link rel="next" href="w03.html"/>
<script src="libs/header-attrs-2.6/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">R tutorials for the course BIOS1140</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a></li>
<li class="part"><span><b>I Introductory R course</b></span></li>
<li class="chapter" data-level="1" data-path="introduction-to-r.html"><a href="introduction-to-r.html"><i class="fa fa-check"></i><b>1</b> Introduction to R</a>
<ul>
<li class="chapter" data-level="" data-path="introduction-to-r.html"><a href="introduction-to-r.html#what-to-expect"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="1.1" data-path="start-using-r.html"><a href="start-using-r.html"><i class="fa fa-check"></i><b>1.1</b> Start using R!</a>
<ul>
<li class="chapter" data-level="1.1.1" data-path="start-using-r.html"><a href="start-using-r.html#install-r-and-rstudio"><i class="fa fa-check"></i><b>1.1.1</b> Install R and Rstudio</a></li>
<li class="chapter" data-level="1.1.2" data-path="start-using-r.html"><a href="start-using-r.html#getting-familiar-with-r-and-the-console"><i class="fa fa-check"></i><b>1.1.2</b> Getting familiar with R and the console</a></li>
<li class="chapter" data-level="1.1.3" data-path="start-using-r.html"><a href="start-using-r.html#scripts"><i class="fa fa-check"></i><b>1.1.3</b> Scripts</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="r-essentials.html"><a href="r-essentials.html"><i class="fa fa-check"></i><b>1.2</b> R essentials</a>
<ul>
<li class="chapter" data-level="1.2.1" data-path="r-essentials.html"><a href="r-essentials.html#assigning-to-objects"><i class="fa fa-check"></i><b>1.2.1</b> Assigning to objects</a></li>
<li class="chapter" data-level="1.2.2" data-path="r-essentials.html"><a href="r-essentials.html#vectors"><i class="fa fa-check"></i><b>1.2.2</b> Vectors</a></li>
<li class="chapter" data-level="1.2.3" data-path="r-essentials.html"><a href="r-essentials.html#strings"><i class="fa fa-check"></i><b>1.2.3</b> Strings</a></li>
<li class="chapter" data-level="1.2.4" data-path="r-essentials.html"><a href="r-essentials.html#logical"><i class="fa fa-check"></i><b>1.2.4</b> Logical values</a></li>
<li class="chapter" data-level="1.2.5" data-path="r-essentials.html"><a href="r-essentials.html#functions"><i class="fa fa-check"></i><b>1.2.5</b> Functions</a></li>
<li class="chapter" data-level="1.2.6" data-path="r-essentials.html"><a href="r-essentials.html#data-frames"><i class="fa fa-check"></i><b>1.2.6</b> Data frames</a></li>
</ul></li>
<li class="chapter" data-level="1.3" data-path="dataimport.html"><a href="dataimport.html"><i class="fa fa-check"></i><b>1.3</b> Data import</a>
<ul>
<li class="chapter" data-level="1.3.1" data-path="dataimport.html"><a href="dataimport.html#data-formats"><i class="fa fa-check"></i><b>1.3.1</b> Data formats</a></li>
<li class="chapter" data-level="1.3.2" data-path="dataimport.html"><a href="dataimport.html#working-directories"><i class="fa fa-check"></i><b>1.3.2</b> Working directories</a></li>
<li class="chapter" data-level="1.3.3" data-path="dataimport.html"><a href="dataimport.html#importing"><i class="fa fa-check"></i><b>1.3.3</b> Importing</a></li>
</ul></li>
<li class="chapter" data-level="1.4" data-path="plotting.html"><a href="plotting.html"><i class="fa fa-check"></i><b>1.4</b> Plotting</a>
<ul>
<li class="chapter" data-level="1.4.1" data-path="plotting.html"><a href="plotting.html#plotting-vectors"><i class="fa fa-check"></i><b>1.4.1</b> Plotting vectors</a></li>
<li class="chapter" data-level="1.4.2" data-path="plotting.html"><a href="plotting.html#customizing-your-plots"><i class="fa fa-check"></i><b>1.4.2</b> Customizing your plots</a></li>
</ul></li>
<li class="chapter" data-level="1.5" data-path="installing-and-loading-packages.html"><a href="installing-and-loading-packages.html"><i class="fa fa-check"></i><b>1.5</b> Installing and loading packages</a></li>
<li class="chapter" data-level="1.6" data-path="going-further.html"><a href="going-further.html"><i class="fa fa-check"></i><b>1.6</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="building-on-your-foundations-going-further-with-r.html"><a href="building-on-your-foundations-going-further-with-r.html"><i class="fa fa-check"></i><b>2</b> Building on your foundations: going further with R</a>
<ul>
<li class="chapter" data-level="" data-path="building-on-your-foundations-going-further-with-r.html"><a href="building-on-your-foundations-going-further-with-r.html#what-to-expect-1"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="2.1" data-path="intro-to-data-manipulation-with-tidyverse.html"><a href="intro-to-data-manipulation-with-tidyverse.html"><i class="fa fa-check"></i><b>2.1</b> Intro to data manipulation with tidyverse</a>
<ul>
<li class="chapter" data-level="2.1.1" data-path="intro-to-data-manipulation-with-tidyverse.html"><a href="intro-to-data-manipulation-with-tidyverse.html#what-is-the-tidyverse"><i class="fa fa-check"></i><b>2.1.1</b> What is the tidyverse?</a></li>
<li class="chapter" data-level="2.1.2" data-path="intro-to-data-manipulation-with-tidyverse.html"><a href="intro-to-data-manipulation-with-tidyverse.html#the-dplyr-package"><i class="fa fa-check"></i><b>2.1.2</b> The dplyr package</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="using-dplyr-to-work-with-your-data.html"><a href="using-dplyr-to-work-with-your-data.html"><i class="fa fa-check"></i><b>2.2</b> Using dplyr to work with your data</a>
<ul>
<li class="chapter" data-level="2.2.1" data-path="using-dplyr-to-work-with-your-data.html"><a href="using-dplyr-to-work-with-your-data.html#the-pipe"><i class="fa fa-check"></i><b>2.2.1</b> The pipe</a></li>
<li class="chapter" data-level="2.2.2" data-path="using-dplyr-to-work-with-your-data.html"><a href="using-dplyr-to-work-with-your-data.html#selecting-colums-with-select"><i class="fa fa-check"></i><b>2.2.2</b> Selecting colums with <code>select()</code></a></li>
<li class="chapter" data-level="2.2.3" data-path="using-dplyr-to-work-with-your-data.html"><a href="using-dplyr-to-work-with-your-data.html#filtering-colums-using-filter"><i class="fa fa-check"></i><b>2.2.3</b> Filtering colums using <code>filter()</code></a></li>
<li class="chapter" data-level="2.2.4" data-path="using-dplyr-to-work-with-your-data.html"><a href="using-dplyr-to-work-with-your-data.html#creating-new-columns-with-mutate"><i class="fa fa-check"></i><b>2.2.4</b> Creating new columns with <code>mutate()</code></a></li>
<li class="chapter" data-level="2.2.5" data-path="using-dplyr-to-work-with-your-data.html"><a href="using-dplyr-to-work-with-your-data.html#grouped-summaries-with-group_by-and-summarise"><i class="fa fa-check"></i><b>2.2.5</b> Grouped summaries with <code>group_by()</code> and <code>summarise()</code></a></li>
<li class="chapter" data-level="2.2.6" data-path="using-dplyr-to-work-with-your-data.html"><a href="using-dplyr-to-work-with-your-data.html#using-everything-weve-learned-in-a-single-pipe-and-a-dplyr-exercise"><i class="fa fa-check"></i><b>2.2.6</b> Using everything we’ve learned in a single pipe, and a <code>dplyr</code> exercise</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="plotting-your-data-with-ggplot2.html"><a href="plotting-your-data-with-ggplot2.html"><i class="fa fa-check"></i><b>2.3</b> Plotting your data with <code>ggplot2</code></a>
<ul>
<li class="chapter" data-level="2.3.1" data-path="plotting-your-data-with-ggplot2.html"><a href="plotting-your-data-with-ggplot2.html#the-three-things-you-need-in-a-ggplot"><i class="fa fa-check"></i><b>2.3.1</b> The three things you need in a ggplot</a></li>
<li class="chapter" data-level="2.3.2" data-path="plotting-your-data-with-ggplot2.html"><a href="plotting-your-data-with-ggplot2.html#storing-ggplots-in-objects"><i class="fa fa-check"></i><b>2.3.2</b> Storing ggplots in objects</a></li>
<li class="chapter" data-level="2.3.3" data-path="plotting-your-data-with-ggplot2.html"><a href="plotting-your-data-with-ggplot2.html#customizing-your-plots-1"><i class="fa fa-check"></i><b>2.3.3</b> Customizing your plots</a></li>
<li class="chapter" data-level="2.3.4" data-path="plotting-your-data-with-ggplot2.html"><a href="plotting-your-data-with-ggplot2.html#saving-your-plots"><i class="fa fa-check"></i><b>2.3.4</b> Saving your plots</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="reshaping-data-with-pivot-longer.html"><a href="reshaping-data-with-pivot-longer.html"><i class="fa fa-check"></i><b>2.4</b> Reshaping data with pivot_longer()</a>
<ul>
<li class="chapter" data-level="2.4.1" data-path="reshaping-data-with-pivot-longer.html"><a href="reshaping-data-with-pivot-longer.html#wide-and-long-format"><i class="fa fa-check"></i><b>2.4.1</b> Wide and long format</a></li>
<li class="chapter" data-level="2.4.2" data-path="reshaping-data-with-pivot-longer.html"><a href="reshaping-data-with-pivot-longer.html#import-example-data"><i class="fa fa-check"></i><b>2.4.2</b> Import example data</a></li>
<li class="chapter" data-level="2.4.3" data-path="reshaping-data-with-pivot-longer.html"><a href="reshaping-data-with-pivot-longer.html#reshape-the-data"><i class="fa fa-check"></i><b>2.4.3</b> Reshape the data</a></li>
<li class="chapter" data-level="2.4.4" data-path="reshaping-data-with-pivot-longer.html"><a href="reshaping-data-with-pivot-longer.html#plot-the-data"><i class="fa fa-check"></i><b>2.4.4</b> Plot the data</a></li>
</ul></li>
<li class="chapter" data-level="2.5" data-path="study-questions.html"><a href="study-questions.html"><i class="fa fa-check"></i><b>2.5</b> Study questions</a></li>
<li class="chapter" data-level="2.6" data-path="going-further-1.html"><a href="going-further-1.html"><i class="fa fa-check"></i><b>2.6</b> Going further</a></li>
</ul></li>
<li class="part"><span><b>II Evolutionary genetics with R</b></span></li>
<li class="chapter" data-level="3" data-path="changes-in-allele-and-genotype-frequency.html"><a href="changes-in-allele-and-genotype-frequency.html"><i class="fa fa-check"></i><b>3</b> Changes in Allele and Genotype Frequency</a>
<ul>
<li class="chapter" data-level="" data-path="changes-in-allele-and-genotype-frequency.html"><a href="changes-in-allele-and-genotype-frequency.html#what-to-expect-2"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="3.1" data-path="r-programming-for-loops.html"><a href="r-programming-for-loops.html"><i class="fa fa-check"></i><b>3.1</b> R programming: for-loops</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="r-programming-for-loops.html"><a href="r-programming-for-loops.html#motivation-why-loop"><i class="fa fa-check"></i><b>3.1.1</b> Motivation: why loop?</a></li>
<li class="chapter" data-level="3.1.2" data-path="r-programming-for-loops.html"><a href="r-programming-for-loops.html#how-a-for-loop-works"><i class="fa fa-check"></i><b>3.1.2</b> How a for-loop works</a></li>
<li class="chapter" data-level="3.1.3" data-path="r-programming-for-loops.html"><a href="r-programming-for-loops.html#indexing"><i class="fa fa-check"></i><b>3.1.3</b> Indexing with for-loops</a></li>
<li class="chapter" data-level="3.1.4" data-path="r-programming-for-loops.html"><a href="r-programming-for-loops.html#solving-our-problem"><i class="fa fa-check"></i><b>3.1.4</b> Solving our problem</a></li>
<li class="chapter" data-level="3.1.5" data-path="r-programming-for-loops.html"><a href="r-programming-for-loops.html#storing-values-from-a-for-loop"><i class="fa fa-check"></i><b>3.1.5</b> Storing values from a for-loop</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="evolutionary-biology.html"><a href="evolutionary-biology.html"><i class="fa fa-check"></i><b>3.2</b> Evolutionary biology</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="evolutionary-biology.html"><a href="evolutionary-biology.html#the-hardy-weinberg-model"><i class="fa fa-check"></i><b>3.2.1</b> The Hardy-Weinberg Model</a></li>
<li class="chapter" data-level="3.2.2" data-path="evolutionary-biology.html"><a href="evolutionary-biology.html#testing-for-deviations-from-the-hardy-weinberg-expectation"><i class="fa fa-check"></i><b>3.2.2</b> Testing for deviations from the Hardy Weinberg Expectation</a></li>
<li class="chapter" data-level="3.2.3" data-path="evolutionary-biology.html"><a href="evolutionary-biology.html#simulating-genetic-drift"><i class="fa fa-check"></i><b>3.2.3</b> Simulating genetic drift</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="study-questions-1.html"><a href="study-questions-1.html"><i class="fa fa-check"></i><b>3.3</b> Study questions</a></li>
<li class="chapter" data-level="3.4" data-path="going-further-2.html"><a href="going-further-2.html"><i class="fa fa-check"></i><b>3.4</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="the-theory-of-natural-selection.html"><a href="the-theory-of-natural-selection.html"><i class="fa fa-check"></i><b>4</b> The Theory of Natural Selection</a>
<ul>
<li class="chapter" data-level="" data-path="the-theory-of-natural-selection.html"><a href="the-theory-of-natural-selection.html#what-to-expect-3"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="" data-path="the-theory-of-natural-selection.html"><a href="the-theory-of-natural-selection.html#getting-started"><i class="fa fa-check"></i>Getting started</a></li>
<li class="chapter" data-level="4.1" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html"><i class="fa fa-check"></i><b>4.1</b> R programming: Making custom functions</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#motivation"><i class="fa fa-check"></i><b>4.1.1</b> Motivation</a></li>
<li class="chapter" data-level="4.1.2" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#function-basics"><i class="fa fa-check"></i><b>4.1.2</b> Function basics</a></li>
<li class="chapter" data-level="4.1.3" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#a-simple-example"><i class="fa fa-check"></i><b>4.1.3</b> A simple example</a></li>
<li class="chapter" data-level="4.1.4" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#some-important-function-properties"><i class="fa fa-check"></i><b>4.1.4</b> Some important function properties</a></li>
<li class="chapter" data-level="4.1.5" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#a-slightly-more-useful-example"><i class="fa fa-check"></i><b>4.1.5</b> A slightly more useful example</a></li>
<li class="chapter" data-level="4.1.6" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#example-calculating-genotype-frequencies"><i class="fa fa-check"></i><b>4.1.6</b> Example: calculating genotype frequencies</a></li>
<li class="chapter" data-level="4.1.7" data-path="r-programming-making-custom-functions.html"><a href="r-programming-making-custom-functions.html#creating-a-function-of-the-drift-simulation"><i class="fa fa-check"></i><b>4.1.7</b> Creating a function of the drift simulation</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="evolutionary-biology-fitness.html"><a href="evolutionary-biology-fitness.html"><i class="fa fa-check"></i><b>4.2</b> Evolutionary biology: fitness</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="evolutionary-biology-fitness.html"><a href="evolutionary-biology-fitness.html#understanding-fitness"><i class="fa fa-check"></i><b>4.2.1</b> Understanding fitness</a></li>
<li class="chapter" data-level="4.2.2" data-path="evolutionary-biology-fitness.html"><a href="evolutionary-biology-fitness.html#one-locus-model-of-viability-selection"><i class="fa fa-check"></i><b>4.2.2</b> One-locus model of viability selection</a></li>
<li class="chapter" data-level="4.2.3" data-path="evolutionary-biology-fitness.html"><a href="evolutionary-biology-fitness.html#directional-selection"><i class="fa fa-check"></i><b>4.2.3</b> Directional selection</a></li>
<li class="chapter" data-level="4.2.4" data-path="evolutionary-biology-fitness.html"><a href="evolutionary-biology-fitness.html#over-and-underdominance"><i class="fa fa-check"></i><b>4.2.4</b> Over and underdominance</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="study-questions-2.html"><a href="study-questions-2.html"><i class="fa fa-check"></i><b>4.3</b> Study questions</a></li>
<li class="chapter" data-level="4.4" data-path="going-further-3.html"><a href="going-further-3.html"><i class="fa fa-check"></i><b>4.4</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="detecting-natural-selection.html"><a href="detecting-natural-selection.html"><i class="fa fa-check"></i><b>5</b> Detecting Natural Selection</a>
<ul>
<li class="chapter" data-level="" data-path="detecting-natural-selection.html"><a href="detecting-natural-selection.html#what-to-expect-4"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="" data-path="detecting-natural-selection.html"><a href="detecting-natural-selection.html#getting-started-1"><i class="fa fa-check"></i>Getting started</a></li>
<li class="chapter" data-level="5.1" data-path="more-on-functions-vectorisation.html"><a href="more-on-functions-vectorisation.html"><i class="fa fa-check"></i><b>5.1</b> More on functions: Vectorisation</a>
<ul>
<li class="chapter" data-level="5.1.1" data-path="more-on-functions-vectorisation.html"><a href="more-on-functions-vectorisation.html#the-apply-function"><i class="fa fa-check"></i><b>5.1.1</b> The <code>apply()</code> function</a></li>
<li class="chapter" data-level="5.1.2" data-path="more-on-functions-vectorisation.html"><a href="more-on-functions-vectorisation.html#ifelse"><i class="fa fa-check"></i><b>5.1.2</b> The <code>ifelse()</code> function</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="understanding-fst-the-fixation-index.html"><a href="understanding-fst-the-fixation-index.html"><i class="fa fa-check"></i><b>5.2</b> Understanding <em>F</em><sub>ST</sub> - the fixation index</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="understanding-fst-the-fixation-index.html"><a href="understanding-fst-the-fixation-index.html#a-worked-example-of-fst-in-humans"><i class="fa fa-check"></i><b>5.2.1</b> A worked example of <em>F</em><sub>ST</sub> in humans</a></li>
<li class="chapter" data-level="5.2.2" data-path="understanding-fst-the-fixation-index.html"><a href="understanding-fst-the-fixation-index.html#writing-a-set-of-fst-functions"><i class="fa fa-check"></i><b>5.2.2</b> Writing a set of <em>F</em><sub>ST</sub> functions</a></li>
<li class="chapter" data-level="5.2.3" data-path="understanding-fst-the-fixation-index.html"><a href="understanding-fst-the-fixation-index.html#applying-functions-to-matrices-and-data-frames"><i class="fa fa-check"></i><b>5.2.3</b> Applying functions to matrices and data frames</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="visualising-fst-along-a-chromosome.html"><a href="visualising-fst-along-a-chromosome.html"><i class="fa fa-check"></i><b>5.3</b> Visualising <em>F</em><sub>ST</sub> along a chromosome</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="visualising-fst-along-a-chromosome.html"><a href="visualising-fst-along-a-chromosome.html#identifying-outliers-in-our-fst-distribution"><i class="fa fa-check"></i><b>5.3.1</b> Identifying outliers in our <em>F</em><sub>ST</sub> distribution</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="study-questions-3.html"><a href="study-questions-3.html"><i class="fa fa-check"></i><b>5.4</b> Study questions</a></li>
<li class="chapter" data-level="5.5" data-path="going-further-4.html"><a href="going-further-4.html"><i class="fa fa-check"></i><b>5.5</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="multilocus-evolution.html"><a href="multilocus-evolution.html"><i class="fa fa-check"></i><b>6</b> Multilocus Evolution</a>
<ul>
<li class="chapter" data-level="" data-path="multilocus-evolution.html"><a href="multilocus-evolution.html#what-to-expect-5"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="" data-path="multilocus-evolution.html"><a href="multilocus-evolution.html#getting-started-2"><i class="fa fa-check"></i>Getting started</a></li>
<li class="chapter" data-level="6.1" data-path="tools-for-investigating-differences-between-groups.html"><a href="tools-for-investigating-differences-between-groups.html"><i class="fa fa-check"></i><b>6.1</b> Tools for investigating differences between groups</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="tools-for-investigating-differences-between-groups.html"><a href="tools-for-investigating-differences-between-groups.html#exploring-group-differences"><i class="fa fa-check"></i><b>6.1.1</b> Exploring group differences</a></li>
<li class="chapter" data-level="6.1.2" data-path="tools-for-investigating-differences-between-groups.html"><a href="tools-for-investigating-differences-between-groups.html#analysis-of-variance-anova"><i class="fa fa-check"></i><b>6.1.2</b> Analysis of variance (ANOVA)</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="performing-a-qtl-analysis-in-bedbugs.html"><a href="performing-a-qtl-analysis-in-bedbugs.html"><i class="fa fa-check"></i><b>6.2</b> Performing a QTL analysis in bedbugs</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="performing-a-qtl-analysis-in-bedbugs.html"><a href="performing-a-qtl-analysis-in-bedbugs.html#bedbugs-pesticide-resistance-and-study-design"><i class="fa fa-check"></i><b>6.2.1</b> Bedbugs, pesticide resistance and study design</a></li>
<li class="chapter" data-level="6.2.2" data-path="performing-a-qtl-analysis-in-bedbugs.html"><a href="performing-a-qtl-analysis-in-bedbugs.html#linking-anova-to-our-qtl-analysis"><i class="fa fa-check"></i><b>6.2.2</b> Linking ANOVA to our QTL analysis</a></li>
<li class="chapter" data-level="6.2.3" data-path="performing-a-qtl-analysis-in-bedbugs.html"><a href="performing-a-qtl-analysis-in-bedbugs.html#testing-the-same-genotype-phenotype-association-with-anova"><i class="fa fa-check"></i><b>6.2.3</b> Testing the same genotype phenotype association with ANOVA</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="study-questions-4.html"><a href="study-questions-4.html"><i class="fa fa-check"></i><b>6.3</b> Study questions</a></li>
<li class="chapter" data-level="6.4" data-path="going-further-5.html"><a href="going-further-5.html"><i class="fa fa-check"></i><b>6.4</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="inferring-evolutionary-processes-from-sequence-data.html"><a href="inferring-evolutionary-processes-from-sequence-data.html"><i class="fa fa-check"></i><b>7</b> Inferring Evolutionary Processes from Sequence Data</a>
<ul>
<li class="chapter" data-level="" data-path="inferring-evolutionary-processes-from-sequence-data.html"><a href="inferring-evolutionary-processes-from-sequence-data.html#what-to-expect-6"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="" data-path="inferring-evolutionary-processes-from-sequence-data.html"><a href="inferring-evolutionary-processes-from-sequence-data.html#getting-started-3"><i class="fa fa-check"></i>Getting started</a></li>
<li class="chapter" data-level="7.1" data-path="working-with-dna-sequence-data.html"><a href="working-with-dna-sequence-data.html"><i class="fa fa-check"></i><b>7.1</b> Working with DNA sequence data</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="working-with-dna-sequence-data.html"><a href="working-with-dna-sequence-data.html#reading-sequence-data-into-r."><i class="fa fa-check"></i><b>7.1.1</b> Reading sequence data into R.</a></li>
<li class="chapter" data-level="7.1.2" data-path="working-with-dna-sequence-data.html"><a href="working-with-dna-sequence-data.html#exploring-dna-sequence-data"><i class="fa fa-check"></i><b>7.1.2</b> Exploring DNA sequence data</a></li>
<li class="chapter" data-level="7.1.3" data-path="working-with-dna-sequence-data.html"><a href="working-with-dna-sequence-data.html#calculating-basic-sequence-statistics"><i class="fa fa-check"></i><b>7.1.3</b> Calculating basic sequence statistics</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="working-with-a-larger-dataset.html"><a href="working-with-a-larger-dataset.html"><i class="fa fa-check"></i><b>7.2</b> Working with a larger dataset</a>
<ul>
<li class="chapter" data-level="7.2.1" data-path="working-with-a-larger-dataset.html"><a href="working-with-a-larger-dataset.html#sample-size-and-sequence-statistics"><i class="fa fa-check"></i><b>7.2.1</b> Sample size and sequence statistics</a></li>
<li class="chapter" data-level="7.2.2" data-path="working-with-a-larger-dataset.html"><a href="working-with-a-larger-dataset.html#inferring-evolutionary-processes-using-tajimas-d"><i class="fa fa-check"></i><b>7.2.2</b> Inferring evolutionary processes using Tajima’s <em>D</em></a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="calculating-statistics-at-the-whole-genome-level.html"><a href="calculating-statistics-at-the-whole-genome-level.html"><i class="fa fa-check"></i><b>7.3</b> Calculating statistics at the whole genome level</a>
<ul>
<li class="chapter" data-level="7.3.1" data-path="calculating-statistics-at-the-whole-genome-level.html"><a href="calculating-statistics-at-the-whole-genome-level.html#reading-in-variant-data"><i class="fa fa-check"></i><b>7.3.1</b> Reading in variant data</a></li>
<li class="chapter" data-level="7.3.2" data-path="calculating-statistics-at-the-whole-genome-level.html"><a href="calculating-statistics-at-the-whole-genome-level.html#calculating-nucleotide-diversity-statistics"><i class="fa fa-check"></i><b>7.3.2</b> Calculating nucleotide diversity statistics</a></li>
<li class="chapter" data-level="7.3.3" data-path="calculating-statistics-at-the-whole-genome-level.html"><a href="calculating-statistics-at-the-whole-genome-level.html#visualising-nucleotide-diversity-along-the-chromosome"><i class="fa fa-check"></i><b>7.3.3</b> Visualising nucleotide diversity along the chromosome</a></li>
<li class="chapter" data-level="7.3.4" data-path="calculating-statistics-at-the-whole-genome-level.html"><a href="calculating-statistics-at-the-whole-genome-level.html#performing-a-sliding-window-analysis"><i class="fa fa-check"></i><b>7.3.4</b> Performing a sliding window analysis</a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="study-questions-5.html"><a href="study-questions-5.html"><i class="fa fa-check"></i><b>7.4</b> Study questions</a></li>
<li class="chapter" data-level="7.5" data-path="going-further-6.html"><a href="going-further-6.html"><i class="fa fa-check"></i><b>7.5</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="speciation-genomics.html"><a href="speciation-genomics.html"><i class="fa fa-check"></i><b>8</b> Speciation Genomics</a>
<ul>
<li class="chapter" data-level="" data-path="speciation-genomics.html"><a href="speciation-genomics.html#what-to-expect-7"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="" data-path="speciation-genomics.html"><a href="speciation-genomics.html#getting-started-4"><i class="fa fa-check"></i>Getting started</a></li>
<li class="chapter" data-level="8.1" data-path="r-programming-visualizing-complex-data.html"><a href="r-programming-visualizing-complex-data.html"><i class="fa fa-check"></i><b>8.1</b> R programming: Visualizing complex data</a>
<ul>
<li class="chapter" data-level="8.1.1" data-path="r-programming-visualizing-complex-data.html"><a href="r-programming-visualizing-complex-data.html#faceting"><i class="fa fa-check"></i><b>8.1.1</b> Faceting</a></li>
</ul></li>
<li class="chapter" data-level="8.2" data-path="evolutionary-genetics-returning-to-the-sparrow-dataset.html"><a href="evolutionary-genetics-returning-to-the-sparrow-dataset.html"><i class="fa fa-check"></i><b>8.2</b> Evolutionary genetics: Returning to the sparrow dataset</a>
<ul>
<li class="chapter" data-level="8.2.1" data-path="evolutionary-genetics-returning-to-the-sparrow-dataset.html"><a href="evolutionary-genetics-returning-to-the-sparrow-dataset.html#reading-in-the-sparrow-vcf"><i class="fa fa-check"></i><b>8.2.1</b> Reading in the sparrow vcf</a></li>
<li class="chapter" data-level="8.2.2" data-path="evolutionary-genetics-returning-to-the-sparrow-dataset.html"><a href="evolutionary-genetics-returning-to-the-sparrow-dataset.html#examining-the-variant-data"><i class="fa fa-check"></i><b>8.2.2</b> Examining the variant data</a></li>
<li class="chapter" data-level="8.2.3" data-path="evolutionary-genetics-returning-to-the-sparrow-dataset.html"><a href="evolutionary-genetics-returning-to-the-sparrow-dataset.html#setting-up-sliding-windows"><i class="fa fa-check"></i><b>8.2.3</b> Setting up sliding windows</a></li>
<li class="chapter" data-level="8.2.4" data-path="evolutionary-genetics-returning-to-the-sparrow-dataset.html"><a href="evolutionary-genetics-returning-to-the-sparrow-dataset.html#calculating-sliding-window-estimates-of-nucleotide-diversity-and-differentiation"><i class="fa fa-check"></i><b>8.2.4</b> Calculating sliding window estimates of nucleotide diversity and differentiation</a></li>
<li class="chapter" data-level="8.2.5" data-path="evolutionary-genetics-returning-to-the-sparrow-dataset.html"><a href="evolutionary-genetics-returning-to-the-sparrow-dataset.html#visualising-the-data"><i class="fa fa-check"></i><b>8.2.5</b> Visualising the data</a></li>
<li class="chapter" data-level="8.2.6" data-path="evolutionary-genetics-returning-to-the-sparrow-dataset.html"><a href="evolutionary-genetics-returning-to-the-sparrow-dataset.html#interlude-relative-vs.-absolute-measures-of-nucleotide-diversity"><i class="fa fa-check"></i><b>8.2.6</b> Interlude: relative vs. absolute measures of nucleotide diversity</a></li>
<li class="chapter" data-level="8.2.7" data-path="evolutionary-genetics-returning-to-the-sparrow-dataset.html"><a href="evolutionary-genetics-returning-to-the-sparrow-dataset.html#investigating-recombination-rate-variation"><i class="fa fa-check"></i><b>8.2.7</b> Investigating recombination rate variation</a></li>
</ul></li>
<li class="chapter" data-level="8.3" data-path="study-questions-6.html"><a href="study-questions-6.html"><i class="fa fa-check"></i><b>8.3</b> Study questions</a></li>
<li class="chapter" data-level="8.4" data-path="going-further-7.html"><a href="going-further-7.html"><i class="fa fa-check"></i><b>8.4</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="reconstructing-the-past.html"><a href="reconstructing-the-past.html"><i class="fa fa-check"></i><b>9</b> Reconstructing the Past</a>
<ul>
<li class="chapter" data-level="" data-path="reconstructing-the-past.html"><a href="reconstructing-the-past.html#what-to-expect-8"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="" data-path="reconstructing-the-past.html"><a href="reconstructing-the-past.html#getting-started-5"><i class="fa fa-check"></i>Getting started</a></li>
<li class="chapter" data-level="9.1" data-path="phylogenetics-in-r.html"><a href="phylogenetics-in-r.html"><i class="fa fa-check"></i><b>9.1</b> Phylogenetics in R</a>
<ul>
<li class="chapter" data-level="9.1.1" data-path="phylogenetics-in-r.html"><a href="phylogenetics-in-r.html#storing-trees-in-r"><i class="fa fa-check"></i><b>9.1.1</b> Storing trees in R</a></li>
<li class="chapter" data-level="9.1.2" data-path="phylogenetics-in-r.html"><a href="phylogenetics-in-r.html#plotting-trees"><i class="fa fa-check"></i><b>9.1.2</b> Plotting trees</a></li>
<li class="chapter" data-level="9.1.3" data-path="phylogenetics-in-r.html"><a href="phylogenetics-in-r.html#manipulating-trees"><i class="fa fa-check"></i><b>9.1.3</b> Manipulating trees</a></li>
<li class="chapter" data-level="9.1.4" data-path="phylogenetics-in-r.html"><a href="phylogenetics-in-r.html#a-simple-example-with-real-data---avian-phylogenetics"><i class="fa fa-check"></i><b>9.1.4</b> A simple example with real data - avian phylogenetics</a></li>
<li class="chapter" data-level="9.1.5" data-path="phylogenetics-in-r.html"><a href="phylogenetics-in-r.html#constructing-trees-with-r"><i class="fa fa-check"></i><b>9.1.5</b> Constructing trees with R</a></li>
</ul></li>
<li class="chapter" data-level="9.2" data-path="population-structure.html"><a href="population-structure.html"><i class="fa fa-check"></i><b>9.2</b> Population structure</a>
<ul>
<li class="chapter" data-level="9.2.1" data-path="population-structure.html"><a href="population-structure.html#village-dogs-as-an-insight-to-dog-domestication"><i class="fa fa-check"></i><b>9.2.1</b> Village dogs as an insight to dog domestication</a></li>
<li class="chapter" data-level="9.2.2" data-path="population-structure.html"><a href="population-structure.html#reading-the-data-into-r"><i class="fa fa-check"></i><b>9.2.2</b> Reading the data into R</a></li>
<li class="chapter" data-level="9.2.3" data-path="population-structure.html"><a href="population-structure.html#performing-a-pca"><i class="fa fa-check"></i><b>9.2.3</b> Performing a PCA</a></li>
<li class="chapter" data-level="9.2.4" data-path="population-structure.html"><a href="population-structure.html#visualising-the-pca"><i class="fa fa-check"></i><b>9.2.4</b> Visualising the PCA</a></li>
</ul></li>
<li class="chapter" data-level="9.3" data-path="study-questions-7.html"><a href="study-questions-7.html"><i class="fa fa-check"></i><b>9.3</b> Study questions</a></li>
<li class="chapter" data-level="9.4" data-path="going-further-8.html"><a href="going-further-8.html"><i class="fa fa-check"></i><b>9.4</b> Going further</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="advancing-further-in-r.html"><a href="advancing-further-in-r.html"><i class="fa fa-check"></i><b>10</b> Advancing Further in R</a>
<ul>
<li class="chapter" data-level="" data-path="advancing-further-in-r.html"><a href="advancing-further-in-r.html#what-to-expect-9"><i class="fa fa-check"></i>What to expect</a></li>
<li class="chapter" data-level="" data-path="advancing-further-in-r.html"><a href="advancing-further-in-r.html#getting-started-6"><i class="fa fa-check"></i>Getting started</a></li>
<li class="chapter" data-level="10.1" data-path="advanced-features-of-rstudio.html"><a href="advanced-features-of-rstudio.html"><i class="fa fa-check"></i><b>10.1</b> Advanced features of RStudio</a>
<ul>
<li class="chapter" data-level="10.1.1" data-path="advanced-features-of-rstudio.html"><a href="advanced-features-of-rstudio.html#projects-projects-projects"><i class="fa fa-check"></i><b>10.1.1</b> Projects, projects, projects</a></li>
<li class="chapter" data-level="10.1.2" data-path="advanced-features-of-rstudio.html"><a href="advanced-features-of-rstudio.html#everyone-has-a-history"><i class="fa fa-check"></i><b>10.1.2</b> Everyone has a history</a></li>
<li class="chapter" data-level="10.1.3" data-path="advanced-features-of-rstudio.html"><a href="advanced-features-of-rstudio.html#tab-complete-and-other-hotkeys"><i class="fa fa-check"></i><b>10.1.3</b> Tab complete and other hotkeys</a></li>
<li class="chapter" data-level="10.1.4" data-path="advanced-features-of-rstudio.html"><a href="advanced-features-of-rstudio.html#code-checking"><i class="fa fa-check"></i><b>10.1.4</b> Code checking</a></li>
</ul></li>
<li class="chapter" data-level="10.2" data-path="control-flow-for-else-if-and-ifelse.html"><a href="control-flow-for-else-if-and-ifelse.html"><i class="fa fa-check"></i><b>10.2</b> Control flow - <code>for</code>, <code>else</code>, <code>if</code> and <code>ifelse</code></a>
<ul>
<li class="chapter" data-level="10.2.1" data-path="control-flow-for-else-if-and-ifelse.html"><a href="control-flow-for-else-if-and-ifelse.html#if-and-else"><i class="fa fa-check"></i><b>10.2.1</b> <code>if</code> and <code>else</code></a></li>
<li class="chapter" data-level="10.2.2" data-path="control-flow-for-else-if-and-ifelse.html"><a href="control-flow-for-else-if-and-ifelse.html#ifelse---a-single-line-alternative"><i class="fa fa-check"></i><b>10.2.2</b> <code>ifelse</code> - a single line alternative</a></li>
<li class="chapter" data-level="10.2.3" data-path="control-flow-for-else-if-and-ifelse.html"><a href="control-flow-for-else-if-and-ifelse.html#for-loops"><i class="fa fa-check"></i><b>10.2.3</b> <code>for</code> loops</a></li>
</ul></li>
<li class="chapter" data-level="10.3" data-path="vectorisation.html"><a href="vectorisation.html"><i class="fa fa-check"></i><b>10.3</b> Vectorisation</a>
<ul>
<li class="chapter" data-level="10.3.1" data-path="vectorisation.html"><a href="vectorisation.html#why-is-vectorisation-preferable-in-r"><i class="fa fa-check"></i><b>10.3.1</b> Why is vectorisation preferable in R?</a></li>
<li class="chapter" data-level="10.3.2" data-path="vectorisation.html"><a href="vectorisation.html#using-sapply-instead-of-for-on-vectors"><i class="fa fa-check"></i><b>10.3.2</b> Using <code>sapply</code> instead of <code>for</code> on vectors</a></li>
<li class="chapter" data-level="10.3.3" data-path="vectorisation.html"><a href="vectorisation.html#using-apply-instead-of-for-on-matrices-and-dataframes"><i class="fa fa-check"></i><b>10.3.3</b> Using apply instead of <code>for</code> on matrices and dataframes</a></li>
</ul></li>
<li class="chapter" data-level="10.4" data-path="returning-to-writing-your-own-functions.html"><a href="returning-to-writing-your-own-functions.html"><i class="fa fa-check"></i><b>10.4</b> Returning to writing your own functions</a></li>
<li class="chapter" data-level="10.5" data-path="concluding-remarks.html"><a href="concluding-remarks.html"><i class="fa fa-check"></i><b>10.5</b> Concluding remarks</a></li>
<li class="chapter" data-level="10.6" data-path="study-questions-8.html"><a href="study-questions-8.html"><i class="fa fa-check"></i><b>10.6</b> Study questions</a></li>
<li class="chapter" data-level="10.7" data-path="going-further-9.html"><a href="going-further-9.html"><i class="fa fa-check"></i><b>10.7</b> Going further</a></li>
</ul></li>
<li class="part"><span><b>III Assignments</b></span></li>
<li class="chapter" data-level="" data-path="w01.html"><a href="w01.html"><i class="fa fa-check"></i>Week 1-2 assignment</a></li>
<li class="chapter" data-level="" data-path="w03.html"><a href="w03.html"><i class="fa fa-check"></i>Week 3 Assignment</a></li>
<li class="chapter" data-level="" data-path="w04.html"><a href="w04.html"><i class="fa fa-check"></i>Week 4 assignment</a></li>
<li class="chapter" data-level="" data-path="w05.html"><a href="w05.html"><i class="fa fa-check"></i>Week 5 assignment</a></li>
<li class="chapter" data-level="" data-path="w06.html"><a href="w06.html"><i class="fa fa-check"></i>Week 6 assignment</a></li>
<li class="chapter" data-level="" data-path="w07.html"><a href="w07.html"><i class="fa fa-check"></i>Week 7 assignment</a></li>
<li class="chapter" data-level="" data-path="w08.html"><a href="w08.html"><i class="fa fa-check"></i>Week 8 assignment</a></li>
<li class="chapter" data-level="" data-path="w09.html"><a href="w09.html"><i class="fa fa-check"></i>Week 9 assignment</a></li>
<li class="chapter" data-level="" data-path="w10.html"><a href="w10.html"><i class="fa fa-check"></i>Week 10 assignment</a></li>
<li class="appendix"><span><b>Appendix</b></span></li>
<li class="chapter" data-level="A" data-path="rmarkdown.html"><a href="rmarkdown.html"><i class="fa fa-check"></i><b>A</b> About assignments and RMarkdown</a>
<ul>
<li class="chapter" data-level="" data-path="rmarkdown.html"><a href="rmarkdown.html#rmarkdown-1"><i class="fa fa-check"></i>RMarkdown</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Made with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">R tutorials in BIOS1140</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="w01" class="section level1 unnumbered">
<h1>Week 1-2 assignment</h1>
<p>Solve all the problems below using R. Some questions have optional parts that are a little more difficult. We encourage you to do these, but feel free to skip them if the mandatory tasks offers sufficient challenge. You can solve the problems in any way you want (as long as you’re using R!) unless otherwise stated.</p>
<div id="basic-r" class="section level4 unnumbered">
<h4>1. Basic R</h4>
<p><strong>1.1</strong></p>
<ol style="list-style-type: lower-alpha">
<li>Assign a variable with the value 6 to an object named after your favorite animal.<br />
</li>
<li>Assign a variable with the value 1 to an object named after your least favorite animal.<br />
</li>
<li>Calculate the mean of your two objects, and assign that number to an object named after an animal you think is mediocre at best.</li>
<li>Combine all your scores in a vector and assign it to an object. Add names to each element of the vector so that the name corresponds to the score.</li>
<li><strong>Optional:</strong> Combine your scores into a data frame with one column for the name and one for the score. Add a column with the average weight of your animals, and another with their top speeds.</li>
</ol>
<div class="sourceCode" id="cb459"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb459-1"><a href="w01.html#cb459-1" aria-hidden="true" tabindex="-1"></a>grevling <span class="ot">&lt;-</span> <span class="dv">6</span></span>
<span id="cb459-2"><a href="w01.html#cb459-2" aria-hidden="true" tabindex="-1"></a>rur <span class="ot">&lt;-</span> <span class="dv">1</span></span>
<span id="cb459-3"><a href="w01.html#cb459-3" aria-hidden="true" tabindex="-1"></a>torsk <span class="ot">&lt;-</span> <span class="fu">mean</span>(<span class="fu">c</span>(grevling, rur))</span>
<span id="cb459-4"><a href="w01.html#cb459-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb459-5"><a href="w01.html#cb459-5" aria-hidden="true" tabindex="-1"></a>dyr <span class="ot">&lt;-</span> <span class="fu">c</span>(grevling, rur, torsk)</span>
<span id="cb459-6"><a href="w01.html#cb459-6" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(dyr) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;grevling&quot;</span>, <span class="st">&quot;rur&quot;</span>, <span class="st">&quot;torsk&quot;</span>)</span>
<span id="cb459-7"><a href="w01.html#cb459-7" aria-hidden="true" tabindex="-1"></a><span class="co">#dyr</span></span></code></pre></div>
<p><strong>1.2</strong></p>
<p>The formula for the volume of a sphere is <span class="math inline">\(\frac{4}{3}\pi r^3\)</span>.</p>
<ol style="list-style-type: lower-alpha">
<li>Calculate the volume of a sphere with a radius of 4.5 cm</li>
<li>Create a vector of radiuses from 1 to 10, in steps of 0.5. Calculate the volume of spheres with all these radiuses. (<em>hint: in R the easiest way to do this is in a single operation without the use of any loops, see if you can figure it out!</em>)</li>
<li><strong>Optional:</strong> make a function that takes one or more radiuses as input, and outputs the volume of a sphere with those radiuses.</li>
</ol>
<p><strong>1.3</strong></p>
<p>You should use logical operators (<code>==</code>, <code>&gt;</code> etc.) to solve the following tasks.</p>
<ol style="list-style-type: lower-alpha">
<li><p>Check if the number <span class="math inline">\(e\)</span> is larger than the number <span class="math inline">\(\pi\)</span>.</p>
<p>Create a vector of random numbers with the following code:</p>
<div class="sourceCode" id="cb460"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb460-1"><a href="w01.html#cb460-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">645</span>)</span>
<span id="cb460-2"><a href="w01.html#cb460-2" aria-hidden="true" tabindex="-1"></a>rnumbers <span class="ot">&lt;-</span> <span class="fu">rpois</span>(<span class="dv">100</span>, <span class="at">lambda =</span> <span class="dv">2</span>)</span></code></pre></div></li>
<li><p>How many of the numbers are equal to or larger than 2?<br />
<em>Hint: you can use the</em> <code>sum()</code> <em>function on a vector of TRUE/FALSE, it will then treat each TRUE as 1 and each FALSE as 0. Example: </em> <code>sum(c(TRUE, FALSE, TRUE))</code> <em>returns</em> <code>2</code></p></li>
<li><p>How many of the numbers are exactly 1?</p></li>
<li><p><strong>Optional:</strong> Check if R considers <span class="math inline">\((\sqrt2 )^2\)</span> to be exactly equal to 2. What is the result? Why is it this way?</p></li>
</ol>
</div>
<div id="data-import-and-plotting" class="section level4 unnumbered">
<h4>2. Data import and plotting</h4>
<p>For this section you will use the dataset “vitruvian.txt”. The data set consists of body measurements from biology students from earlier years in the course BIO2150. Make sure you download it from Canvas, and take note of where it’s stored on your computer.</p>
<p>For this part of the exercise, in addition to the course’s computing exercises, <a href="https://evengar.github.io/short-tutorials">these short tutorials</a> may be helpful. Go through them before continuing if you’d like, or use them as reference material when you’re unsure about something.</p>
<p><strong>2.1</strong></p>
<p>Importing and inspecting the data</p>
<ol style="list-style-type: lower-alpha">
<li>Open the file “vitruvian.txt” in a plain text editor (like notepad for PC or TextEdit for mac). What kind of file is this? How are the data entries separated?</li>
<li>Import the file into R.</li>
<li>Use tools in R to figure out the following:
<ol style="list-style-type: lower-roman">
<li>are there any missing values (NAs) in this data set?</li>
<li>how many observations (rows) and variables (columns) are there?</li>
<li>which years did the measuring take place?</li>
<li>what is the sex ratio for biology students in this period?</li>
</ol></li>
<li>Remove any rows that contain NA in your data set, leaving you with only complete cases. How many rows did you lose?</li>
</ol>
<p><strong>2.2</strong></p>
<p>Summarizing the data</p>
<ol style="list-style-type: lower-alpha">
<li><p>What is the mean body length of all bio students from all years? What is the standard deviation? What is the median?</p>
<p>A popular claim/myth is that the ratio of body length to the length from foot to navel is approximately the golden ratio (<span class="math inline">\(\approx\)</span> 1.618)</p></li>
<li><p>Check if this is true for the bio students by getting the mean and standard deviation of the ratio between <code>body.length</code> and <code>foot.navel</code>.</p></li>
<li><p>Calculate the mean height of each of the sexes from all years.</p></li>
<li><p>Subset your data to everyone taller than the average for their respective sex. How many males and females are taller than average?</p></li>
<li><p><strong>Optional:</strong> use R’s <code>lm()</code> function to model <code>body.length</code> from <code>foot.navel</code>. What is the slope of the regression? What is the intercept, and how do you interpret that?</p></li>
</ol>
<p><strong>2.3</strong></p>
<p>Before plotting the vitruvian data, we take a detour with some basic plot exercises.</p>
<ol style="list-style-type: lower-alpha">
<li>Make a vector of the numbers 1 through 50, and one with the numbers 51 through 100, call them x and y respectively. Plot them against each other as points.</li>
<li>Plot x against x<sup>2</sup> as a line with any color other than black.</li>
<li>Plot a line of x against <span class="math inline">\(\frac{y^2}{4}\)</span> on top of your previous plot in a different color.</li>
<li>Add a title to your plot, stating how pretty it is.</li>
</ol>
<p><strong>2.4</strong></p>
<p>Now we return to our vitruvian data set. We will more or less be plotting the same things we explored in task 2.2.</p>
<ol style="list-style-type: lower-alpha">
<li>Make a histogram of <code>body.length</code>. What can you say about the resulting distribution?</li>
<li>Plot <code>foot.navel</code> against <code>body.length</code> (so that body length is on the y-axis), and color the points by sex. Add a line with a slope of 1.618. How well does the data fit the golden ratio? Add a title, legend and informative axis labels to your plot.</li>
<li>Make a boxplot of students’ heights for the two sexes. Are there any outliers?</li>
<li>Make a subset of your data that only contains observations from 2012. Make a boxplot of neck size for this year, grouped by sex.</li>
<li><strong>Optional</strong> Use your model from 2.2 e. to add a line of best fit to your plot of <code>foot.navel</code> against <code>body.length</code>. How does it compare to the golden ratio line? Add a confidence interval if you figure out how.</li>
</ol>
</div>
<div id="interpreting-code-and-errors" class="section level4 unnumbered">
<h4>3. Interpreting code and errors</h4>
<p><strong>3.1</strong></p>
<p>Look at the code below:</p>
<div class="sourceCode" id="cb461"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb461-1"><a href="w01.html#cb461-1" aria-hidden="true" tabindex="-1"></a>a <span class="ot">=</span> <span class="dv">5</span></span>
<span id="cb461-2"><a href="w01.html#cb461-2" aria-hidden="true" tabindex="-1"></a>a <span class="sc">==</span> <span class="dv">6</span></span></code></pre></div>
<ol style="list-style-type: lower-alpha">
<li>In your own words, what is the difference between <code>=</code> and <code>==</code> in the code above?</li>
<li>What will the code output? Why?</li>
<li>What is the value of <code>a</code> after executing this code?</li>
</ol>
<p><strong>3.2</strong></p>
<p>You tried to import the file “example.csv”, and got the following error:</p>
<pre><code>## Warning in file(file, &quot;rt&quot;): cannot open file &#39;example.csv&#39;: No such file or
## directory</code></pre>
<pre><code>## Error in file(file, &quot;rt&quot;): cannot open the connection</code></pre>
<p>What is the most probable cause of the error message? How can you fix it?</p>
<p><strong>3.4</strong></p>
<p>You tried to extract the <code>Sepal.Length</code> column from the iris data set as follows, but got an error:</p>
<div class="sourceCode" id="cb464"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb464-1"><a href="w01.html#cb464-1" aria-hidden="true" tabindex="-1"></a>iris[,Sepal.Length]</span></code></pre></div>
<pre><code>## Error in `[.data.frame`(iris, , Sepal.Length): object &#39;Sepal.Length&#39; not found</code></pre>
<p>What caused the error? How can you fix it?</p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="going-further-9.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="w03.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
